%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%DIF LATEXDIFF DIFFERENCE FILE
%DIF DEL v2.0.1/main.tex     Wed Sep 30 22:24:14 2015
%DIF ADD Document/main.tex   Wed Sep 30 22:05:01 2015
% Masters/Doctoral Thesis
% LaTeX Template
% Version 1.43 (17/5/14)
%
% This template has been downloaded from:
% http://www.LaTeXTemplates.com
%
% Original authors:
% Steven Gunn
% http://users.ecs.soton.ac.uk/srg/softwaretools/document/templates/
% and
% Sunil Patel
% http://www.sunilpatel.co.uk/thesis-template/
%
% License:
% CC BY-NC-SA 3.0 (http://creativecommons.org/licenses/by-nc-sa/3.0/)
%
% Note:
% Make sure to edit document variables in the Thesis.cls file
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%----------------------------------------------------------------------------------------
%	PACKAGES AND OTHER DOCUMENT CONFIGURATIONS
%----------------------------------------------------------------------------------------
\documentclass[11pt, oneside]{Thesis} % The default font size and one-sided printing (no margin offsets)

\graphicspath{{Pictures/}} % Specifies the directory where pictures are stored

\usepackage[square,sort,comma,numbers]{natbib} % Use the natbib reference package - read up on this to edit the reference style; if you want text (e.g. Smith et al., 2012) for the in-text references (instead of numbers), remove 'numbers'
\hypersetup{urlcolor=black, colorlinks=false} % Colors hyperlinks in blue - change to black if annoying
\title{\ttitle} % Defines the thesis title - don't touch this

\input{ushyphex}
%DIF 36c36-37
%DIF < \usepackage[compact]{titlesec}
%DIF -------
 %DIF > 
\usepackage[lining,semibold]{libertine} %DIF > 
%DIF -------
\usepackage[T1]{fontenc}
%DIF 38a39-47
\usepackage{textcomp} %DIF > 
\renewcommand*\ttdefault{txtt} %DIF > 
\usepackage{amsthm} %DIF > 
\usepackage[libertine,cmintegrals,bigdelims,vvarbb]{newtxmath} %DIF > 
\usepackage[scr=rsfso]{mathalfa} %DIF > 
\usepackage{bm} %DIF > 
\useosf %DIF > 
 %DIF > 
\usepackage[compact]{titlesec} %DIF > 
%DIF -------
\usepackage{algorithm}
\usepackage{algpseudocode}
\usepackage{array}
\usepackage{booktabs}
\usepackage{caption}
\usepackage{datetime}
%DIF 44c54
%DIF < \usepackage{lmodern}
%DIF -------
\usepackage{icomma} %DIF > 
%DIF -------
\usepackage{mathtools}
\usepackage{paralist}
\usepackage{pdflscape}
\usepackage{rnamacros}
\usepackage{seqsplit}
\usepackage{silence}
\usepackage{subcaption}
\usepackage{tabularx}
%DIF 53a63
\usepackage{textcomp} %DIF > 
%DIF -------
\usepackage{url}
\usepackage{xfrac}
\usepackage{xpatch}

\WarningFilter{latex}{Text page}
\WarningFilter{latex}{Float too large for page by}

%DIF 60a71-74
\newcommand{\completionDate}{\formatdate{1}{12}{2015}} %DIF > 
\newdateformat{monthyeardate}{\monthname[\THEMONTH] \THEYEAR} %DIF > 
\newdateformat{yeardate}{\THEYEAR} %DIF > 
 %DIF > 
%DIF -------
\makeatletter
\xpatchcmd{\algorithmic}{\itemsep\z@}{\itemsep=.75ex plus1pt}{}{}

% \def\env@cases{
%   \let\@ifnextchar\new@ifnextchar
%   \left\lbrace
%   \def\arraystretch{.7}
%   \array{@{}l@{\quad}l@{}}
% }
%DIF < % \makeatother
%DIF -------
\makeatother %DIF > 
%DIF PREAMBLE EXTENSION ADDED BY LATEXDIFF
%DIF UNDERLINE PREAMBLE %DIF PREAMBLE
\RequirePackage[normalem]{ulem} %DIF PREAMBLE
\RequirePackage{color}\definecolor{RED}{rgb}{1,0,0}\definecolor{BLUE}{rgb}{0,0,1} %DIF PREAMBLE
\providecommand{\DIFadd}[1]{{\protect\color{blue}\uwave{#1}}} %DIF PREAMBLE
\providecommand{\DIFdel}[1]{{\protect\color{red}\sout{#1}}}                      %DIF PREAMBLE
%DIF SAFE PREAMBLE %DIF PREAMBLE
\providecommand{\DIFaddbegin}{} %DIF PREAMBLE
\providecommand{\DIFaddend}{} %DIF PREAMBLE
\providecommand{\DIFdelbegin}{} %DIF PREAMBLE
\providecommand{\DIFdelend}{} %DIF PREAMBLE
%DIF FLOATSAFE PREAMBLE %DIF PREAMBLE
\providecommand{\DIFaddFL}[1]{\DIFadd{#1}} %DIF PREAMBLE
\providecommand{\DIFdelFL}[1]{\DIFdel{#1}} %DIF PREAMBLE
\providecommand{\DIFaddbeginFL}{} %DIF PREAMBLE
\providecommand{\DIFaddendFL}{} %DIF PREAMBLE
\providecommand{\DIFdelbeginFL}{} %DIF PREAMBLE
\providecommand{\DIFdelendFL}{} %DIF PREAMBLE
%DIF END PREAMBLE EXTENSION ADDED BY LATEXDIFF

\begin{document}

\DIFdelbegin %DIFDELCMD < \frontmatter %%%
%DIF <  Use roman page numbering style (i, ii, iii, iv...) for the pre-content pages
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend \setstretch{1.3} % Line spacing of 1.3

% Define the page headers using the FancyHdr package and set up for one-sided printing
\fancyhead{} % Clears all page headers and footers
\rhead{\thepage} % Sets the right side header to show the page number
\lhead{} % Clears the left side page header

\pagestyle{fancy} % Finally, use the "fancy" page style to implement the FancyHdr headers

\DIFdelbegin %DIFDELCMD < \newcommand{\HRule}{\rule{\linewidth}{0.5mm}} %%%
\DIFdelend \DIFaddbegin \newcommand{\HRule}{\rule{\linewidth}{0.25mm}} \DIFaddend % New command to make the lines in the title page

% PDF meta-data
\hypersetup{pdftitle={\ttitle}}
\hypersetup{pdfsubject=\subjectname}
\hypersetup{pdfauthor=\authornames}
\hypersetup{pdfkeywords=\keywordnames}

%----------------------------------------------------------------------------------------
%	TITLE PAGE
%----------------------------------------------------------------------------------------
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend \begin{titlepage}
\begin{center}

\textsc{\LARGE \univname}\\[1.5cm] % University name
\textsc{\Large Doctoral Thesis}\\[0.5cm] % Thesis type

\HRule \\[0.4cm] % Horizontal line
{\large \DIFdelbegin %DIFDELCMD < \bfseries \ttitle%%%
\DIFdelend \DIFaddbegin \textsc{\ttitle}\DIFaddend }\\[0.225cm] % Thesis title
\HRule \\[1.5cm] % Horizontal line

\begin{minipage}{0.4\textwidth}
\begin{flushleft} \large
\emph{Author:}\\
\href{http://www.evansenter.com}{\authornames} % Author name - remove the \href bracket to remove the link
\end{flushleft}
\end{minipage}
\begin{minipage}{0.4\textwidth}
\begin{flushright} \large
\emph{Supervisor:} \\
\href{http://clavius.bc.edu/~clote/}{\supname} % Supervisor name - remove the \href bracket to remove the link
\end{flushright}
\end{minipage}\\[3cm]

\large \textit{A thesis submitted in \DIFaddbegin \DIFadd{partial }\DIFaddend fulfilment of the requirements\\ for the degree of \degreename}\\[0.3cm] % University requirement text
\textit{in the}\\[0.4cm]
\groupname\\\deptname\DIFaddbegin \\\subcollege\DIFaddend \\[2cm] % Research group name and department name

{\large \DIFdelbegin %DIFDELCMD < \today%%%
\DIFdelend \DIFaddbegin \monthyeardate\completionDate\DIFaddend }\\[4cm] % Date
%\includegraphics{Logo} % University/department logo - uncomment to place it

\vfill
\DIFdelbegin %DIFDELCMD < \end{center}
%DIFDELCMD < %%%
\DIFdelend 

\DIFaddbegin \end{center}
\DIFaddend \end{titlepage}

\DIFaddbegin \begin{bottompar}
  \DIFadd{\textcopyright}\xspace \DIFadd{by }\authornames \\
  \yeardate\completionDate
\end{bottompar}

\DIFaddend %----------------------------------------------------------------------------------------
%DIF < 	DECLARATION PAGE
%DIF < 	Your institution may give you a different text to place here
%DIF >  ABSTRACT
%----------------------------------------------------------------------------------------
%DIF >  \addtotoc{Abstract} % Add the "Abstract" page entry to the Contents

\DIFdelbegin %DIFDELCMD < \Declaration{
%DIFDELCMD < 

%DIFDELCMD < \addtocontents{toc}{\vspace{1em}} % Add a gap in the Contents, for aesthetics
%DIFDELCMD < 

%DIFDELCMD < I, \authornames, declare that this thesis titled, `\ttitle' and the work presented in it are my own. I confirm that:
%DIFDELCMD < 

%DIFDELCMD < \begin{itemize}
%DIFDELCMD < \item This work was done wholly or mainly while in candidature for a research degree at this University.
%DIFDELCMD < \item Where any part of this thesis has previously been submitted for a degree or any other qualification at this University or any other institution, this has been clearly stated.
%DIFDELCMD < \item Where I have consulted the published work of others, this is always clearly attributed.
%DIFDELCMD < \item Where I have quoted from the work of others, the source is always given. With the exception of such quotations, this thesis is entirely my own work.
%DIFDELCMD < \item I have acknowledged all main sources of help.
%DIFDELCMD < \item Where the thesis is based on work done by myself jointly with others, I have made clear exactly what was done by others and what I have contributed myself.\\
%DIFDELCMD < \end{itemize}
%DIFDELCMD < 

%DIFDELCMD < Signed:\\
%DIFDELCMD < \rule[1em]{25em}{0.5pt} % This prints a line for the signature
%DIFDELCMD < 

%DIFDELCMD < Date:\\
%DIFDELCMD < \rule[1em]{25em}{0.5pt} % This prints a line to write the date
%DIFDELCMD < }
%DIFDELCMD < %%%
\DIFdelend \DIFaddbegin \abstract{
RNA folding pathways play an important role in various biological processes, such as
\begin{inparaenum}[\em 1\upshape)]
\item the conformational switch in spliced leader RNA from {\em Leptomonas collosoma},
which controls trans-splicing of a portion of the $5$' exon, and
\item \rbs--portions of the $5$' untranslated region of mRNA that regulate genes by allostery.
\end{inparaenum}
Since RNA folding pathways are determined by the thermodynamic landscape, we have developed a number
of novel algorithms---including \fftbor and \ffttwo---which efficiently compute the coarse-grained
energy landscape for a given RNA sequence.
These energy landscapes can then be used to produce a model for RNA folding kinetics
that can compute both the \mfpt (MFPT) and \eqt in a deterministic and efficient manner,
using a new software package we call \hermes. The speed of the software provided within
\hermes---namely \fftmfpt and \ffteq---present what we believe to be the first suite of kinetic analysis
tools for RNA sequences that are suitable for high throughput usage, something
we believe to be of interest in the field of synthetic design.
}
\DIFaddend 

\clearpage % Start a new page

%DIF < ----------------------------------------------------------------------------------------
%DIF < 	QUOTATION PAGE
%DIF < ----------------------------------------------------------------------------------------
\DIFaddbegin \frontmatter %DIF >  Use roman page numbering style (i, ii, iii, iv...) for the pre-content pages
\pagestyle{fancy} %DIF >  The page style headers have been "empty" all this time, now use the "fancy" headers as defined before to bring them back
\DIFaddend 

%DIF <  \pagestyle{empty} % No headers or footers for the following pages
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \null\vfill % Add some space to move the quote down the page a bit
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \textit{``Thanks to my solid academic training, today I can write hundreds of words on virtually any topic without possessing a shred of information, which is how I got a good job in journalism."}
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \begin{flushright}
%DIF <  Dave Barry
%DIF <  \end{flushright}
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \vfill\vfill\vfill\vfill\vfill\vfill\null % Add some space at the bottom to position the quote just right
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \clearpage % Start a new page
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend %----------------------------------------------------------------------------------------
%DIF < 	ABSTRACT PAGE
%DIF < ----------------------------------------------------------------------------------------
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \addtotoc{Abstract} % Add the "Abstract" page entry to the Contents
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \abstract{\addtocontents{toc}{\vspace{1em}} % Add a gap in the Contents, for aesthetics
%DIF <  Abstract
%DIF <  }
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \clearpage % Start a new page
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF < ----------------------------------------------------------------------------------------
\DIFdelend %	ACKNOWLEDGEMENTS
%----------------------------------------------------------------------------------------
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend \setstretch{1.3} % Reset the line-spacing to 1.3 for body text (if it has changed)

\DIFdelbegin %DIFDELCMD < \acknowledgements{\addtocontents{toc}{\vspace{1em}} % Add a gap in the Contents, for aesthetics
%DIFDELCMD < First and foremost, I would like to thank my advisor, Prof. Peter Clote, for the
%DIFDELCMD < countless hours spent explaining the nuances of the computational RNA field.
%DIFDELCMD < With all the guidance and support he provided over my graduate school career,
%DIFDELCMD < I could not have asked for a better advisor.
%DIFDELCMD < 

%DIFDELCMD < Additionally, I'd like to thank my family and friends for their support and
%DIFDELCMD < encouragement, and the members of my thesis committee for their insights
%DIFDELCMD < along the way.
%DIFDELCMD < }
%DIFDELCMD < %%%
\DIFdelend \DIFaddbegin \acknowledgements{
First and foremost, I would like to thank my advisor, Prof. Peter Clote, for the
countless hours spent explaining the nuances of the computational RNA field.
With all the guidance and support he provided over my graduate school career,
I could not have asked for a better advisor.

Additionally, I'd like to thank my family and friends for their support and
encouragement, and the members of my thesis committee for their insights
along the way.
}

\DIFaddend \clearpage % Start a new page

%----------------------------------------------------------------------------------------
%	LIST OF CONTENTS/FIGURES/TABLES PAGES
%----------------------------------------------------------------------------------------

\DIFdelbegin %DIFDELCMD < \pagestyle{fancy} %%%
%DIF <  The page style headers have been "empty" all this time, now use the "fancy" headers as defined before to bring them back
%DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend \lhead{\emph{Contents}} % Set the left side page header to "Contents"
\tableofcontents % Write out the Table of Contents

\lhead{\emph{List of Figures}} % Set the left side page header to "List of Figures"
\listoffigures % Write out the List of Figures

\lhead{\emph{List of Tables}} % Set the left side page header to "List of Tables"
\listoftables % Write out the List of Tables

%----------------------------------------------------------------------------------------
%DIF < 	ABBREVIATIONS
%DIF < ----------------------------------------------------------------------------------------
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \clearpage % Start a new page
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \setstretch{1.5} % Set the line spacing to 1.5, this makes the following tables easier to read
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \lhead{\emph{Abbreviations}} % Set the left side page header to "Abbreviations"
%DIF <  \listofsymbols{ll} % Include a list of Abbreviations (a table of two columns)
%DIF <  {
%DIF <  \textbf{LAH} & \textbf{L}ist \textbf{A}bbreviations \textbf{H}ere \\
%DIF <  %\textbf{Acronym} & \textbf{W}hat (it) \textbf{S}tands \textbf{F}or \\
%DIF <  }
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF < ----------------------------------------------------------------------------------------
%DIF < 	DEDICATION
%DIF < ----------------------------------------------------------------------------------------
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \setstretch{1.3} % Return the line spacing back to 1.3
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \pagestyle{empty} % Page style needs to be empty for this page
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \dedicatory{For/Dedicated to/To my\ldots} % Dedication text
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF <  \addtocontents{toc}{\vspace{2em}} % Add a gap in the Contents, for aesthetics
%DIFDELCMD < 

%DIFDELCMD < %%%
%DIF < ----------------------------------------------------------------------------------------
\DIFdelend %	THESIS CONTENT - CHAPTERS
%----------------------------------------------------------------------------------------
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend % \setstretch{2}
\setstretch{1.3}

\mainmatter % Begin numeric (1,2,3...) page numbering

\pagestyle{fancy} % Return the page headers back to the "fancy" style

\setlength{\abovedisplayskip}{0pt}
\setlength{\abovedisplayshortskip}{0pt}

%!TEX root = ../main.tex

\chapter{Introduction}
\label{ch:intro}

\lhead{Introduction}

Introduced in \DIFdelbegin \DIFdel{1958}\DIFdelend \DIFaddbegin \DIFadd{$1958$}\DIFaddend , the central dogma of biology has been an excellent model for
the biological flow of information, much as Newtonian classical mechanics stood the
test of time for over \DIFdelbegin \DIFdel{200 }\DIFdelend \DIFaddbegin \DIFadd{$200$ }\DIFaddend years. But just as Einstein’s revolutionary principle of
relativity has upended our understanding of space in a way unheard of since
Copernicus, recent research has gone to confirm that for all our scientific
progress, the cell still holds fundamental mysteries, and even the central dogma
isn't sacred. Even Francis Crick himself indicated in \citep{crick:1970wb} that
ribonucleic acids (RNAs) likely had a role beyond the traditional {\em messenger}
intermediary between
DNA and proteins, as evidenced by viral RNAs \citep{coffin:1997ws}. Though recent
research continues to suggest that the genome is pervasively transcribed, current
estimates indicate that only 1.2\% of the mammalian genome constitutes
protein-coding sequences \citep{berretta:2009tq,clark:2011cc,jensen:2013vb}. The
study of the transcriptome has led to the identification of a wide variety of
non-coding RNAs (ncRNAs) that highlight the diversity of roles for which RNA can
be put to use \citep{costa:2005ug}. Now understood to be much more than the
intermediary step between DNA and proteins, RNAs have been implicated in a variety
of regulatory and enzymatic activities, including gene knockdown and silencing
\citep{fire:1998tv,mccaffrey:2002tf,hannon:2002vn,he:2004uk},
transcriptional and translational regulation \citep{nudler:2004vm,mandal:2004vh},
intronic splicing \citep{kruger:1982wk,cech:1990tn}, cite-specific cleavage
\citep{doherty:2001wq}, and more.
A prevailing theory now suggests that self-replicating RNA molecules were the
predecessors to all life on Earth---the RNA world hypothesis \citep{gilbert:1986td}.
As our appreciation of RNA diversity has increased,
significant effort has been put forth by the scientific community to understand
and characterize the properties of these molecules.

Unlike DNA, RNA is generally single stranded and thus able to interact with itself
to form interesting shapes with various functional characteristics, akin to
proteins. RNA is a polymer \DIFdelbegin \DIFdel{comprised }\DIFdelend \DIFaddbegin \DIFadd{composed }\DIFaddend of four monomer {\em building blocks}:
the purines
adenine (A) and guanine (G); and the pyrimidines cytosine (C) and uracil (U). These
nucleotides can form planar base pairs \DIFdelbegin \DIFdel{comprised }\DIFdelend \DIFaddbegin \DIFadd{composed }\DIFaddend of energetically favorable
hydrogen bonds, the stacking of which produces a stable {\em helix} structure
\citep{yakovchuk:2006bm}. There are only a select set of possible
\DIFaddbegin \DIFadd{commonly occuring }\DIFaddend base pairs; the
Watson-Crick pairs (A-U, G-C) or the G-U wobble pair. Given an arbitrary RNA
sequence $\seq = \seqN$, where $s_i \in \{\text{A,\,U,\,G,\,C}\}$, we can define
a secondary structure \str for \seq as the set of index tuples indicating those
bases involved in a base pair within \seq. Again like proteins, RNA molecules tend
to fold into a `native' conformation, usually that which minimizes free energy.
While protein folding is predominantly motivated by hydrophobic
interactions, RNA structure is driven by stacking base pair interactions, and
therefore secondary structure tends to be a much better predictor for the function
of the molecule in question than is the case with proteins, whose function is
largely determined by 3D `tertiary' structure.

From a computational perspective, the history of RNA folding is far too long for
proper treatment within this introduction. Instead we will touch upon just some
of the major milestones to get a flavor for what progress has been made.
In \DIFdelbegin \DIFdel{1978, Michael Waterman presented the first }\DIFdelend \DIFaddbegin \DIFadd{$1960$, the first modern model of RNA secondary structure was presented
\mbox{%DIFAUXCMD
\citep{fresco:1960vg}}%DIFAUXCMD
, followed by
Michael Waterman's }\DIFaddend graph-theoretic model of single stranded
nucleic acids such as RNA \DIFaddbegin \DIFadd{in $1978$ }\DIFaddend \citep{waterman:1978va}. This was followed in \DIFdelbegin \DIFdel{1980 }\DIFdelend \DIFaddbegin \DIFadd{$1980$ }\DIFaddend by
the work of Ruth Nussinov and Ann Jacobson, who together presented an algorithm for
determining the maximally matching secondary structure \str for
a given RNA sequence \citep{nussinov1980}, using dynamic programming
\citep{bellman:1952vza}. In the following years Michael Zuker and Patrick Stiegler
developed an algorithm and accompanying software for the \mfe
formulation of the problem \citep{zuker:1981tf,zuker:1989im}. In \DIFdelbegin \DIFdel{1990}\DIFdelend \DIFaddbegin \DIFadd{$1990$}\DIFaddend , John
McCaskill showed how dynamic programming could be used to compute the
partition function for an RNA molecule, and even compute the probability that an
arbitrary base is bound \citep{mccaskill}. Alongside these early developments,
more robust energy models were experimentally derived \citep{turner,turner:2009vy},
further improving the accuracy of computational models.

Fast-forwarding to today, there is now a huge collection of software aimed at
computing various properties of RNAs, be it folding, inverse folding, kinetics,
design, and more. The work that we present here intends to be a contribution
to the diverse toolset that researchers have at their disposal for the analysis
and design of both existing and novel RNA sequences.

\section{Thesis Content}
\label{sec:intro:thesiscontent}

The work of this thesis is based on the following four journal articles, alongside
unpublished data and observations. The journal articles constituting the
primary body of research include:

\begin{itemize}
\setstretch{1.3}
\item Senter, E., Sheikh, S., Dotu, I., Ponty, Y., \& Clote, P. (2012). Using the Fast Fourier Transform to Accelerate the Computational Search for RNA Conformational Switches. PloS One, 7(12), e50506. \url{http://doi.org/10.1371/journal.pone.0050506}
\item Ding, Y., Lorenz, W. A., Dotu, I., Senter, E., \& Clote, P. (2014). Computing the Probability of RNA Hairpin and Multiloop Formation. Journal of Computational Biology : a Journal of Computational Molecular Cell Biology, 21(3), 201–218. \url{http://doi.org/10.1089/cmb.2013.0148}
\item Senter, E., Dotu, I., \& Clote, P. (2014). RNA folding pathways and kinetics using \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend energy landscapes. Journal of Mathematical Biology, 70(1-2), 173–196. \url{http://doi.org/10.1007/s00285-014-0760-4}
\item Senter, E., \& Clote, P. (2015). Fast, Approximate Kinetics of RNA Folding. Journal of Computational Biology : a Journal of Computational Molecular Cell Biology, 22(2), 124–144. \url{http://doi.org/10.1089/cmb.2014.0193}
\end{itemize}

Text, figures, and tables from these papers are used throughout this thesis without
additional notice.

\section{Thesis Organization}
\label{sec:intro:thesisorg}

The remainder of this thesis is organized in the following fashion. We begin in
\Chref{rfinder} with the presentation of \rfinder, a pipeline of
software intended to facilitate the detection of full \rb sequences
alongside their corresponding `on' and `off' structures in genomic data. In
\Chref{fftbor} we introduce the program \fftbor, which computes---for
each integer $k$---the Boltzmann probability \pk of the subensemble of structures
whose \bpd to an input reference structure \str is $k$.
In \Chref{ffttwo} we extend this idea to simultaneously consider two
reference structures \strST and produce as a result the coarse-grained \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend energy
landscape where---for each integer pair $x,y$---we compute the Boltzmann
probability $p_{x,y}$
of those structures whose \bpd from \strS [resp. \strTresp] is $x$
[resp. $y$]. This program---\ffttwo---allows for the efficient approximation of
kinetic characteristics of RNA molecules, presented in \Chref{hermes}
through the software package \hermes. Finally in \Chref{disc} we conclude
with a summary of this work as a whole, and consider its place in the greater
ecosystem of computational RNA tools.


%DIF <  \input{Chapters/Background}
%!TEX root = ../main.tex

\newcommand{\fnaRetrievalTime}{\formatdate{25}{11}{2014} at \formattime{9}{14}{0}}

\chapter{Ribofinder}
\label{ch:rfinder}

\lhead{Ribofinder: A \Rb Detection Pipeline}

\section{Introduction}
\label{sec:rfinder:intro}

In this chapter, we present the \rfinder program---a pipeline to facilitate the
detection of putative \grbs across genomic data. The \rfinder
tool operates in three stages. First we use \infernal
\citep{infernal,nawrocki:2013hk} and \tthp \citep{ermolaeva:2000cl} to detect
putative aptamers and expression platforms, two distinct components of
\rbs described in \Secref{sec:rfinder:bkgrnd}. After coalescing
this data into a pool of candidate \rbs, we use \rfold \citep{lorenz.amb11}
with constraints based on experimental data to compute the two distinct structural
conformations---gene `on' and gene `off'. In the third and final stage, we
leverage \foldalign \citep{gorodkin:1997tr,havgaard:2007ca} to measure the similarity between our
candidate pool and a
canonical \grb well studied in the literature, the
xanthine phosphoribosyltransferase (xpt) \grb from {\em Bacillus subtilis}.
\DIFdelbegin \DIFdel{At the
time of this writing, Prof. Dr. Mario
M\"orl at Universit\"at Leipzig is overseeing preliminary structural
validation of predicted gene `on' and off structures for a number of
computationally predicted }%DIFDELCMD < \grbs %%%
\DIFdel{in }%DIFDELCMD < {\em %%%
\DIFdel{Bacillus megaterium}%DIFDELCMD < }%%%
\DIFdel{. We have already
prepared a draft of the paper describing this work, and are presently awaiting
biochemical validation before submitting it with co-authors M. M\"orl and
coworkers.
}\DIFdelend 

\subsection{Organization}
\label{subsec:rfinder:org}

This chapter is organized in the following fashion. After providing background
on the structural components of a \rb alongside their biological
significance, we outline the deficiencies in the `state of the art' software
as it relates specifically to \rb detection. We then move on to outline
the three stages of \rfinder: candidate selection, structural prediction, and
candidate curation. Having described the approach of the software, we move on
to present our findings in using \rfinder to detect \grbs across
the bacterial RefSeq database. Finally, we provide brief commentary on possible
extensions of the algorithm to locate other flavors of \rbs, of which
adenine-sensitive aptamers are a straightforward extension.

\section{Background}
\label{sec:rfinder:bkgrnd}

\Rbs are regulatory mRNA elements that modulate gene expression via
structural changes induced by the direct sensing of a small-molecule metabolite.
Most often found in bacteria, \rbs regulate diverse pathways including the
metabolism and transport of purines, methionine, and thiamin amongst others. The
structure of a \rb includes an aptamer domain---involved in the direct
sensing of the small-molecule---and a downstream expression platform whose
structure changes upon the aptamer binding the metabolite. Because of the
discriminatory nature of metabolite sensing, groups have had great success in
finding representative examples of aptamers across a diverse collection of
bacterial species; Rfam \DIFdelbegin \DIFdel{12.0 currently contains 26 }\DIFdelend \DIFaddbegin \DIFadd{$12.0$ currently contains $26$ }\DIFaddend different families of aptamers
involved in different metabolic pathways.

Tools such as \Rb finder \citep{bengert2004} and \ms{RiboSW}
\citep{chang:2009de} have used the conserved
structural characteristics of the aptamer as search criteria with promising
results, while the webserver \ms{RibEx} relies solely on sequence conservation
to detect putative aptamer domains in genomic data \citep{abreugoodger:2005hb}.
Meanwhile other groups have used covariance model (CM)-based approaches to find
aptamer domains, most notably \ms{CMfinder} \citep{yao2006}\DIFdelbegin \DIFdel{and Rfam
\mbox{%DIFAUXCMD
\citep{nawrocki:2014uy}}%DIFAUXCMD
}\DIFdelend .

Whereas there exists strong sequence and
structural similarity within the aptamer of a \rb family, the expression
platform is highly variable, and thus challenging to capture using traditional
SCFG-based approaches. For this reason databases such as Rfam only contain the aptamer
portion of the \rb, and there exists no database providing sequences
including expression platforms, necessary for capturing the `on' and `off'
conformations of this regulatory element. We have developed a new
pipeline---called \rfinder---which can detect putative \rbs including their
expression platforms and likely conformational structures across a wide collection
of genomic sequences.

\section{The \rfinder pipeline}
\label{sec:rfinder:pipeline}

At the time of our retrieval (\fnaRetrievalTime), the RefSeq database hosted by
NCBI comprised \DIFdelbegin \DIFdel{5,121 }\DIFdelend \DIFaddbegin \DIFadd{$5,121$ }\DIFaddend complete bacterial genomes with corresponding genomic
annotations. In order to both detect putative full \rbs across this
collection of data as well as filter the candidates down to a number tractable for
experimental validation, we developed a novel pipeline which takes a three-tiered
approach to candidate selection. Our approach is to
\begin{inparaenum}[\em 1\upshape)]
\item identify a pool of candidate \rbs across genomic data;
\item perform a coarse-grained filtering of the candidate pool based on structural
characteristics; and finally
\item fine-grained curation of the candidates based on a collection of measures
and pairwise similarity.
\end{inparaenum} \Figref{rfinder:flowchart} outlines this approach as a
flowchart.

\begin{figure}[!ht]
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/ribofinderOverview.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption{%
{%DIFAUXCMD
\DIFdelFL{Outline of the approach for the }%DIFDELCMD < \rfinder %%%
\DIFdelFL{pipeline.}}
%DIFAUXCMD
\DIFdelendFL \DIFaddbeginFL \caption[Outline of the approach for the \rfinder pipeline]
{\DIFaddFL{Outline of the approach for the }\rfinder \DIFaddFL{pipeline.}}
\DIFaddendFL \label{fig:rfinder:flowchart}
\end{figure}

In the following discussion, we describe the application of \rfinder to identify
unannotated guanine purine \rbs; guanine-sensing cis-regulatory elements
which modulate the expression of genes involved in purine biosynthesis.

\subsection{Step 1: Candidate selection}
\label{subsec:rfinder:selection}

The RefSeq data we used for analysis (downloaded on \fnaRetrievalTime)
contains \DIFdelbegin \DIFdel{5,121 }\DIFdelend \DIFaddbegin \DIFadd{$5,121$ }\DIFaddend annotated bacterial genomes
across \DIFdelbegin \DIFdel{2,732 }\DIFdelend \DIFaddbegin \DIFadd{$2,732$ }\DIFaddend different organisms, totaling over \DIFdelbegin \DIFdel{$9.5 * 10^9$ }\DIFdelend \DIFaddbegin \DIFadd{$9.5 \times 10^9$ }\DIFaddend bases. We used the
program \infernal to determine the coordinates of putative aptamer structures
within the RefSeq genomes, and \tthp to locate candidate rho-independent
transcription terminators.

\subsubsection{Detecting aptamers with \infernal}
\label{subsubsec:rfinder:infernal}

\infernal \DIFdelbegin \DIFdel{\mbox{%DIFAUXCMD
\citep{infernal}}%DIFAUXCMD
, \mbox{%DIFAUXCMD
\citep{nawrocki:2013hk} }%DIFAUXCMD
}\DIFdelend \DIFaddbegin \DIFadd{\mbox{%DIFAUXCMD
\citep{infernal,nawrocki:2013hk} }%DIFAUXCMD
}\DIFaddend uses a stochastic context-free
grammar (SCFG) with a user-provided multiple sequence alignment (MSA) to
efficiently scan genomic data for RNA homologs, taking into consideration both
sequence and structural conservation. Using the purine aptamer MSA from Rfam \DIFdelbegin \DIFdel{12.0
(RF00167}\DIFdelend \DIFaddbegin \DIFadd{$12.0$
(RF$00167$}\DIFaddend ), \infernal (\DIFdelbegin \DIFdel{v1.1.1}\DIFdelend \DIFaddbegin \DIFadd{v$1.1.1$}\DIFaddend , default options) detects \DIFdelbegin \DIFdel{1,537 }\DIFdelend \DIFaddbegin \DIFadd{$1,537$ }\DIFaddend significant hits
having E-value $<= 0.01$. Because \infernal leverages the concept of a
`local end'---a large insertion or deletion in the alignment at reduced cost---it
is possible for the software to return a significant hit whose aligned structure
does not have the canonical three-way junction observed in all purine
\rbs. \rfinder prunes these truncated \infernal hits by converting the
alignment structure into a parse tree, and only permitting trees of sufficient
complexity to contain a multiloop (described further in
\ref{subsubsec:rfinder:shapes}). The pyrimidine residue abutted next to the P1
stem in the J3--1 junction differentiates between guanine and adenine-sensing
\rbs by binding the complimentary purine ligand; for our interest in \grbs
exclusively we require the presence of a cytidine at this residue
(\Figref{rfinder:aptamerDiagram}). In total, using
\infernal with these additional filters yields \DIFdelbegin \DIFdel{1,280 }\DIFdelend \DIFaddbegin \DIFadd{$1,280$ }\DIFaddend guanine aptamers across \DIFdelbegin \DIFdel{555
}\DIFdelend \DIFaddbegin \DIFadd{$555$
}\DIFaddend unique organisms.

\begin{figure}[!ht]
\centering
\includegraphics[width=.4\textwidth]{Figures/Ribofinder/NC_000964_3_2320197_2320054_APTAMER.pdf}
\caption[Diagram of the aptamer portion of the \Bsxpt \grb]{Diagram of the aptamer portion of the \Bsxpt \grb, with annotations
to indicate the P1, P2, and P3 stems of the multiloop, junctions between the
hairpins, and the ligand binding site (in yellow). This structural diagram was
generated using VARNA \citep{darty:2009gt}.}
\label{fig:rfinder:aptamerDiagram}
\end{figure}

\subsubsection{Detecting expression platforms with \tthp}
\label{subsubsec:rfinder:tthp}

\tthp \citep{ermolaeva:2000cl} detects rho-independent terminators in bacterial
genomes in a context-sensitive fashion by leveraging the protein annotations
available in NCBI Protein Table (PTT) data. These terminator sequences canonically have a stable
hairpin loop structure immediately preceding a run of $5+$ uracil residues, the
combination of which causes \DIFdelbegin \DIFdel{the ribosomal machinery }\DIFdelend \DIFaddbegin \DIFadd{RNA polymerase }\DIFaddend to stall and dissociate from
the transcript. \tthp performs a genomic scan to determine candidate loci with
this motif, and returns scored hits. The scoring system considers both structural
homology and the genomic contextual information available in the PTT file. Across
our collection of bacterial genomes acquired from NCBI RefSeq data, \tthp
identified \DIFdelbegin \DIFdel{2,752,469 }\DIFdelend \DIFaddbegin \DIFadd{$2,752,469$ }\DIFaddend rho-independent terminators using the default filters.

Due to the spatially-mediated structural regulation of purine \rbs,
whereby ligand interaction with the aptamer domain induces local structural
rearrangement in the expression platform, we paired aptamers with corresponding
terminators by minimizing the genomic distance, with an upper bound of \DIFdelbegin \DIFdel{200
}\DIFdelend \DIFaddbegin \DIFadd{$200$
}\DIFaddend nucleotides between the end of the aptamer domain and start of the terminator.
This approach yields \DIFdelbegin \DIFdel{577 }\DIFdelend \DIFaddbegin \DIFadd{$577$ }\DIFaddend candidate \rbs, \DIFdelbegin \DIFdel{81 }\DIFdelend \DIFaddbegin \DIFadd{$81$ }\DIFaddend of which have multiple
rho-independent terminators within range of a putative aptamer produced by
\infernal. For these, we simply pair the closest \tthp hit with the aptamer domain.

\begin{figure}[!ht]
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/refseqGenomeSizes.pdf}
\caption[Histogram displaying the distribution of genome sizes across RefSeq]{Histogram displaying the distribution of genome sizes across the RefSeq
data analyzed, comprising \DIFdelbeginFL \DIFdelFL{5,172 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$5,172$ }\DIFaddendFL bacterial genomes. Genome size is shown using a
$\log_{10}$ scale, and appears to have a bimodal distribution.}
\label{fig:rfinder:genomeSizes}
\end{figure}

\subsection{Step 2: Structural prediction}
\label{subsec:rfinder:strpred}

Until this point we have been focused on the generation of candidate sequences
from our RefSeq dataset, without yet focusing on the specifics of underlying
secondary structures for these candidates. In the following section, we explain
how constraint folding is used to generate putative `on' and `off' conformations
for each candidate.

\subsubsection{Notation for representing abstract RNA shapes}
\label{subsubsec:rfinder:shapes}

Given an RNA sequence $\seq = \seqN$, where positions $s_i$ are drawn from the
collection of single-letter nucleotide codes, i.e.
$s_i \in \{\text{A,\,U,\,G,\,C}\}$, it is possible to describe a corresponding
secondary structure \strS compatible with \seq using the dot-bracket notation.
In this notation, each nucleotide $a_i$ has a corresponding state $s_i$, where
$s_i$ is denoted as a `.' if unpaired and a `(' [resp. `)'] if the left [resp.
right] base in a base pair. Given any two base pairs $(i,j)$ and $(k,l)$ in \strS,
then $i < k < j \iff i < l < j$; pseudoknots are not permitted in the structure. A
secondary structure taking this form is said to have balanced parentheses, and can
additionally be represented using a context-free grammar such as the
following, \DIFdelbegin \footnote{\DIFdel{Note that in the grammar from }%DIFDELCMD < \eqnref{rfinder:strCfg}%%%
\DIFdel{,
$\bullet^{\theta}$ should be interpreted to mean $\theta$ repetitions
of the $\bullet$ symbol, where $\theta$ is generally taken to be 3.}}
%DIFAUXCMD
\addtocounter{footnote}{-1}%DIFAUXCMD
\DIFdelend derived from \citep{fusy:2012ka}:

\begin{align}
\label{eq:rfinder:strCfg}
\DIFdelbegin %DIFDELCMD < \begin{split}
%DIFDELCMD < S &\rightarrow \bullet \;|\; S \bullet \,|\; T \;|\; S\,T \\
%DIFDELCMD < T &\rightarrow \bpL \bullet^{\theta} \bpR \;|\;
%DIFDELCMD < \bpL S\,\bullet^{\theta} \bpR \;|\; \bpL T \bpR \;|\; \bpL S\,T \bpR \\
%DIFDELCMD < \end{split}
%DIFDELCMD < %%%
\DIFdelend \DIFaddbegin \DIFadd{S }&\DIFadd{\rightarrow \bullet \;|\; S \bullet \,|\; }\bpL \DIFadd{S }\bpR \DIFadd{\;|\; S\, }\bpL \DIFadd{S }\bpR
\DIFaddend \end{align}

The grammar from \eqnref{rfinder:strCfg}\DIFaddbegin \DIFadd{,
where the minimum number of unpaired bases $\theta$ in a hairpin loop
is taken to be $1$ for expository clarity,
}\DIFaddend can be used to generate a parse tree
\tree for \strS. The benefit of working with \tree over \strS is that the parse
tree offers an abstract representation of secondary structure shape independent of
sequence length, permitting us to classify and eventually constrain a large
collection of sequences having variable length which are all expected to have the
same abstract tree shape \citep{voss:2006iq}. This is analogous to what
the Giegerich lab refers to as
their `type 5' structural abstraction using the \rshapes tool, and can be
described using the following grammar \citep{Lorenz:2008gz}\DIFaddbegin \DIFadd{, which generates
all combinations of matched brackets}\DIFaddend :

\begin{align}
\label{eq:rfinder:shapeCfg}
\begin{split}
S &\rightarrow \brL T \brR\, S \;|\; \brL T \brR \\
T &\rightarrow \brL T \brR \;|\; \epsilon \\
\end{split}
\end{align}

Every node in \tree
represents a helix in \strS, and internally tracks the indices of both its
beginning $(i,j)$ and closing $(k,l)$ base pair. We use a level-order naming
convention to refer to helices within the parse tree, whereby a position
\treePos{p}{1} references the first child of the root node, \treePos{p}{1,2}
references the second child of \textdown{\ms{p}}{1}, and generally
\treePos{p}{$i_1,i_2,\cdots,i_n$} refers to the $i_n$\textsuperscript{th} child of
\treePos{p}{$i_1,i_2,\cdots,i_{n-1}$}. To reference specific nucleotides in the
context of their location relative to a helix, we use the opening and closing base
pairs $(i,j)$ and $(k,l)$ as landmarks. Thus, \treeIdx{p}{1}{l} is the index in
\strS of the right-hand side closing base pair of \treePos{p}{1}. We use the
notation \treePos{t}{$i$} to refer to the subtree of \tree whose root is
\treePos{p}{$i$}.

Finally, we introduce the concept of a tree signature. The tree signature for a
tree \tree is a list of the node depths when traversed in a depth-first pre-order
fashion. To provide a concrete example, consider the following experimentally
validated xpt \grb from {\em Bacillus subtilis} subsp. subtilis str. \DIFdelbegin \DIFdel{168
}\DIFdelend \DIFaddbegin \DIFadd{$168$
}\DIFaddend (NC\_\DIFdelbegin \DIFdel{000964.3 2320197--2320054}\DIFdelend \DIFaddbegin \DIFadd{$000964.3$ $2320197$--$2320054$}\DIFaddend ) with corresponding gene `off' structure as seen in
\Figref{rfinder:xptOff}.
\medskip

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_000964_3_2320197_2320054_OFF.pdf}
\end{subfigure} \\
\medskip
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_000964_3_2320197_2320054_ON.pdf}
\end{subfigure}
\DIFdelbeginFL %DIFDELCMD < \caption[The xanthine phosphoribosyltransferase (xpt) \grb from
%DIFDELCMD < {\em B. subtilis} subsp. subtilis str. 168 (NC\_000964.3 2320197--2320054),
%DIFDELCMD < and corresponding gene `off' and `on' structures]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[The xanthine phosphoribosyltransferase (xpt) \grb from
{\em B. subtilis} subsp. subtilis str. $168$ (NC\_$000964.3$ $2320197$--$2320054$),
and corresponding gene `off' and `on' structures]{\DIFaddendFL {\em (Top)} The xanthine phosphoribosyltransferase (xpt) \grb from
{\em B. subtilis} subsp. subtilis str. \DIFdelbeginFL \DIFdelFL{168 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$168$ }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{000964.3 2320197--2320054}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$000964.3$ $2320197$--$2320054$}\DIFaddendFL ),
and corresponding gene `off' structure derived from inline probing studies both in
the presence and absence of guanine \citep{mandal2003}. {\em (Bottom)} The
experimentally
derived gene `on' structure for \Bsxpt. These structural diagrams were generated
using VARNA \citep{darty:2009gt}.}
\label{fig:rfinder:xptOff}
\end{figure}

The \rshapes \citep{janssen:2015cq} `type 5' representation for this structure is
\ms{[[][]][][]} (note the coalesced left bulge in the hairpin immediately
downstream the closing multiloop stem, at helix \treePos{p}{2}) and the tree
signature for this parse tree of the structure is \ms{[0,1,2,2,1,1]}.

We leverage the notion of abstract structural filtering initially to ensure that
all \infernal aptamer hits have a tree signature of \ms{[0,1,2,2]}, which
represents a three-way junction, and that the binding site for the guanine ligand
$\treeIdx{p}{1}{l - 1} = \text{C}$. These filters, in combination with the
proximal terminator hairpins produced by \tthp yield the aforementioned \DIFdelbegin \DIFdel{577
}\DIFdelend \DIFaddbegin \DIFadd{$577$
}\DIFaddend candidate \grbs for which we then try to produce reasonable gene `on' and off
structures.

\subsubsection{Constrained folding to predict switch structures}
\label{subsubsec:rfinder:consfold}

To restrict our search to unannotated \grbs, and further ensure that we are not
re-detecting sequences based off the Rfam covariance model provided to \infernal,
we constrain our search to those RefSeq organisms not represented in the Rfam seed
alignment. \DIFdelbegin \DIFdel{503 of the 577 }\DIFdelend \DIFaddbegin \DIFadd{$503$ of the $577$ }\DIFaddend candidates, or \DIFdelbegin \DIFdel{87.18}\DIFdelend \DIFaddbegin \DIFadd{$87.18$}\DIFaddend \% represent putative unannotated
\rbs not represented by \DIFdelbegin \DIFdel{RF00167}\DIFdelend \DIFaddbegin \DIFadd{RF$00167$}\DIFaddend .

The gene `off' structure \strOff for a \grb is the easier of the two to find
computationally, since the terminator stem is exceptionally thermodynamically
stable. In the gene `on' conformation \strOn, the P1 stem of the multiloop partially
dissociates and an anti-terminator stem forms between the region immediately 3' of
the P1 stem and what was the left-hand side of the terminator stem. This truncated
P1 stem, which closes the three-way junction in the aptamer, is exceptionally
unstable based on present energy models available for structural folding, and
requires special treatment to reconstitute in our final structures.

The software \rfold (\DIFdelbegin \DIFdel{v2.1.8}\DIFdelend \DIFaddbegin \DIFadd{v$2.1.8$}\DIFaddend ) allows for the folding of RNA molecules with `loose'
constraints. In this model of constrained folding, the resulting structure
produced by the software guarantees not to explicitly invalidate any user-provided
constraints, but does not guarantee all constraints will be satisfied in the
resulting structure. For each of the candidate \grbs, having \treeFor{\infernal}
and \treeFor{\tthp}, we build the following constraint masks:

{\large Structural constraints for both conformations of the \grb aptamer:}
\begin{enumerate}
\setstretch{1.3}
\item Prohibit base pairing upstream of \treeIdx{p}{1}{i} and
downstream of \treeIdx{p}{2}{l}. \\[1.5ex] {\em Do not permit any possible disruptive pairing
interactions 5' of the aptamer or 3' of the terminator stem.}
\item Force base pairs and unpaired regions in \treePos{t}{1}, with
the exception of \treePos{p}{1}. \\[1.5ex] {\em Since the aptamer structure is well
conserved and we have the \infernal-provided alignment with the
covariance model, force this structure to form as aligned.}
\item Explicitly prohibit formation of \treePos{p}{1} stem, which closes the
three-way junction. \\[1.5ex] {\em The only exception to above is the closing of the
P1 multiloop stem. In our experience, since \rfold uses soft constraints
(meaning that constrained base pairs are only allowed to pair with each other
or not at all), in practice we rarely see the P1 stem form as we would like.
Instead, restrict it from forming at all, so that it can be added in after the
fact without disrupting any other base pairs.}
\end{enumerate}
{\large Constraints exclusive to the gene `off' structure:}
\begin{enumerate}
\setstretch{1.3}
\item Force base pairs and unpaired regions in \treePos{t}{2}. \\[1.5ex] {\em This simply
forces the formation of the terminator stem, as predicted by \tthp.}
\end{enumerate}
{\large Constraints exclusive to the gene `on' structure:}
\begin{enumerate}
\setstretch{1.3}
\item Require $m$ nucleotides starting from \treeIdx{p}{1}{l + 3} to pair to the
right, where $m = \textit{length}(\treePos{p}{2})$, and require the left-hand side of the
\treePos{p}{2} helix to pair to the left. \\[1.5ex] {\em The formation of the
anti-terminator stem involves the partial disruption of the \treePos{p}{1} stem.
Though there is no consensus for the length of the anti-terminator stem,
experimental data suggests that the left side of the terminator stem}
(\treeIdx{p}{2}{i}--\treeIdx{p}{2}{k}) {\em alternatively base pairs to the left,
thus forming the anti-terminator hairpin and permitting transcription to proceed
\citep{mandal:2004ja}.}
\item Disallow pairing downstream of \treeIdx{p}{2}{j}. \\[1.5ex] {\em Avoid
disruptive pairing downstream of the newly formed anti-terminator stem.}
\end{enumerate}

These constraint masks are run using the command-line flags
\ms{-d 0 -P rna\_turner1999.par} to disable dangles and use the Turner \DIFdelbegin \DIFdel{1999
}\DIFdelend \DIFaddbegin \DIFadd{$1999$
}\DIFaddend energies respectively \citep{mathews:1999jw}. We choose to disable dangles (\ms{-d 0}) and use
the Turner \DIFdelbegin \DIFdel{1999 }\DIFdelend \DIFaddbegin \DIFadd{$1999$ }\DIFaddend energy model (\ms{-P rna\_turner1999.par})
based on visual inspection of the structures output by \rfold with
constraints---these flags appear to yield conformations most frequently consistent
with the known structures for the \Bsxpt \grb.
Experimental evidence using inline probing and
crystallographic analysis suggests that
the `on' conformation of the \grb has a reduced P1 stem length of \DIFdelbegin \DIFdel{3 }\DIFdelend \DIFaddbegin \DIFadd{$3$ }\DIFaddend base pairs
\citep{mandalboesebarrickwinklerbreaker,serganov:2004dq};
in practice we were unable to force \rfold to respect this constraint regardless
of command-line options specified. For this reason we reconstitute the P1 stem in
both structures after constrained folding, having length equivalent to it the
\infernal P1 stem [resp. \DIFdelbegin \DIFdel{3 }\DIFdelend \DIFaddbegin \DIFadd{$3$ }\DIFaddend base pairs] in the gene `off' [resp. gene `on']
structure.

\subsection{Step 3: Candidate curation}
\label{subsec:rfinder:curation}

Until now, we have described our approach for generating the \DIFdelbegin \DIFdel{503 }\DIFdelend \DIFaddbegin \DIFadd{$503$ }\DIFaddend \grb candidates
in RefSeq, alongside their gene `on' and off structures. Unfortunately the
experimental validation of all \DIFdelbegin \DIFdel{503 }\DIFdelend \DIFaddbegin \DIFadd{$503$ }\DIFaddend candidates is not tractable, so it was
necessary to reduce this collection again to a more manageable size, while only
keeping the most promising candidates. Our original approach involved using
\foldalign \citep{gorodkin:1997tr,havgaard:2007ca} alongside the \ms{needleall} tool from EMBOSS
\citep{rice:2000wr}, to simultaneously
select sequences which closely approximate the more thermodynamically stable
gene `off' conformation of the experimentally known \Bsxpt \grb, while minimizing
sequence similarity between candidates selected for experimental validation. Due to
expense, we elected to instead choose a small number
($n = 2$) of organisms easily available which had multiple promising hits as our
experimental candidate pool.

In \Figref{rfinder:histogramFoldalignCandidatesVsXpt}, we display a histogram of scores produced
by \foldalign for the \DIFdelbegin \DIFdel{503 }\DIFdelend \DIFaddbegin \DIFadd{$503$ }\DIFaddend candidates, when aligned with the \Bsxpt \grb. \foldalign
is based off a simplification of Sankoff's algorithm \citep{sankoff:1985wc}, and
is a dynamic programming
algorithm for simultaneous folding and alignment that runs in \On{4} time. Because three of the sequences from our pool of \DIFdelbegin \DIFdel{503 }\DIFdelend \DIFaddbegin \DIFadd{$503$ }\DIFaddend candidates have
no global alignment with the \Bsxpt sequence, we have pruned them from our
dataset and only consider those remaining \DIFdelbegin \DIFdel{500 }\DIFdelend \DIFaddbegin \DIFadd{$500$ }\DIFaddend sequences for which \foldalign
scores are produced. The \foldalign scores produce have a mean of $153.798$ with a
minimum [resp. maximum] score of $-2698$ [resp. \DIFdelbegin \DIFdel{1908}\DIFdelend \DIFaddbegin \DIFadd{$1908$}\DIFaddend ]. Running \foldalign with the
\Bsxpt sequence aligned with itself produces a theoretical maximum score of \DIFdelbegin \DIFdel{2419.
}\DIFdelend \DIFaddbegin \DIFadd{$2419$.
}\DIFaddend 

\begin{figure}[!ht]
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/histogramFoldalignCandidatesVsXpt.pdf}
\caption[Histogram displaying the distribution of scores produced by \foldalign]{Histogram displaying the distribution of scores produced by \foldalign
\DIFdelbeginFL \DIFdelFL{2.1.1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$2.1.1$ }\DIFaddendFL using flags \ms{-global -summary -format commandline} when folding each of
the \DIFdelbeginFL \DIFdelFL{503 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$503$ }\DIFaddendFL candidates against the \Bsxpt sequence NC\_\DIFdelbeginFL \DIFdelFL{000964.3 2320197--2320054}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$000964.3$ $2320197$--$2320054$}\DIFaddendFL .
Three of the sequences run against \foldalign (NC\_\DIFdelbeginFL \DIFdelFL{010674.1 1516712--1516868}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$010674.1$ $1516712$--$1516868$}\DIFaddendFL ,
NC\_\DIFdelbeginFL \DIFdelFL{010723.1 1487041--1487197}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$010723.1$ $1487041$--$1487197$}\DIFaddendFL , and NC\_\DIFdelbeginFL \DIFdelFL{020291.1 4599412--4599258}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$020291.1$ $4599412$--$4599258$}\DIFaddendFL ) have no global
alignment with the \Bsxpt sequence, and thus the histogram represents \DIFdelbeginFL \DIFdelFL{500 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$500$ }\DIFaddendFL of the
original \DIFdelbeginFL \DIFdelFL{503 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$503$ }\DIFaddendFL sequences.}
\label{fig:rfinder:histogramFoldalignCandidatesVsXpt}
\end{figure}

Of these \DIFdelbegin \DIFdel{500 sequences, 335 }\DIFdelend \DIFaddbegin \DIFadd{$500$ sequences, $335$ }\DIFaddend have a \foldalign score $s > 0$, representing \DIFdelbegin \DIFdel{227
}\DIFdelend \DIFaddbegin \DIFadd{$227$
}\DIFaddend unique accession numbers, with a distribution of candidates per accession number
as shown in \Figref{rfinder:candidateHistogramGroupedByAccession}. From
the perspective of experimental validation, we have tried to maximize the chance
of success per organism by selecting those having multiple candidates within the
same genome. Only \DIFdelbegin \DIFdel{25 }\DIFdelend \DIFaddbegin \DIFadd{$25$ }\DIFaddend of the candidate organisms have more than three hits within
their genome (only two have five hits). Our approach for selecting the initial
two organisms for experimental validation was to take this pool of \DIFdelbegin \DIFdel{25 }\DIFdelend \DIFaddbegin \DIFadd{$25$ }\DIFaddend organisms,
sort by descending average score $s$, and select the first two which are available
via DSMZ (\url{https://www.dsmz.de/}), the warehouse for microorganisms used by
our collaborators. Prof. Dr. Mario M\"orl at Universit\"at Leipzig is presently
overseeing Dr. Regula Aregger, a post-doc who is using the SHAPE protocol
\citep{wilkinson:2006vd} to
validate the computationally predicted structure of these candidates.

\begin{figure}[!ht]
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/candidateHistogramGroupedByAccession.pdf}
\caption[Histogram displaying the number of candidate \rbs we observe
per organism]{Histogram displaying the number of candidate \rbs we observe
per organism. From this data, it is clear that the majority of organisms have
only one putative \rb, however two organisms have five candidates each:
{\em Clostridium botulinum} B str. Eklund \DIFdelbeginFL \DIFdelFL{17B }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$17$B }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{010674.1}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$010674.1$}\DIFaddendFL ) and
{\em Clostridium botulinum} E3 str. Alaska \DIFdelbeginFL \DIFdelFL{E43 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{E$43$ }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{010723.1}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$010723.1$}\DIFaddendFL )}
\label{fig:rfinder:candidateHistogramGroupedByAccession}
\end{figure}

Proceeding in this fashion, we have selected {\em B. megaterium} QM \DIFdelbegin \DIFdel{B1551
(}%DIFDELCMD < \href{http://www.ncbi.nlm.nih.gov/nuccore/NC_014019.1}{NC\_014019.1}%%%
\DIFdel{,
}%DIFDELCMD < \href{http://www.dsmz.de/catalogues/details/culture/DSM-1804.html}{DSM1804}%%%
\DIFdelend \DIFaddbegin \DIFadd{B$1551$
(}\href{http://www.ncbi.nlm.nih.gov/nuccore/NC_014019.1}{NC\_$014019.1$}\DIFadd{,
}\href{http://www.dsmz.de/catalogues/details/culture/DSM-1804.html}{DSM$1804$}\DIFaddend )
and {\em B. megaterium} \DIFdelbegin \DIFdel{DSM319
(}%DIFDELCMD < \href{http://www.ncbi.nlm.nih.gov/nuccore/NC_014103.1}{NC\_014103.1}%%%
\DIFdel{,
}%DIFDELCMD < \href{http://www.dsmz.de/catalogues/details/culture/DSM-319.html}{DSM319}%%%
\DIFdelend \DIFaddbegin \DIFadd{DSM$319$
(}\href{http://www.ncbi.nlm.nih.gov/nuccore/NC_014103.1}{NC\_$014103.1$}\DIFadd{,
}\href{http://www.dsmz.de/catalogues/details/culture/DSM-319.html}{DSM$319$}\DIFaddend )
for initial validation. These organisms have four
candidate \grbs each, outlined in Table \ref{table:rfinderCandidateLocs}.

\begin{table}[!ht]
\centering
\begin{tabularx}{\linewidth}{*{1}{L} *{2}{C}}
\toprule
\small{Downstream gene function} & \DIFdelbeginFL %DIFDELCMD < \small{{\em B. megaterium} QM B1551} %%%
\DIFdelendFL \DIFaddbeginFL \small{{\em B. megaterium} QM B$1551$} \DIFaddendFL & \DIFdelbeginFL %DIFDELCMD < \small{{\em B. megaterium} DSM319} %%%
\DIFdelendFL \DIFaddbeginFL \small{{\em B. megaterium} DSM$319$} \DIFaddendFL \\
\cmidrule(lr){1-3}
\small{xpt} & \DIFdelbeginFL \DIFdelFL{1427313--1427501 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1427313$--$1427501$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1413696--1413884 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1413696$--$1413884$ }\DIFaddendFL \\[1ex]
\small{GMP synthase} & \DIFdelbeginFL \DIFdelFL{231630--231806 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$231630$--$231806$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{230059--230235 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$230059$--$230235$ }\DIFaddendFL \\[1ex]
\small{guanine permease} & \DIFdelbeginFL \DIFdelFL{233482--233680 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$233482$--$233680$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{231911--232108 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$231911$--$232108$ }\DIFaddendFL \\[1ex]
\small{N5-carboxyaminoimidazole} & \DIFdelbeginFL \DIFdelFL{240759--240970 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$240759$--$240970$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{239188--239400 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$239188$--$239400$ }\DIFaddendFL \\
\bottomrule
\end{tabularx}
\DIFdelbeginFL %DIFDELCMD < \caption[The genomic coordinates for the four candidate \grbs in both
%DIFDELCMD < {\em B. megaterium} QM B1551 and {\em B. megaterium} DSM319]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[The genomic coordinates for the four candidate \grbs in both
{\em B. megaterium} QM B$1551$ and {\em B. megaterium} DSM$319$]{\DIFaddendFL The genomic coordinates for the four candidate \grbs in both
{\em B. megaterium} QM \DIFdelbeginFL \DIFdelFL{B1551 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{B$1551$ }\DIFaddendFL and {\em B. megaterium} \DIFdelbeginFL \DIFdelFL{DSM319}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{DSM$319$}\DIFaddendFL . Note that the
\grbs are located upstream of the same genes, and that these two strains of
{\em B. megaterium} are highly similar. These structures are pictured in
\Secref{sec:rfinder:grbValidationVarna}, plotted using VARNA \citep{darty:2009gt}.}
\label{table:rfinderCandidateLocs}
\end{table}

\section{Extending beyond \grbs}
\label{sec:rfinder:ext}

We believe that the \rfinder pipeline allows for the detection of both structural
conformations of \rbs beyond the \grb. The investigation of
adenine-sensitive purine \rbs is a small extension of the existing
implementation. As indicated in \Secref{subsubsec:rfinder:infernal}, adenine
\rbs have a complimentary uradine residue at the ligand binding site in
the J3--1 junction within the aptamer. Beyond differences in ligand specificity,
the adenine \rb anti-terminator stem is incorporated into the aptamer structure
itself, and thus stabilized with the base pairing of the adenine ligand. As a
result, the adenine \rb permits transcription when bound, unlike the \grb.
As a result of the extensive overlap between the anti-terminator stem and adenine
\rb aptamer, the formation of the terminator stem completely dissociates both the
P3 and P1 stems \citep{mandal2004a}.

From a computational perspective, these changes are simple to handle within the
\rfinder pipeline, and provide some indication to how we believe the framework
could be more generally applied in the future. Rather than filter for the
discriminatory cytidine residue in the \rb aptamer (\Secref{subsubsec:rfinder:infernal}) we can only select those hits from \infernal having a uridine at the
ligand binding site. Structural on and off conformations are known from
experimental data for the {\em B. subtilis} ydhL gene \citep{mandal2004a} and can
be used as templates for the constraint masks used in \Secref{subsec:rfinder:strpred}.

In general, we believe those \rbs using rho-independent transcription
termination as a mode of regulation---for which an aptamer alignment exists and
some experimental knowledge of the expression platform's structural
conformation is known---are well suited for more robust structural
prediction using \rfinder.

\section{Guanine \rbs for experimental validation}
\label{sec:rfinder:grbValidationVarna}

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014019_1_1427313_1427501_OFF.pdf}
\end{subfigure} \\
\medskip
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014019_1_1427313_1427501_ON.pdf}
\end{subfigure}
\DIFdelbeginFL %DIFDELCMD < \caption[Structures for the putative \rb located upstream of the xpt gene in {\em B. megaterium} QM B1551]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Structures for the putative \rb located upstream of the xpt gene in {\em B. megaterium} QM B$1551$]{\DIFaddendFL {\em Top:} the computationally predicted gene `off' conformation of
sequence NC\_\DIFdelbeginFL \DIFdelFL{014019.1 1427313--1427501}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$014019.1$ $1427313$--$1427501$}\DIFaddendFL , using \rfold from the ViennaRNA \DIFdelbeginFL \DIFdelFL{2.1.8
}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$2.1.8$
}\DIFaddendFL suite, with dangles disabled and the Turner \DIFdelbeginFL \DIFdelFL{1999 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1999$ }\DIFaddendFL energies. This sequence is
located upstream of the xpt gene in {\em B. megaterium} QM \DIFdelbeginFL \DIFdelFL{B1551}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{B$1551$}\DIFaddendFL . {\em Bottom:}
the gene `on' conformation.}
\label{fig:figure:NC_014019_1_1427313_1427501}
\end{figure}
\medskip

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014019_1_231630_231806_OFF.pdf}
\end{subfigure} \\
\medskip
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014019_1_231630_231806_ON.pdf}
\end{subfigure}
\DIFdelbeginFL %DIFDELCMD < \caption[Structures for the putative \rb located upstream of the GMP synthase gene in {\em B. megaterium} QM B1551]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Structures for the putative \rb located upstream of the GMP synthase gene in {\em B. megaterium} QM B$1551$]{\DIFaddendFL The computationally predicted \rb located upstream of the GMP synthase
gene in {\em B. megaterium} QM \DIFdelbeginFL \DIFdelFL{B1551 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{B$1551$ }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{014019.1 231630--231806}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$014019.1$ $231630$--$231806$}\DIFaddendFL ).
{\em Top:} The gene `off' conformation. {\em Bottom:} The gene `on' conformation.}
\label{fig:figure:NC_014019_1_231630_231806}
\end{figure}
\medskip

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014019_1_233482_233680_OFF.pdf}
\end{subfigure} \\
\medskip
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014019_1_233482_233680_ON.pdf}
\end{subfigure}
\DIFdelbeginFL %DIFDELCMD < \caption[Structures for the putative \rb located upstream of the guanine permease gene in {\em B. megaterium} QM B1551]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Structures for the putative \rb located upstream of the guanine permease gene in {\em B. megaterium} QM B$1551$]{\DIFaddendFL The computationally predicted \rb located upstream of the guanine permease
gene in {\em B. megaterium} QM \DIFdelbeginFL \DIFdelFL{B1551 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{B$1551$ }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{014019.1 233482--233680}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$014019.1$ $233482$--$233680$}\DIFaddendFL ).
{\em Top:} The gene `off' conformation. {\em Bottom:} The gene `on' conformation.}
\label{fig:figure:NC_014019_1_233482_233680}
\end{figure}
\medskip

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014019_1_240759_240970_OFF.pdf}
\end{subfigure} \\
\medskip
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014019_1_240759_240970_ON.pdf}
\end{subfigure}
\DIFdelbeginFL %DIFDELCMD < \caption[Structures for the putative \rb located upstream of the N5- carboxy\-amino\-imidazole gene in {\em B. megaterium} QM B1551]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Structures for the putative \rb located upstream of the N5-carboxy\-amino\-imidazole gene in {\em B. megaterium} QM B$1551$]{\DIFaddendFL The computationally predicted \rb located upstream of the
\DIFdelbeginFL \DIFdelFL{N5- carboxy}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{N5-carboxy}\DIFaddendFL \-amino\-imidazole
gene in {\em B. megaterium} QM \DIFdelbeginFL \DIFdelFL{B1551 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{B$1551$ }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{014019.1 240759--240970}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$014019.1$ $240759$--$240970$}\DIFaddendFL ).
{\em Top:} The gene `off' conformation. {\em Bottom:} The gene `on' conformation.}
\label{fig:figure:NC_014019_1_240759_240970}
\end{figure}
\medskip

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014103_1_1413696_1413884_OFF.pdf}
\end{subfigure} \\
\medskip
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014103_1_1413696_1413884_ON.pdf}
\end{subfigure}
\DIFdelbeginFL %DIFDELCMD < \caption[Structures for the putative \rb located upstream of the xpt gene in {\em B. megaterium} DSM319]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Structures for the putative \rb located upstream of the xpt gene in {\em B. megaterium} DSM$319$]{\DIFaddendFL The computationally predicted \rb located upstream of the xpt
gene in {\em B. megaterium} \DIFdelbeginFL \DIFdelFL{DSM319 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{DSM$319$ }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{014103.1 1413696--1413884}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$014103.1$ $1413696$--$1413884$}\DIFaddendFL ).
{\em Top:} The gene `off' conformation. {\em Bottom:} The gene `on' conformation.}
\label{fig:figure:NC_014103_1_1413696_1413884}
\end{figure}
\medskip

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014103_1_230059_230235_OFF.pdf}
\end{subfigure} \\
\medskip
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014103_1_230059_230235_ON.pdf}
\end{subfigure}
\DIFdelbeginFL %DIFDELCMD < \caption[Structures for the putative \rb located upstream of the GMP synthase gene in {\em B. megaterium} DSM319]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Structures for the putative \rb located upstream of the GMP synthase gene in {\em B. megaterium} DSM$319$]{\DIFaddendFL The computationally predicted \rb located upstream of the GMP synthase
gene in {\em B. megaterium} \DIFdelbeginFL \DIFdelFL{DSM319 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{DSM$319$ }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{014103.1 230059--230235}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$014103.1$ $230059$--$230235$}\DIFaddendFL ).
{\em Top:} The gene `off' conformation. {\em Bottom:} The gene `on' conformation.}
\label{fig:figure:NC_014103_1_230059_230235}
\end{figure}
\medskip

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014103_1_231911_232108_OFF.pdf}
\end{subfigure} \\
\medskip
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014103_1_231911_232108_ON.pdf}
\end{subfigure}
\DIFdelbeginFL %DIFDELCMD < \caption[Structures for the putative \rb located upstream of the guanine permease gene in {\em B. megaterium} DSM319]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Structures for the putative \rb located upstream of the guanine permease gene in {\em B. megaterium} DSM$319$]{\DIFaddendFL The computationally predicted \rb located upstream of the guanine permease
gene in {\em B. megaterium} \DIFdelbeginFL \DIFdelFL{DSM319 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{DSM$319$ }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{014103.1 231911--232108}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$014103.1$ $231911$--$232108$}\DIFaddendFL ).
{\em Top:} The gene `off' conformation. {\em Bottom:} The gene `on' conformation.}
\label{fig:figure:NC_014103_1_231911_232108}
\end{figure}
\medskip

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014103_1_239188_239400_OFF.pdf}
\end{subfigure} \\
\medskip
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.9\textwidth]{Figures/Ribofinder/NC_014103_1_239188_239400_ON.pdf}
\end{subfigure}
\DIFdelbeginFL %DIFDELCMD < \caption[Structures for the putative \rb located upstream of the N5- carboxy\-amino\-imidazole gene in {\em B. megaterium} DSM319]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Structures for the putative \rb located upstream of the N5-carboxy\-amino\-imidazole gene in {\em B. megaterium} DSM$319$]{\DIFaddendFL The computationally predicted \rb located upstream of the
\DIFdelbeginFL \DIFdelFL{N5- carboxy}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{N5-carboxy}\DIFaddendFL \-amino\-imidazole
gene in {\em B. megaterium} \DIFdelbeginFL \DIFdelFL{DSM319 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{DSM$319$ }\DIFaddendFL (NC\_\DIFdelbeginFL \DIFdelFL{014103.1 239188--239400}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$014103.1$ $239188$--$239400$}\DIFaddendFL ).
{\em Top:} The gene `off' conformation. {\em Bottom:} The gene `on' conformation.}
\label{fig:figure:NC_014103_1_239188_239400}
\end{figure}

%!TEX root = ../main.tex

\chapter{FFTbor}
\label{ch:fftbor}

\lhead{FFTbor: Coarse-Grained Energy Landscapes}

\section{Introduction}
\label{sec:fftbor:intro}

In this chapter, we present the \fftbor algorithm and accompanying software.
\fftbor is a novel algorithm developed with the intent of efficiently computing
the Boltzmann probability of those structures which, for a given input RNA
sequence \seq, differ by $k$ base pairs. By leveraging polynomial interpolation
via the \fft, this algorithm runs in \On{4} time and
\On{2} space, a significant improvement over its predecessor. The accompanying
software which implements this algorithm has been used to evaluate the
correlation between kinetic folding speed and landscape ruggedness.

\subsection{Organization}
\label{subsec:fftbor:org}

This chapter is organized in the following fashion. First, we provide
background on
the problem which \fftbor aims to address, as well as a brief overview of
existing approaches. We follow by a formal explanation of the problem, and
proceed to describe how the energy landscape is \DIFdelbegin \DIFdel{coarsified }\DIFdelend \DIFaddbegin \DIFadd{partitioned }\DIFaddend into discrete bins.
We then develop the recursions for the parameterized partition function using
the Nussinov-Jacobson energy model, which allows us to highlight the novel aspects
of the algorithm. After developing the recursions, we indicate how they can be
reformulated as a polynomial whose coefficients $z_k=\bfZ{k}{1,n}$. We then
describe how the \fft can be employed to efficiently compute the coefficients
$z_k$, finishing our description of the underlying algorithm. Then we proceed
to present an application of \fftbor, in the area of RNA folding kinetics.

\section{Background}
\label{sec:fftbor:bkgrnd}

In \citep{freyhult.b07}, a dynamic programming algorithm
\rnabor---pronounced {\em RNA neighbor}---was developed which simultaneously
computes for
each integer $k$, the Boltzmann probability $\pk = \frac{\bfZ{k}{}}{\fullZ}$
of the subensemble of structures
whose \bpd to a given {\em initial}, or
{\em reference}, structure \strSt is $k$.
\footnote{As later
explained, \fullZ denotes the partition function, defined as the sum of
all Boltzmann factors \boltzf{\str}, over all secondary structures \str
of a given RNA sequence, and $R$ denotes the universal
gas constant and $T$ absolute temperature. Similarly \bfZ{k}{} denotes the
sum of all Boltzmann factors of all structures \str, whose \bpd
to the initial structure \strSt is exactly $k$.}
\rnabor stores the value of the (partial)
partition functions \bfZ{k}{i,j} for all $1 \leq i \leq j \leq n$ and
$0 \leq k \leq n$, each of which requires quadratic time to compute.
Thus it follows that \rnabor runs in time \On{5} and space
\On{3}, which severely limits its applicability to genomic annotation.
This restriction is somewhat mitigated by the fact that
in \citep{cloteloulorenz}, we showed how to use sampling
\citep{ding.nar03} to efficiently approximate
\rnabor in cubic time \On{3} and quadratic space \On{2},
{\em provided} that the starting structure \strSt is the \mfe
(MFE) structure. We expect that a more efficient version of
\rnabor could be used in applications in genomics and synthetic
biology, to detect potential conformational switches---
RNA sequences containing two or more (distinct) metastable structures.

In this chapter, we describe a radically different algorithm, \fftbor
\citep{senter.po12},
prounounced {\em FFT neighbor},
that uses polynomial interpolation to compute the
coefficients $p_0,\ldots,p_{n-1}$ of the polynomial defined in
\eqnref{fftbor:pOfX},
where \pk is defined by $\pk = \frac{\bfZ{k}{}}{\fullZ}$.
Due to severe numerical instability issues in both the Lagrange
interpolation formula and in Gaussian elimination, we employ
the \fft (FFT) to compute the \idft (DFT) on values $y_0,\ldots,y_{n-1}$,
where $y_k = p(\omega^k)$ and
$\omega = \pRoU$ is the principal \nRoU and
$p(x)$ is defined in \eqnref{fftbor:pOfX}. This
gives rise to an improved version of \rnabor, denoted \fftbor,
which runs in time \On{4} and space \On{2}.
%DIF <  Once two metastable structures \strST are identified, we can
%DIF <  subsequently evaluate the feasibility of transition between
%DIF <  structures \strS and \strT,
%DIF <  by computing the {\em barrier energy} using algorithms, such as that
%DIF <  described in Dotu et al. \citep{dotu.nar10} or Flamm et al.
%DIF <  \citep{flamm.r01}.

\section{Formalization of the problem}
\label{sec:fftbor:formal}

\fftbor aims to compute the coefficients $p_0,\dots,p_{n-1}$ of the polynomial

\begin{align}
\label{eq:fftbor:pOfX}
p(x) = p_0 + p_1 x + p_2 x^2 + \dots + p_{n-1} x^{n-1},
\end{align}

where \pk is defined as $\pk = \frac{\bfZ{k}{}}{\fullZ}$. We employ the \fft to compute
the \idft on values $y_0,\dots,y_{n-1}$, where
$y_k = p(\omega^k)$ and $\omega = \pRoU$ is the principal \nRoU and $p(x)$ is defined in
\eqnref{fftbor:pOfX}. By leveraging
\nRoUs in conjunction with the \idft the we subvert numeric instability
issues observed with both Lagrange interpolation and Gaussian elimination.

Consider an RNA sequence $\seq = \seqN$, where
$s_i \in \{\text{A,\,U,\,G,\,C}\}$, i.e. a sequence of nucleotides. We can describe a
secondary structure \str which is compatible with \seq as a collection of
base pair tuples $(i,j)$, where $1 \le i \le i+\theta < j \le n$ and
$\theta \ge 0$ (generally taken to be \DIFdelbegin \DIFdel{3}\DIFdelend \DIFaddbegin \DIFadd{$3$}\DIFaddend ), the minimum number of unpaired bases
in a hairpin loop due to steric constraints.

To more simply develop the underlying recursions for \fftbor, we introduce a
number of constraints on the base pairs within \str. Firstly, we require that
each base pair is either a Watson-Crick or G-U wobble, i.e. base pair $(i,j)$
for sequence \seq has corresponding nucleotides $(s_i,s_j)$, which are
restricted to the set

\begin{align}
\label{eq:fftbor:validBP}
\bpSet =
\{\text{(A,\,U),\,(U,\,A),\,(G,\,C),\,(C,\,G),\,(G,\,U),\,(U,\,G)}\}.
\end{align}

With
this constraint satisfied we say that \str is {\em compatible} with \seq, and
for the remainder of this chapter will only consider those structures which are
compatible with \seq.
Secondly, we insist that given two base pairs $(i,j), (x,y)$ from \str,
$i=x \iff j=y$ (bases have at most one partner). Finally, we require that
$i<x<j \iff i<y<j$ (no pseudoknots are allowed). While pseudoknots have been
shown to be present in some biologically relevant RNAs, their inclusion greatly
complicates the recursive decomposition of the structure, and thus it is common
to ignore them. \DIFaddbegin \DIFadd{As a result, sequences known to contain
pseudoknots---such as RNase P \mbox{%DIFAUXCMD
\citep{haas:1991vl}}%DIFAUXCMD
---are
not well suited for }\fftbor\DIFadd{, and whenever possible investigators should take
care to consider the likely structure for their sequences before proceeding with
computational analysis.
}\DIFaddend 

Provided two secondary structures \strST, we can define a notion of
distance between them. There are a number of different definitions of distance
used across the literature; we will use {\em \bpd} for \fftbor.
\Bpd is defined as the symmetric difference between the sets
\strST:

\begin{align}
\label{eq:fftbor:dBP}
\dBP{\str}{\strT} = |\str \cup \strT| - |\str \cap \strT|.
\end{align}

Given this definition of distance, two structures \str and \strT are said to
be \kNbrs if $\dBP{\str}{\strT} = k$. It is important to note that
the notion of \bpd is also applicable to restrictions of secondary structures
on the subsequence $\seq_{i,j}$,
i.e. $\str_{[i,j]} = \{ (x,y) \,:\, i \leq  x < y \leq j,  (x,y) \in \str \}$.

For a restriction of base pairs for a given structure $\str_{[i,j]}$,
$\strT_{[i,j]}$ is said to be a \kNbr of $\str_{[i,j]}$ if

\begin{align}
\label{eq:fftbor:dBPonRestriction}
\dBP{\str_{[i,j]}}{\strT_{[i,j]}} =
|\{ (x,y): i \leq x<y\leq j,
(x,y) \in \str - \strT \text{or} (x,y) \in \strT - \str \}| = k.
\end{align}

\section{Derivation of the \fftbor algorithm}
\label{sec:fftbor:math}

Given an RNA sequence $\seq=\seqN$ and compatible secondary structure
\strSt, let \bfZ{k}{} denote the sum of the Boltzmann factors
\boltzf{\str} of all \kNbrs \str of \strSt; i.e.

\begin{align}
\bfZ{k}{} = \bfZ{k}{1,n} =
\sum_{\mathclap{\substack{\str \text{ such that } \rule[-.5ex]{0pt}{0pt} \\
 \dBP{\str}{\strSt}=k}}}\;
\boltzF{\str}
\end{align}

where $E(\str)$ denotes the Turner (nearest neighbor)
energy \DIFdelbegin \DIFdel{\mbox{%DIFAUXCMD
\citep{}
}%DIFAUXCMD
}\DIFdelend \DIFaddbegin \DIFadd{\mbox{%DIFAUXCMD
\citep{xia:RNA}
}%DIFAUXCMD
}\DIFaddend of \str, \DIFdelbegin \DIFdel{$R = 0.00198$ kcal/mol }\DIFdelend \DIFaddbegin \DIFadd{$R = \kmolk{0.00198}$ }\DIFaddend denotes the universal
gas constant and $T$ denotes absolute temperature. From this, it follows that
the full partition function is defined as

\begin{align}
\label{eq:fftbor:partFunc}
\fullZ = \bfZ{}{1,n} = \sum_{k=0}^n \bfZ{k}{1,n}
\end{align}

since the \bpd between \strSt and \str is at most

\begin{align}
\label{eq:fftbor:maxDist}
\dBP{\strSt}{\str} \leq |\strSt| + \lfloor \frac{n-\theta}{2} \rfloor \leq n.
\end{align}

We can then define the Boltzmann probability of all \kNbrs of \strSt as

\begin{align}
\label{eq:fftbor:probK}
p(k) =\frac{\bfZ{k}{1,n}}{\bfZ{}{1,n}}.
\end{align}

By visualizing the probabilities \pk as a function of $k$, we generate a
coarse-grained view of the one-dimensional energy landscape of \seq with
respect to \strSt. When \strSt is taken to be the \mfes for example, one would
anticipate to see a peak at $k=0$, with additional peaks implying additional
metastable structures; local energy minima which could suggest an energetic
trap while folding.

\subsection{Definition of the partition function
\texorpdfstring{\bfZ{k}{1,n}}{}}
\label{subsec:fftbor:recursions}

For the rest of the chapter, we consider both \seq as well as the
secondary structure \strSt on \seq to be fixed. We now recall the
recursions from Freyhult et al. \citep{Freyhult.ab05} to determine
the partition function \bfZ{k}{i,j} with
respect to the Nussinov-Jacobson
energy $E_0$ model \citep{nussinovjacobson}, defined by
$-1$ times the number of base pairs; i.e. $E_0(S) = -1 \cdot |S|$.
Although we describe here the recursions for the Nussinov-Jacobson
model, for the sake of
simplicity of exposition, both \rnabor
\citep{Freyhult.ab05} as well as our current software \fftbor,
concern the Turner energy model (described in \Secref{sec:fftbor:turner}), consisting of free energy parameters for
stacked bases, hairpins, bulges, internal loops and multiloops.

% The full
% recursions for \fftbor are described for the
% the Turner energy model in the appendix.

The base case for \bfZ{k}{i,j} is given by

\begin{align}
\label{eq:fftbor:initZpart1}
\bfZ{0}{i,j} = 1, \text{ for } i \le j,
\end{align}

since the only \DIFdelbegin \DIFdel{0-neighbor }\DIFdelend \DIFaddbegin \DIFadd{$0$-neighbor }\DIFaddend to a structure \strSt
is the structure \strSt itself, and

\begin{align}
\label{eq:fftbor:initZpart2}
\bfZ{k}{i,j} = 0, \text{ for } k > 0, i \le j \leq i + \theta,
\end{align}

since the empty structure is the only possible structure for a
sequence shorter than $\theta + 2$ nucleotides, and so there are no
\kNbrs for $k>0$. The recursion used to compute
\bfZ{k}{i,j} for $k > 0$ and $j > i+\theta$ is

\begin{align}
\label{eq:fftbor:bfZkij}
\bfZ{k}{i,j} = \bfZ{k-b_0}{i,j-1}\enspace +
\sum_{\substack{(s_r,s_j) \in \bpSet, \\ i \leq r<j}}\qquad
\sum_{\mathclap{w+w'=k-b(r)}}\enspace
\exp(-E_0(r,j)/RT) \cdot \bfZ{w}{i,r-1} \bfZ{w'}{r+1,j-1},
\end{align}

where $E_0(r,j) = -1$ if positions $r,j$ can pair in sequence \seq,
and otherwise $E_0(r,j) = +\infty$. Additionally,
$b_0 = 1$ if $j$ is base-paired
in $\strSt_{[i,j]}$ and \DIFdelbegin \DIFdel{0 }\DIFdelend \DIFaddbegin \DIFadd{$0$ }\DIFaddend otherwise, and
$b(r)=\dBP{\strSt_{[i,j]}}{\strSt_{[i,r-1]} \cup \strSt_{[r+1,j-1]} \cup\{(r,j)\}}$.
This holds since in a secondary
structure \DIFdelbegin \DIFdel{$T_{[i,j]}$ }\DIFdelend \DIFaddbegin \DIFadd{$\strT_{[i,j]}$ }\DIFaddend on \seqIJ that is a \kNbr of
$\strSt_{[i,j]}$,
either nucleotide $j$ is unpaired in $[i,j]$ or it is
paired to a nucleotide $r$ such that $i \leq r < j$. In this
latter case it is enough to study the smaller sequence segments
$[i,r-1]$ and $[r+1,j-1]$ noting that, except for $(r,j)$,
base pairs outside of these regions are not allowed, since there
are no pseudoknots. In addition,
for \DIFdelbegin \DIFdel{$\dBP{\strSt_{[i,j]}}{T_{[i,j]}} = k$ }\DIFdelend \DIFaddbegin \DIFadd{$\dBP{\strSt_{[i,j]}}{\strT_{[i,j]}} = k$ }\DIFaddend to hold,
it is necessary for $w+w' = k -b(r)$ to hold, where \DIFdelbegin \DIFdel{$w =
\dBP{\strSt_{[i,r-1]}}{T_{[i,r-1]}}$ and $w' =
\dBP{\strSt_{[r+1,j-1]}}{T_{[r+1,j-1]}}$}\DIFdelend \DIFaddbegin \DIFadd{$w =
\dBP{\strSt_{[i,r-1]}}{\strT_{[i,r-1]}}$ and $w' =
\dBP{\strSt_{[r+1,j-1]}}{\strT_{[r+1,j-1]}}$}\DIFaddend , since $b(r)$ is the
number of base pairs that differ between $\strSt_{[i,j]}$ and a
structure \DIFdelbegin \DIFdel{$T_{[i,j]}$}\DIFdelend \DIFaddbegin \DIFadd{$\strT_{[i,j]}$}\DIFaddend , due to the introduction of the base pair
$(r,j)$.

Given RNA sequence \seq and compatible initial structure \strSt,
we define the {\em polynomial}

\begin{align}
\label{eq:fftbor:zOfX}
\fullZx = \sum_{k=0}^n z_k x^k
\end{align}

where coefficients $z_k=\bfZ{k}{1,n}$. Moreover, because of
\eqnref{fftbor:maxDist} and the fact that the minimum number of
unpaired bases in a hairpin loop $\theta$ is \DIFdelbegin \DIFdel{3}\DIFdelend \DIFaddbegin \DIFadd{$3$}\DIFaddend , we know that $z_n=0$,
so that \fullZx is a polynomial of degree strictly less than $n$.
If we evaluate the polynomial \fullZx for $n$ distinct values

\begin{align}
\label{eq:fftbor:solutionsForAlpha}
\emZof{}{a_1} = y_1, \dots, \emZof{}{a_n} = y_n,
\end{align}

then the Lagrange polynomial interpolation formula guarantees that
$\fullZx = \sum_{k=1}^n y_k P_k(x)$, where the polynomials $P_k(x)$ have degree
at most $n-1$ and are given by the Lagrange formula

\begin{align}
\label{eq:fftbor:lagrangeInterpolation}
P_k(x) = \frac{\prod_{i\ne k} (x-x_i)}{\prod_{i \ne k} (x_k-x_i)}.
\end{align}

Since the polynomials $P_k(x)$ can be explicitly computed, it follows that
we can compute the coefficients $z_k$ of polynomial \fullZx. As we describe
below, the evaluation of \fullZx for a fixed value of $x$ can be done in
time \On{3} and space \On{2}.  It follows that the coefficients
$z_k=\bfZ{k}{1,n}$ can be computed after
$n$ evaluations of \fullZx, where the space for each evaluation of \fullZx
is re-used; hence these evaluations can be performed in time \On{4} and space
\On{2}. Finally,
Lagrange interpolation is clearly computable in time \On{3}.
Although this approach is theoretically sound, there are severe
numerical stability issues related to the interpolation method
\citep{highambarycentricinterpolation},
the choice of values $a_1,\dots,a_{n}$ in the interpolation,
and floating point arithmetic (round-off error) related to the
astronomically large values of the partition functions
\bfZ{k}{1,n}, for $0 \leq k < n$. After many unsuccessful
approaches including scaling we obtained excellent results by
interpolating the polynomial $p(x)$, defined in \eqnref{fftbor:pOfX},
rather than the polynomial \fullZx, defined in \eqnref{fftbor:zOfX},
and performing interpolation with the \fft (FFT) \citep{cormen}
where \alphaN are
chosen to be \nRoUs,
$\alpha_k = \kRoU$.
One
advantage of the FFT is that interpolation can be performed in $O(n \log n)$
time, rather than the cubic time required by using the Lagrange formula
shown in \eqnref{fftbor:lagrangeInterpolation} or by Gaussian elimination. Fewer
numerical operations implies increased numerical stability in our application.

\subsection{Recursions to compute the polynomial
\texorpdfstring{\emZ{i,j}}{}}
\label{subsec:fftbor:polynomial}

Given an initial secondary structure \strSt of a
given RNA sequence \seq, our goal is to compute

\begin{align}
\label{eq:fftbor:defZofK}
\bfZ{k}{1,n} = \sum_{\mathclap{\substack{\str \text{ such that }\\ \dBP{\str}{\strSt}=k}}}\;
\boltzNuss{\str}
\end{align}

where \str can be any structure compatible with \seq.
As previously mentioned, the recurrence relation for \rnabor
with respect to the Nussinov energy model $E_0$ is

\begin{align}
\label{eq:fftbor:rnaborNuss}
\bfZ{k}{i,j} = \bfZ{k-b_0}{i,j-1}\enspace +
\sum_{\substack{(s_r,s_j) \in \bpSet, \\ i \le r<j}}
\left(
\boltzNuss{r,j}\enspace \sum_{\mathclap{w+w'=k-b(r)}}\quad
\bfZ{w}{i,r-1} \bfZ{w'}{r+1,j-1}
\right)
\end{align}

where $E_0(r,j)=-1$ if $r$ and $j$ can base-pair and otherwise
$+\infty$, and
$b_0 = 1$ if $j$ is base paired in $\strSt_{[i,j]}$ and \DIFdelbegin \DIFdel{0 }\DIFdelend \DIFaddbegin \DIFadd{$0$ }\DIFaddend otherwise, and
$b(r)=\dBP{\strSt_{[i,j]}}{\strSt_{[i,r-1]} \cup \strSt_{[r+1,j-1]} \cup\{(r,j)\}}$.
The following theorem shows that an analogous recursion can be used to compute
the {\em polynomial} $\emZ{i,j}$ defined by

\begin{align}
\label{eq:fftbor:polynomialZij}
\emZ{i,j} = \sum_{k=0}^n z_k(i,j)\,x^k
\end{align}

where

\begin{align}
z_k(i,j)\>=\>
\bfZ{k}{i,j}\enspace=\enspace
\sum_{\mathclap{\substack{\str \text{ such that } \\
\dBP{\str}{\strSt_{[i,j]}}=k}}}\enspace
\boltzNuss{\str}.
\end{align}

Here, in the summation, \str runs over structures on \seqIJ, which
are \kNbrs of the restriction $\strSt_{[i,j]}$ of initial structure
\strSt to interval $[i,j]$, and
$E_0(S)=-1 \cdot |S|$ denotes the Nussinov-Jacobson energy of \str.

\begin{theorem}
\label{thm:fftbor:recursions}
Let \seqN be a given RNA sequence.
For any integers $1 \leq i \leq j \leq n$, let

\begin{align}
\emZ{i,j} = \sum_{k=0}^n z_k\,x^k
\end{align}

where

\begin{align}
z_k(i,j) = \bfZ{k}{i,j}.
\end{align}

Then for $i\leq j \leq i+\theta$, $\emZ{i,j}=1$ and for
$j>i+\theta$ we have the recurrence relation

\begin{align}
\label{eq:fftbor:fftborNussPoly}
\emZ{i,j} = \emZ{i,j-1} \cdot x^{b_0} +
\sum_{\substack{(s_r,s_j) \in \bpSet, \\ i\le r<j}}
\left(
\boltzNuss{r,j} \cdot \emZ{i,r-1} \cdot \emZ{r+1,j-1} \cdot x^{b(r)}
\right).
\end{align}

where $b_0 = 1$ if $j$ is base-paired in $\strSt_{[i,j]}$ and \DIFdelbegin \DIFdel{0 }\DIFdelend \DIFaddbegin \DIFadd{$0$ }\DIFaddend otherwise, and
$b(r) =
\dBP{\strSt_{[i,j]}}{\strSt_{[i,r-1]} \cup \strSt_{[r+1,j-1]} \cup \{(r,j)\}}$.
\end{theorem}

\begin{proof}
First, some notation is necessary.
Recall that if $F$ is an arbitrary
polynomial [resp. analytic] function, then $[x^k]F(x)$
denotes the coefficient of $x^k$ [resp. the $k$th Taylor coefficient in the
Taylor expansion of $F(x)$]. For instance, in
\eqnref{fftbor:pOfX}, $[x^k]p(x) = \pk$, and in
\eqnref{fftbor:zOfX}, $[x^k]\fullZx = z_k$.

By definition, it is clear that $\emZ{i,j}=1$ if $i \leq j \leq i + \theta$,
where we recall that $\theta = 3$ is the minimum number of unpaired bases in
a hairpin loop. For $j > i + \theta$, we have

\begin{align}
\begin{split}
[x^k] \emZ{i,j} &= z_k(i,j) = \bfZ{k}{i,j} \\
&= \bfZ{k-b_0}{i,j-1} +
\sum_{r=i}^{j-1}\hspace{2.25em} \sum_{\mathclap{k_0+k_1=k-b(r)}}\hspace{1.5em}
\left( \boltzNuss{r,j} \cdot \bfZ{k_0}{i,r-1} \cdot \bfZ{k_1}{r+1,j-1} \right) \\
&= [x^{k-b_0}] \emZ{i,j-1} \\
&+ \sum_{r=i}^{j-1}\hspace{2.25em} \sum_{\mathclap{k_0+k_1=k-b(r)}}\hspace{1.5em}
\left( \boltzNuss{r,j} \cdot \left( [x^{k_0}] \emZ{i,r-1} \right) \cdot
\left( [x^{k_1}] \emZ{r+1,j-1} \right) \right) \\
&= [x^{k-b_0}] \emZ{i,j-1} \\
&+ \sum_{r=i}^{j-1}\hspace{2.25em} \sum_{\mathclap{k_0+k_1=k-b(r)}}\hspace{1.5em}
\left( \boltzNuss{r,j} \cdot [x^{k_0+k_1}]
\left( \emZ{i,r-1} \cdot \emZ{r+1,j-1} \right) \right).\\
\end{split}
\end{align}

By induction, the proof of the theorem now follows.
\end{proof}

Notice that if one were to compute all terms of the polynomial $\emZ{1,n}$
by explicitly performing polynomial multiplications,
then the computation would require \On{5} time and \On{3} space.
Instead of explicitly performing polynomial expansion in {\em variable} $x$,
we instantiate $x$ to a fixed complex number $\alpha \in \mathbb{C}$, and apply
the following recursion for this instantiation:

\begin{align}
\label{eq:fftbor:fftborNussPolyAlpha}
\emZof{i,j}{\alpha} = \emZof{i,j-1}{\alpha} \cdot \alpha^{b_0} +
\sum_{\substack{(s_r,s_j) \in \bpSet, \\ i \le r<j}}
\left(
\boltzNuss{r,j} \cdot
\emZof{i,r-1}{\alpha} \cdot \emZof{r+1,j-1}{\alpha} \cdot \alpha^{b(r)}
\right).
\end{align}

In this fashion, we can compute $\emZof{}{\alpha}=\emZof{1,n}{\alpha}$ in
\On{3} time and \On{2} space. For $n$ distinct complex values
\alphaN, we can compute and save only the
values $\emZof{}{\alpha_0},\dots,\emZof{}{\alpha_{n-1}}$, each time re-using the
\On{2} space for the next computation of $\emZof{}{\alpha_k}$. It follows that
the computation resources used to determine the (column) vector

\begin{align}
\label{eq:fftbor:yColumn}
\bfY = (y_0,\dots,y_{n-1})^{\text T} =
\left(
\begin{array}{l}
y_0 \\
y_1 \\
\vdots \\
y_{n-1} \\
\end{array}
\right)
\end{align}

where
\DIFdelbegin \DIFdel{$y_0=\emZof{}{\alpha_0},\dots,y_{n-1}=\emZof{}{\alpha_{n-1}})$ }\DIFdelend \DIFaddbegin \DIFadd{$y_0=\emZof{}{\alpha_0},\dots,y_{n-1}=\emZof{}{\alpha_{n-1}}$ }\DIFaddend is thus quartic time \On{4} and quadratic space \On{2}.

\subsection{Polynomial interpolation to evaluate
\texorpdfstring{\emZ{i,j}}{}}
\label{subsec:fftbor:fft}

Let $\omega = \pRoU$ be the principal \nRoU.
Recall that the Vandermonde matrix $V_n$ is defined to be the
$n \times n$ matrix, whose $i,j$ entry is $\omega^{i \cdot j}$; i.e.

\begin{align}
\label{eq:fftbor:vandermonde}
V_n =
\left(
\begin{array}{rrrrr}
1 & 1 & 1 & \dots & 1 \\
1 & \omega & \omega^2 & \dots & \omega^{n-1} \\
1 & \omega^2 & \omega^4 & \dots & \omega^{2(n-1)} \\
1 & \omega^3 & \omega^6 & \dots & \omega^{3(n-1)} \\
\vdots & \vdots & \vdots & \vdots & \vdots \\
1 & \omega^{n-1} & \omega^{2(n-1)} & \dots & \omega^{(n-1)(n-1)} \\
\end{array}
\right)
\end{align}

The \fft is defined to be the $O(n \log n)$
algorithm to compute the Discrete Fourier Transform (DFT), defined
as the matrix product $\bfY = V_n {\bf A}$:

\begin{align}
\label{eq:fftbor:dftMatrix}
\left(
\begin{array}{l}
y_0 \\
y_1 \\
y_2 \\
\vdots \\
y_{n-1} \\
\end{array}
\right)
= V_n \cdot
\left(
\begin{array}{l}
a_0 \\
a_1 \\
a_2 \\
\vdots \\
a_{n-1} \\
\end{array}
\right)
\end{align}

On page \DIFdelbegin \DIFdel{837 }\DIFdelend \DIFaddbegin \DIFadd{$837$ }\DIFaddend of \citep{cormen}, it is shown that the
$(i,j)$ entry of $V_n^{-1}$ is $\frac{\omega^{-j i}}{n}$
and that

\begin{align}
\label{eq:fftbor:aFromY}
a_j = \frac{1}{n} \sum_{k=0}^{n-1} y_k\,\omega^{-kj}
\end{align}

for $j=0,\dots,n-1$.

Since we defined \bfY in \eqnref{fftbor:yColumn} by $\bfY =
(y_0,\dots,y_{n-1})^{\text T}$, where
\DIFdelbegin \DIFdel{$y_0=\emZof{}{\alpha_0},\dots,y_{n-1}=\emZof{}{\alpha_{n-1}})$
}\DIFdelend \DIFaddbegin \DIFadd{$y_0=\emZof{}{\alpha_0},\dots,y_{n-1}=\emZof{}{\alpha_{n-1}}$
}\DIFaddend and $\alpha_k = \omega^k \kRoU$, it follows that the coefficients
$z_k=\bfZ{k}{1,n}$ in the polynomial
$\fullZx = z_0 + z_1 x + \dots + z_{n-1} x^{n-1}$ defined in
\eqnref{fftbor:zOfX} can be computed, at least in principle,
by using the \fft. It turns out, however, that the values of
\bfZ{k}{1,n} are so astronomically large, that the ensuing numerical
instability makes even this approach infeasible for values of $n$
that exceed \DIFdelbegin \DIFdel{56 }\DIFdelend \DIFaddbegin \DIFadd{$56$ }\DIFaddend (data not shown).
Nevertheless, our approach can be modified as follows.
Define \bfY by $\bfY = (y_1,\dots,y_n)^{\text T}$, where
$y_1=\frac{\emZof{}{\alpha_1}}{\fullZ},\dots,
y_{n}=\frac{\emZof{}{\alpha_n}}{\fullZ}$, and
\fullZ is the partition function defined in \eqnref{fftbor:partFunc}.
Using the \fft to compute the \idft, it follows from
\eqnref{fftbor:aFromY} that we can compute the probabilities $p_0,\dots,p_{n-1}$
that are coefficients of the polynomial
$p(x) = p_0 + p_1 x + \dots + p_{n-1}x^{n-1}$
defined in \eqnref{fftbor:pOfX}. For genomics applications, we are
only interested in the $m$ most significant digits of each \pk, as described
in the pseudocode on the following page.
\medskip

\begin{figure}[!ht]
\hrule \rule[0ex]{0pt}{0pt}
\begin{center}
{\large Pseudocode for \fftbor} \\
\end{center}
\begin{tabular*}{\textwidth}{ll}
{\sc Purpose:} & Computes the $m$ most significant digits
of probabilities \DIFdelbeginFL \DIFdelFL{$\pk = \sfrac{\bfZ{k}{1,n}}{\fullZ}$ }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$\pk = \rfrac{\bfZ{k}{1,n}}{\fullZ}$ }\DIFaddendFL \rule[-1.5ex]{0pt}{0pt} \\
{\sc Input:} & RNA sequence $\seq = \seqN$, secondary
structure \strSt of \seq, integer $m$ \rule[-1.5ex]{0pt}{0pt} \\
{\sc Output:} & Probabilities \DIFdelbeginFL \DIFdelFL{$\pk = \sfrac{\bfZ{k}{1,n}}{\fullZ}$ }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$\pk = \rfrac{\bfZ{k}{1,n}}{\fullZ}$ }\DIFaddendFL to $m$ significant digits for $k=0,\dots,n-1$ \rule[-1.75em]{0pt}{0pt} \\
\hline \rule[0ex]{0pt}{0pt}
\end{tabular*}
\begin{algorithmic}[1]
\Function{FFTbor}{\seq, \strSt, $m$}
\State $n \gets \textit{length}(\seq)$
\For{$k \gets 0, n-1$}
\Comment{Compute all \nRoUs}
\State $\omega_k \gets \exp(\frac{2 \pi i k}{n})$
\EndFor
\For{$k \gets 0, n-1$}
\Comment{Note that $\emZof{}{\omega_0} = \fullZ$}
\State $y_k \gets 10^m \cdot \frac{\emZof{}{\omega_k}}{\emZof{}{\omega_0}}$
\EndFor
\For{$k \gets 0, n-1$}
\Comment{Compute IDFT from \eqnref{fftbor:aFromY}}
\State $a_k \gets \frac{1}{n} \sum_{j=0}^{n-1} y_j\, \omega^{-kj}$
\State $\pk \gets 10^{-m} \cdot \lfloor a_k \rfloor$
\Comment{Truncate to $m$ significant digits}
\EndFor
\State \textbf{return} $p_0,\dots,p_{n-1}$
\Comment{Return all \pk for $0 \leq k < n$,
from \eqnref{fftbor:probK}}
\EndFunction
\rule[-0.35ex]{0pt}{0pt}
\end{algorithmic}
\caption[Pseudocode for \fftbor]{The function {\sc FFTbor} computes the $m$ most significant digits
of $p_0,\dots,p_{n-1}$, where $\pk = \frac{\bfZ{k}{}}{\fullZ}$. This algorithm
operates in \On{4} time and \On{2} space, a significant improvement over its
predecessor \rnabor.}
\label{fig:fftbor:algo}
\rule[0ex]{0pt}{1.5em} \hrule
\end{figure}

\section{Benchmarking and performance considerations}
\label{sec:fftbor:benchmarking}

In this subsection, we show that we need only evaluate the polynomial
\fullZx, as defined in
\eqnref{fftbor:zOfX}, for $n/2$ of the \nRoUs.
 It is first necessary to recall the definition of complex
conjugate.
Recall that the complex conjugate of $z$ is denoted by $\overline{z}$;
i.e. if $z=\aPbi$ where $a,b \in \mathbb{R}$ are real numbers and
$i = \sqrt{-1}$,  then $\overline{z} = \aMbi$.

\begin{lemma}
\label{lem:fftbor:compconj}

If \fullZx is the complex polynomial defined in
\eqnref{fftbor:zOfX}, then for any \nRoU
 $\alpha$, it is the case that $\emZof{}{\overline{\alpha}} =
\overline{\emZof{}{\alpha}}$. In other words, if $\alpha$ is a \nRoU
 of the form \aPbi, where $a,b \in \mathbb{R}$ and $b>0$, and
if $\emZof{}{\aPbi} = A + Bi$ where $A,B \in \mathbb{R}$, then it is the case that

\begin{align}
\emZof{}{\aMbi} = A - Bi.
\end{align}

\end{lemma}

\begin{proof}
Letting $i = \sqrt{-1}$, if
$\theta = \frac{2 \pi}{n}$, then
$\omega = e^{i \theta} = \cos(\theta) + i \sin(\theta)$
is the principal \nRoU, and
$e^{(0) \cdot i \theta} = 1 = \omega^{0}, \dots ,
e^{(n-1) \cdot i \theta} = \omega^{n-1}$ together
constitute the complete collection of all
\nRoUs---i.e. the $n$ unique solutions of
of the equation $x^n -1 = 0$ over the field $\mathbb{C}$ of complex numbers.
Clearly, for any $1 \leq k < n$,
$e^{-i k \theta} = 1 \cdot e^{-i k \theta} =
e^{2 \pi i} \cdot e^{-i k \theta} = e^{i(2 \pi - k \theta)} =
e^{i(n \theta - k \theta)} = e^{i \theta (n - k)}$.
Moreover, if $\omega^k = e^{i k \theta} = a + b i$ where
$b>0$, then we have $e^{-i k \theta} = \aMbi$. It follows that for any
\nRoU of the form \aPbi, where $b>0$, the number \aMbi
is also an \nRoU.

Recall that $\fullZx = \sum_{k=0}^n z_k x^k$, where
$z_k \in \mathbb{R}$ are real numbers representing the partition function
\bfZ{k}{1,n} over
all secondary structures of a given RNA sequence \seqN,
whose \bpd from initial structure
\strSt is $k$. Thus, in order to prove the lemma, it suffices to show
that for all values $k=0,\dots,n-1$, if \aPbi is a
\nRoU, where $a,b \in \mathbb{R}$
and $b>0$, and if $(\aPbi)^k = C+Di$ where $C,D \in \mathbb{R}$, {\em then}
$(\aMbi)^k = C-Di$. Indeed, we have the following.

\begin{align}
\begin{split}
(\aPbi)^m &= \sum_{k=0}^m \binom{m}{k}\; a^{m-k} \cdot (bi)^k \\
(bi)^k  &=
\begin{cases}
b^k    & \text{if $k \equiv 0 \bmod 4$} \\
i b^k  & \text{if $k \equiv 1 \bmod 4$} \\
-b^k   & \text{if $k \equiv 2 \bmod 4$} \\
-i b^k & \text{if $k \equiv 3 \bmod 4$} \\
\end{cases}
\end{split}
\end{align}

\begin{align}
\begin{split}
(\aMbi)^m &= \sum_{k=0}^m \binom{m}{k}\; a^{m-k} \cdot (-bi)^k \\
(-bi)^k &=
\begin{cases}
b^k   & \text{if $k \equiv 0 \bmod 4$} \\
-ib^k & \text{if $k \equiv 1 \bmod 4$} \\
-b^k  & \text{if $k \equiv 2 \bmod 4$} \\
ib^k  & \text{if $k \equiv 3 \bmod 4$} \\
\end{cases}
\end{split}
\end{align}

It follows that each term of the form
$a^{m-k} \cdot (bi)^k$, for $k=0,\dots,m$, is the complex conjugate of
$a^{m-k} \cdot (-bi)^k$, and thus $(\aPbi)^m$ is the complex conjugate of
$(\aMbi)^m$. Since \emZof{}{\aPbi} is a sum of terms of the form $z_k (\aPbi)^k$,
it follows that \emZof{}{\aMbi} is the complex conjugate of \emZof{}{\aPbi}.
This completes the proof of the lemma.

\end{proof}

Lemma \ref{lem:fftbor:compconj} immediately entails that we need only to evaluate \fullZx on $n/2$
many of the \nRoUs---namely, those of the form
\aPbi, where $b \geq 0$. The remaining values of \fullZx are obtained by
taking conplex conjugates of the first $n/2$ values. This, along with a
precomputation of powers of the \nRoUs, leads to an
enormous performance speed-up in our implementation of \fftbor.

\section{Coarse-grained kinetics with \fftbor}
\label{sec:fftbor:kinetics}

The output of \fftbor, as shown in
\Figref{fftbor:tppDistributions}, is a probability distribution,
where the $x$-axis represents the \bpd from an arbitrary,
but fixed secondary structure \strSt, and the $y$-axis represents the
Boltzmann probability $p(k) = \frac{\bfZ{k}{}}{\fullZ}$ that a secondary structure
has \bpd $k$ from \strSt. Arguably, this probability distribution
is an accurate one-dimensional projection of the rugged, high dimensional energy
landscape near structure \strSt,
of the sort artistically rendered in the well-known
energy landscape depicted in Figure 1 of \citep{wolynes.ptam05}.
A hypothesis behind theoretical work in biomolecular folding theory in
\citep{bryngelson.p95}
is that kinetic folding slows down as the energy landscape becomes more
{\em rugged}. This is borne out in our computational experiments for RNA
using \fftbor, as reported
in \Figref{fftbor:tppDistributions}.

We randomly chose two TPP \rb
aptamers from the seed alignment for
Rfam family \DIFdelbegin \DIFdel{RF00059}\DIFdelend \DIFaddbegin \DIFadd{RF$00059$}\DIFaddend . The first sequence \DIFaddbegin \DIFadd{from }{\em \DIFadd{B. bacteriovorus}} \DIFaddend has EMBL accession code
\DIFdelbegin \DIFdel{BX842649.1/277414--277318 and is comprised of the 97 }\DIFdelend \DIFaddbegin \DIFadd{BX$842649.1$ $277414$--$277318$ and is composed of the $97$ }\DIFaddend nt sequence
\seqsplit{%
  ACCUGACGCUAGGGGUGUUGGUGAAUUCACCGACUGAGAAUAACCCUUU%
  GAACCUGAUAGAGAUAAUGCUCGCGCAGGGAAGCAAGAAUAGAAAGAU%
}\DIFdelbegin \DIFdel{, while the second sequence }\DIFdelend \DIFaddbegin \DIFadd{. The second sequence from the marine metagenome }\DIFaddend has
EMBL accession code \DIFdelbegin \DIFdel{AACY022101973.1/389--487 and is comprised of the 99
}\DIFdelend \DIFaddbegin \DIFadd{AACY$022101973.1$ $389$--$487$ and is composed of the $99$
}\DIFaddend nt sequence
\seqsplit{%
  UAUAAGUCCAAGGGGUGCCAAUUGGCUGAGAUGGUUUUAACCAAUCCCUU%
  UGAACCUGAUCCGGUUAAUACCGGCGUAGGAAUGGAUUUUCUCUACAGC%
}.
Rfam consensus and \mfess for both sequences are
depicted in \Figref{fftbor:tppConsensusAndMfe}.
Despite the fact that there is no sequence \DIFdelbegin \DIFdel{homology }\DIFdelend \DIFaddbegin \DIFadd{identity }\DIFaddend according to
pairwise BLAST \citep{blast}, this figure clearly demonstrates that
consensus and
\mfess closely resemble each other, and that the
structures of both TPP \rb aptamers are quite similar, with the
exception of the leftmost hairpin loop [resp. multiloop].
The MFE structures differ
from the consensus structures principally by the addition of base pairs not
determined by covariation in the Rfam alignment.
Indeed, if we let \DIFdelbegin \DIFdel{$\str_0,\str_1$
}\DIFdelend \DIFaddbegin \DIFadd{$\str_0,\,\str_1$
}\DIFaddend denote the Rfam consensus structure [resp. MFE structure] for the \DIFdelbegin \DIFdel{97 }\DIFdelend \DIFaddbegin \DIFadd{$97$ }\DIFaddend nt
sequence with EMBL accession code \DIFdelbegin \DIFdel{BX842649.1/277414--277318}\DIFdelend \DIFaddbegin \DIFadd{BX$842649.1$ $277414$--$277318$}\DIFaddend , then
$\str_0 \setminus \str_1$ has \DIFdelbegin \DIFdel{4 }\DIFdelend \DIFaddbegin \DIFadd{$4$ }\DIFaddend base pairs, and $\str_1 \setminus \str_0$
 has \DIFdelbegin \DIFdel{7
}\DIFdelend \DIFaddbegin \DIFadd{$7$
}\DIFaddend base pairs. If we let \DIFdelbegin \DIFdel{$\strT_0,\strT_1$
}\DIFdelend \DIFaddbegin \DIFadd{$\strT_0,\,\strT_1$
}\DIFaddend denote the Rfam consensus structure [resp. MFE structure] for the \DIFdelbegin \DIFdel{99 }\DIFdelend \DIFaddbegin \DIFadd{$99$ }\DIFaddend nt
sequence with EMBL accession code
\DIFdelbegin \DIFdel{AACY022101973.1/389--487}\DIFdelend \DIFaddbegin \DIFadd{AACY$022101973.1$ $389$--$487$}\DIFaddend , then
$\strT_0 \setminus \strT_1$ has \DIFdelbegin \DIFdel{1 }\DIFdelend \DIFaddbegin \DIFadd{$1$ }\DIFaddend base pair, and $\strT_1 \setminus \strT_0$ has \DIFdelbegin \DIFdel{5
}\DIFdelend \DIFaddbegin \DIFadd{$5$
}\DIFaddend base pairs.

\begin{figure}[!ht]
\centering
\includegraphics[width=.9\textwidth]{Figures/FFTbor/tppConsensusAndMfe.pdf}
\caption[Rfam consensus structures (Rfam) and \mfe (MFE)
secondary structures for two thiamine pyrophosphate (TPP) \rb aptamers]{Rfam consensus structures (Rfam) and \mfe (MFE)
secondary structures for two thiamine pyrophosphate (TPP) \rb aptamers,
chosen at random from \DIFdelbeginFL \DIFdelFL{RF00059 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{RF$00059$ }\DIFaddendFL Rfam family seed alignment
\citep{Gardner.nar11}. Using pairwise BLAST \citep{blast}, there is no
sequence similarity, although the secondary structures are very similar,
as shown in this figure. From left to right:
{\em (A)} MFE structure for \DIFdelbeginFL \DIFdelFL{BX842649.1/277414--277318}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{BX$842649.1$ $277414$--$277318$}\DIFaddendFL .
{\em (B)} Rfam consensus structure for \DIFdelbeginFL \DIFdelFL{BX842649.1/277414--277318}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{BX$842649.1$ $277414$--$277318$}\DIFaddendFL .
{\em (C)} MFE structure for \DIFdelbeginFL \DIFdelFL{AACY022101973.1/389--487}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AACY$022101973.1$ $389$--$487$}\DIFaddendFL .
{\em (D)} Rfam consensus structure for \DIFdelbeginFL \DIFdelFL{AACY022101973.1/389--487}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AACY$022101973.1$ $389$--$487$}\DIFaddendFL .
}
\label{fig:fftbor:tppConsensusAndMfe}
\end{figure}

We ran \fftbor on each of the TPP \rb aptamer
sequences, with the MFE structure of each
sequence taken as the initial structure \strSt for that sequence. For the
first sequence, \DIFdelbegin \DIFdel{BX842649.1/277414--277318}\DIFdelend \DIFaddbegin \DIFadd{BX$842649.1$ $277414$--$277318$}\DIFaddend , the \fftbor output
suggests that there are low energy structures
at a distance from the MFE structure, which might compete with the MFE
structure and hence slow the kinetics of folding. In contrast, for the
second sequence, \DIFdelbegin \DIFdel{AACY022101973.1/389--487}\DIFdelend \DIFaddbegin \DIFadd{AACY$022101973.1$ $389$--$487$}\DIFaddend , the \fftbor output suggests
that there are no such competing low energy structures, hence
the second sequence should fold more quickly than the first.

\begin{figure}[!ht]
\centering
\DIFdelbeginFL %DIFDELCMD < \includegraphics[width=.45\textwidth]{Figures/FFTbor/FFTbor_AACY022101973_1.pdf}
%DIFDELCMD < \quad
%DIFDELCMD < %%%
\DIFdelendFL \includegraphics[width=.45\textwidth]{Figures/FFTbor/FFTbor_BX842649_1.pdf}
\DIFaddbeginFL \quad
\includegraphics[width=.45\textwidth]{Figures/FFTbor/FFTbor_AACY022101973_1.pdf}
\DIFaddendFL \caption[Output from \fftbor on two randomly selected
thiamine pyrophosphate \rb (TPP) aptamers]{Output from \fftbor on two randomly selected
thiamine pyrophosphate \rb (TPP) aptamers, taken from the Rfam database
\citep{Gardner.nar11}. The $x$-axis represents \bpd from the
\mfes for each given sequence; the $y$-axis represents
Boltzmann probabilities \DIFdelbeginFL \DIFdelFL{$p(k) = \sfrac{\bfZ{k}{}}{\fullZ}$}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$p(k) = \frac{\bfZ{k}{}}{\fullZ}$}\DIFaddendFL , where
\bfZ{k}{} denotes the sum of Boltzmann factors or all secondary structures,
whose \bpd from the MFE structure is exactly $k$.
{\em (Left)}
The \DIFdelbeginFL \DIFdelFL{97 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$97$ }\DIFaddendFL nt sequence \DIFdelbeginFL \DIFdelFL{BX842649.1/277414--277318 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{BX$842649.1$ $277414$--$277318$ }\DIFaddendFL appears to have a rugged energy
landscape near its \mfes, with distinct
low energy structures that may compete with the MFE structure during the
folding process.
{\em (Right)}
The \DIFdelbeginFL \DIFdelFL{99 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$99$ }\DIFaddendFL nt sequence, \DIFdelbeginFL \DIFdelFL{AACY022101973.1/389--487 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AACY$022101973.1$ $389$--$487$ }\DIFaddendFL appears to have a smooth energy
landscape near its MFE structure, with no distinct low energy structures
to might compete with the MFE structure.
Based on the \fftbor output or {\em structural profile} near MFE
structure \strSt, one might expect
folding time for the first sequence to increase due to competition from
metastable structures, while one might expect the second sequence to have
rapid folding time.
Computational Monte Carlo folding experiments bear out this fact.
\kinfold \citep{flamm} simulations clearly show that the second
sequence folds
at least four times more quickly than the first sequence. See section
\ref{sec:fftbor:kinetics} for
details.}
\label{fig:fftbor:tppDistributions}
\end{figure}

To test the hypothesis that folding is slower for rugged energy landscapes,
we ran the kinetic folding software, \kinfold \citep{flamm},
on each of the two TPP \rb aptamer sequences,
\DIFdelbegin \DIFdel{BX842649.1/277414--277318 and AACY022101973.1/389--487}\DIFdelend \DIFaddbegin \DIFadd{BX$842649.1$ $277414$--$277318$ and AACY$022101973.1$ $389$--$487$}\DIFaddend ,
to determine the \mfpt (MFPT) to
fold into the MFE structure, when starting from the empty structure.
In this computational
experiment, we took MFPT to be the average number of Monte Carlo steps
taken by \kinfold---each step consisting of the addition or removal
of a single base pair---to fold the
empty structure into the MFE
structure, where the average was taken over \DIFdelbegin \DIFdel{30 }\DIFdelend \DIFaddbegin \DIFadd{$30$ }\DIFaddend runs, with an absolute
maximum number of Monte Carlo steps taken to be \DIFdelbegin \DIFdel{500,000.
}\DIFdelend \DIFaddbegin \DIFadd{$500,000$.
}\DIFaddend The first sequence, \DIFdelbegin \DIFdel{BX842649.1/277414--277318}\DIFdelend \DIFaddbegin \DIFadd{BX$842649.1$ $277414$--$277318$}\DIFaddend , converged within \DIFdelbegin \DIFdel{500,000
}\DIFdelend \DIFaddbegin \DIFadd{$500,000$
}\DIFaddend steps only for \DIFdelbegin \DIFdel{20 out of 30 }\DIFdelend \DIFaddbegin \DIFadd{$20$ out of $30$ }\DIFaddend runs. Assigning the maximum step count of
\DIFdelbegin \DIFdel{500,000 for the 10 }\DIFdelend \DIFaddbegin \DIFadd{$500,000$ for the $10$ }\DIFaddend runs that did not converge, we found a \mfpt
of $311,075.06$ steps for this sequence.
The second sequence, \DIFdelbegin \DIFdel{AACY022101973.1/389--487}\DIFdelend \DIFaddbegin \DIFadd{AACY$022101973.1$ $389$--$487$}\DIFaddend , converged within \DIFdelbegin \DIFdel{500,000
steps in 29 out of 30 }\DIFdelend \DIFaddbegin \DIFadd{$500,000$
steps in $29$ out of $30$ }\DIFaddend runs, and we found a \mfpt of
$61,575.69$ steps for this sequence. From computational experiments of this
type, it is suggestive that \fftbor may prove useful in synthetic
biology,
where one would like to design rapidly folding RNA molecules that
fold into a designated target structure.

In order to more systematically determine the relation between kinetic
folding speed and the ruggedness of an energy landscape near the MFE structure,
we need to numerically quantify ruggedness. To this end, in the following
we define the notion of \ebpd to a designated
structure. Let \strSt be an arbitrary secondary structure of the RNA sequence
$\seq = \seqN$.
The expected \bpd to \strSt is defined by

\begin{align}
E[ \{ \dBP{\str}{\strSt} : \str \in \mathbb{S}(\seqN)\} ] =
\sum_{\str} P(\str) \cdot \dBP{\str}{\strSt}
\end{align}

where
$\mathbb{S}(\seqN)$ denotes the set of secondary structures for
$\seq = \seqN$, $P(\str) = \frac{\boltzf{\str}}{\fullZ}$ is the Boltzmann
probability of \str, and
\dBP{\str}{\strSt} denotes \bpd between \str and \strSt.
If we run \fftbor on an input sequence \seq and secondary structure
\strSt, then clearly
$E[ \{ \dBP{\str}{\strSt} : \str \in \mathbb{S}(\seqN)\} ] =
\sum_{k} k \cdot p(k)$, where $p(k)=\frac{\bfZ{k}{}}{\fullZ}$, obtained from the
program output.  If \strSt is the empty structure, then \fftbor output
is simply the probability distribution of the number of base pairs per
secondary structure, taken over the Boltzmann ensemble of all structures.

For the benchmarking assay, we took all \DIFdelbegin \DIFdel{61 }\DIFdelend \DIFaddbegin \DIFadd{$61$ }\DIFaddend selenocysteine insertion sequence
(SECIS) sequences from the seed alignment of Rfam family \DIFdelbegin \DIFdel{RF00031
}\DIFdelend \DIFaddbegin \DIFadd{RF$00031$
}\DIFaddend \citep{Gardner.nar11}. Average length was $64.32 \pm 2.83$ nt.
For each sequence, we ran both \fftbor (when starting
from the empty structure rather than the MFE structure) and a Monte Carlo
folding algorithm, developed by E. Freyhult and P. Clote (unpublished).
Using the Monte Carlo algorithm, we
determined the \mfpt (MFPT), defined as the average
taken over \DIFdelbegin \DIFdel{50 }\DIFdelend \DIFaddbegin \DIFadd{$50$ }\DIFaddend runs, of the number of Monte Carlo steps taken to fold
the empty structure into the MFE structure, where an absolute upper bound
of \DIFdelbegin \DIFdel{5 }\DIFdelend \DIFaddbegin \DIFadd{$5$ }\DIFaddend million steps was allowed in the simulation.

Surprisingly, we found that there is a significant
correlation of $0.4847$ with one-tailed
$p$-value of $0.0002$ between the
standard deviation of the \fftbor output (when starting from the
empty structure) and logarithm base \DIFdelbegin \DIFdel{10 }\DIFdelend \DIFaddbegin \DIFadd{$10$ }\DIFaddend of the \mfpt.
As described above, \fftbor output is simply
the probability distribution
for the number of base pairs per structure, taken over the ensemble
of all secondary structure for the input RNA
sequence. The standard deviation of this probability distribution corresponds
to a notion of the \DIFdelbegin %DIFDELCMD < {\em %%%
\DIFdel{width }%DIFDELCMD < } %%%
\DIFdelend \DIFaddbegin \DIFadd{width }\DIFaddend of the distribution. It is possible that those
sequences having distributions tightly centered around the
mean have faster folding times than those with a wider distribution
due to other local minima causing the RNA to get trapped while folding.

\begin{table}[!ht]
\centering
\begin{tabularx}{\linewidth}{c *{6}{R}}
  \toprule
  ~ & \small{$\mu$} & \small{$\sigma$} & \DIFdelbeginFL %DIFDELCMD < \small{$\sfrac{\sigma}{\mu}$} %%%
\DIFdelendFL \DIFaddbeginFL \small{$\rfrac{\sigma}{\mu}$} \DIFaddendFL & \small{$n$} & \small{MFE} & \small{$\log_{10}(\text{MFPT})$} \\
  \cmidrule(l){2-7}
  \small{$\mu$} & \DIFdelbeginFL \DIFdelFL{1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$ }\DIFaddendFL & & & & \\
  \small{$\sigma$} & \DIFdelbeginFL \DIFdelFL{-0.4372 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$-0.4372$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$ }\DIFaddendFL & & & \\
  \DIFdelbeginFL %DIFDELCMD < \small{$\sfrac{\sigma}{\mu}$} %%%
\DIFdelendFL \DIFaddbeginFL \small{$\rfrac{\sigma}{\mu}$} \DIFaddendFL & \DIFdelbeginFL \DIFdelFL{-0.6914 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$-0.6914$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.9437 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.9437$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$ }\DIFaddendFL & & & \\
  \small{$n$} & \DIFdelbeginFL \DIFdelFL{0.7077 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7077$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{-0.1590 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$-0.1590$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{-0.3646 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$-0.3646$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$ }\DIFaddendFL & &  \\
  \small{MFE} & \DIFdelbeginFL \DIFdelFL{-0.5695 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$-0.5695$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7395 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7395$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7596 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7596$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{-0.3685 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$-0.3685$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$ }\DIFaddendFL & \\
  \small{$\log_{10}(\text{MFPT})$} & \DIFdelbeginFL \DIFdelFL{-0.0363 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$-0.0363$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.4844 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.4844$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.3762 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.3762$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.4059 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.4059$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.3990 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.3990$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$ }\DIFaddendFL \\
  % puts data.chomp.split(/\\/).map { |line| line.split("&")[1..-1].map { |item| item.gsub(/\s+/, "") }.reject { |item| item.empty? }.map { |number| "%.4f" % number.to_f }.join(" & ") + " \\\\" }.join(?\n)
  \bottomrule
\end{tabularx}
\DIFdelbeginFL %DIFDELCMD < \caption[Pearson correlation between various aspects of selenocysteine
%DIFDELCMD < insertion sequences from the seed alignment of Rfam family
%DIFDELCMD < RF00031]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Pearson correlation between various aspects of selenocysteine
insertion sequences from the seed alignment of Rfam family
RF$00031$]{\DIFaddendFL \small Pearson correlation between various aspects of selenocysteine
insertion sequences from the seed alignment of Rfam family
\DIFdelbeginFL \DIFdelFL{RF00031 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{RF$00031$ }\DIFaddendFL \citep{Gardner.nar11}.
For each of the \DIFdelbeginFL \DIFdelFL{61 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$61$ }\DIFaddendFL RNA sequences, we ran \fftbor, starting
from empty initial structure \strSt, and we ran a Monte Carlo
folding algorithm, developed by E. Freyhult and P. Clote (unpublished).
Using the Monte Carlo algorithm, we
determined the \mfpt (MFPT), defined as the average
taken over \DIFdelbeginFL \DIFdelFL{50 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$50$ }\DIFaddendFL runs, of the number of Monte Carlo steps taken to fold
the empty structure into the MFE structure, where an absolute upper bound
of \DIFdelbeginFL \DIFdelFL{5 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$5$ }\DIFaddendFL million steps was allowed in the simulation.  From the output of
\fftbor, we computed
{\em 1})\, the mean number ($\mu$) of base pairs per structure, taken over
the ensemble of all secondary structures for the given sequence;
{\em 2})\, the standard deviation ($\sigma$) of the number of base pairs per
structure;
{\em 3})\, the coefficient of variation $\frac{\sigma}{\mu}$;
{\em 4})\, the RNA sequence length $n$; and
{\em 5})\, the \mfe (MFE).
Additionally, we computed the logarithm base \DIFdelbeginFL \DIFdelFL{10 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$10$ }\DIFaddendFL of \mfpt
($\log_{10}(\text{MFPT})$), taken over \DIFdelbeginFL \DIFdelFL{50 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$50$ }\DIFaddendFL Monte Carlo runs per sequence
(log base \DIFdelbeginFL \DIFdelFL{10 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$10$ }\DIFaddendFL of the standard deviation of number of Monte Carlo
steps per run was approximately
9\% of $\log_{10}(\text{MFPT})$ on average). The table shows the correlation between each of these aspects.
Some correlations are obvious---for example,
{\em i})\;
the standard deviation $\sigma$ is highly correlated with the
coefficient of variation $\frac{\sigma}{\mu}$;
{\em ii})\;
the mean $\mu$ is negatively correlated with the
coefficient of variation $\frac{\sigma}{\mu}$;
{\em iii})\;
the mean $\mu$ is negatively correlated with the
\mfe (MFE) --- if most low energy structures in the ensemble
have many base pairs, then it is likely that the \mfe is very
low (i.e. since MFE is negative, the absolute value of MFE increases); and
{\em iv})\;
sequence length is negatively correlated with MFE --- as sequence length
increases, the \mfe (MFE) decreases.
However, it may appear surprising that
{\em v})\; the
mean $\mu$ number of base pairs per structure is independent of MFPT
(correlation $-0.0363$), although
{\em vi})\; MFE is correlated with MFPT
(correlation $0.3990$) --- i.e. from {\em (iii)},
lower MFE is correlated with a larger average $\mu$ number of base pairs per
structure, from ({\em vi})
higher MFE is correlated with longer folding time, but
from ({\em v}) the average $\mu$  number of base pairs per structure is
independent of folding time.
The most important insight from this table is that
{\em vii})\;
standard deviation $\sigma$ is correlated with \mfpt---the correlation is statistically significant, with one-tailed
$p$-value of $0.0002$.}
\label{table:correlationFFTborEmpty}
\end{table}

\begin{figure}[!ht]
\centering
\includegraphics[width=.3\textwidth]{Figures/FFTbor/AB030643_1_4176-4241.pdf}
\includegraphics[width=.3\textwidth]{Figures/FFTbor/AL645723_11_192421-192359.pdf}
\includegraphics[width=.3\textwidth]{Figures/FFTbor/FFTborOutputFromEmptyStrForTPPriboswitches.pdf}
\caption[Example of the graphical output of \fftbor, when the empty structure is chosen as initial structure \strSt]{This figure represents the
graphical output of \fftbor, when the empty structure is chosen as
initial structure \strSt.
The $x$-axis represents the number of base pairs per structure,
taken over the ensemble of all secondary structures for the given RNA
sequence; the $y$-axis represents Boltzmann probability
\DIFdelbeginFL \DIFdelFL{$p(k) = \sfrac{\bfZ{k}{}}{\fullZ}$}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$p(k) = \frac{\bfZ{k}{}}{\fullZ}$}\DIFaddendFL ,
where \fullZ is the partition function for all secondary structures
having exactly $k$ base pairs.
{\em (Left)}
For the \DIFaddbeginFL {\em \DIFaddFL{M. musculus}} \DIFaddendFL selenocysteine (SECIS) element \DIFdelbeginFL \DIFdelFL{AB030643.1/4176--4241 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AB$030643.1$ $4176$--$4241$ }\DIFaddendFL from Rfam family
\DIFdelbeginFL \DIFdelFL{RF00031}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{RF$00031$}\DIFaddendFL , the standard deviation $\sigma$ of the number of base pairs,
taken over the ensemble of all secondary structures, is
$0.7276$, while the logarithm base \DIFdelbeginFL \DIFdelFL{10 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$10$ }\DIFaddendFL of the \mfpt ($\log_{10}(\text{MFPT})$)
is $4.75$.
{\em (Center)}
For the \DIFaddbeginFL {\em \DIFaddFL{M. musculus}} \DIFaddendFL selenocysteine (SECIS) element
\DIFdelbeginFL \DIFdelFL{AL645723.11/192421--192359 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AL$645723.11$ $192421$--$192359$ }\DIFaddendFL from Rfam family
\DIFdelbeginFL \DIFdelFL{RF00031}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{RF$00031$}\DIFaddendFL , the standard deviation $\sigma$ of the number of base pairs,
taken over the ensemble of all secondary structures, is
$2.6794$, while $\log_{10}(\text{MFPT})$ is $5.69$.
Among the \DIFdelbeginFL \DIFdelFL{61 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$61$ }\DIFaddendFL sequences in the seed alignment of \DIFdelbeginFL \DIFdelFL{RF00031}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{RF$00031$}\DIFaddendFL ,
\DIFdelbeginFL \DIFdelFL{AB030643.1/4176--4241 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AB$030643.1$ $4176$--$4241$ }\DIFaddendFL was the fastest folder, while
\DIFdelbeginFL \DIFdelFL{AL645723.11/192421--192359 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AL$645723.11$ $192421$--$192359$ }\DIFaddendFL was the slowest folder.
{\em (Right)}
Superimposition of output of \fftbor for two TPP \rb aptamers: the
\DIFdelbeginFL \DIFdelFL{97 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$97$ }\DIFaddendFL nt sequence \DIFdelbeginFL \DIFdelFL{BX842649.1/277414--277318 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{BX$842649.1$ $277414$--$277318$ from }{\em \DIFaddFL{B. bacteriovorus}} \DIFaddendFL and the
\DIFdelbeginFL \DIFdelFL{99 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$99$ }\DIFaddendFL nt sequence \DIFdelbeginFL \DIFdelFL{AACY022101973.1/389--487}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AACY$022101973.1$ $389$--$487$ from the marine metagenome}\DIFaddendFL , both obtained when
taking the empty structure for the initial structure \strSt.
The mean $\mu$ for the \fftbor structural profile near the empty
structure is $23.0203$  [resp. $27.5821$], the
standard deviation $\sigma$ for the \fftbor structural profile
is $2.2253$  [resp. $1.9857$], and the \kinfold MFPT is
$311,075.06$ [resp. $61,575.69$] for the TPP \rb aptamer
\DIFdelbeginFL \DIFdelFL{AB030643.1/4176--4241 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{BX$842649.1$ $277414$--$277318$ }\DIFaddendFL [resp. \DIFdelbeginFL \DIFdelFL{AL645723.11/192421--192359}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AACY$022101973.1$ $389$--$487$}\DIFaddendFL ].
The right panel of this figure should be compared with Figure
\ref{fig:fftbor:tppDistributions}.
These anecdotal results bear up the correlation between standard deviation
$\sigma$ and $\log_{10}(\text{MFPT})$ described in \DIFaddbeginFL \DIFaddFL{Table }\DIFaddendFL \ref{table:correlationFFTborEmpty}.
}
\label{fig:fftbor:correlationFFTborEmpty}
\end{figure}

In the right panel of \Figref{fftbor:correlationFFTborEmpty}, we
applied \fftbor to each of the two randomly chosen TPP \rb
aptamers \DIFdelbegin \DIFdel{BX842649.1/277414--277318 and AACY022101973.1/389--487}\DIFdelend \DIFaddbegin \DIFadd{BX$842649.1$ $277414$--$277318$ from }{\em \DIFadd{B. bacteriovorus}}
\DIFadd{and AACY$022101973.1$ $389$--$487$ from the marine metagenome}\DIFaddend , starting
from the empty reference structure \DIFdelbegin \DIFdel{$\strSt=\emptyset$}\DIFdelend \DIFaddbegin \DIFadd{$\strSt=\varnothing$}\DIFaddend .
The mean for the \fftbor structural profile near the empty
structure is $\mu_1=23.0203$  [resp. $\mu_2=27.5821$], the
standard deviation $\sigma$ for the \fftbor structural profile
is $\sigma_1=2.2253$  [resp. $\sigma_2=1.9857$], and the \kinfold MFPT is
$311,075.06$ [resp. $61,575.69$] for the TPP \rb aptamer
\DIFdelbegin \DIFdel{AB030643.1/4176--4241 }\DIFdelend \DIFaddbegin \DIFadd{BX$842649.1$ $277414$--$277318$ }\DIFaddend [resp. \DIFdelbegin \DIFdel{AL645723.11/192421--192359}\DIFdelend \DIFaddbegin \DIFadd{AACY$022101973.1$ $389$--$487$}\DIFaddend ]. This
anecdotal evidence supports the hypothesis that small standard deviation
in \fftbor distribution is correlated with fast folding.

\DIFdelbegin \DIFdel{Additionally, we }\DIFdelend \DIFaddbegin \DIFadd{We }\DIFaddend randomized the TPP \rbs \DIFdelbegin \DIFdel{BX842649.1/277414--277318 and
AACY022101973.1/389--487 }\DIFdelend \DIFaddbegin \DIFadd{BX$842649.1$ $277414$--$277318$ and
AACY$022101973.1$ $389$--$487$ }\DIFaddend by using our implementation of the
Altschul-Erikson dinucleotide shuffle algorithm
\citep{altschulErikson:dinucleotideShuffle}, and then applied \fftbor to these
sequences, starting from the empty structure.  The mean $\mu_1$ and standard
deviation $\sigma_1$ for the \fftbor distribution for randomized \DIFdelbegin \DIFdel{BX842649 }\DIFdelend \DIFaddbegin \DIFadd{BX$842649.1$ }\DIFaddend are
respectively $\mu_1=19.93$ and $\sigma_1=2.88$, while those for randomized
\DIFdelbegin \DIFdel{AACY022101973 }\DIFdelend \DIFaddbegin \DIFadd{AACY$022101973.1$ }\DIFaddend are $\mu_2=24.39$ and $\sigma_2=24.00$. Running \kinfold, with a
maximum of \DIFdelbegin \DIFdel{500,000 steps with 30 }\DIFdelend \DIFaddbegin \DIFadd{$500,000$ steps with $30$ }\DIFaddend replicates (as explained in the text), we
found
that for randomized \DIFdelbegin \DIFdel{BX842649, all 30 }\DIFdelend \DIFaddbegin \DIFadd{BX$842649$, all $30$ }\DIFaddend runs converged yielding a \mfpt
(MFPT) of $13,022.58$ with standard deviation of $15,221.78$. In contrast for
randomized \DIFdelbegin \DIFdel{AACY022101973, only 15 out of 30 }\DIFdelend \DIFaddbegin \DIFadd{AACY$022101973.1$, only $15$ out of $30$ }\DIFaddend runs converged within \DIFdelbegin \DIFdel{500,000 }\DIFdelend \DIFaddbegin \DIFadd{$500,000$ }\DIFaddend steps,
and discounting these nonconvergent data, we obtain an average \mfpt
(MFPT) of $94,446.93$ with standard deviation of $157,107.43$. This additional
test provides more anecdotal evidence supporting our hypothesis that small
standard deviation $\sigma$ in \fftbor probability density is correlated with fast folding, as measured by MFPT.

This notion of correlation between the coarse-grained energy landscape and
kinetics is what motivates the work described in Chapters \ref{ch:ffttwo}
and \ref{ch:hermes}, where a more detailed explanation of kinetics is provided,
and additional evidence is provided to support this claim.

\section{Performance characteristics of \fftbor and \rnabor}
\label{sec:fftbor:speed}

As visible from the defining recursions, the algorithmic time complexity of
\rnabor is \On{5} and space complexity is \On{3}, where $n$ is
the length of input RNA sequence. In contrast, the time complexity of
\fftbor is \On{4} and space complexity is \On{2}.
While \fftbor saves an order of magnitude in performance and memory,
\rnabor has the benefit of producing suboptimal structures $\text{MFE}_k$
for all $0 \leq k \leq n$, whose free energy is minimal across all structures
having \bpd $k$ from input structures \strSt.
\Figref{fftbor:benchmarking} displays run time curves for both
\rnabor and \fftbor, when the initial structure \strSt is
taken to be either the empty structure or the \mfe
(MFE) structure.

Here, we compare the run time of \rnabor \citep{freyhult.b07} and
the (unparallelized version of) \fftbor, using
a Dell Power Edge \DIFdelbegin \DIFdel{1950, 2 }\DIFdelend \DIFaddbegin \DIFadd{$1950$, $2$ }\DIFaddend x Intel Xeon \DIFdelbegin \DIFdel{E5430 }\DIFdelend \DIFaddbegin \DIFadd{E$5430$ }\DIFaddend Quad
core with \DIFdelbegin \DIFdel{2.80 GHz and 16 }\DIFdelend \DIFaddbegin \DIFadd{$2.80$ GHz and $16$ }\DIFaddend GB RAM. For $n = 20,40,60,\dots,300$, in step
size of \DIFdelbegin \DIFdel{20 }\DIFdelend \DIFaddbegin \DIFadd{$20$ }\DIFaddend nt, we generated $n$ random RNA sequences of length $n$ with equal
probability for each nucleotide A,C,G,U (i.e. a \DIFdelbegin \DIFdel{$0$th }\DIFdelend \DIFaddbegin \DIFadd{0th }\DIFaddend order Markov chain).
For values of $n \leq 200$, \DIFdelbegin \DIFdel{100 }\DIFdelend \DIFaddbegin \DIFadd{$100$ }\DIFaddend random sequences of length
$n$ were generated, while for values of $220 \leq n \leq 300$, only
\DIFdelbegin \DIFdel{10 }\DIFdelend \DIFaddbegin \DIFadd{$10$ }\DIFaddend sequences of length $n$ were generated.
RNA sequences larger than \DIFdelbegin \DIFdel{300 }\DIFdelend \DIFaddbegin \DIFadd{$300$ }\DIFaddend nt were not tested,
due to \On{3} memory constraints required by \rnabor.
For each RNA sequence, \rnabor and \fftbor were both run,
each starting with empty initial structure \strSt, and also
with initial sequence \strSt taken to be the MFE structure.
Each data point in the table comprises the average run time for three
independent evaluations.

\begin{figure}[!ht]
\centering
\includegraphics[width=.9\textwidth]{Figures/FFTbor/rnaborfftborRunTimeEvan.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption[Run times in seconds for \rnabor and \fftbor, on random RNA
%DIFDELCMD < of length $20,40,60,\dots,300$ in step size of 20 nt]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Run times in seconds for \rnabor and \fftbor, on random RNA
of length $20$--$300$ in step size of $20$ nt]{\DIFaddendFL Run times in seconds for \rnabor and \fftbor, on random RNA
of length $20,40,60,\dots,300$ in step size of \DIFdelbeginFL \DIFdelFL{20 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$20$ }\DIFaddendFL nt. Each algorithm
was run with the empty initial structure \strSt, see rows
\rnabor (empty), \fftbor (empty), and with the \mfes as the initial structure
\strSt, see rows
\rnabor (MFE) and \fftbor (MFE). Note that for both \rnabor
and \fftbor, the run time increases when \strSt is the MFE structure,
rather than the empty structure. Notice the radical improvement in the
run time of \fftbor over that of \rnabor.
}
\label{fig:fftbor:benchmarking}
\end{figure}

\subsection{OpenMP parallelization of \fftbor}
\label{subsec:fftbor:openmp}

OpenMP is a simple and flexible
multi-platform shared-memory parallel programming environment, that supports
parallelizations of C/C++ code---see \url{http://openmp.org/}.
Using OpenMP primitives, we created multiple threads to evaluate the polynomial
\fullZx on different \nRoUs. \Figref{fftbor:benchmarkingParallel}
presents benchmarks, executed on
a \DIFdelbegin \DIFdel{24-core AMD Opteron 6172 with 2.10GHz and 64GB }\DIFdelend \DIFaddbegin \DIFadd{$24$-core AMD Opteron $6172$ with $2.10$GHz and $64$GB }\DIFaddend RAM, for the speedup
of \fftbor as a function of the number of cores.
The data in Table \ref{table:fftborBenchmarkingParallel} describes average
run time in seconds ($\pm$ one standard deviation) for running \fftbor
on random RNA of length $200,250,300,400,450,500$ with either \DIFdelbegin \DIFdel{1 or 2 }\DIFdelend \DIFaddbegin \DIFadd{$1$ or $2$ }\DIFaddend cores.
\Figref{fftbor:benchmarkingParallel}
presents similar data for running
\fftbor on $2,3,6,4,12,15,20$ cores.

\begin{table}[!ht]
\centering
\begin{tabularx}{\linewidth}{c *{2}{R}}
\toprule
\small{$n$} & \small{Single core} & \small{Two cores} \\
\cmidrule(lr){1-3}
\DIFdelbeginFL \DIFdelFL{200 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$200$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{123.2 $\pm$ 16.2 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$123.2 \pm 16.2$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{61.8 $\pm$ 8.0 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$61.8 \pm 8.0$ }\DIFaddendFL \\
\DIFdelbeginFL \DIFdelFL{250 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$250$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{331.1 $\pm$ 27.2 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$331.1 \pm 27.2$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{166.1 $\pm$ 13.7 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$166.1 \pm 13.7$ }\DIFaddendFL \\
\DIFdelbeginFL \DIFdelFL{300 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$300$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{723.4 $\pm$ 59.9 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$723.4 \pm 59.9$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{365.2 $\pm$ 30.1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$365.2 \pm 30.1$ }\DIFaddendFL \\
\DIFdelbeginFL \DIFdelFL{350 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$350$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1,380.8 $\pm$ 95.2 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1,380.8 \pm 95.2$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{698.4 $\pm$ 46.9 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$698.4 \pm 46.9$ }\DIFaddendFL \\
\DIFdelbeginFL \DIFdelFL{400 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$400$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{2,239.1 $\pm$ 210.9 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$2,239.1 \pm 210.9$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1,129.5 $\pm$ 104.3 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1,129.5 \pm 104.3$ }\DIFaddendFL \\
\DIFdelbeginFL \DIFdelFL{450 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$450$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{3,635.0 $\pm$ 857.4 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$3,635.0 \pm 857.4$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1,980.9 $\pm$ 126.5 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1,980.9 \pm 126.5$ }\DIFaddendFL \\
\DIFdelbeginFL \DIFdelFL{500 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$500$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{5,076.7 $\pm$ 1,292.1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$5,076.7 \pm 1,292.1$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{3,389.8 $\pm$ 788.4 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$3,389.8 \pm 788.4$ }\DIFaddendFL \\
\bottomrule
\end{tabularx}
\caption[Table showing parallel run times in seconds
for \fftbor, using OpenMP]{Table showing parallel run times in seconds
for \fftbor, using OpenMP---\url{http://openmp.org/}.
For each sequence length $200,\dots,500$,
five random RNAs were generated using equal probability for each nucleotide
A,C,G,U. Run time in seconds, plus or minus one standard deviation, are
given for a \DIFdelbeginFL \DIFdelFL{24-core
}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$24$-core
}\DIFaddendFL AMD Opteron \DIFdelbeginFL \DIFdelFL{6172 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$6172$ }\DIFaddendFL running at \DIFdelbeginFL \DIFdelFL{2.10}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$2.10$}\DIFaddendFL GHz with \DIFdelbeginFL \DIFdelFL{64GB }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$64$GB }\DIFaddendFL RAM, with only \DIFdelbeginFL \DIFdelFL{1 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$ }\DIFaddendFL [resp. \DIFdelbeginFL \DIFdelFL{2}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$2$}\DIFaddendFL ] cores
used.}
\label{table:fftborBenchmarkingParallel}
\end{table}

\begin{figure}[!ht]
\centering
\includegraphics[width=.75\textwidth]{Figures/FFTbor/fftborParallelRunTimes.pdf}
\caption[]{Graph showing parallel run time of \fftbor as a function of
sequence length, running on an AMD Opteron \DIFdelbeginFL \DIFdelFL{6172 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$6172$ }\DIFaddendFL running at \DIFdelbeginFL \DIFdelFL{2.10}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$2.10$}\DIFaddendFL GHz with \DIFdelbeginFL \DIFdelFL{64GB }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$64$GB }\DIFaddendFL RAM,
using respectively \DIFdelbeginFL \DIFdelFL{1,2,3,4,6,9,12,15,20 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1,2,3,4,6,9,12,15,20$ }\DIFaddendFL cores.
}
\label{fig:fftbor:benchmarkingParallel}
\end{figure}

%DIF <  \input{Chapters/RNAparametric}
%!TEX root = ../main.tex

\chapter{FFTbor2D}
\label{ch:ffttwo}

\DIFdelbegin %DIFDELCMD < \lhead{FFTbor2D: 2D Coarse-Grained Energy Landscapes}
%DIFDELCMD < %%%
\DIFdelend \DIFaddbegin \lhead{FFTbor\twoD: \twoD Coarse-Grained Energy Landscapes}
\DIFaddend 

\section{Introduction}
\label{sec:ffttwo:intro}

In this chapter, we present the \ffttwo algorithm and accompanying software.
\ffttwo, like \fftbor described in \Chref{fftbor}, is an algorithm
which computes the paramerized partition function for an input RNA sequence
\seq. \ffttwo computes the two-dimensional coarse-grained energy landscape for \seq
given two compatible input secondary structures \strA and \strB, where position
$(x,y)$ on the discrete energy landscape corresponds to the Boltzmann
probability for those structures \str which have $\dBP{\str}{\strA}=x$ and
$\dBP{\str}{\strB}=y$ (where $d_{\text{BP}}$ is as defined in
equation \ref{eq:fftbor:dBP}). By again leveraging the \fft, \ffttwo runs in \On{5}
time and only uses \On{2} space---a significant improvement over previous
approaches. This permits the output energy landscape to be used in a
high-throughput fashion to analyze folding kinetics; a topic covered in detail
in \Chref{hermes}.

\subsection{Organization}
\label{subsec:ffttwo:org}

This chapter is organized in the following fashion. Because the history for
this work arises naturally from the history provided in
\Secref{sec:fftbor:bkgrnd}, we provide only a brief background
and immediately fall into
a technical discussion of the underlying algorithm. We first develop the
recursions for the Nussinov energy model for expository clarity, the
underlying implementation uses the more complicated and robust Turner energy
model. Recursions in place, we then move to show how these lead to
a single variable polynomial $P(x)$ whose coeffecients can be computed by
the \idft, and map to the \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend energy landscape. We describe two improvements over
the straight-forward use of the \fft to compute $P(x)$,
a parity condition and complex conjugates, which together reduce the
runtime by a factor of \DIFdelbegin \DIFdel{4. }\DIFdelend \DIFaddbegin \DIFadd{$4$. }\DIFaddend Finally, we contrast this software against \rnatwofold,
and outline the performance characteristics of both softwares and highlight
the benefits and drawbacks of both. We elect to refrain from describing
applications of \ffttwo until \Chref{hermes}, where the software is
applied to quickly approximate \mfpt and \eqt for the folding of RNA molecules
between any two distinct, user provided structures \strAB.

\section{Background}
\label{sec:ffttwo:bkgrnd}

RNA folding pathways play an important role in biological processes.
For instance, in the {\em hok/sok}
(host-killing/suppression of killing) system \citep{gerdes.arg97},
the transition between two metastable RNA structures determines the
fate of a cell as follows.
The {\em hok} gene of {\em E. coli} and other bacteria
codes a small (\DIFdelbegin \DIFdel{52 }\DIFdelend \DIFaddbegin \DIFadd{$52$ }\DIFaddend amino acid) toxin causing irreversible damage to the cell
membrane. It has been shown that {\em hok}-mRNA is
constitutively expressed from a weak promoter, while
the rapidly degraded {\em sok}-RNA is constitutively expressed from
a strong promoter.  The {\em hok}-mRNA is initially
inactive, since a foldback sequesters the
Shine-Dalgarno sequence; however, slow exonucleolytic processing
digests the last $\approx 40$ nt of the $3'$ end of {\em hok}-mRNA,
transforming the molecule into its active form in which
the Shine-Dalgarno sequence is no longer sequestered.
If R1 plasmids of {\em E. coli} are present in
sufficient copy number, then a portion of the \DIFdelbegin \DIFdel{64 }\DIFdelend \DIFaddbegin \DIFadd{$64$ }\DIFaddend nt
{\em sok}-RNA, which is complementary to {\em hok}-mRNA leader
region, binds to the active conformation of {\em hok}-mRNA, thus
causing degradation of the complex by RNase III \citep{gerdes.arg97}.
If plasmids are not present in sufficient copy number, then the
cell is killed by {\em hok} toxin, thus ensuring fitness of the daughter
cells.

In the case of spliced leader (SL) RNA from certain trypanosomes and nematodes,
a portion of the $5'$ exon is donated to
another mRNA by \DIFdelbegin \DIFdel{trans splicing}\DIFdelend \DIFaddbegin \DIFadd{trans-splicing}\DIFaddend .
Intermediate structures appear to be important in the process of splicing,
as shown by LeCuyer and Crothers \citep{lecuyercrothers}, who performed
stopped-flow rapid-mixing and temperature-jump measurements
of the kinetics for the structural transition between two low
energy structures of SL RNA from {\em Leptomonas collosoma}.
Conformational switches are thought not only to play a role in such
\DIFdelbegin \DIFdel{trans splicing}\DIFdelend \DIFaddbegin \DIFadd{trans-splicing}\DIFaddend , but as well in transcriptional and
translational regulation, protein synthesis, and mRNA splicing.

For these reasons, substantial experimental and computational work
has been done on folding pathways. In \citep{lorenz}, \rnatwofold---a dynamic
programming method that operates in
\On{7} time and \On{4} space---is presented, a generalization of \rnabor
\citep{freyhult.b07} that computes a \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend projection of the energy landscape,
discretized by \bpd to two input structures. Our work presented
here, called \ffttwo, produces a similar energy landscape in \On{5} time and
\On{2} space---efficient enough for the analysis of folding pathways of large
RNA sequences not tractable using \rnatwofold.

\section{Derivation of the \ffttwo algorithm}
\label{sec:ffttwo:math}

For expository clarity, we describe \ffttwo \citep{senter.jmb14}
and all recursions
in terms of the Nussinov energy model \citep{nussinovjacobson}
(same as in \Chref{fftbor}), where
the energy $E_0(i,j)$ of a base pair $(i,j)$ is defined to be $-1$, and the
energy $E(\str)$ of a secondary structure \str is $-1$ times the number $|\str|$
of base pairs in structure \str.  Nevertheless, the implementation of
\ffttwo involves the full Turner energy model \citep{xia:RNA}, where
free energy $E(\str)$ depends on negative, stabilizing energy contributions
from base stacking, and positive, destabilizing energy contributions due to
loss of entropy in loops.

\subsection{Definition of the partition function
\texorpdfstring{\bfZ{x,y}{1,n}}{}}
\label{subsec:ffttwo:recursions}

Given reference secondary structures \strAB of a
given RNA sequence $\seq=s_1,\dots,s_n$, our goal is to compute

\begin{align}
\label{eq:ffttwo:sumBoltzFactors}
\bfZ{x,y}{1,n}\;=\;
\sum_{\mathclap{\substack{
\str \text{ such that } \rule[-.5ex]{0pt}{0pt} \\
\dBP{\str}{\strA} = x,\,\dBP{\str}{\strB} = y}}}\enspace
\boltzF{\str}
\end{align}

for all $0 \leq x,y < n$, where $R$ is the universal gas constant, $T$
is absolute temperature, $E(\str)$ denotes the free energy of \str,
and \str ranges
over all secondary structures that are compatible with \seq. As mentioned,
we emphasize that for expository reasons alone, the Nussinov energy model is
used in the recursions in this chapter, although full recursions and
the implementation of \ffttwo, like \fftbor, involve the Turner energy model.

For any secondary structure \str of \seq, and any values
$1 \leq i \leq j \leq n$, the restriction $\str_{[i,j]}$ is defined to be the
collection of base pairs of \str, lying within interval $[i,j]$; i.e.
$\str_{[i,j]} = \{ (k, \ell) : i \leq k < \ell \leq j \}$.
In \citep{hofacker:RNAbor2D}, Lorenz et al. generalized
the dynamic programming recursions of our earlier work \citep{freyhult.b07},
to yield recursions
for the partition function $\bfZ{x,y}{i,j}$ in
\eqnref{ffttwo:sumBoltzFactors}. In the context of the Nussinov model,
$\bfZ{x,y}{i,j}$ is equal to

\begin{align}
\label{eq:ffttwo:bfZabij}
\begin{split}
& \bfZ{x-\alpha_0,y-\beta_0}{i,j-1}\enspace + \\
& \sum_{\substack{(s_k,s_j) \in \bpSet, \\ i \le k<j}}
\left(
\boltzNuss{k,j}\quad
\sum_{\mathclap{u+u'=x-\alpha(k)}}\hspace{3.75em}
\sum_{\mathclap{v+v'=y-\beta(k)}}\quad
\bfZ{u,v}{i,k-1} \cdot \bfZ{u',v'}{k+1,j-1}
\right)
\end{split}
\end{align}

where $\alpha_0 = 1$ if $j$ is base paired in $\strA_{[i,j]}$ and \DIFdelbegin \DIFdel{0 }\DIFdelend \DIFaddbegin \DIFadd{$0$ }\DIFaddend otherwise,
$\beta_0 = 1$ if $j$ is base paired in $\strB_{[i,j]}$ and \DIFdelbegin \DIFdel{0 }\DIFdelend \DIFaddbegin \DIFadd{$0$ }\DIFaddend otherwise,
$E_0(k,j)=-1$ if $k,j$ can base-pair
(see equation \ref{eq:fftbor:validBP}), and otherwise $E_0(k,j)=0$, and
$\alpha(k) =
\dBP{\strA_{[i,j]}}{\strA_{[i,k-1]} \cup \strA_{[k+1,j-1]} \cup \{ (k,j) \}}$,
and
$\beta(k) =
\dBP{\strB_{[i,j]}}{\strB_{[i,k-1]} \cup \strB_{[k+1,j-1]} \cup \{ (k,j) \}}$.

\subsection{Recursions to compute the polynomial
\texorpdfstring{\emZ{i,j}}{}}
\label{subsec:ffttwo:polynomial}

Given RNA sequence $\seq = s_1,\dots,s_n$
and two arbitrary, but fixed reference
structures \strAB, we define the {\em polynomial}

\begin{align}
\label{eq:ffttwo:zOfX}
\fullZx = \sum_{r=0}^{n-1} \sum_{s=0}^{n-1}\; z_{rn+s} x^{rn+s}
\end{align}

where (constant) coefficients

\begin{align}
z_{rn+s} = \bfZ{r,s}{1,n}\;=
\enspace\sum_{\mathclap{\substack{
\str \text{ such that } \rule[-.5ex]{0pt}{0pt} \\
\dBP{\str}{\strA} = r,\,\dBP{\str}{\strB} = s
}}}\enspace
\boltzF{\str}
\end{align}

where $E(\str)$ denotes the free energy of \str.
If we evaluate the polynomial \fullZx at $n^2$ distinct pairs of values
$a_0,\dots,a_{n^2-1}$ in

\begin{align}
\label{eq:ffttwo:solutionsForAlpha}
\emZof{}{a_0} = y_{0}, \dots, \emZof{}{a_{n^2-1}} = y_{n^2-1},
\end{align}

then Lagrange polynomial interpolation
(\DIFdelbegin %DIFDELCMD < \eqnref{fftbor:lagrangeInterpolation}%%%
\DIFdelend \DIFaddbegin \DIFadd{equation \ref{eq:fftbor:lagrangeInterpolation}}\DIFaddend )
guarantees that we can determine the coefficients $z_{rn+s}$ of \fullZx,
for $0 \leq r,s < n$. Due to technical difficulties concerning numerical
robustness observered while working on the \fftbor software
(\Chref{fftbor}), we will perform polynomial interpolation
by using Vandermonde matrices and the \fft (FFT).

The following theorem shows that a
recursion, analogous to \eqnref{ffttwo:bfZabij},
can be used to compute
the {\em polynomial} $\emZ{i,j}$ defined by

\begin{align}
\label{eq:ffttwo:emZij}
\emZ{i,j} &= \sum_{r=0}^{n-1} \sum_{s=0}^{n-1}\;
z_{rn+s}(i,j) \cdot x^{rn+s}\; =
\sum_{k=0}^{n^2-1} z_{k}(i,j) \cdot x^{k}
\end{align}

where

\begin{align}
z_{rn+s}(i,j) \;=\; \bfZ{r,s}{i,j}\;=
\enspace\sum_{\mathclap{\substack{
\str \text{ such that } \rule[-.5ex]{0pt}{0pt} \\
\dBP{\str}{\strA} = r,\,\dBP{\str}{\strB} = s
}}}\enspace
\boltzF{\str}.
\end{align}

Here, in the summation, \str runs over structures on $s_i,\dots,s_j$, which
are $r$-neighbors of the restriction $\strA_{[i,j]}$ of reference structure
\strA to interval $[i,j]$, and simultaneously
\DIFdelbegin %DIFDELCMD < \str%%%
\DIFdelend \DIFaddbegin \DIFadd{$s$}\DIFaddend -neighbors of the restriction $\strB_{[i,j]}$ of reference structure
\strB to interval $[i,j]$.

\begin{theorem}
\label{thm:ffttwo:recursions}
Let $s_1,\dots,s_n$ be a given RNA sequence.
For any integers $1 \leq i < j \leq n$, let

\begin{align}
\emZ{i,j} = \sum_{r=0}^{n-1} \sum_{s=0}^{n-1}\; z_{rn+s} x^{rn+s}
\end{align}

where

\begin{align}
z_{rn+s}(i,j) = \bfZ{r,s}{i,j}.
\end{align}

Inductively we define $\emZ{i,j}$ to equal

\begin{align}
\label{eq:ffttwo:polyRecNussJac}
\begin{split}
&\emZof{i,j-1}{x} \cdot x^{\alpha_0n+\beta_0} + \\
&\sum_{\substack{(s_k,s_j) \in \bpSet,\\i\le k<j}}
\left(\boltzNuss{k,j}\cdot
\emZof{i,k-1}{x} \cdot \emZof{k+1,j-1}{x}\cdot x^{\alpha(k)n+\beta(k)} \right)
\end{split}
\end{align}

where $\alpha_0 = 1$ if $j$ is base-paired in $\strA_{[i,j]}$ and \DIFdelbegin \DIFdel{0 }\DIFdelend \DIFaddbegin \DIFadd{$0$ }\DIFaddend otherwise,
$\beta_0 = 1$ if $j$ is base-paired in $\strB_{[i,j]}$ and \DIFdelbegin \DIFdel{0 }\DIFdelend \DIFaddbegin \DIFadd{$0$ }\DIFaddend otherwise, and
$\alpha(k) =
\dBP{\strA_{[i,j]}}{\strA_{[i,k-1]} \cup \strA_{[k+1,j-1]} \cup \{ (k,j) \}}$,
$\beta(k) =
\dBP{\strB_{[i,j]}}{\strB_{[i,k-1]} \cup \strB_{[k+1,j-1]} \cup \{ (k,j) \}}$.

The proof is given in \Secref{sec:ffttwo:recursions}.
\end{theorem}

Note that if one were to compute all terms of the polynomial $\emZ{1,n}$
by explicitly performing polynomial multiplications,
then the computation would require \On{7} time and \On{4} space, the
same time complexity of \rnatwofold \citep{hofacker:RNAbor2D}.
Instead of explicitly performing polynomial expansion in {\em variable} $x$,
we instantiate $x$ to a
complex number $\rho \in \mathbb{C}$, and apply
the following recursion, by setting $\emZof{i,j}{\rho}$ equal to

\begin{align}
\label{eq:ffttwo:polyRecNussJacInit}
\begin{split}
&\emZof{i,j-1}{\rho} \cdot \rho^{\alpha_0n+\beta_0} + \\
&\sum_{\substack{(s_k,s_j) \in \bpSet,\\i\le k<j}}
\left(\boltzNuss{k,j}\cdot
\emZof{i,k-1}{\rho} \cdot \emZof{k+1,j-1}{\rho}
\cdot \rho^{\alpha(k)n+\beta(k)} \right)
\end{split}
\end{align}

Note that this approach is similar to what we do in \fftbor---specifically
\eqnref{fftbor:fftborNussPolyAlpha}---however notationally we will use the
variable $\rho$ instead of $\alpha$, to avoid confusion. In this fashion, we
can compute $\emZof{}{\rho}=\emZof{1,n}{\rho}$ in
\On{3} time and \On{2} space. For $n^2$ distinct complex numbers
$\rho_i$ where $0 \leq i \leq n^2-1$, we can compute and save only the
values $\emZof{}{\rho_0},\dots,\emZof{}{\rho_{n^2-1}}$, each time re-using the
\On{2} space for the next computation of $\emZof{}{\rho_i}$.
It follows that
the computation resources used to determine the (column) vector
\DIFaddbegin 

\DIFaddend \begin{align}
\label{eq:ffttwo:yColumn}
\bfY = (y_0,\dots,y_{n^2-1})^{\text T} =
\left(
\begin{array}{l}
y_0 \\
y_1 \\
\vdots \\
y_{n^2-1} \\
\end{array}
\right)
\end{align}

where $y_0=\emZof{}{\rho_0},\dots,y_{n^2-1}=\emZof{}{\rho_{n^2-1}})$ are thus
quintic time \On{5} and quadratic space \On{2}.

\subsection{Polynomial interpolation}
Our plan is to determine the coefficients of the polynomial
\fullZx in \eqnref{ffttwo:zOfX} by
polynomial interpolation.  For reasons of numerical stability,
we instead determine the coefficients of the polynomial $p(x)$,
defined by

\begin{align}
\label{eq:ffttwo:pOfX}
p(x) =
\sum_{r=0}^{n-1} \sum_{s=0}^{n-1} p_{rn+s}
 x^{rn+s} =
\sum_{r=0}^{n-1} \sum_{s=0}^{n-1} \frac{\bfZ{rn+s}{1,n}}{\fullZ}
 x^{rn+s},
\end{align}

where the \fft (FFT) is used to implement the
interpolation of
the coefficients using the \idft (DFT), as
described in \Secref{subsec:ffttwo:fft}.  The following pseudocode
describes how
to compute the $m$ most significant digits
for probabilities
$p_{rn+s} = \frac{\bfZ{r,s}{1,n}}{\fullZ}$. It is well-known that
the FFT requires $O(N \log N)$ time to solve the \idft for a polynomial
of degree $N$. In our case,
$N=n^2$, and so the computation involving the FFT requires time $O(n^2 \log n)$.

The pseudocode for the algorithm to compute $p(x)$ is given in
\Figref{ffttwo:algo}.
In the next section, we explain a highly non-trivial improvement of
this algorithm to reduce time by a factor of \DIFdelbegin \DIFdel{4.
}\DIFdelend \DIFaddbegin \DIFadd{$4$.
}\DIFaddend \medskip

\begin{figure}[!ht]
\hrule \rule[0ex]{0pt}{0pt}
\begin{center}
{\large Pseudocode for \ffttwo} \\
\end{center}
\begin{tabular*}{\textwidth}{ll}
{\sc Purpose:} & Computes the $m$ most significant digits
of probabilities \DIFdelbeginFL \DIFdelFL{$p_{rn+s}=\sfrac{\bfZ{r,s}{1,n}}{\fullZ}$
}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$p_{rn+s}=\rfrac{\bfZ{r,s}{1,n}}{\fullZ}$
}\DIFaddendFL \rule[-1.5ex]{0pt}{0pt} \\
{\sc Input:} & RNA sequence $\seq = \seqN$, secondary
structures \strA, \strB of \seq, integer $m$ \rule[-1.5ex]{0pt}{0pt} \\
{\sc Output:} & \DIFdelbeginFL \DIFdelFL{$p_{rn+s}=\sfrac{\bfZ{r,s}{1,n}}{\fullZ}$ }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$p_{rn+s}=\rfrac{\bfZ{r,s}{1,n}}{\fullZ}$ }\DIFaddendFL to
$\lfloor \log_{10}(2^m) \rfloor$ significant digits for $r,s=0,\dots,n-1$
\rule[-1.75em]{0pt}{0pt} \\
\hline \rule[0ex]{0pt}{0pt}
\end{tabular*}
\begin{algorithmic}[1]
\Function{FFTbor2D}{\seq, \strA, \strB, $m$}
\State $n \gets \textit{length}(\seq)$
\For{$k \gets 0, n^2-1$}
\Comment{Compute all complex $n^2$-roots of unity}
\State $\omega_k \gets \exp(\frac{2 \pi i k}{n^2})$
\EndFor
\For{$k \gets 0, n^2-1$}
\Comment{Note that $\emZof{}{\omega_0} = \fullZ$}
\State $y_k \gets 2^m \cdot \frac{\emZof{}{\omega_k}}{\emZof{}{\omega_0}}$
\EndFor
\For{$k \gets 0, n^2-1$}
\Comment{Compute IDFT from \eqnref{ffttwo:aFromY}}
\State $a_k \gets \frac{1}{n} \sum_{j=0}^{n^2-1} a_j\, \omega^{-kj}$
\State $\pk \gets 2^{-m} \cdot \lfloor a_k \rfloor$
\Comment{Truncate to $m$ significant digits}
\EndFor
\State \textbf{return} $p_0,\dots,p_{n^2-1}$
\Comment{Return all \pk for $0 \leq k < n^2$}
\EndFunction
\rule[-0.35ex]{0pt}{0pt}
\end{algorithmic}
\DIFdelbeginFL %DIFDELCMD < \caption{
%DIFDELCMD < %%%
\DIFdelendFL \DIFaddbeginFL \caption[Pseudocode for \ffttwo]{\DIFaddendFL Pseudocode to compute the $m$
 most significant digits for probabilities
\DIFdelbeginFL \DIFdelFL{$p_{rn+s} = \sfrac{\bfZ{r,s}{1,n}}{\fullZ}$}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$p_{rn+s} = \frac{\bfZ{r,s}{1,n}}{\fullZ}$}\DIFaddendFL . In our implementation,
due to numerical stability issues in the FFT engine, precision parameter
$m$ has an upper bound of \DIFdelbeginFL \DIFdelFL{27---only }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$27$---only }\DIFaddendFL the $\lfloor \log_{10}(2^m) \rfloor =8$
most significant digits
are computed with \ffttwo.
It is well-known that
the FFT requires $O(N \log N)$ time to solve the inverse discrete
Fourier transform for a polynomial of degree $N$. In our case,
$N=n^2$, and so the FFT requires time $O(n^2 \log n)$.
}
\label{fig:ffttwo:algo}
\rule[0ex]{0pt}{1.5em} \hrule
\end{figure}

\section{Acceleration of the \ffttwo algorithm}
\label{sec:ffttwo:perf}

Recall that if \aPbi is a complex number, where $a,b$ are real values and $i$
denotes $\sqrt{-1}$, then the complex conjugate of \aPbi, denoted by
$\overline{\aPbi}$ is defined to be \aMbi. Recall that a \nRoU is a number
whose $n$th power equals one. Moreover,
\pRoU is the {\em principal} \nRoU; i.e.
$\{ \kRoU : k=0,\dots,n-1 \}$ is a set of pairwise distinct
\nRoUs. We have the following.

\begin{lemma}
\label{lem:ffttwo:lemma1}
Let \strAB denote two distinct, arbitrary but fixed,
secondary structures of RNA sequence \seq, let \str range over
all secondary structures of \seq, and let $d_0$ denote \dBP{\strA}{\strB}.
If $x=\dBP{\strA}{\str}$ and $y=\dBP{\str}{\strB}$, then
$y \in \{ d_0-x+2k: k=0,\dots,x\}$.

It follows that if $x=\dBP{\strA}{\str}$ and $y=\dBP{\str}{\strB}$,
then the only possible values for
$(x,y)$ are $(0,d_0), (1,d_0-1), (1,d_0+1),
(2,d_0-2), (2,d_0), (2, d_0+2),
(3,d_0-3), (3,d_0-1), (3, d_0+1), (3,d_0+3), \dots$.
As a corollary, we have the \DIFdelbegin %DIFDELCMD < {\em %%%
\DIFdel{parity condition}%DIFDELCMD < }%%%
\DIFdelend \DIFaddbegin \DIFadd{parity condition}\DIFaddend , that

\begin{align}
\label{eq:ffttwo:parity}
\dBP{\strA}{\str} + \dBP{\str}{\strB} \equiv \dBP{\strA}{\strB} \bmod 2
\end{align}

first noticed in \citep{hofacker:RNAbor2D}, as well as
the triangle inequality
$\dBP{\strA}{\str} + \dBP{\str}{\strB} \geq \dBP{\strA}{\strB}$
for \bpd, probably folklore.
Lorenz et al. \citep{hofacker:RNAbor2D} exploited the parity condition
and the triangle inequality by using sparse matrix methods to improve on
the efficiency of the na\"{\i}ve implementation of the
\On{7} time and \On{4} space algorithm to compute
the partition function, $\bfZ{r,s}{1,n}$,
and \mfes, $\text{MFE\,}^{r,s}_{1,n}$, over all
structures having \bpd $r$ to \strA and \str to \strB.
\end{lemma}

\begin{lemma}
\label{lem:ffttwo:lemma2}
If \fullZx is the complex polynomial defined in
\eqnref{ffttwo:zOfX}, then for any \nRoU
$\alpha$, it is the case that $\emZof{}{\overline{\alpha}} =
\overline{\emZof{}{\alpha}}$.
\end{lemma}

\begin{lemma}
\label{lem:ffttwo:lemma3}
Let \fullZx be defined by \eqnref{ffttwo:zOfX}, and
let $\alpha \in \mathbb{C}$ be any complex number.
If the \bpd between reference
structures \strAB is even, then $\emZof{}{-\alpha} = \emZof{}{\alpha}$,
while if
the distance is odd, then $\emZof{}{-\alpha} = -\emZof{}{\alpha}$.
\end{lemma}

\begin{lemma}
\label{lem:ffttwo:lemma4}
Suppose that $M$ is evenly divisible by 4,
$\nu = \exp(\frac{2 \pi i}{M})$ is the principal $M$-root of unity, and
$\frac{M}{4} < k \leq \frac{M}{2}$. Then

\begin{align}
\nu^k = -(\nu^{-(M_0-k)}) = -\overline{\nu^{M_0-k}}.
\end{align}
\end{lemma}

Lemma \ref{lem:ffttwo:lemma1} is proved by induction;
Lemma \ref{lem:ffttwo:lemma2}
is proved in \Secref{lem:fftbor:compconj} by a computation involving
binomial coefficients;
Lemma \ref{lem:ffttwo:lemma3} is proved in \Secref{sec:ffttwo:lemma3proof}
by the parity observation above, resulting from Lemma \ref{lem:ffttwo:lemma1};
Lemma \ref{lem:ffttwo:lemma4} is proved in \Secref{sec:ffttwo:lemma4proof},
relying on Euler's formula and trigonometric addition formulas.

Lemma \ref{lem:ffttwo:lemma1}
entails that either all even coefficients, or all odd coefficients
of \fullZx are zero, and so by a variable change described in detail below,
we require only half the number of evaluations of \fullZx, in order to perform
polynomial interpolation.
Lemma \ref{lem:ffttwo:lemma2}
entails that we require only half again the number of evaluations of
\fullZx, since the remainder can be inferred by taking the complex conjugate.
Lemma \ref{lem:ffttwo:lemma1} and Lemma \ref{lem:ffttwo:lemma2},
along with a
precomputation of powers of the complex roots of unity, lead to a
large performance speed-up in our implementation of \ffttwo---by a factor
of \DIFdelbegin \DIFdel{4 }\DIFdelend \DIFaddbegin \DIFadd{$4$ }\DIFaddend or more.

\subsection{Optimization due to parity condition}
\label{subsec:ffttwo:parity}

Let $n$ denote the length of RNA sequence \seq, and let $N$ denote the
least {\em even} integer greater than or equal to $n$. Since $N$ is even,
we have $(r+s) \equiv (r\cdot(N+1)+s) \bmod 2$. For distinct
fixed structures \strAB, let
$\pi_1(k) = \lfloor \frac{k}{N+1} \rfloor$, and
$\pi_2(k) = k \bmod (N+1)$, and define the polynomial

\begin{align}
\label{eq:ffttwo:polyRecEvenN}
\begin{split}
\fullZx &= \sum_{r=0}^{N} \sum_{s=0}^{N}\; z_{rN+s} x^{rN+s} \\
&=
\sum_{k=0}^{\mathclap{(N+1)^2-1}}\; z_{\pi_1(k)\cdot (N+1) + \pi_2(k)}
x^{\pi_1(k)\cdot (N+1) + \pi_2(k)} \\
&=
\sum_{k=0}^{\mathclap{(N+1)^2-1}}\; z_k x^k
\end{split}
\end{align}

where for the last equality, we have used the fact that
$k = \pi_1(k)\cdot (N+1)+\pi_2(k)$, well-known from
row major order of a \DIFdelbegin \DIFdel{0-indexed 2-dimensional }\DIFdelend \DIFaddbegin \DIFadd{$0$-indexed $2$-dimensional }\DIFaddend array.

Consider the coefficients of the polynomial

\begin{align}
\label{eq:ffttwo:zOfXEvenN}
\fullZx = \sum_{r=0}^{N} \sum_{s=0}^{N}\; z_{rN+s} x^{rN+s}
\enspace=\enspace\sum_{k=0}^{\mathclap{(N+1)^2-1}}\; z_k x^k.
\end{align}

Since $N$ is even, the parity of $r+s$ equals the parity of
$r(N+1)+s$, hence  it follows from the parity condition that either
\begin{inparaenum}[\em 1\upshape)]
\item all coefficients $z_1,z_3,z_5,\dots$ of odd parity are zero;
or \item all coefficients $z_0,z_2,z_4,\dots$ of even parity are zero.
\end{inparaenum}
To simplify notation, in the remainder of this subsection, let $M$ be
the least integer greater than or equal to $(N+1)^2$ that is evenly divisible
by 4, and let $M_0=M/2$. We will assume that $\fullZx = \sum_{k=0}^{M-1}
z_k x^k$, whereupon coefficients $z_k=0$ for $k>(N+1)^2$.

\medskip
{\sc Case 1}: All coefficients $z_k$ of odd parity in
\eqnref{ffttwo:zOfXEvenN} are zero.
\smallskip

In this case, we have
$\fullZx = \sum_{k=0}^{M_0-1} z_{2k} x^{2k}$. But then
$\fullZx = \emYof{}{u} = \sum_{k=0}^{M_0-1} b_k u^k$,
where we have made a variable change $u=x^2$, and coefficient changes
$b_k = a_{2k}$.
By evaluating $M_0=\frac{M}{2}$ many complex
$M_0$-roots of unity, we can
use polynomial interpolation to determine all coefficients $b_k$ of
the polynomial

\begin{align}
\emYof{}{u} = \sum_{k=0}^{M_0-1} b_k u^k = \sum_{k=0}^{M_0-1} z_{2k} x^{2k}.
\end{align}

Since $\emYof{}{x^2}=\fullZx$, we have
$\emYof{}{\exp(\frac{2\pi i k}{M/2})} =
\emYof{}{\exp(\frac{4\pi i k}{M})} =
\emZof{}{\exp(\frac{2\pi i k}{M})}$, hence we can
use the previous recursions from \eqnref{ffttwo:polyRecNussJac}
to evaluate \emZof{}{\exp(\frac{2\pi i k}{M})}.  Instead of
performing $M$ evaluations of \fullZx at $M$-roots of unity,
this requires only $M_0=M/2$ evaluations of $\emYof{}{u}$ at
$M_0$-roots of unity; i.e. only half the number of
evaluations of \fullZx are necessary to
obtain the coefficients of $\emYof{}{x}$. But then, we immediately obtain the
full polynomial \fullZx, since its coefficients of odd parity are zero.

\medskip
{\sc Case 2}: All coefficients $z_k$ of even parity in
\eqnref{ffttwo:zOfXEvenN} are zero.
\smallskip

In this case, $z_0,z_2,z_4,\dots$ are zero, so
$\fullZx = \sum_{k=0}^{M_0-1} z_{2k+1} x^{2k+1}$. But then
$\fullZx = x \cdot \emYof{}{u}$, where
$\emYof{}{u}= \sum_{k=0}^{M_0-1} b_k u^k$, where we have made
a variable change $u=x^2$, and coefficient changes
$b_k = z_{2k+1}$. Similarly to Case 1,
we can interpolate the $M_0$ coefficients of the polynomial
$\emYof{}{u} = \sum_{k=0}^{M_0-1} b_k u^k$
by evaluating $M_0$ many complex $M_0$-roots
of unity.  Since $\fullZx=x \cdot \emYof{}{x^2}$, $\emYof{}{x^2} = x^{-1} \cdot \fullZx$,
so
$\emYof{}{\exp(\frac{2\pi i k}{M/2})} = \emYof{}{\exp(\frac{4\pi ki}{M})} =
\exp(\frac{-2\pi i k}{M}) \cdot
\emZof{}{\exp(\frac{2\pi i k}{M})}$, employing
the previous recursions from \eqnref{ffttwo:polyRecNussJac}
to evaluate \emZof{}{\exp(\frac{2\pi i k}{M})}.  Note, that unlike the
Case 1, since $\fullZx = x \cdot \emYof{}{x^2}$, we have
$\emYof{}{x^2}=\frac{\fullZx}{x}$, which explains the presence of additional factor
$\exp(\frac{-2\pi i k}{M}))$ in Case \DIFdelbegin \DIFdel{2.  }\DIFdelend \DIFaddbegin \DIFadd{$2$.  }\DIFaddend Thus, instead of
performing $M$ evaluations of \fullZx at $M$-roots of unity,
we perform only $M_0=\frac{M}{2}$ evaluations of $\emYof{}{u}$ at
$M_0$-roots of unity; i.e. only half the number of
evaluations of \fullZx are necessary to
obtain the coefficients of $\emYof{}{x}$. But then, we immediately obtain the
full polynomial \fullZx, since $\fullZx = x \cdot \emYof{}{x^2}$, and the
coefficients of \fullZx of even parity are zero.

In the following, we will need the observation, that if the parity of
\bpd \dBP{\strA}{\strB} between \strAB is
even, then

\begin{align}
\label{eq:ffttwo:evenY}
\emYof{}{x^2} = \fullZx
\end{align}

while if the parity is odd, then

\begin{align}
\label{eq:ffttwo:oddY}
\emYof{}{x^2} = \frac{1}{x} \cdot \fullZx.
\end{align}

\subsection{Optimization due to complex conjugates}
\label{subsec:ffttwo:compconj}

As before, let $M$ be the the least number evenly divisible by 4, which is
greater than or equal to $(N+1)^2$, let $M_0=\frac{M}{2}$, let $\nu = \exp(\frac{2 \pi i}{M})$
and $\omega=\nu^2 = \exp(\frac{2 \pi \cdot 2i}{M}) =
\exp(\frac{2 \pi i}{M_0})$. Clearly, $\nu$ is a principal complex
$M$-root of unity, while $\omega$ is a principal complex $M_0$-root
of unity. Evaluate $\mathcal Z$ for each $M_0$-root of unity
that belongs to the first quadrant, and apply Lemma \ref{lem:ffttwo:lemma2}
 to infer the values
of $\mathcal Z$ for each $M_0$-root of unity that belongs to the
fourth quadrant. More precisely,
we compute $\emZof{}{\nu^k}$, for $k=0,\dots,\frac{M_0}{2}$, and by Lemmas
\ref{lem:ffttwo:lemma2}, \ref{lem:ffttwo:lemma3}, \ref{lem:ffttwo:lemma4} infer
that for $k=\frac{M_0}{2}+1,\dots,M_0-1$, we have
$\emZof{}{\nu^{k}}= -1^{d_0} \cdot \overline{\emZof{}{\nu^{M_0-k}}}$,
where
$d_0 = \dBP{\strA}{\strB}$. This is justified in the following.

By induction on $k=\frac{M_0}{2}+1,\dots,M_0-1$, we have

\newlength{\maxLengthCases}
\settowidth{\maxLengthCases}{$\nu^{-k} \cdot \emZof{}{-\overline{\nu^{(M_0-k)}}}$}

\begin{align}
\begin{split}
\emYof{}{\omega^k} &= \emYof{}{\nu^{2k}} \\ &=
\begin{cases}
\makebox[\maxLengthCases][l]{\emZof{}{\nu^{k}}}
& \text{if $\dBP{\strA}{\strB} = 0 \bmod 2$} \\
\frac{1}{\nu^{k}} \cdot \emZof{}{\nu^{k}}
& \text{if $\dBP{\strA}{\strB} = 1 \bmod 2$} \\
\end{cases}
\\ &=
\begin{cases}
\emZof{}{-\overline{\nu^{(M_0-k)}}}
& \text{if $\dBP{\strA}{\strB} = 0 \bmod 2$} \\
\nu^{-k} \cdot \emZof{}{-\overline{\nu^{(M_0-k)}}}
& \text{if $\dBP{\strA}{\strB} = 1 \bmod 2$} \\
\end{cases}
\\ &=
\begin{cases}
\makebox[\maxLengthCases][l]{\emZof{}{\overline{\nu^{(M_0-k)}}}}
& \text{if $\dBP{\strA}{\strB} = 0 \bmod 2$} \\
\nu^{-k} \cdot \emZof{}{\overline{\nu^{(M_0-k)}}}
& \text{if $\dBP{\strA}{\strB} = 1 \bmod 2$} \\
\end{cases}
\\ &=
\begin{cases}
\overline{\emZof{}{\nu^{(M_0-k)}}}
& \text{if $\dBP{\strA}{\strB} = 0 \bmod 2$} \\
-\nu^{-k} \cdot \overline{\emZof{}{\nu^{(M_0-k)}}}
& \text{if $\dBP{\strA}{\strB} = 1 \bmod 2$} \\
\end{cases}
\end{split}
\end{align}

Line \DIFdelbegin \DIFdel{1 }\DIFdelend \DIFaddbegin \DIFadd{$1$ }\DIFaddend follows by definition, since $\omega=\nu^2$;
line \DIFdelbegin \DIFdel{2 }\DIFdelend \DIFaddbegin \DIFadd{$2$ }\DIFaddend follows by equations (\ref{eq:ffttwo:evenY}) and
(\ref{eq:ffttwo:oddY});
line \DIFdelbegin \DIFdel{3 }\DIFdelend \DIFaddbegin \DIFadd{$3$ }\DIFaddend follows by Lemma \ref{lem:ffttwo:lemma4};
line \DIFdelbegin \DIFdel{4 }\DIFdelend \DIFaddbegin \DIFadd{$4$ }\DIFaddend follows by Lemma \ref{lem:ffttwo:lemma3}.
Thus if \dBP{\strA}{\strB} is even, then

\begin{align}
\label{eq:ffttwo:trickForRoU1}
y_k = \emYof{}{\omega^k} =
\begin{cases}
\emZof{}{\nu^k}
& \text{ for } k=0,\dots,\frac{M_0}{2} \\
\overline{\emZof{}{\nu^{M_0-k}}}
& \text{ for } k=\frac{M_0}{2}+1,\dots,M_0-1 \\
\end{cases}
\end{align}

while if \dBP{\strA}{\strB} is odd, then

\begin{align}
\label{eq:ffttwo:trickForRoU2}
y_k = \emYof{}{\omega^k} =
\begin{cases}
\nu^{-k} \cdot \emZof{}{\nu^k}
& \text{ for } k=0,\dots,\frac{M_0}{2} \\
-\nu^{-k} \cdot \overline{\emZof{}{\nu^{M_0-k}}}
& \text{ for } k=\frac{M_0}{2}+1,\dots,M_0-1 \\
\end{cases}
\end{align}

It follows that values $y_0,\dots,y_{M_0-1}$ can be obtained by only
$\frac{M}{4}$ evaluations of \fullZx.

\subsection{Polynomial interpolation to evaluate
\texorpdfstring{\emZ{i,j}}{}}
\label{subsec:ffttwo:fft}

Now let $M_0=\frac{M}{2}$, let
$\nu=\exp(\frac{2\pi i}{M})$ be the principal $M$-root of unity, and
$\omega=\nu^2=\exp(\frac{2\pi i}{M/2})=\exp(\frac{2\pi \cdot 2i}{M})$ be
the principal $M_0$-root of unity. Recall that the
Vandermonde matrix $V_{M_0}$ is defined to be the
$M_0 \times M_0$ matrix, whose $i,j$ entry is
$\omega^{i \cdot j} = \nu^{2 i \cdot j}$;
i.e.

\begin{align}
\label{eq:ffttwo:vandermonde}
V_{M_0} = \left(
\begin{array}{rrrrr}
1 & 1 & 1 & \dots & 1 \\
1 & \omega & \omega^2 & \dots & \omega^{M_0-1} \\
1 & \omega^2 & \omega^4 & \dots & \omega^{2(M_0-1)} \\
1 & \omega^3 & \omega^6 & \dots & \omega^{3(M_0-1)} \\
\vdots & \vdots & \vdots & \vdots & \vdots \\
1 & \omega^{M_0-1} & \omega^{2(M_0-1)} & \dots & \omega^{(M_0-1)(M_0-1)} \\
\end{array}
\right)
\end{align}

As described in \Chref{fftbor}, the \fft is the $O(n \log n)$
algorithm, which computes the Discrete Fourier Transform (DFT), defined
as the matrix product $\bfY = V_{M_0} {\bf A}$:

\begin{align}
\label{eq:ffttwo:dftMatrix}
\left(
\begin{array}{l}
y_0 \\
y_1 \\
y_2 \\
\vdots \\
y_{M_0-1} \\
\end{array}
\right)
= V_n \cdot
\left(
\begin{array}{l}
a_0 \\
a_1 \\
a_2 \\
\vdots \\
a_{M_0-1} \\
\end{array}
\right)
\end{align}

The $(i,j)$ entry of $V_{M_0}^{-1}$ is $\frac{\omega^{-j i}}{M_0}$
and that

\begin{align}
\label{eq:ffttwo:aFromY}
a_j &=\frac{1}{M_0} \sum_{k=0}^{M_0-1} y_k\, \omega^{-kj}
=\frac{1}{M_0} \sum_{k=0}^{M_0-1} y_k\, \nu^{-2kj}
\end{align}

for $j=0,\dots,M_0-1$.

Since we defined \bfY in \eqnref{ffttwo:yColumn} by $\bfY =
(y_0,\dots,y_{M_0-1})^{\text T}$, where
$y_0=\emZof{}{\alpha_0},\dots,\allowbreak y_{M_0-1}=\emZof{}{\alpha_{M_0-1}}$
and $\alpha_k = \omega^k
\exp(\frac{k \cdot 2\pi i}{M_0})$,
it follows that the coefficients
$z_k=\bfZ{\pi_1(k),\pi_2(k)}{1,n}$ in the polynomial
$\fullZx = z_0 + z_1 x + \dots + z_{M} x^{M}$ defined in
\eqnref{ffttwo:zOfX} can be computed using the FFT.
However, in practice we encounter the same issues of numerical instability
observed in \Secref{subsec:fftbor:fft}, and adopt a similar approach
to compute the $m$ most significant
digits of $\frac{\bfZ{\pi_1(k),\pi_2(k)}{1,n}}{\fullZ}$,
where the partition function $\fullZ = \sum_{S} \boltzf{\str}$ satisfies
${\fullZ} = \sum_{x,y} \bfZ{x,y}{1,n}$. This leads to numerical stability,
allowing \ffttwo to compute the
$m$ most significant digits of $p(x,y) = \frac{\bfZ{x,y}{1,n}}{\fullZ}$.
Pseudocode for \ffttwo which includes the performance enhancements described in
\Secref{sec:ffttwo:perf} follows below.
\medskip

\begin{figure}[!ht]
\hrule \rule[0ex]{0pt}{0pt}
\begin{center}
{\large Pseudocode for improved \ffttwo} \\
\end{center}
\begin{tabular*}{\textwidth}{ll}
{\sc Purpose:} & Computes the $m$ most significant digits
of probabilities \DIFdelbeginFL \DIFdelFL{$p_{r\cdot(N + 1) + s}=\sfrac{\bfZ{r,s}{1,n}}{\fullZ}$
}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$p_{r\cdot(N + 1) + s}=\rfrac{\bfZ{r,s}{1,n}}{\fullZ}$
}\DIFaddendFL \rule[-1.5ex]{0pt}{0pt} \\
{\sc Input:} & RNA sequence $\seq = \seqN$, secondary
structures \strA, \strB of \seq, integer $m$ \rule[-1.5ex]{0pt}{0pt} \\
{\sc Output:} & \DIFdelbeginFL \DIFdelFL{$p_{r\cdot(N + 1) + s}=\sfrac{\bfZ{r,s}{1,n}}{\fullZ}$ }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$p_{r\cdot(N + 1) + s}=\rfrac{\bfZ{r,s}{1,n}}{\fullZ}$ }\DIFaddendFL to
$\lfloor \log_{10}(2^m) \rfloor$ significant digits for $r,s=0,\dots,N$
\rule[-1.75em]{0pt}{0pt} \\
\hline \rule[0ex]{0pt}{0pt}
\end{tabular*}
\begin{algorithmic}[1]
\DIFdelbeginFL %DIFDELCMD < \Function{FFTbor2D Improved}{\seq, \strA, \strB, $m$}
%DIFDELCMD < %%%
\DIFdelendFL \DIFaddbeginFL \Function{FFTbor\twoD Improved}{\seq, \strA, \strB, $m$}
\DIFaddendFL \State $n \gets \textit{length}(\seq)$
\State $N \gets n + (n \bmod 2)$
\State $M \gets (N + 1)^2 + ((N + 1)^2 \bmod 4)$
\State $M_0 \gets \frac{M}{2}$
\For{$k \gets 0, (N + 1)^2-1$}
\Comment{Note that $k \gets \pi_1(k) \cdot M + \pi_2(k)$}
\State $\pi_1(k) \gets \lfloor \frac{k}{N + 1} \rfloor$
\State $\pi_2(k) \gets k \bmod (N + 1)$
\EndFor
\For{$k \gets 0, M-1$}
\Comment{Compute all complex $M$ and $M_0$-roots of unity}
\State $\nu_k \gets \exp(\frac{2 \pi i k}{M})$
\If{$k < M_0$}
\State $\omega_k \gets \exp(\frac{2 \pi i k}{M_0})$
\EndIf
\EndFor
\For{$k \gets 0, M_0-1$}
\If{$\dBP{\strA}{\strB} \bmod 2 = 0$}
\Comment{From \eqnref{ffttwo:trickForRoU1}}
\If{$k \leq \frac{M_0}{2}$}
\State $y_k \gets \emZof{}{\nu^k}$
\Else
\State $y_k \gets \overline{\emZof{}{\nu^{M_0-k}}}$
\EndIf
\Else
\Comment{From \eqnref{ffttwo:trickForRoU2}}
\If{$k \leq \frac{M_0}{2}$}
\State $y_k \gets \nu^{-k} \cdot \emZof{}{\nu^k}$
\Else
\State $y_k \gets -\nu^{-k} \cdot \overline{\emZof{}{\nu^{M_0-k}}}$
\EndIf
\EndIf
\EndFor
\algstore{ffttwoImproved}
\end{algorithmic}
\rule[0ex]{0pt}{1.5em} \hrule
\end{figure}
\clearpage

\begin{figure}[!ht]
\begin{algorithmic}[1]
\algrestore{ffttwoImproved}
\For{$k \gets 0, M_0-1$}
\Comment{Note that $\emZof{}{\nu_0} = \fullZ$}
\State $y_k \gets 2^m \cdot \frac{y_k}{\emZof{}{\nu_0}}$
\EndFor
\For{$k \gets 0, M_0-1$}
\Comment{Compute IDFT from \eqnref{ffttwo:aFromY}}
\State $a_k \gets \frac{1}{M_0} \sum_{k=0}^{M_0-1} y_k \omega^{-kj}$
\EndFor
\For{$k \gets 0, M-1$}
\Comment{Change the polynomial back to degree $M-1$}
\If{$\dBP{\strA}{\strB} \bmod 2 = 0$}
\If{$k \bmod 2 = 0$}
\State $p_{\pi_1(k) \cdot (N+1) + \pi_2(k)} \gets a_{k/2}$
\Else
\State $p_{\pi_1(k) \cdot (N+1) + \pi_2(k)} \gets 0$
\EndIf
\Else
\If{$k \bmod 2 = 0$}
\State $p_{\pi_1(k) \cdot (N+1) + \pi_2(k)} \gets 0$
\Else
\State $p_{\pi_1(k) \cdot (N+1) + \pi_2(k)} \gets a_{\frac{k-1}{2}}$
\EndIf
\EndIf
\EndFor
\For{$k \gets 0, (N + 1)^2 - 1$}
\Comment{Truncate to $m$ significant digits}
\State $\pk \gets 2^{-m} \cdot \lfloor \pk \rfloor$
\EndFor
\State \textbf{return} $p_0,\dots,p_{(N + 1)^2 - 1}$
\Comment{Return all $p(r,s) = p_{r \cdot (N + 1) + s} = \frac{\bfZ{r,s}{1,n}}{\fullZ}$}
\EndFunction
\rule[-0.35ex]{0pt}{0pt}
\end{algorithmic}
\caption[Pseudocode for improved \ffttwo]{
Pseudocode to compute the $m$ most significant digits
for probabilities
\DIFdelbeginFL \DIFdelFL{$p_{r\cdot(N + 1)+s} = \sfrac{\bfZ{r,s}{1,n}}{\fullZ}$}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$p_{r\cdot(N + 1)+s} = \frac{\bfZ{r,s}{1,n}}{\fullZ}$}\DIFaddendFL . In our implementation,
due to numerical stability issues in the FFT engine, precision parameter
$m$ has an upper bound of \DIFdelbeginFL \DIFdelFL{27---only }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$27$---only }\DIFaddendFL the $\lfloor \log_{10}(2^m) \rfloor =8$ most significant digits
are computed with \ffttwo. It is well-known that
the FFT requires $O(N \log N)$ time to solve the inverse discrete
Fourier transform for a polynomial of degree $N$. In our case,
$N=n^2$, and so the FFT requires time $O(n^2 \log n)$.
}
\label{fig:ffttwo:algoImproved}
\rule[0ex]{0pt}{1.5em} \hrule
\end{figure}

\section{Performance characteristics of \ffttwo}
\label{sec:ffttwo:speed}

To perform comparative benchmarking between \rnatwofold and \ffttwo,
we took precision parameter $m=8$, and proceeded as follows.
For each sequence length $n = 20,25,30,\ldots,300$, we generated
\DIFdelbegin \DIFdel{100 }\DIFdelend \DIFaddbegin \DIFadd{$100$ }\DIFaddend random sequences using probability $0.25$ for each nucleotide
A,\,C,\,G,\,U.
For a given RNA sequence \seq, the metastable structure \strA was
taken to be the MFE structure of \seq.
Using \rnabor, we determined that value $k_0 \geq 10$, for which
partition function $\bfZ{k_0}{}$ constitutes a visible peak in the graphical
\DIFdelbegin \DIFdel{output --- see }\DIFdelend \DIFaddbegin \DIFadd{output---see }\DIFaddend Figure 2 and 3 of \citep{freyhult.b07} for an example.
Subsequently, metastable structure \strB was taken to be that structure
having \mfe over all structures, whose \bpd from
\strA was $k_0$.

For all $0 \leq x,y \leq n$, \rnatwofold and \ffttwo were
benchmarked in the computation of all Boltzmann probabilities
$p(x,y)= \frac{\bfZ{x,y}{}}{\fullZ}$, where $x$ [resp. $y$] represents base pair
distance to metastable structure \strA [resp. \strBresp]. Care was taken for
both software to employ the same energy model (\DIFdelbegin \DIFdel{Turner99 }\DIFdelend \DIFaddbegin \DIFadd{Turner$99$ }\DIFaddend energy model,
no dangles, suppression of \mfes computations
for \rnatwofold) and the same number of parallel threads (8 threads
using OpenMP). Nonetheless, there are slight differences in the energy
models --- namely, \rnatwofold includes
mismatch penalties for multiloop stems and for exterior loops, while
\ffttwo does not. Even in the computation of the partition function
$Z$, for spliced leader RNA from {\em L. collosoma} of length \DIFdelbegin \DIFdel{56 }\DIFdelend \DIFaddbegin \DIFadd{$56$ }\DIFaddend nt,
\rnatwofold \ms{-d0} obtains a value of
\DIFdelbegin \DIFdel{$-9.6604$ kcal/mol}\DIFdelend \DIFaddbegin \kmolk{-9.6604}\DIFaddend , while \ffttwo obtains
\DIFdelbegin \DIFdel{$-9.6605$ kcal/mol}\DIFdelend \DIFaddbegin \kmolk{-9.6605}\DIFaddend ; similarly, for attenuator RNA of length \DIFdelbegin \DIFdel{73 }\DIFdelend \DIFaddbegin \DIFadd{$73$ }\DIFaddend nt,
\rnatwofold \ms{-d0} obtains a value of
\DIFdelbegin \DIFdel{$-22.1718$ kcal/mol}\DIFdelend \DIFaddbegin \kmolk{-22.1718}\DIFaddend , while \ffttwo obtains
\DIFdelbegin \DIFdel{$-22.1732$ kcal/mol}\DIFdelend \DIFaddbegin \kmolk{-22.1732}\DIFaddend .
Note that the straightforward calculation of the partition function, following McCaskill's algorithm \citep{mccaskill} makes no use of the FFT engine, and thus the differences cannot be due to floating point or precision issues.

For benchmarking purposes,
to allow for a fair comparison of \ffttwo
with \rnatwofold, we restricted the range of $x,y$ in the same manner as done in the source code of \rnatwofold.
In that code, parameters $K$ [resp. $L$] are defined respectively to be the
sum of the number of base pairs in reference structure \strA
[resp. \strBresp] plus the number of base pairs in
the maximum matching (Nussinov) structure which contains
no base pair of \strA [resp. \strBresp].
For $x \geq K, y \geq L$, both \rnatwofold and \ffttwo
set $p(x,y)=0$.  For the benchmarking results displayed in
Figures \ref{fig:ffttwo:ffttwoRtwofoldTiming},
\ref{fig:ffttwo:ffttwoRtwofoldLogScale},
\ref{fig:ffttwo:ffttwoRtwofoldStdevMinMax},
the values $x,y$ are restricted in
\ffttwo to $0 \leq x,y \leq \max(K,L)$, while $0 \leq x \leq K$
and $0 \leq y \leq L$ in \rnatwofold.

\Figref{ffttwo:ffttwoRtwofoldTiming} depicts average run time of \rnatwofold
and \ffttwo as a function of RNA sequence length, for random RNA
sequences of lengths \DIFdelbegin \DIFdel{20--200 }\DIFdelend \DIFaddbegin \DIFadd{$20$--$200$ }\DIFaddend and their metastable structures \strAB, as
previously explained.  We see that both programs have
roughly comparable run times for sequences of length up to approximately
\DIFdelbegin \DIFdel{80 }\DIFdelend \DIFaddbegin \DIFadd{$80$ }\DIFaddend nt, while \ffttwo is demonstrably faster for longer sequences.
\Figref{ffttwo:ffttwoRtwofoldLogScale}
presents $\log_{10}$ run time as a function of sequence length, in order
to more clearly determine the crossover point in performance.
\rnatwofold is marginally faster for sequences of
length up to roughly \DIFdelbegin \DIFdel{80 }\DIFdelend \DIFaddbegin \DIFadd{$80$ }\DIFaddend nt, though the difference is in the millisecond
range.
\Figref{ffttwo:ffttwoRtwofoldStdevMinMax}
shows that the standard deviation of run times on random sequences is
tiny for \ffttwo compared with \rnatwofold, where standard
deviation increases rapidly as a function of sequence length. This figure
shows that run time of \rnatwofold depends on sequence details, as
well as sequence length, while the run time of \ffttwo depends only
on sequence length.

\begin{figure}[!ht]
\centering
\includegraphics[width=.9\textwidth]{Figures/FFTbor2D/ffttwoRtwofoldTiming.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption[Run time in seconds for \rnatwofold and \ffttwo on random
%DIFDELCMD < RNA sequences of length 20--200 nt]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Run time in seconds for \rnatwofold and \ffttwo on random
RNA sequences of length $20$--$200$ nt]{\DIFaddendFL 
Run time in seconds for \rnatwofold and \ffttwo on random
RNA sequences of length \DIFdelbeginFL \DIFdelFL{20--200 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$20$--$200$ }\DIFaddendFL nt, where sequence generation and
choice of metastable structures \strAB is described in the text.
Beyond a length of approximately \DIFdelbeginFL \DIFdelFL{80 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$80$ }\DIFaddendFL nt, \ffttwo is demonstrably
faster.
}
\label{fig:ffttwo:ffttwoRtwofoldTiming}
\end{figure}

\begin{figure}[!ht]
\centering
\includegraphics[width=.9\textwidth]{Figures/FFTbor2D/ffttwoRtwofoldLogScale.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption[Logarithm of run time in seconds for \rnatwofold and \ffttwo
%DIFDELCMD < on random RNA sequences of length less than 200 nt]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Logarithm of run time in seconds for \rnatwofold and \ffttwo
on random RNA sequences of length less than $200$ nt]{\DIFaddendFL 
Logarithm of run time in seconds for \rnatwofold and \ffttwo
on random RNA sequences of length less than \DIFdelbeginFL \DIFdelFL{200 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$200$ }\DIFaddendFL nt, for same data as that
in \Figref{ffttwo:ffttwoRtwofoldTiming}.
By taking $\log_{10}$ of the run times,
crossover points are apparent,
where \ffttwo is faster than \rnatwofold. For very small
sequences, \rnatwofold is faster, though since both programs converge
in a fraction of a second, this difference is of no practical consequence.
}
\label{fig:ffttwo:ffttwoRtwofoldLogScale}
\end{figure}

\begin{figure}[!ht]
\centering
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.65\textwidth]{Figures/FFTbor2D/ffttwoRtwofoldStdev.pdf}
\end{subfigure} \\
\begin{subfigure}[h]{\textwidth}
\centering
\includegraphics[width=.65\textwidth]{Figures/FFTbor2D/ffttwoRtwofoldMinMax.pdf}
\end{subfigure}
\caption[{\em (Top)}
Standard deviation of run times of \rnatwofold and \ffttwo
as a function of sequence length $n$.
{\em (Bottom)}
Minimum and maximum run times for \rnatwofold and \ffttwo]{
{\em (Top)}
Standard deviation of run times of \rnatwofold and \ffttwo
as a function of sequence length $n$.
{\em (Bottom)}
Minimum and maximum run times for \rnatwofold and \ffttwo.
For each collection of \DIFdelbeginFL \DIFdelFL{100 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$100$ }\DIFaddendFL random sequences of length $n$, the minimum
and maximum run time for a sequence of that length was computed.
Taken together, these figures clearly show the
run time dependence of \rnatwofold on particular sequences, while
the run time of \ffttwo depends only on sequence length, rather than
sequence details.
}
\label{fig:ffttwo:ffttwoRtwofoldStdevMinMax}
\end{figure}

\begin{figure}[!ht]
\centering
\includegraphics[width=.7\textwidth]{Figures/FFTbor2D/ffttwoProbContour.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption[2D projection of energy landscape for Spliced Leader (SL) RNA
%DIFDELCMD < from {\em L. collosoma}, as computed by \ffttwo]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[\twoD projection of energy landscape for Spliced Leader (SL) RNA
from {\em L. collosoma}, as computed by \ffttwo]{\DIFaddendFL 
\DIFdelbeginFL \DIFdelFL{2D }\DIFdelendFL \DIFaddbeginFL \twoD \DIFaddendFL projection of energy landscape for Spliced Leader (SL) RNA
from {\em L. collosoma}, having
sequence and metastable secondary structures: \rule[-1.25em]{0pt}{0pt} \\
\DIFaddbeginFL \DIFaddFL{\hspace{.4em}}\DIFaddendFL $\seq\,=$ \DIFdelbeginFL %DIFDELCMD < \ms{AACUAAAACAAUUUUUGAAGAACAGUUUCUGUACUUCAUUGGUAUGUAGAGACUUC} %%%
\DIFdelendFL \DIFaddbeginFL \strfont{AACUAAAACAAUUUUUGAAGAACAGUUUCUGUACUUCAUUGGUAUGUAGAGACUUC} \DIFaddendFL \\
                $\strA =$ \DIFdelbeginFL %DIFDELCMD < \ms{..((...((((((..(((((.((((...)))).)))))..))).)))..)).....} %%%
\DIFdelendFL \DIFaddbeginFL \strfont{..((...((((((..(((((.((((...)))).)))))..))).)))..)).....} \DIFaddendFL \\
  \DIFaddbeginFL \DIFaddFL{\hspace{.4em}}\DIFaddendFL $\strB =$ \DIFdelbeginFL %DIFDELCMD < \ms{.......................((((((((((((.....)))))..)))))))..}
%DIFDELCMD < %%%
\DIFdelendFL \DIFaddbeginFL \strfont{.......................((((((((((((.....)))))..)))))))..}
\DIFaddendFL \rule[-1.25em]{0pt}{0pt} \\
The $x$-axis [resp. $y$-axis]
represents \bpd between metastable structure \strA [resp.
\strBresp],
while the $z$-axis represents the ensemble free energy $-RT \log \bfZ{x,y}{}$,
where \bfZ{x,y}{} is computed in \ffttwo by $\bfZ{x,y}{}=p(x,y) \cdot \fullZ$.
Low energy positions $(x,y)$ correspond to high Boltzmann probability positions.
The left panel depicts a heat map of the ensemble free energy,
while the right panel depicts a contour map with level curves. In analogy
with mountain climbing, one expects an optimal path to follow along the
valley regions in traversing the landscape from \strA to \strB. Data produced
with \ffttwo; graphics produced using Mathematica.
}
\label{fig:ffttwo:ffttwoProbContour}
\end{figure}

\begin{figure}[!ht]
\centering
\includegraphics[width=.45\textwidth]{Figures/FFTbor2D/rtwofoldProbContour.pdf}
\quad
\includegraphics[width=.45\textwidth]{Figures/FFTbor2D/rtwofoldEnsFreeEnergyContour.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption[2D projection of energy landscape for Spliced Leader (SL) RNA
%DIFDELCMD < from {\em L. collosoma}, as in
%DIFDELCMD < \Figref{ffttwo:ffttwoProbContour},
%DIFDELCMD < except that in the right panel, ensemble free energy $-RT \log \bfZ{x,y}{}$
%DIFDELCMD < is computed from the values of \bfZ{x,y}{} output by \rnatwofold,
%DIFDELCMD < while in the left panel, ensemble free energy is computed from
%DIFDELCMD < the values $\bfZ{x,y}{}=p(x,y) \cdot \fullZ$, where values $p(x,y)$ are output
%DIFDELCMD < by \rnatwofold]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[\twoD projection of energy landscape for Spliced Leader (SL) RNA
from {\em L. collosoma}, as in
\Figref{ffttwo:ffttwoProbContour},
except that in the right panel, ensemble free energy $-RT \log \bfZ{x,y}{}$
is computed from the values of \bfZ{x,y}{} output by \rnatwofold,
while in the left panel, ensemble free energy is computed from
the values $\bfZ{x,y}{}=p(x,y) \cdot \fullZ$, where values $p(x,y)$ are output
by \rnatwofold]{\DIFaddendFL 
\DIFdelbeginFL \DIFdelFL{2D }\DIFdelendFL \DIFaddbeginFL \twoD \DIFaddendFL projection of energy landscape for Spliced Leader (SL) RNA
from {\em L. collosoma}, as in
\Figref{ffttwo:ffttwoProbContour},
except that in the right panel, ensemble free energy $-RT \log \bfZ{x,y}{}$
is computed from the values of \bfZ{x,y}{} output by \rnatwofold,
while in the left panel, ensemble free energy is computed from
the values $\bfZ{x,y}{}=p(x,y) \cdot \fullZ$, where values $p(x,y)$ are output
by \rnatwofold.
The loss of detail in the \DIFdelbeginFL \DIFdelFL{2D }\DIFdelendFL \DIFaddbeginFL \twoD \DIFaddendFL energy landscape is caused uniquely by
working with probabilities $p(x,y)$, rather than partition function
values \bfZ{x,y}{}.
Data produced with \rnatwofold; graphics produced using Mathematica.
}
\label{fig:ffttwo:rtwofoldProbContour}
\end{figure}

An important advantage of
\rnatwofold over \ffttwo is that the former can additionally
compute the structures $M_{x,y}$ having \mfe over all
structures that are $x$-neighbors of metastable \strA and simultaneously
$y$-neighbors of metastable \strB. (There is a similar advantage of \rnabor
\citep{freyhult.b07} over the faster \fftbor \citep{senter.po12}.)
As well, \rnatwofold directly computes the partition function values
\bfZ{x,y}{}, while \ffttwo estimates \bfZ{x,y}{} by computing
$p(x,y) \cdot \fullZ$. This difference entails a significant loss of precision,
when depicting the energy landscape.

%!TEX root = ../main.tex

\chapter{Hermes}
\label{ch:hermes}

\lhead{Hermes: Efficient Kinetic Analysis of RNA Molecules}

\section{Introduction}
\label{sec:hermes:intro}

In this chapter, we present the \hermes software suite---a collection of
programs aimed at evaluating the kinetic properties of RNA molecules.
Provided a coarse-grained energy landscape generated by \ffttwo (described
in \Chref{ffttwo}), we present software which computes both the \mfpt
and \eqt for this discretized energy landscape. We also provide software which
computes the exact kinetics for an RNA molecule, however since this requires
exhaustive enumeration of all secondary structures---which is known to be an
exponential quantity for the length of the RNA in consideration---the full
kinetics are not expected to be practical for anything beyond a sequence of
trivial length. The software in \hermes presents a practical application of
the energy landscapes computed by the \ffttwo algorithm. Contrasted against
the other kinetics software in the field, \hermes offers similar accuracy
with unparalleled performance which opens up the possibility for large-scale
kinetic analysis {\em in silico}, which we expect to be of use for synthetic
design.

\subsection{Organization}
\label{subsec:hermes:org}

This chapter is organized in the following fashion. We begin by providing
background on the state-of-the-art approaches for kinetic analysis of RNAs.
From there, we move into a technical discussion of two traditional approaches
for kinetics, computation of the \mfpt and the \eqt. With this foundation in
place, we proceed to discuss the high-level organization of the \hermes
software package, and describe in detail each of the four underlying programs
which comprise the kinetics suite. We then move on to present comparitive
benchmarking of \hermes against other methods, before finally concluding with
some remarks on the accuracy and applicability of \hermes to computational
RNA design.

\section{Background}
\label{sec:hermes:bkgrnd}

Remarkable results in RNA synthetic biology have recently been
obtained by various groups. In \citep{hochrein.jacs13}, small
conditional RNAs have been engineered to silence a gene Y by using the
RNA interference machinery, only if a gene X is transcribed. In
\citep{wachsmuth.nar13} a novel theophylline \rb has been
computationally designed to transcriptionally regulate a gene in
{\em E. coli}, and in \citep{synthetichammerheads} a purely computational
approach was used to design functionally active hammerhead ribozymes.
Computational design of synthetic RNA molecules invariably uses some
form of thermodynamics-based algorithm; indeed, \ms{NUPACK-Design}
\citep{zadeh.jcc11} was used to design small conditional RNAs
\citep{hochrein.jacs13}, Vienna RNA Package \citep{gruber08} was used in
the design of the synthetic theophylline \rb, and the
\ms{RNAiFold} inverse folding software
\citep{garcia.jbcbb13,garciamartin.nar13} was used to design the
synthetic hammerhead ribozymes. The next step in the computational
design of synthetic RNA molecules is to control the kinetics of
folding---such control could be important in engineering
conformational switches. Kinetics is already used as a design feature
for synthetic design in proteins \citep{bujotzek.jcam11,fasting.acie12}.

In this chapter we introduce a new software suite, called \hermes
\citep{senter:2015fq},
which efficiently computes RNA secondary structure
folding kinetics by using a coarse-grained
method to model RNA transitions that add or remove a single base pair.
Since our motivation in developing \hermes is to provide a new
tool to aid in engineering synthetic RNA molecules with desired
kinetic properties, \hermes does not model {\em co-transcriptional
folding}, but only the {\em refolding} of RNA sequences.

There is a rich history of both experimental and computational work on
RNA folding pathways and kinetics. Experimental approaches to
determine the kinetics of RNA folding include temperature jump
experiments \citep{lecuyercrothers}, using fluorophores
\citep{hobartner.jmb03}, using mechanical tension at the single
molecule level \citep{vieregg.mp06}, etc. and will not be further
discussed. Computational approaches to folding kinetics commonly model
stepwise transitions between secondary structures,
involving the addition or removal of a single base
pair, as first considered in \citep{flammhofacker}.
Nevertheless, it should be noted that there are a number of
methods that concern the addition or removal of an entire helical
region, e.g. \citep{huang.bb14,zhao.jcp11}.

Most computational approaches involve either
\DIFdelbegin \DIFdel{(1) }\DIFdelend \DIFaddbegin \begin{inparaenum}[\em 1\upshape)]
\item \DIFaddend algorithms to determine optimal or near-optimal folding pathways\DIFdelbegin \DIFdel{, (2) }\DIFdelend \DIFaddbegin \DIFadd{;
}\item \DIFaddend explicit solutions of the master equation\DIFdelbegin \DIFdel{, (3) }\DIFdelend \DIFaddbegin \DIFadd{; or
}\item \DIFaddend repeated simulations to fold an initially empty secondary structure to
the target \mfe (MFE) structure.
\DIFaddbegin \end{inparaenum}
\DIFaddend Examples of methods to determine optimal or
near-optimal folding pathways include the greedy approach of Morgan
and Higgs \citep{morganhiggsbarrier}, the exact, optimal, exponential
time program \barriers \citep{flammhofacker}, the program
\ms{Findpath} \citep{flamm.r01}, which uses bounded look-ahead
breadth-first search, a genetic algorithm \citep{shapiro.jmb01}, the
program \ms{RNAtabupath} that uses local search to determine
near-optimal folding pathways, the Basin Hopping Graph
(BHG, \citep{kucharik:2014ft}) which uses a greedy algorithm to determine local minima
and approximate their barrier energies, etc. The program \barriers
\citep{flammhofacker} is the only current method guaranteed to
produce optimal folding pathways. Since it has been shown that
the problem of determining optimal folding pathways is
NP-complete \citep{thachuk.psb10}, it is now understandable why
\ms{barriers} can take exponential time to converge, depending
on the RNA sequence. For this reason, near-optimal solutions
provided by heuristic methods, such as \ms{RNAtabupath}, are very useful.

Methods that employ the master equation include \ms{Treekin}
\citep{wolfingerstadler:kinetics}, which uses the programs
\ms{RNAsubopt} \citep{wuchtyfontanahofackerschuster} and \barriers
\citep{flammhofacker} to determine {\em macrostates}, defined as basins
of attraction near a locally optimal structure. The resulting
coarse-grained Markov chain is then sufficiently small to allow explicit
solution of the master equation. In \citep{tang.jcb05}, a moderate
number of RNA structures were sampled according to different
strategies, from which a robotic motion planning graph was defined
to connect each sampled structure to its $k$ nearest sampled
neighbors. Again, the resulting coarse-grained Markov chain is
sufficiently small for an explicit solution of the master equation to
be given.

We now come to simulation approaches to estimate RNA folding kinetics.
The program \kinfold \citep{flammphd,flamm} is an implementation
of Gillespie's algorithm \citep{gillespiestochasticsimulation1},
directly related to the master equation, hence is considered by many
to be the {\em gold standard} for RNA kinetics. A recent extension of the
\kinfold algorithm was reported in \citep{aviram.amb12}.
\ms{KFOLD} \citep{dykeman:2015bxa} also implements Gillespie's algorithm,
but leverages memoization of transition rates to operate much more efficiently
than \kinfold. Meanwhile \ms{Kinefold} \citep{xayaphoummine.nar05} uses
stochastic simulations of
the nucleation and dissociation of helical regions to predict
secondary structure and folding pathways. In contrast to the
previously mentioned methods, both \ms{RNAKinetics}
\citep{danilova.jbcb06} and \ms{Kinwalker} \citep{geis.jmb08} model
co-transcriptional folding, known to be necessary when simulating
{\em in vivo} folding of long RNA molecules \citep{lai.r13}. As well,
\ms{Kinefold} can simulate both refolding and co-transcriptional folding
pathways. Finally, unlike all the previous simulation results, which
depend on thermodynamic free energy parameters \citep{turner.nar10},
the program \ms{Oxfold} \citep{anderson.b13} performs kinetic folding
of RNA using stochastic context-free grammars and evolutionary
information.

In contrast to the previous methods, \hermes computes the \mfpt (MFPT)
and \DIFdelbegin %DIFDELCMD < {\em %%%
\DIFdel{equilibrium time}%DIFDELCMD < } %%%
\DIFdelend \DIFaddbegin \eqt \DIFaddend for a
coarse-grained Markov chain consisting of the ensemble of all
secondary structures having \bpd $x$ [resp. $y$] from
reference structures \strA [resp. \strBresp] of a given RNA sequence.
\Mfpt (MFPT) is computed exactly by matrix inversion,
and equilibrium time is computed using spectral decomposition of the rate
matrix for the coarse-grained master equation.

\section{Traditional approaches for kinetics}
\label{sec:hermes:kinetics}

To better understand the underlying algorithms behind the software,
we describe two traditional approaches in kinetics, \mfpt and \eqt.

\subsection{\Mfpt}
\label{subsec:hermes:mfpt}

Consider a physical process,
which when monitored over time, yields the stochastic sequence
$q_0,q_1,q_2,\dots$ of discrete, observed states. If the transition from
state $q_t$ to $q_{t+1}$ depends only on $q_t$ at time
$t$ and not on the historical sequence of prior states visited,
as often assumed in the case of protein or
RNA folding, then a Markov chain provides a reasonable
mathematical model to simulate the process.

A first-order, time-homogeneous
{\em Markov chain} $\mChain = (Q,\pi,P)$ is given by a finite
set $Q = \{ 1,\dots,n \}$ of states, an initial probability distribution
$\pi = (\pi_1,\dots,\pi_n)$, and the $n \times n$
{\em transition probability matrix} $P = (p_{i,j})$.
At time $t=0$, the initial state of the system is
$q_t=i$ with probability $\pi_i$, and at discrete time
$t=1,2,3,\dots$, the system makes a transition from state
$i$ to state $j$ with probability $p_{i,j}$; i.e.
the conditional probability $Pr[ q_{t+1} = j | q_t=i] = p_{i,j}$.
Define the {\em population occupancy frequency} of visiting state
$i$ at time $t$ by $p_{i}(t) = Pr[q_t=i]$.  Denote
$p_{i,j}^{(t)} = Pr[q_t = j | q_0 = i]$ and notice that
the $(i,j)$th entry of the $t$th power $P^t$ of matrix $P$  equals
$p_{i,j}^{(t)}$.

The {\em \mfpt} (MFPT) or
{\em \hit} for the Markov chain \mChain,
starting from initial state \xStart and proceeding
to the target state \xEnd, is defined
as the sum, taken over all paths $\rho$
from \xStart to \xEnd, of the path length $\textit{length}(\rho)$
times the probability
of path $\rho$, where $\textit{length}(\rho_0,\dots,\rho_n)$ is defined to
be $n$. In other words,
${\text MFPT} = \sum_{\rho} Pr[\rho] \cdot \textit{length}(\rho)$, where
the sum is taken over sequences $\rho=\rho_0,\dots,\rho_n$ of states
where $\rho_0=\xStart$ and $\rho_n=\xEnd$, and no state is visited more
than once in the path $\rho$.

Given the target state \xEnd, MFPT can be exactly
determined by computing the inverse
$(I - P^{-}_{\xEnd})^{-1}$, where $I$ is the $(n-1)\times (n-1)$ identity
matrix and $P^{-}_{\xEnd}$
denotes the $(n-1)\times (n-1)$ matrix obtained from the
Markov chain transition probability matrix $P$,
by deleting the row and column with index \xEnd.
Letting ${\bf e}$ denote the
$(n-1) \times 1$ column vector consisting entirely of 1's, it can be
shown that \mfpt from state \xStart to state \xEnd
is the \xStart-th coordinate
of column vector
$(I - P^{-}_{\xEnd})^{-1} \cdot {\bf e}$ \citep{meyermfpt}.

The {\em stationary} probability distribution \statDist
is a row vector such that $\pSt \cdot P = \pSt$; i.e.
$\pSt_j = \sum_{i}^n \pSt_i\, \cdot\, p_{i,j}$, for all $1\leq j \leq n$.
It can be shown that the stationary probability $\pSt_i$ is the limit,
as $m$ tends to infinity, of the frequency of visiting state $i$ in a
random walk of length $m$ on Markov chain \mChain.
It is well-known that the stationary distribution exists and is unique
for any finite aperiodic irreducible Markov chain \citep{cloteBackofen:book}.

The Metropolis Monte Carlo algorithm \citep{metropolis:MonteCarlo} can
be used to simulate a random walk from initial state \xStart to target state
\xEnd, when energies are associated with the states, as is the case in
macromolecular folding, where free energies can be determined for
protein and RNA conformations from mean field theory, quantum theory,
or experimental measurements. In such cases, a {\em move set}
defines the set $N_x$ of conformations reachable in unit time
from conformation $x$, and the transition probability matrix
$P = (p_{x,y})$ is defined as follows:

\begin{align}
\label{eq:hermes:mfptXitionProb}
p_{x,y} =
\begin{cases}
\frac{1}{|N_x|} \cdot \min\left( 1, \boltzYmX \right)
& \text{if $y \in N_x$} \\
1 - \sum_{k \in N_x} p_{x,k} & \text{if $x=y$} \\
0 & \text{if $y \not\in N_x$, and $x \ne y$} \\
\end{cases}
\end{align}

If $\pSt_x \cdot p_{x,y} = \pSt_y \cdot p_{y,x}$ holds for all distinct
$x,y \in Q$, then
{\em detailed balance} is said to hold, or equivalently the Markov
chain \mChain is said to be {\em reversible}.
If transitional probabilities
are defined as in \eqnref{hermes:mfptXitionProb}, and if neighborhood
size is constant ($|N_x|=|N_y|$ for all $x,y$), then it is well-known
that the stationary probability distribution
\statDist is the Boltzmann distribution; i.e.

\begin{align}
\label{eq:hermes:stationaryBoltzmannProb}
\pSt_k = \frac{\boltzf{k}}{\fullZ}
\end{align}

where $E(k)$ is the energy of conformation $k$ at temperature $T$,
$R$ is the universal gas constant,
$T$ is absolute temperature, and the \DIFdelbegin %DIFDELCMD < {\em %%%
\DIFdel{partition function
}%DIFDELCMD < }
%DIFDELCMD < %%%
\DIFdel{$Z = \sum_{k=1}^n \boltzf{k}$ }\DIFdelend \DIFaddbegin \DIFadd{partition function
$\fullZ = \sum_{k=1}^n \boltzf{k}$ }\DIFaddend is a normalization constant
\citep{waterman:book,cloteBackofen:book}. If neighborhood size is not
constant, as in the case where states are RNA secondary structures and transitions
are restricted to the addition or removal of a single base pair, then
by {\em Hasting's trick}, an equivalent Markov chain can be defined which
satisfies detailed balance---see \eqnref{hermes:xitionProbWithHastings}.

Following Anfinsen's experimental work on the denaturation and
refolding of bovine pancreatic ribonuclease \citep{anfinsen},
the native conformation is assumed to be the ground state having
\mfe. These results justify the use of the Monte Carlo
Algorithm \ref{fig:hermes:mcmc} in macromolecular kinetics and
structure prediction.
\medskip

\begin{figure}[!ht]
\hrule \rule[0ex]{0pt}{0pt}
\begin{center}
{\large Pseudocode for the Metropolis Monte Carlo algorithm} \\
\end{center}
\begin{tabular*}{\textwidth}{ll}
{\sc Purpose:} & Compute discrete time $t$ to move from
\xStart to \xEnd, with upper limit of \tMax \rule[-1.5ex]{0pt}{0pt} \\
{\sc Input:} & Starting state \xStart, target state \xEnd and
upper bound on time \tMax \rule[-1.5ex]{0pt}{0pt} \\
{\sc Output:} & Discrete time $t$ taken to reach \xEnd from \xStart
\rule[-1.75em]{0pt}{0pt} \\
\hline \rule[0ex]{0pt}{0pt}
\end{tabular*}
\begin{algorithmic}[1]
\Function{Metropolis}{\xStart, \xEnd, \tMax}
\State $t \gets 0$
\State $x \gets \xStart$
\While{$x \neq \xEnd$ AND $t < \tMax$}
\State $y \in N_x$
\If{$E(y) < E(x)$}
\Comment{Greedy move}
\State $x \gets y$
\Else
\Comment{Metropolis move}
\State $z \in (0,1)$
\If{$z < \boltzYmX$}
\State $x \gets y$
\EndIf
\EndIf
\State $t \gets t+1$
\EndWhile
\State \textbf{return} $t$
\Comment{Return discrete time $t$ to travel from \xStart to \xEnd}
\EndFunction
\rule[-0.35ex]{0pt}{0pt}
\end{algorithmic}
\caption[Pseudocode for the Metropolis Monte Carlo algorithm]{The function {\sc Metropolis} implements a discrete time simulation
of folding trajectories for a Markov chain.}
\label{fig:hermes:mcmc}
\rule[0ex]{0pt}{1.5em} \hrule
\end{figure}

The \mfpt from state $x$ to state $y$ can
be approximated by repeated runs of the Monte Carlo algorithm.
In particular, \v{S}ali, Shakhnovich, and Karplus used such Monte Carlo
simulations to investigate the Levinthal paradox of how a protein
can fold to its native state within milliseconds to seconds.
By repeated Monte Carlo simulations using a protein lattice model,
\v{S}ali et al. observed that a large energy difference between
the ground state
and the first misfolded state appears to be correlated with fast folding.

\subsection{\Eqt}
\label{subsec:hermes:eq}

A continuous time {\em Markov process} $\mChain=(Q,\pi,P(t))$
is given by a finite set $Q= \{1,\dots,n\}$ of states, the
initial probability distribution $\pi$,
and the $n\times n$ matrix $P(t)=(p_{i,j}(t))$ of
probability transition functions.

Letting $q_t$ denote the state at (continuous)
time $t$, the probability that the initial state $q_0$ at time \DIFdelbegin \DIFdel{0 }\DIFdelend \DIFaddbegin \DIFadd{$0$ }\DIFaddend is
$k$ is $\pi_k$, while

\begin{align}
\label{eq:hermes:markovProcess}
p_{i,j}(t) = Pr[q_{t} = j| q_0 = i].
\end{align}

The matrix $P'(t)$ of derivatives, defined by

\begin{align}
  P'(t) = \left(
\begin{array}{ccc}
\frac{d p_{1,1}}{d t}(t) & \dots  & \frac{d p_{1,n}}{d t}(t)\\
\vdots & \ddots & \vdots \\
 \frac{d p_{n,1}}{d t}(t) & \dots  & \frac{d p_{n,n}}{d t}(t)\\
\end{array}
\right),
\end{align}

can be shown to satisfy

\begin{align}
P'(t) = P(t) \cdot R
\end{align}

where $R = (r_{i,j})$ is an $n \times n$ {\em rate matrix} with the
property that each diagonal entry is $-1$ times the row sum

\begin{align}
r_{i,i} = - \sum_{j\ne i} r_{i,j}.
\end{align}

Define the {\em population occupancy} distribution
$p(t) = (p_1(t),\dots,p_n(t))$ by

\begin{align}
\label{eq:hermes:markovProcessPopulationFreq}
p_i(t) = Pr[q(t) = i] = \sum_{k=1}^n\pi_k p_{k,i}(t)
\end{align}

where $q(t)$ denotes the state of the Markov process
at (continuous) time $t$.

In the case of macromolecular folding, where Markov process states
are molecular conformations and conformational energies are available,
it is typical to define the rate matrix
$R = (r_{x,y})$ as follows:

\begin{align}
\label{eq:hermes:eqXitionRate}
r_{x,y} =
\begin{cases}
\min\left( 1, \boltzYmX \right)
& \text{if $y \in N_x$} \\
-\sum_{k \in N_x} p_{x,k} & \text{if $x=y$} \\
0 & \text{if $y \not\in N_x$, and $x \ne y$} \\
\end{cases}
\end{align}

The {\em master equation} is defined by the matrix differential equation

\begin{align}
\label{eq:hermes:masterEquationMatrix}
\frac{d p(t)}{dt} = p(t) \cdot R
\end{align}

or equivalently, for all $1 \leq x \leq n$,

\begin{align}
\label{eq:hermes:masterEquation}
\frac{d p_x(t)}{dt} = \sum_{y=1}^n\, (p_y(t) \cdot r_{y,x} - p_x(t) \cdot
r_{x,y})\; = \sum_{x \ne y}\, (p_y(t) \cdot r_{y,x} - p_x(t) \cdot r_{x,y}).
\end{align}

As in the case of Markov chains,
\statDist is defined to be
the stationary distribution if
$\pSt \cdot P(t)= \pSt$; i.e. $\pSt_k = Pr[ q(0) = k ]$ implies that
$Pr[ q(t) = k] = \pSt_k$ for all $t \in \mathbb{R}$ and
$1 \leq k \leq n$.
Define the {\em equilibrium distribution} \statDist
to be the unique solution for $p(t)$, when the master equation
(\ref{eq:hermes:masterEquation}) is set to equal zero; i.e.

\begin{align}
\label{eq:hermes:equilibriumTime}
\sum_{x \ne y} \pSt_x \cdot r_{x,y} = \sum_{x \ne y} \pSt_y \cdot r_{y,x}.
\end{align}

If the equilibrium distribution exists, then necessarily it is equal to
the stationary distribution. A Markov process is said
to satisfy detailed balance if $\pSt_x \cdot r_{x,y} = \pSt_y \cdot r_{y,x}$,
for all $1 \leq x,y \leq n$, where the rate matrix $R = (r_{x,y})$.

The rate equation $R$ for is usually defined
as in (\ref{eq:hermes:eqXitionRate}) for
Markov processes which model macromolecular folding,
hence it is easy to see that
such Markov processes satisfy detailed balance and moreover that
the equilibrium distribution is the
Boltzmann distribution; i.e.  $\pSt_x = \boltzf{x}$ for
all $1 \leq x \leq n$.
Since detailed balance ensures that the eigenvalues of the rate matrix $R$ are
real, one can solve the matrix differential equation
(\ref{eq:hermes:masterEquation}) by diagonalizing the rate matrix, and thus obtain
the solution

\begin{align}
\label{eq:hermes:populationOccupancyEquation}
p(t) = \sum_{k=1}^n c_k {\bf v}_k\, e^{\lambda_k t}
\end{align}

where $p(t) = (p_1(t),\dots,p_n(t))$, and the
values $c_k$ are determined by the initial population occupancy
distribution $p(0)$ at time \DIFdelbegin \DIFdel{0. }\DIFdelend \DIFaddbegin \DIFadd{$0$. }\DIFaddend Here
${\bf v}_k$ denotes the $k$th eigenvector and
$\lambda_k$ the $k$th eigenvalue. In particular,
\DIFdelbegin \DIFdel{$c_k = \left(p(0) \cdot T^{-1})\right)_k$}\DIFdelend \DIFaddbegin \DIFadd{$c_k = \left(p(0) \cdot T^{-1}\right)_k$}\DIFaddend , where the $j$th row of $T$
is the $j$th left eigenvector of $R$,
and $p(0)$ is the population occupancy distribution
at time $t=0$. If the eigenvalues are labeled in decreasing
order, then $\lambda_1 \geq \lambda_2 \geq \cdots \geq \lambda_n$,
and the largest eigenvalue $\lambda_1=0$ has eigenvector \pSt, corresponding
to the equilibrium population occupancy distribution, which in this case
is the Boltzmann distribution. The
remaining $n-1$ eigenvalues are negative, and their corresponding eigenvectors
correspond to nonequilibrium kinetic {\em relaxation modes}.

In our software \hermes, we prefer to work with column vectors and
right eigenvectors due to our usage of the GSL scientific computing library,
and so the population occupancy frequency
$p(t)$ is defined to be the column vector
$p(t)=(p_1(t),\dots,p_n(t))^{\text T}$. Let
$\mathbbP = (p_1(0),\dots,p_n(0))^{\text T}$ be the column vector of initial
population occupancy probabilities,
$\mathbbt^1,\dots,\mathbbt^n$ be the right eigenvectors and
$\lambda_1,\dots,\lambda_n$ be the corresponding right eigenvalues of the
transpose $R^{\text T}$ of the rate matrix.
Letting $T$ be the $n\times n$ matrix, whose columns are
$\mathbbt^1,\dots,\mathbbt^n$, using standard matrix algebra
\citep{matrixtheory}, it can be shown that

\begin{align}
\label{eq:hermes:solutionMasterEquation}
p(t) = \sum_{j=1}^n\; (T^{-1}
\mathbbP(0))_j\, \mathbbt^j\, e^{\lambda_j t}.
\end{align}

or for a single state $i$,

\begin{align}
\label{eq:hermes:solutionMasterEquationForSingleState}
p_i(t) = \sum_{j=1}^n \sum_{k=1}^n\, \left(
T_{ij}^{\vphantom{-1}}\, T_{jk}^{-1}\, \mathbbP_k(0)
\right)_j e^{\lambda_j t}.
\end{align}

If we assume that the initial population starts in a single state $i'$,
i.e. $\mathbbP_{i'}(0) = 1$, equation
\ref{eq:hermes:solutionMasterEquationForSingleState} can be simplified to

\begin{align}
\label{eq:hermes:solutionMasterEquationForSingleStateFixedStartPop}
p_i(t) = \sum_{j=1}^n\, T_{ij}^{\vphantom{-1}}\, T_{ji'}^{-1}\, e^{\lambda_j t}.
\end{align}

The {\em \eqt} can be directly computed by using a
nonlinear solver to solve for $t$ in

\begin{align}
\label{eq:hermes:solutionEquilibriumTime}
\pSt = \sum_{j=1}^n\; (T^{-1}
\mathbbP(0))_j\, \mathbbt^j\, e^{\lambda_j t}
\end{align}

where $\statDist^{\text T}$ and $\pSt_k = \frac{\boltzf{k}}{\fullZ}$.
However, we have found it more
expedient to compute the {\em \eqt} as the smallest $t_0$,
such that for $t \in \{t_0+1,t_0+2,t_0+3,t_0+4\}$, the absolute difference
$|p(t)[\xEnd] - p(t_0)[\xEnd]| < \epsilon$, for $\epsilon =
10^{-4}$, where \xEnd is the target RNA structure (usually taken
to be the \mfes, though this is not necessary for
the software). We provide a more detailed explanation of the reasoning behind
this decision in \Secref{subsubsec:hermes:eqestimate}.

In \citep{gillespiestochasticsimulation1} Gillespie described
a very influential algorithm to simulate a finite Markov process.
The pseudocode, is given in Algorithm \ref{fig:hermes:gillespie}.
Though Gillespie's original application was for chemical kinetics,
Flamm et al. adapted the method for the kinetics of RNA secondary structure
folding, as implemented in \kinfold \citep{flammphd,flamm}.
\medskip

\begin{figure}[!ht]
\centering
\hrule \rule[0ex]{0pt}{0pt}
\begin{center}
{\large Pseudocode for the Gillespie algorithm} \\
\end{center}
\begin{tabular*}{\textwidth}{ll}
{\sc Purpose:} & Compute continuous time $t$ to move from
\xStart to \xEnd, with a limit of \tMax \rule[-1.5ex]{0pt}{0pt} \\
{\sc Input:} & Starting state \xStart, target state \xEnd and
upper bound on time \tMax \rule[-1.5ex]{0pt}{0pt} \\
{\sc Output:} & Continuous time $t$ taken to reach \xEnd from \xStart
\rule[-1.75em]{0pt}{0pt} \\
\hline \rule[0ex]{0pt}{0pt}
\end{tabular*}
\begin{algorithmic}[1]
\Function{Gillespie}{\xStart, \xEnd, \tMax}
\State $t \gets 0$
\State $x \gets \xStart$
\While{$x \neq \xEnd$ AND $t < \tMax$}
\State $\Phi \gets 0$
\Comment{$\Phi$ is the {\em flux} out of $x$}
\For{$y \in N_x$}
\State $r_{x,y} \gets \min\left(1,\boltzYmX\right)$
\State $\Phi \gets \Phi + r_{x,y}$
\EndFor
\For{$y \in N_x$}
\State $r_{x,y} \gets \frac{r_{x,y}}{\Phi}$
\EndFor
\State $z_1 \in (0,1)$
\State $t \gets t + (-\frac{1}{\Phi} \ln(z_1))$
\Comment{Update $t$ to reflect time spent in state $x$}
\State $z_2 \in (0,1)$
\State $s \gets 0$
\For{$y \in N_x$}
\Comment{Use roulette wheel to select new state for $x$}
\State $s \gets s + r_{x,y}$
\If{$z_2 \leq s$}
\State $x \gets y$
\State \textbf{break}
\EndIf
\EndFor
\EndWhile
\State \textbf{return} $t$
\Comment{Return continuous time $t$ to travel from \xStart to \xEnd}
\EndFunction
\rule[-0.35ex]{0pt}{0pt}
\end{algorithmic}
\caption[Pseudocode for the Gillespie algorithm]{The function {\sc Gillespie} implements a continuous time simulation of
folding trajectories for a Markov process.}
\label{fig:hermes:gillespie}
\rule[0ex]{0pt}{1.5em} \hrule
\end{figure}

\section{Software within the \hermes suite}
\label{sec:hermes:layout}
The \hermes software package was developed on the Macintosh OS X
operating system (\DIFdelbegin \DIFdel{10.9.2--10.11}\DIFdelend \DIFaddbegin \DIFadd{$10.9.2$--$10.11$}\DIFaddend ) and should work with any Unix-like platform
(Ubuntu, Debian, and CentOS were tested). We make the source code freely
available under the MIT License in two locations.
Our lab hosts the latest stable version of the code at
\url{http://bioinformatics.bc.edu/clotelab/Hermes} and a fully
version-controlled copy at \url{https://github.com/evansenter/hermes}.
The data and figures presented in this article were generated with the
source code hosted at the first URL, and we make no guarantee as to
the stability of development branches in our Git repository.

External dependencies for the software include a C [resp. C++]
compiler supporting the \DIFdelbegin \DIFdel{GNU99 }\DIFdelend \DIFaddbegin \DIFadd{GNU$99$ }\DIFaddend language specification [resp. C++\DIFdelbegin \DIFdel{98}\DIFdelend \DIFaddbegin \DIFadd{$98$}\DIFaddend ],
FFTW implementation of \fft \citep{fftw05} ($\geq
3.3.4$) \url{http://www.fftw.org/}, Gnu Scientific Library GSL ($\geq
1.15$) \url{http://www.gnu.org/software/gsl/}, Vienna RNA Package
\citep{lorenz.amb11} ($\geq 2.0.7$) \url{http://www.viennarna.at}, and
any corresponding sub-packages included with the aforementioned
software. For a more detailed explanation of both external
dependencies and installation instructions, refer to the
`DOCS.pdf' file at the web site
outlining the configuration and compilation process for the \hermes
suite.

\hermes is organized into three independent directories:
\begin{inparaenum}[\em 1\upshape)]
\item \ffttwo; \item \rnamfpt; and \item \rnaeq (see
\Figref{hermes:organizationHermes}).
\end{inparaenum}
These packages compile into both
standalone executables and archive files. The archives provide an
API which allow the development of novel applications using source
from across the \hermes package without having to copy-and-paste
relevant functions. We provide two such examples of this in the
\ms{ext} subdirectory: \fftmfpt and \ffteq. These applications
are simple C++ wrappers that use functions from \ffttwo, \rnamfpt and
\rnaeq to replicate a pipeline of executable calls without having to
deal with intermediary data transformation, I/O between calls or
slow-down due to a scripting language wrapper such as Python, Perl or R.
We believe
the API that \hermes provides via its archives to be one of the exciting
aspects of this software for other developers---the C++ sourcecode comprising
\fftmfpt is only \DIFdelbegin \DIFdel{56 }\DIFdelend \DIFaddbegin \DIFadd{$56$ }\DIFaddend lines of code (LOC) and \ffteq is only \DIFdelbegin \DIFdel{58 }\DIFdelend \DIFaddbegin \DIFadd{$58$ }\DIFaddend LOC.

\begin{figure}[!ht]
\centering
\includegraphics[width=.9\textwidth]{Figures/Hermes/softwareOrg.pdf}
\caption[Overall organization of the \hermes software suite]{Overall
organization of
\hermes. \ffttwo, \rnamfpt, and \rnaeq are three distinct software
packages we have developed, which compile into both standalone
executables and archive files, providing an API that allow novel
applications development using source from each of the packages,
without having to copy-and-paste relevant functions. The applications
\fftmfpt and \ffteq are C++ wrappers that use data structures and functions from
\ffttwo, \rnamfpt and \rnaeq. \fftmfpt computes the
\mfpt (MFPT) for an RNA secondary structure to fold from an
initial structure, such as the empty structure or a given metastable
structure, into a target structure, such as the \mfe
(MFE) structure or possibly the Rfam \citep{Gardner.nar11} consensus
structure. \ffteq uses spectral decomposition to compute the \eqt and the
fraction of the population of RNA structures that are
equal to a given target structure, as a function of time.
}
\label{fig:hermes:organizationHermes}
\end{figure}

\subsection{Exact \mfpt with \rnamfpt}
\label{subsec:hermes:rnamfpt}

\rnamfpt computes the \mfpt (MFPT), sometimes
referred to as the \hit of a Markov chain, by using
matrix inversion \citep{meyermfpt}---see
\Secref{subsec:hermes:mfpt}. The program takes as input a
comma separated value (CSV) file containing the non-zero positions and
values of a \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend probability grid; i.e. a CSV format file having columns {\em i},
{\em j}, and {\em p}. The first two columns, {\em i} and {\em j}
correspond to the \DIFdelbegin \DIFdel{0-indexed }\DIFdelend \DIFaddbegin \DIFadd{$0$-indexed }\DIFaddend row-ordered position in the probability grid, and
the final column {\em p} is the stationary probability $p_{i,j}$.
From this input, the probability transition matrix is constructed using
\eqnref{hermes:xitionProbFfttwoWithoutHastings} and the
\mfpt is computed by matrix inversion.

Given RNA sequence \seq and secondary structure $x$,
let $N(x)$ denote the set of neighboring secondary structures of
\seq, whose \bpd with $x$ is one. Define the Markov
chain $\mathcal{M}(\seq)=(\str,P)$, where \str denotes the
set of all secondary structures of \seq, $\pSt(x)$ is the stationary
probability of structure $x$, defined by $\pSt(x)=\boltzf{x}/\fullZ$,
$\fullZ=\sum_{x} \boltzf{x}$, and the transition probability matrix $P
= (p_{x,y})$ is defined either with or without the Hastings
modification as follows.

{\em With the Hastings modification},

\begin{align}
\label{eq:hermes:xitionProbWithHastings}
p_{x,y}=
\begin{cases}
\frac{1}{|N(x,y)|} \cdot \min\left(1, \frac{\pSt(y)}{\pSt(x)} \cdot
\frac{N(x)}{N(y)}\right) & \text{if $y \in N(x)$} \\
0 &\text{if $x \ne y, y \not\in N(x)$} \\
1 - \sum_{k \in N(x)} p_{x,k} & \text{if $x=y$} \\
\end{cases}
\end{align}

{\em Without the Hastings modification},

\begin{align}
\label{eq:hermes:xitionProbWithoutHastings}
p_{x,y}=
\begin{cases}
\frac{1}{|N(x,y)|} \cdot \min\left(1, \frac{\pSt(y)}{\pSt(x)}\right)
& \text{if $y \in N(x)$} \\
0 &\text{if $x \ne y, y \not\in N(x)$} \\
1 - \sum_{k \in N(x)} p_{x,k} & \text{if $x=y$} \\
\end{cases}
\end{align}

The exact value of \mfpt (MFPT) can be computed as
follows. Let $x_0$ [resp. $x_{\infty}$] denote the empty structure
[resp. MFE structure] for sequence \seq (here we have implicitly
identified integer indices with secondary structures). Let
$\widetilde{P}_{x_{\infty}}$ be the matrix obtained from $P$ by
removal of the row and column with index $x_{\infty}$, and $I$ denote
the $(n-1)\times(n-1)$ identity matrix, where $n=|\str|$ is the
number of secondary structures of \seq. Let $e$ denote the vector of
size $n-1$, each of whose coordinates is \DIFdelbegin \DIFdel{1. }\DIFdelend \DIFaddbegin \DIFadd{$1$. }\DIFaddend It is well-known
\citep{meyermfpt} that for each $k\ne x_{\infty}$, the $k$th coordinate
of the vector $(I - \widetilde{P}_{x_{\infty}})^{-1} \cdot {\bf e}$ is
exactly equal to the \mfpt from the structure with
index $k$ to the target structure $x_{\infty}$. In particular, the
MFPT from the empty structure to the MFE structure is computable by
applying matrix inversion using a computational library such as the GSL.

Since this computation of the
\mfpt is mathematically exact, we consider that MFPT
to be the {\em gold standard} value for RNA folding kinetics.

Because this program was designed with the original intent
of handling \DIFdelbegin \DIFdel{2D-probability }\DIFdelend \DIFaddbegin \twoD \DIFadd{probability }\DIFaddend grids, all vertexes are uniquely identified
by index tuples (which conceptually correspond to positioning in a \DIFdelbegin \DIFdel{2D
}\DIFdelend \DIFaddbegin \twoD
\DIFaddend array). However, it is trivial to use this code with both
1D-probability grids such as those produced by \fftbor
\citep{senter.po12} or arbitrary transition matrices without any change
to the underlying implementation. The software additionally provides
many options for defining the format of the input as well as the
structure of the graph underlying the
Markov chain.

Due to our interest in efficiently estimating MFPT using coarse-grained \DIFdelbegin \DIFdel{2D
}\DIFdelend \DIFaddbegin \twoD
\DIFaddend energy landscapes, the default input for \rnamfpt (and \rnaeq) is a \DIFdelbegin \DIFdel{2D
}\DIFdelend \DIFaddbegin \twoD
\DIFaddend probability grid, as described above. We additionally allow as alternative input
an energy landscape (whereby transition probabilities are defined as in
equation \ref{eq:hermes:mfptXitionProb}).
Some of the options to influence the underlying transition matrix
include the option to force a fully
connected graph (useful in cases where there is no non-zero path
between the start / end state) or to enforce detailed balance.
Finally, \rnamfpt also accepts as input the probability transition matrix,
a stochastic matrix with row sums equal to 1, and computes the mean
first passage time for the corresponding Markov chain.

\subsection{Approximate \mfpt with \fftmfpt}
\label{subsec:hermes:fftmfpt}

\fftmfpt approximates the \mfpt of a given RNA
sequence folding from input structure \strA to \strB, by
computing {\em exactly} the \mfpt from state $(0,d_0)$ to state
$(d_0,0)$ in the \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend probability grid obtained from running
\ffttwo. Here, $d_0$ is the \bpd between structures
\strAB, and the MFPT is computed for the Markov chain, whose states are
the non-empty \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend probability grid points, and whose transition
probabilities are defined by $p_{(x,y),(x',y')} =
\frac{P(x',y')}{P(x,y)}$.

More formally, given an RNA sequence \seq, starting structure \strA, and
target structure \strB, we proceed in the following fashion. Let \dBP{\strA}{\strB} denote the \bpd between structures
\strAB. Then $\bfZ{}{x,y} = \sum_\str \boltzf{\str}$, where the sum is over
structures \str, such that $\dBP{\strA}{\str} = x$, and
$\dBP{\strB}{\str} = y$; as
well the partition function $\fullZ = \sum_\str \boltzf{\str}$, where the sum
is over all secondary structures \str of the sequence \seq.

Let $d_0= \dBP{\strA}{\strB}$, the \bpd between initial structure
\strA and target structure \strB.  Let $n$ denote the length of sequence \seq.
Define the Markov chain $\mathcal{M}_1(\seq) = (Q_1,P_1)$, where

\begin{align}
\label{eq:hermes:markovChainDef}
\begin{split}
Q_1 =\;& \{ (x,y) : 0 \leq x,y \leq n, (x+y \bmod 2)
= (d_0 \bmod 2), \\
& (d_0 \leq x+y), (x \leq d_0 + y), (y \leq d_0 + x) \}.
\end{split}
\end{align}

For reference, we say that the {\em parity condition} holds for
$(x,y)$ if

\begin{align}
\label{eq:hermes:parityCondition} (x+y \bmod 2) = (d_0 \bmod 2).
\end{align}

We say that the {\em triangle inequality} holds for $(x,y)$ if

\begin{align}
\label{eq:hermes:triangleInequality} (d_0 \leq x+y), (x \leq d_0 + y), (y
\leq d_0 + x)
\end{align}

Since we allow transitions between secondary structures that differ by
exactly one base pair, Markov chain transitions are allowed to occur
only between states $(x,y),(u,v) \in Q_1$, such that $u=x\pm 1$, $v =
y \pm 1$. However, we have found that for some RNA sequences, there is
no non-zero probability path from $(0,d_0)$ to $(d_0,0)$
(corresponding to a folding pathway from structure \strA to
\strB). Since \ffttwo computes \DIFdelbegin %DIFDELCMD < {\em %%%
\DIFdel{probabilities }%DIFDELCMD < } %%%
\DIFdelend \DIFaddbegin \DIFadd{probabilities }\DIFaddend $p(x,y)$ by
polynomial interpolation using the \fft, any
probability less than $10^{-8}$ is set to \DIFdelbegin \DIFdel{0. }\DIFdelend \DIFaddbegin \DIFadd{$0$. }\DIFaddend Also with \rnatwofold,
it may arise that there is no non-zero probability path
from structure $(0,d_0)$ to $(d_0,0)$.

To address this situation, we proceed as follows. Let $\epsilon =
10^{-8}$ and for all $(x,y) \in Q_1$, modify probabilities $p(x,y)$ by

\begin{align}
\label{eq:hermes:mfptNormalize} p(x,y) = \frac{p(x,y) +
\epsilon/|Q_1|}{1+\epsilon }.
\end{align}

This operation corresponds to adding the negligeable value of
$\epsilon=10^{-8}$ to the total probability, thus ensuring that there
are paths of non-zero probability between any two states. After this
$\epsilon$-modification and renormalization, {\em when using the
Hastings modification}, the transition probabilities $P((u,v)|(x,y))$
are given by

\begin{align}
\label{eq:hermes:xitionProbFfttwoWithHastings}
P((u,v)|(x,y))=
\begin{cases}
\frac{1}{|N(x,y)|} \cdot \min\left(1, \frac{p(u,v)}{p(x,y)} \cdot
\frac{N(x,y)}{N(u,v)}\right) & \text{if $(u,v) \in N(x,y)$} \\
0 &\text{if $(u,v) \ne (x,y), (u,v) \not\in N(x,y)$} \\
1 - \sum_{(k,\ell) \in N(x,y)} P((k,\ell)|(x,y)) & \text{if $(u,v)=(x,y)$} \\
\end{cases}
\end{align}

Here the set $N(x,y)$ of adjacent neighbors is defined by $N(x,y) = \{
(u,v) \in Q_1 : u = x \pm 1, v = y \pm 1 \}$, and the stationary
probability $p(x,y)$ is obtained from \ffttwo.

{\em Without the Hastings modification}, the transition probabilities
$P((u,v)|(x,y))$ are instead given by

\begin{align}
\label{eq:hermes:xitionProbFfttwoWithoutHastings}
P((u,v)|(x,y))=
\begin{cases}
\frac{1}{|N(x,y)|} \cdot \min\left(1, \frac{p(u,v)}{p(x,y)}\right)
& \text{if $(u,v) \in N(x,y)$} \\
0 &\text{if $(u,v) \ne (x,y), (u,v) \not\in N(x,y)$} \\
1 - \sum_{(k,\ell) \in N(x,y)} P((k,\ell)|(x,y)) & \text{if $(u,v)=(x,y)$} \\
\end{cases}
\end{align}

As we report in \Secref{subsec:hermes:corr},
given an RNA
sequence \seq, if \strA is the empty structure and \strB the MFE
structure of \seq, then \fftmfpt output is well correlated with the
exact MFPT in folding the empty structure to the MFE structure, where
transitions between structures involve the addition or removal of a
single base pair.

\subsection{Exact \eqt with \rnaeq}
\label{subsec:hermes:rnaeq}

\rnaeq computes the population proportion of a user-provided structure
over arbitrary time units. Like \rnamfpt, this program takes as input a
comma separated value (CSV) file containing the non-zero positions and
values of a \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend probability grid. From this input a rate matrix
is constructed for the underlying Markov process. Alternatively,
\rnaeq can accept as input an arbitrary rate matrix. Performing
spectral decomposition of the column-ordered rate matrix that
underlies the corresponding Markov process, \rnaeq computes either
the \eqt or population occupancy frequencies.

Define the continuous Markov process
$\mathcal{M}=(\str,R)$, where \str is defined as in
\Secref{subsec:hermes:rnamfpt}, $R$ is the {\em rate matrix} defined by

\begin{align}
\label{eq:hermes:xitionRateWithoutHastings}
k_{x,y}=
\begin{cases}
\min\left(1, \frac{\pSt(y)}{\pSt(x)}\right)
& \text{if $y \in N(x)$} \\
0 &\text{if $x \ne y, y \not\in N(x)$} \\
- \sum_{k \in N(x)} p_{x,k} & \text{if $x=y$} \\
\end{cases}
\end{align}

Clearly the rate matrix satisfies \DIFdelbegin %DIFDELCMD < {\em %%%
\DIFdel{detailed balance}%DIFDELCMD < }%%%
\DIFdelend \DIFaddbegin \DIFadd{detailed balance}\DIFaddend ; i.e. $\pSt(x)
\cdot k_{x,y} = \pSt(y) \cdot k_{y,x}$ for all distinct $x,y \in
\str$. In fact, the rate matrix for Markov processes is usually
defined as above, precisely to ensure detailed balance, which then
implies that all eigenvalues of the rate matrix are real, thus
permitting explicit solution of the population occupancy frequency for
all states. We additionally considered a Hastings correction
for the rate matrix, where $k_{x,y} = \min(1, \frac{\pSt(y)}{\pSt(x)}
\cdot \frac{N(x)}{N(y)})$. The correlation in
Table \ref{table:correlationHermes} for equilibrium time computed from this
modified rate matrix is somewhat better than without the Hastings
correction. However, the Hastings correction is never used for rate
matrices, hence we only consider the usual definition of rate matrix
given in \eqnref{hermes:xitionRateWithoutHastings}.

\rnaeq can additionall call the
Vienna RNA Package program \rnasub \citep{wuchty.b99}, with a
user-specified upper bound to the energy difference with the minimum
free energy. With this option, the rate matrix is
constructed for the Markov process, whose states consist of all
the structures returned by \rnasub, and the \eqt or
population occupancy frequencies are computed. Due to the time
and memory required for this option, we do not expect it to be used except for
small sequences.

\subsection{Approximate \eqt with \ffteq}
\label{subsec:hermes:ffteq}

\ffteq allows an investigator to efficiently estimate population
kinetics for a sequence folding between two arbitrary, but fixed,
structures. The transition rate matrix underlying the Markov process
necessary for eigendecomposition is derived from the \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend energy
landscape.
Indices $(x,y)$ in the rate matrix correspond to all strucutres having
\bpd $x$ [resp. $y$] to input structure \strA [resp. \strBresp],
which makes the graph size more tractable than structural
sampling with \rnasub using \rnaeq, even with constraints.

This method consists of computing the equilibrium time
from the master equation for the coarse-grain Markov process
$\mathcal{M}=(Q_1,R)$, where $Q_1$ is defined in
\eqnref{hermes:markovChainDef}, and the rate matrix
$R = (k((x,y),(u,v)))$ is defined by

\begin{align}
\label{eq:hermes:xitionRateFfttwoWithoutHastings}
k((u,v),(x,y))=
\begin{cases}
\min\left(1, \frac{p(u,v)}{p(x,y)}\right) & \text{if $(u,v) \in N(x,y)$} \\
0 &\text{if $(u,v) \ne (x,y), (u,v) \not\in N(x,y)$} \\
- \sum_{(k,\ell) \in N(x,y)} P((k,\ell)|(x,y)) & \text{if $(u,v)=(x,y)$} \\
\end{cases}
\end{align}

Equilibrium time is then computed for this Markov process using
\DIFdelbegin \DIFdel{equation \ref{eq:hermes:populationOccupancyEquation}}\DIFdelend \DIFaddbegin \eqnref{hermes:populationOccupancyEquation}\DIFaddend .

\subsubsection{Population occupancy curves with \ffteq}
\label{subsubsec:hermes:popcurve}

Population occupancy curves provide valuable visual insight into the folding
kinetics of RNA molecules.
\DIFdelbegin \DIFdel{Equation \ref{eq:hermes:solutionMasterEquationForSingleState} describes a
}\DIFdelend \DIFaddbegin \DIFadd{A }\DIFaddend computation for the probability distribution
\DIFdelbegin \DIFdel{$p(t)=(p_1(t),\dots,p_n(t))^{\text T}$ }\DIFdelend \DIFaddbegin \DIFadd{$p(t) = (p_1(t),\dots,p_n(t))^{\text T}$ }\DIFaddend of all states $1 \leq x \leq n$ at
time $t$ \DIFaddbegin \DIFadd{is described in }\eqnref{hermes:solutionMasterEquationForSingleState}\DIFaddend .
Rather than plot $p(t)$ as a function of $t$, we find it more
visually impactful to plot the log-logistic function $p(\log_{10}(t))$,
yielding the curvess seen in \Figref{hermes:populationOccupancySplicedLeader}.

\begin{figure}[!ht]
\centering
\includegraphics[width=0.7\textwidth]{Figures/Hermes/populationOccupancyCurves.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption[Population occupancy curves computed with \ffteq for the
%DIFDELCMD < 56 nt conformational switch {\em L. collosoma} spliced leader RNA]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Population occupancy curves computed with \ffteq for the
$56$ nt conformational switch {\em L. collosoma} spliced leader RNA]{\DIFaddendFL 
Population occupancy curves computed with \ffteq for the
\DIFdelbeginFL \DIFdelFL{56 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$56$ }\DIFaddendFL nt conformational switch {\em L. collosoma} spliced leader RNA,
with sequence \ms{AACUAAAACAAUUUUUGAAGAACAGUUUCUGUACUUCAUUGGUAUGUAGA}
GACUUC. The dot bracket format for the MFE structure, as computed by
version \DIFdelbeginFL \DIFdelFL{2.1.7 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$2.1.7$ }\DIFaddendFL of \ms{RNAfold -d0}, is
\DIFdelbeginFL %DIFDELCMD < \ms{.......................((((((((((((.....)))))..)))))))..} %%%
\DIFdelendFL \DIFaddbeginFL \strfont{.......................((((((((((((.....)))))..)))))))..} \DIFaddendFL with free
energy \DIFdelbeginFL \DIFdelFL{$-8.6$ kcal/mol}\DIFdelendFL \DIFaddbeginFL \kmolk{-8.6}\DIFaddendFL , while that of the the alternate suboptimal
structure is
\DIFdelbeginFL %DIFDELCMD < \ms{..((...((((((..(((((.((((...)))).)))))..))).)))..)).....} %%%
\DIFdelendFL \DIFaddbeginFL \strfont{..((...((((((..(((((.((((...)))).)))))..))).)))..)).....} \DIFaddendFL with free
energy \DIFdelbeginFL \DIFdelFL{$-7.5$ kcal/mol}\DIFdelendFL \DIFaddbeginFL \kmolk{-7.5}\DIFaddendFL . In the case of the MFE structure, the
equilibrium occupancy $P(t_{\infty})$, which \hermes approximates
as $0.1789$ should equal the Boltzmann probability $0.1791$,
since the MFE structure is the only structure at distance $x_0$ [resp.
$y_0$] from the reference structures \strA (empty
structure) [resp. \strB (MFE structure)]. As well, if there are few
other low energy structures at the same \bpd $x_1$
[resp. $y_1$] from \strA [resp. \strBresp] as that of the alternate suboptimal
structure, then we expect that the occupancy probability $0.0300$
for the $(x_1,y_1)$ be approximately the Boltzmann probability
$0.0301$ of the alternate structure.}
\label{fig:hermes:populationOccupancySplicedLeader}
\end{figure}

\subsubsection{Approximating \eqt from occupancy curves}
\label{subsubsec:hermes:eqestimate}

The computation of an \eqt value from the eigendecomposition of the rate matrix
is a rather thorny issue. While in principle a nonlinear solver can be used to
compute the \eqt $t$ (from equation \ref{eq:hermes:solutionEquilibriumTime}),
in practice due to what we believe to be numeric instability issues the approach
was intractable. For this reason, we estimate the \eqt from the population
occupancy curve, an approach also taken by \treekin
\citep{wolfingerstadler:kinetics}. Our first
approximation of \eqt involved using a sliding window approach, where we
compute the smallest $t_0$,
such that for $t \in \{t_0+\tD,t_0+2\tD,t_0+3\tD,\dots,t_0+(w-1)\cdot\tD\}$, the absolute difference
$|p(t)[\xEnd] - p(t_0)[\xEnd]| < \epsilon$, for $\epsilon =
10^{-4}$ and step size \tD across a window of length $w$. In systems where there are local energy minima, this approach has the
drawback of prematurely indicating equilibrium, as seen in
\Figref{hermes:eqEstFromRnaEq}. To address this concern, we introduce the
concept of $\delta,\epsilon$ equilibrium.

Given a starting state $i'$, target state $i$, and user definable parameters
$w = 5, \tD = 10^{-3}, \delta = 10^{-3}$ and $\epsilon = 10^{-4}$, we compute the estimated
equilirium using the following algorithm.
\medskip

\begin{figure}[!ht]
\hrule \rule[0ex]{0pt}{0pt}
\begin{center}
{\large Pseudocode for estimating equilibrium time} \\
\end{center}
\begin{tabular*}{\textwidth}{ll}
{\sc Purpose:} & Estimate time $\log_{10}(t)$ at which state \xStart appears
within $\epsilon$ of equilibrium \rule[-1.5ex]{0pt}{0pt} \\
{\sc Input:} & States \xStart, \xEnd; \tMin, \tMax;
$w = 5, \tD = 10^{-3}, \delta = 10^{-3}, \epsilon = 10^{-4}$ \rule[-1.5ex]{0pt}{0pt} \\
{\sc Output:} & Time $\log_{10}(t)$ at which $p_{\xEnd}(10^t)$ varies by no
more than $\epsilon$ within a \\ & window of size $w$, having step size \tD
\rule[-1.75em]{0pt}{0pt} \\
\hline \rule[0ex]{0pt}{0pt}
\end{tabular*}
\begin{algorithmic}[1]
\Function{EstimateEquilibrium}{\xStart, \xEnd, \tMin, \tMax, \tD, $\delta$, $\epsilon$, $w$}
\State $t \gets \Call{SoftBound}{\xEnd, \tMin, \tMax, \tD, \delta, s}$
\If{$\Call{WindowEq}{\xEnd, t, \tD, w, \epsilon}$}
\While{$t > \tMin$ AND $\Call{WindowEq}{\xEnd, t, \tD, w, \epsilon}$}
\State $t \gets t - \tD$
\EndWhile
\EndIf
\While{$t < \tMax$ AND NOT $\Call{WindowEq}{\xEnd, t, \tD, w, \epsilon}$}
\State $t \gets t + \tD$
\EndWhile
\If{$t = \tMin$}
\Comment{$x$ doesn't leave equilibrium after \tMin}
\State \textbf{return} $-\infty$
\ElsIf{$t = \tMax$}
\Comment{$x$ doesn't reach equilibrium before \tMax}
\State \textbf{return} $\infty$
\Else
\State \textbf{return} $t$
\EndIf
\EndFunction
\algstore{eqEst}
\end{algorithmic}
\rule[0ex]{0pt}{1.5em} \hrule
\end{figure}
\clearpage

\begin{figure}[!ht]
\begin{algorithmic}[1]
\algrestore{eqEst}
\Function{SoftBound}{$i$, \tMin, \tMax, \tD, $\delta$, $s$}
\State $i_{\infty} \gets p_i(10^{\tMax})$
\DIFdelbeginFL %DIFDELCMD < \Comment{From equation \ref{eq:hermes:solutionMasterEquationForSingleStateFixedStartPop}}
%DIFDELCMD < %%%
\DIFdelendFL \DIFaddbeginFL \Comment{From \eqnref{hermes:solutionMasterEquationForSingleStateFixedStartPop}}
\DIFaddendFL \State $t \gets \tMin + \frac{\tMax - \tMin}{2}$
\State $t' \gets \tMax$
\While{$|t - t'| > \tD$}
\Comment{Binary search for time $t$ within $\delta$ of \tMax}
\State $t_{\text{temp}} \gets t$
\If{$|p_i(10^t) - p_i(10^{t'})| > \delta$}
\State $t \gets t + s * \frac{|t - t'|}{2}$
\DIFdelbeginFL %DIFDELCMD < \Comment{$s$ is 1 [resp. $-1$] for the right [resp. left] bound}
%DIFDELCMD < %%%
\DIFdelendFL \DIFaddbeginFL \Comment{$s$ is $1$ [resp. $-1$] for the right [resp. left] bound}
\DIFaddendFL \Else
\State $t \gets t - s * \frac{|t - t'|}{2}$
\EndIf
\State $t' \gets t_{\text{temp}}$
\EndWhile
\State \textbf{return} $t$
\EndFunction
\Function{WindowEq}{$x$, $t$, \tD, $w$, $\epsilon$}
\Comment{Is $x$ within $\epsilon$ over $[t, t+(w-1)\cdot\tD]$}
\For{$i \gets 1, w - 1$}
\If{$|p_x(10^{t+i\cdot\tD}) - p_x(10^t)| > \epsilon$}
\State \textbf{return} \textbf{false}
\EndIf
\EndFor
\State \textbf{return} \textbf{true}
\EndFunction
\rule[-0.35ex]{0pt}{0pt}
\end{algorithmic}
\caption[Pseudocode for estimating equilibrium time]{The function {\sc EstimateEquilibrium} computes the smallest $t_0 > t'$,
such that for $t \in \{t_0+\tD,t_0+2\tD,t_0+3\tD,\dots,t_0+(w-1)\cdot\tD\}$,
the absolute difference $|p(t)[\xEnd] - p(t_0)[\xEnd]| < \epsilon$ and
$|p(t')[\xEnd] - p(\tMax)[\xEnd]| \approx \delta|$. If $t'$ is already in
equilibrium, we relax the constraint that $t_0 > t'$ and instead find the
first $t_0 < t'$ that satisfies the equilibrium requirements within the window
$w$. {\sc SoftBound} is a helper function that uses binary search to find the
starting time $t'$ from which the window starts. {\sc WindowEq} returns a
boolean value if the state of interest $x$ varies by no more than $\epsilon$
across the window $w$ starting at time $t$.}
\label{fig:hermes:eqEst}
\rule[0ex]{0pt}{1.5em} \hrule
\end{figure}

In using the approach outlined in \Figref{hermes:eqEst}, the root-mean-square
deviation (RMSD)
between $p_{\xEnd}(t_0)$ and $\frac{\boltzf{\xEnd}}{\fullZ}$ is $0.0041$, where
RMSD is defined as

\begin{align}
\label{eq:hermes:rmsd}
\sqrt{\frac{\sum_{i=1}^{n} (\widehat{x_i} - x_i)^2}{n}}.
\end{align}

Visual inspection of the estimated equilibrium times
(\Figref{hermes:eqEstFromRnaEq}) are much better than the
na\"{\i}ve approach used by \treekin, which uses a simple sliding window for
all $x \in Q$.
We also allow defining the \eqt to be the
smallest $t_0$, such that for
$t \in \{t_0+\tD,t_0+2\tD,t_0+3\tD,\dots,t_0+(w-1)\cdot\tD\}$, the
absolute difference $|p(t)[x] - p(t_0)[x]| < \epsilon$ for all $x \in
Q$ using Algorithm \ref{fig:hermes:eqEst};
however, results suggest that this definition is inferior
to the consideration of a single single target state $i$,
perhaps due to numerical instability issues when
$Q$ is taken to be the set of all secondary structures for sequences in
the benchmarking set described later.

\begin{figure}[!ht]
\centering
\includegraphics[width=0.45\textwidth]{Figures/Hermes/eqEstFromRnaEqSeq146.pdf}
\quad
\includegraphics[width=0.45\textwidth]{Figures/Hermes/eqEstFromRnaEqSeq427.pdf}
\caption[Example of the differences in using a sliding window approach for estimating \eqt versus a more robust approach, outlined in Algorithm \ref{fig:hermes:eqEst}]{Example of the differences in using the simple sliding window
approach (green) versus the approach outlined in Algorithm \ref{fig:hermes:eqEst}
(blue). There are two representative sequences shown from the benchmarking
dataset described in \Secref{sec:hermes:benchdata}, sequence \#\DIFdelbeginFL \DIFdelFL{146
}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$146$
}\DIFaddendFL {\em (Left)} and \#\DIFdelbeginFL \DIFdelFL{427 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$427$ }\DIFaddendFL {\em (Right)}. Note that the $p(t)$ for the approach from
Algorithm \ref{fig:hermes:eqEst} is much closer to the computed Boltzmann
probability---the root-mean-square deviation (RMSD, equation
\ref{eq:hermes:rmsd}) across the dataset of \DIFdelbeginFL \DIFdelFL{1,000 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1,000$ }\DIFaddendFL sequences is $0.0041$
[resp. $0.0491$] for the approach described in Algorithm \ref{fig:hermes:eqEst}
[resp. simple sliding window].}
\label{fig:hermes:eqEstFromRnaEq}
\end{figure}

\section{Benchmarking data for computational comparison}
\label{sec:hermes:benchdata}

In this section, we describe a
benchmarking set of \DIFdelbegin \DIFdel{1,000 }\DIFdelend \DIFaddbegin \DIFadd{$1,000$ }\DIFaddend small RNAs used to benchmark the previously
described kinetics methods in a comparative study. To ensure that mean
first passage time can be computed from
$(I - P^{-}_{\xEnd})^{-1} \cdot {\bf e}$ by using matrix
inversion, that spectral decomposition of the rate matrix is possible,
and to ensure that \kinfold simulations would provide sufficient
statistics, we generated a collection of \DIFdelbegin \DIFdel{1,000 }\DIFdelend \DIFaddbegin \DIFadd{$1,000$ }\DIFaddend random RNA sequences of
length \DIFdelbegin \DIFdel{20 }\DIFdelend \DIFaddbegin \DIFadd{$20$ }\DIFaddend nt, each having expected compositional frequency of $1/4$
for A,C,G,U, and each having at most \DIFdelbegin \DIFdel{2,500 }\DIFdelend \DIFaddbegin \DIFadd{$2,500$ }\DIFaddend distinct secondary
structures, such that the \mfe is less than or equal to
\DIFdelbegin \DIFdel{$-5.5$ kcal/mol}\DIFdelend \DIFaddbegin \kmolk{-5.5}\DIFaddend .

For example, one of the \DIFdelbegin \DIFdel{1,000 }\DIFdelend \DIFaddbegin \DIFadd{$1,000$ }\DIFaddend sequences is \ms{ACGCGACGUGCACCGCACGU} with
\mfes \DIFdelbegin %DIFDELCMD < \ms{.....((((((...))))))} %%%
\DIFdelend \DIFaddbegin \strfont{.....((((((...))))))} \DIFaddend having free
energy of \DIFdelbegin \DIFdel{$-6.4$ kcal/mol}\DIFdelend \DIFaddbegin \kmolk{-6.4}\DIFaddend . Statistics for the free energies of the
\DIFdelbegin \DIFdel{2,453 }\DIFdelend \DIFaddbegin \DIFadd{$2,453$ }\DIFaddend secondary structures of this \DIFdelbegin \DIFdel{20-mer }\DIFdelend \DIFaddbegin \DIFadd{$20$-mer }\DIFaddend are the following: mean is
$10.695$, standary deviation is $4.804$, maximum is $25.00$, minimum
is $-6.40$. A histogram for the free energy of all secondary
structures of \ms{ACGCGACGUGCACCGCACGU} is depicted in
the left panel of \Figref{hermes:plmv}. The right panel of the
same figure depicts the \mfes of the
\DIFdelbegin \DIFdel{54 }\DIFdelend \DIFaddbegin \DIFadd{$54$ }\DIFaddend nt hammerhead type III ribozyme from Peach Latent Mosaic Viroid
(PLMVd), discussed later. This secondary structure is identical
to the consensus structure from Rfam \DIFdelbegin \DIFdel{11.0 }\DIFdelend \DIFaddbegin \DIFadd{$11.0$ }\DIFaddend \citep{Gardner.nar11}.

\begin{figure}[!ht]
\centering
\includegraphics[width=0.525\textwidth]{Figures/Hermes/plmvDist.pdf}
\quad
\includegraphics[width=0.375\textwidth]{Figures/Hermes/plmvStr.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption[{\em (Left)} Histogram of free energies of secondary structures of
%DIFDELCMD < an example 20-mer from the benchmarking dataset.
%DIFDELCMD < {\em (Right)} Minimum free energy structure of the 54 nt Peach Latent Mosaic
%DIFDELCMD < Viroid (PLMVd) AJ005312.1/282-335]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[{\em (Left)} Histogram of free energies of secondary structures of
an example $20$-mer from the benchmarking dataset.
{\em (Right)} Minimum free energy structure of the $54$ nt Peach Latent Mosaic
Viroid (PLMVd) AJ$005312.1$ $282$--$335$]{\DIFaddendFL 
{\em (Left)} Histogram of free energies of secondary structures of
\ms{ACGCGACGUGCACCGCACGU}, which range from $-6.5$ to \DIFdelbeginFL \DIFdelFL{$+25$ kcal/mol}\DIFdelendFL \DIFaddbeginFL \kmolk{+25}\DIFaddendFL , with
mean of \DIFdelbeginFL \DIFdelFL{$10.695$ kcal/mol}\DIFdelendFL \DIFaddbeginFL \kmolk{10.695}\DIFaddendFL .
{\em (Right)} Minimum free energy structure of the \DIFdelbeginFL \DIFdelFL{54 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$54$ }\DIFaddendFL nt Peach Latent Mosaic
Viroid (PLMVd) \DIFdelbeginFL \DIFdelFL{AJ005312.1/282-335}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{AJ$005312.1$ $282$--$335$}\DIFaddendFL , which is identical to the consensus
structure from Rfam \DIFdelbeginFL \DIFdelFL{11.0 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$11.0$ }\DIFaddendFL \citep{Gardner.nar11}. \rfold from
Vienna RNA Package \DIFdelbeginFL \DIFdelFL{2.1.7 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$2.1.7$ }\DIFaddendFL with energy parameters from the Turner \DIFdelbeginFL \DIFdelFL{1999
}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1999$
}\DIFaddendFL model were used, since the \mfes determined by
the more recent Turner \DIFdelbeginFL \DIFdelFL{2004 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$2004$ }\DIFaddendFL energy parameters
does {\em not} agree with the Rfam consensus structure---see
\citep{synthetichammerheads}. Positional entropy, a measure
of divergence in the base pairing status at each positions for the
low energy ensemble of structures, is indicated by color, using the
RNA Vienna Package utility script \ms{relplot.pl}.
}
\label{fig:hermes:plmv}
\end{figure}

\Figref{hermes:kinfoldMeanStdev} displays the mean
and standard deviation for \kinfold simulations of folding time
for each of the \DIFdelbegin \DIFdel{1,000 }\DIFdelend \DIFaddbegin \DIFadd{$1,000$ }\DIFaddend RNA sequences from our benchmarking data. For
each sequence, the mean and standard deviation of the time required to
fold the empty structure to the MFE structure were computed from
\DIFdelbegin \DIFdel{10,000 }\DIFdelend \DIFaddbegin \DIFadd{$10,000$ }\DIFaddend \kinfold runs, each run with an upper bound of $10^8$
Monte Carlo steps, thus ensuring that all simulations converged. The
sequences were then sorted by increasing folding time mean. Standard
deviation exceeded the mean in $83.9\%$ of the \DIFdelbegin \DIFdel{1,000 }\DIFdelend \DIFaddbegin \DIFadd{$1,000$ }\DIFaddend cases, indicating
the enormous variation between separate \kinfold runs, even for
\DIFdelbegin \DIFdel{20 }\DIFdelend \DIFaddbegin \DIFadd{$20$ }\DIFaddend nt RNA sequences having at most \DIFdelbegin \DIFdel{2,500 }\DIFdelend \DIFaddbegin \DIFadd{$2,500$ }\DIFaddend secondary structures.
Elementary considerations from statistics indicate that for our benchmarking set
of \DIFdelbegin \DIFdel{20-mers}\DIFdelend \DIFaddbegin \DIFadd{$20$-mers}\DIFaddend , the minimum sample size $n = \left( \frac{z_{\alpha/2}
\cdot \sigma}{E} \right)^2$ ranges from \DIFdelbegin \DIFdel{937,712 to 23,289,310
}\DIFdelend \DIFaddbegin \DIFadd{$937,712$ to $23,289,310$
}\DIFaddend to have a confidence level of \DIFdelbegin \DIFdel{95}\DIFdelend \DIFaddbegin \DIFadd{$95$}\DIFaddend \% that the average of $n$
\kinfold runs differs from the real folding time by at most \DIFdelbegin \DIFdel{100 }\DIFdelend \DIFaddbegin \DIFadd{$100$ }\DIFaddend steps.
In our opinion, \kinfold is an expertly crafted implementation of
Gillespie's algorithm for an event driven Monte Carlo simulation of
one-step RNA secondary structure folding. From the standpoint of
biophysics and physical chemistry, there is no more reliable
simulation method, except of course the exact computation of mean
first passage time using linear algebra. Nevertheless, the enormous
time required for reliable \kinfold estimations and the large
standard deviations observed point out the need for a faster method to
approximate folding time.

With this in mind, it is natural to turn to \DIFdelbegin %DIFDELCMD < {\em %%%
\DIFdelend coarse-grained \DIFdelbegin %DIFDELCMD < } %%%
\DIFdelend models,
as done by Wolfinger et al. \citep{wolfingerstadler:kinetics} and by
Tang et al. \citep{tang.jmb08}. The software of Tang et al. appears not
to be available. Concerning the method of Wolfinger et al. (called
\DIFdelbegin %DIFDELCMD < \ms{BarriersEq} %%%
\DIFdelend \DIFaddbegin \barrierseq \DIFaddend in our benchmarking), there is now a web server available,
which runs \rnasub \citep{wuchty.b99} to generate all secondary structures within
a user-specified energy range, then runs \barriers \citep{flammhofacker}
to generate basins of attraction around a user-specified number of locally
optimal structures, and then runs \treekin on the output of \barriers.
The program \treekin performs some of the same operations as \hermes,
by computing population occupation frequencies by spectral decomposition.
Nevertheless it would require a user to write scripts and perform
several manual steps, in order to determine the equilibrium time for
an input RNA sequence, with respect to the macrostate Markov process
of \citep{wolfingerstadler:kinetics}. In addition, because \barriers computes
basins of attraction by utilizing the output of \rnasub, estimating kinetics
for the refolding of an RNA molecule from the empty structure requires
exhaustive enumeration of all suboptimal structures having non-positive
free energy.

\begin{figure}[!ht]
\centering
\includegraphics[width=0.45\textwidth]{Figures/Hermes/kinfoldTimeDist.pdf}
\quad
\includegraphics[width=0.45\textwidth]{Figures/Hermes/kinfoldMeanStdev.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption[Analysis of \kinfold folding times for 20-mer
%DIFDELCMD < \ms{CCGAUUGGCGAAAGGCCACC}]{%%%
\DIFdelendFL \DIFaddbeginFL \caption[Analysis of \kinfold folding times for $20$-mer
\ms{CCGAUUGGCGAAAGGCCACC}]{\DIFaddendFL {\em (Left)} Histogram of \kinfold folding times for \DIFdelbeginFL \DIFdelFL{20-mer
}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$20$-mer
}\DIFaddendFL \ms{CCGAUUGGCGAAAGGCCACC}. The mean [resp. standard deviation] of \DIFdelbeginFL \DIFdelFL{10,000 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$10,000$ }\DIFaddendFL runs of \kinfold for this \DIFdelbeginFL \DIFdelFL{20-mer }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$20$-mer }\DIFaddendFL is \DIFdelbeginFL \DIFdelFL{538.37 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$538.37$ }\DIFaddendFL [resp. \DIFdelbeginFL \DIFdelFL{755.65}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$755.65$}\DIFaddendFL ]. Note the close fit to the exponential distribution, {\em (Right)} Mean minus standard deviation ($\mu -\sigma$), mean ($\mu$), and mean plus standard deviation ($\mu + \sigma$) of the logarithm of \kinfold folding times, taken over \DIFdelbeginFL \DIFdelFL{10,000 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$10,000$ }\DIFaddendFL runs for each of the \DIFdelbeginFL \DIFdelFL{1,000 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1,000$ }\DIFaddendFL sequences from the benchmarking set of \DIFdelbeginFL \DIFdelFL{20-mers}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$20$-mers}\DIFaddendFL . For graphical illustration, we have sorted the log folding times in increasing order.}
\label{fig:hermes:kinfoldMeanStdev}
\end{figure}

\subsection{Pearson correlation coefficients for various kinetics packages}
\label{subsec:hermes:corr}

In this section, we display the correlation between
\begin{inparaenum}[\em 1\upshape)]
\item the {\em gold standard} method \rnamfpt, both with and without the Hastings
modification using equations (\ref{eq:hermes:xitionProbWithHastings}) and
(\ref{eq:hermes:xitionProbWithoutHastings});
\item the {\em platinum standard} method \rnaeq, using
\eqnref{hermes:xitionRateWithoutHastings};
\item the {\em silver standard} method \kinfold;
\item \fftmfpt with and without the Hastings modification using equations
(\ref{eq:hermes:xitionProbFfttwoWithHastings}) and
(\ref{eq:hermes:xitionProbFfttwoWithoutHastings});
\item \ffteq which computes \eqt for the \DIFdelbegin \DIFdel{2D-grid }\DIFdelend \DIFaddbegin \twoD \DIFadd{grid }\DIFaddend using
\eqnref{hermes:xitionRateWithoutHastings}; and finally
\item \rnatwofold with and without the Hastings modification using equations
(\ref{eq:hermes:xitionProbFfttwoWithHastings}) and
(\ref{eq:hermes:xitionProbFfttwoWithoutHastings}).
\end{inparaenum}
Correlations
with [resp. without] the Hastings modification are summarized in the
lower [resp. upper] triangular portion of
Table \ref{table:correlationHermes}. It is clear that correlations between
the mathematically exact methods \rnamfpt, \rnaeq, and
approximation methods \kinfold, \fftmfpt, \ffteq, \rnatwofold are
improved when using the Hastings correction.

Figures \ref{fig:hermes:scatterplotRnaEqVsRnaMfptHastingsAndKinfold},
\ref{fig:hermes:scatterplotRnaMfptHastingsVsKinfoldAndFftMfpt},
\ref{fig:hermes:scatterplotFftMfptVsKinfoldAndFftEq}
depict scatterplots for kinetics obtained by some of the algorithms
above. The left panel of
\Figref{hermes:scatterplotRnaEqVsRnaMfptHastingsAndKinfold}
shows a scatter plots for gold standard \rnamfpt versus platinum
standard \rnaeq, with correlation value \DIFdelbegin \DIFdel{0.5652}\DIFdelend \DIFaddbegin \DIFadd{$0.5652$}\DIFaddend . The right
panel of the same figure shows a scatter plot for \kinfold versus
\rnaeq, with correlation \DIFdelbegin \DIFdel{0.7814}\DIFdelend \DIFaddbegin \DIFadd{$0.7814$}\DIFaddend . Note the persence of two
clusters in this and some of the other scatter plots. Cluster A
consists of RNA sequences whose folding time, as determined by \rnamfpt
or \rnaeq, is rapid---specifically, the natural
logarithm of the MFPT is at most \DIFdelbegin \DIFdel{7.5}\DIFdelend \DIFaddbegin \DIFadd{$7.5$}\DIFaddend . Cluster B consists of the
remaining RNA sequences, whose folding time is longer than that of
cluster A. There are no significant differences between RNA sequences
in clusters A and B with respect to GC-content, sequence logo, minimum
free energy, number of secondary structures, etc.  The left panel of
\Figref{hermes:scatterplotRnaMfptHastingsVsKinfoldAndFftMfpt}
shows the scatter plot for \rnamfpt versus \kinfold, with
correlation \DIFdelbegin \DIFdel{0.7933}\DIFdelend \DIFaddbegin \DIFadd{$0.7933$}\DIFaddend , and the right panel shows the scatter plot for
\rnamfpt versus \fftmfpt, with correlation \DIFdelbegin \DIFdel{0.6035}\DIFdelend \DIFaddbegin \DIFadd{$0.6035$}\DIFaddend .
\Figref{hermes:scatterplotFftMfptVsKinfoldAndFftEq}
shows scatter plots for \fftmfpt versus \kinfold {\em (Left)} and
for \fftmfpt versus \ffteq {\em (Right)}, with respective
correlation values \DIFdelbegin \DIFdel{0.7608 and 0.9589}\DIFdelend \DIFaddbegin \DIFadd{$0.7608$ and $0.9589$}\DIFaddend . \kinfold obviously provides
a better correlation with the exact value of \mfpt;
however, since the standard deviation of \kinfold runs is as
large as the mean, accurate kinetics estimates
from \kinfold require prohibitively large computational time---indeed, in
\citep{wolfingerstadler:kinetics} reliable kinetics for phe-tRNA from
yeast were obtained by \DIFdelbegin \DIFdel{9,000 }\DIFdelend \DIFaddbegin \DIFadd{$9,000$ }\DIFaddend \kinfold simulations, each for $10^8$
steps, requiring \DIFdelbegin \DIFdel{3 }\DIFdelend \DIFaddbegin \DIFadd{$3$ }\DIFaddend months of CPU time on an Intel Pentium \DIFdelbegin \DIFdel{4 running at
2.4 }\DIFdelend \DIFaddbegin \DIFadd{$4$ running at
$2.4$ }\DIFaddend GHz under Linux. Although the correlation value of \DIFdelbegin \DIFdel{0.6035 }\DIFdelend \DIFaddbegin \DIFadd{$0.6035$ }\DIFaddend between
\rnamfpt and \fftmfpt is much less than that obtained by \kinfold,
the runtime required by our method \fftmfpt is measured
in seconds, even for moderate to large RNAs. For this reason, we
advocate the use of \fftmfpt in synthetic biology screens to
design RNA sequences having certain desired kinetic properties. Once
promising candidates are found, it is possible to devote additional
computational time to \kinfold simulations for more accurate
kinetics.

\begin{figure}[!ht]
\centering
\includegraphics[width=0.45\textwidth]{Figures/Hermes/rnaMfptHastingsRnaEq.pdf}
\quad
\includegraphics[width=0.45\textwidth]{Figures/Hermes/rnaEqKinfold.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption{%
{%DIFAUXCMD
\DIFdelFL{Scatter plots of the natural logarithm of times from }%DIFDELCMD < \rnamfpt
%DIFDELCMD < %%%
\DIFdelFL{versus }%DIFDELCMD < \msresp{RNAeq} {\em %%%
\DIFdelFL{(Left)}%DIFDELCMD < } %%%
\DIFdelFL{and for }%DIFDELCMD < \kinfold %%%
\DIFdelFL{versus
}%DIFDELCMD < \msresp{RNAeq} {\em %%%
\DIFdelFL{(Right)}%DIFDELCMD < }%%%
\DIFdelFL{. }}
%DIFAUXCMD
\DIFdelendFL \DIFaddbeginFL \caption[Scatter plots of the natural logarithm of times from \rnamfpt
versus \msresp{RNAeq} {\em (Left)} and for \kinfold versus
\msresp{RNAeq} {\em (Right)}]
{\DIFaddFL{Scatter plots of the natural logarithm of times from }\rnamfpt
\DIFaddFL{versus }\msresp{RNAeq} {\em \DIFaddFL{(Left)}} \DIFaddFL{and for }\kinfold \DIFaddFL{versus
}\msresp{RNAeq} {\em \DIFaddFL{(Right)}}\DIFaddFL{.}}
\DIFaddendFL \label{fig:hermes:scatterplotRnaEqVsRnaMfptHastingsAndKinfold}
\end{figure}

\begin{figure}[!ht]
\centering
\includegraphics[width=0.45\textwidth]{Figures/Hermes/rnaMfptHastingsKinfold.pdf}
\quad
\includegraphics[width=0.45\textwidth]{Figures/Hermes/rnaMfptHastingsFftMfptHastings.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption{%
{%DIFAUXCMD
\DIFdelFL{Scatter plots of the natural logarithm of times from }%DIFDELCMD < \rnamfpt
%DIFDELCMD < %%%
\DIFdelFL{versus }%DIFDELCMD < \msresp{Kinfold} {\em %%%
\DIFdelFL{(Left)}%DIFDELCMD < } %%%
\DIFdelFL{and for }%DIFDELCMD < \rnamfpt %%%
\DIFdelFL{versus }%DIFDELCMD < \msresp{FFTmfpt} {\em %%%
\DIFdelFL{(Right)}%DIFDELCMD < }%%%
\DIFdelFL{. }}
%DIFAUXCMD
\DIFdelendFL \DIFaddbeginFL \caption[Scatter plots of the natural logarithm of times from \rnamfpt
versus \msresp{Kinfold} {\em (Left)} and for \rnamfpt versus \msresp{FFTmfpt} {\em (Right)}]
{\DIFaddFL{Scatter plots of the natural logarithm of times from }\rnamfpt
\DIFaddFL{versus }\msresp{Kinfold} {\em \DIFaddFL{(Left)}} \DIFaddFL{and for }\rnamfpt \DIFaddFL{versus }\msresp{FFTmfpt} {\em \DIFaddFL{(Right)}}\DIFaddFL{.}}
\DIFaddendFL \label{fig:hermes:scatterplotRnaMfptHastingsVsKinfoldAndFftMfpt}
\end{figure}

\begin{figure}[!ht]
\centering
\includegraphics[width=0.45\textwidth]{Figures/Hermes/kinfoldFftMfptHastings.pdf}
\quad
\includegraphics[width=0.45\textwidth]{Figures/Hermes/fftMfptFftEq.pdf}
\DIFdelbeginFL %DIFDELCMD < \caption{ %%%
\DIFdelendFL \DIFaddbeginFL \caption[Scatter plots of the natural logarithm of times from \kinfold
versus \msresp{FFTmfpt} {\em (Left)} and for \fftmfpt versus \msresp{FFTeq}
{\em (Right)}]{\DIFaddendFL Scatter plots of the natural logarithm of times from \kinfold
versus \msresp{FFTmfpt} {\em (Left)} and for \fftmfpt versus \msresp{FFTeq}
{\em (Right)}.}
\label{fig:hermes:scatterplotFftMfptVsKinfoldAndFftEq}
\end{figure}

\begin{landscape}
\begin{table}[!ht]
\centering
\begin{tabularx}{\linewidth}{c *{8}{L}}
  \toprule
  \small{Hastings (Yes\textbackslash No)} & \small{\rnamfpt} & \small{\rnaeq} & \small{\kinfold} & \small{\fftmfpt} & \small{\rnatwofold} & \small{\fftbor} & \DIFdelbeginFL %DIFDELCMD < \small{\ms{BarriersEq}} %%%
\DIFdelendFL \DIFaddbeginFL \small{\barrierseq} \DIFaddendFL & \small{\ffteq} \\
  \cmidrule(l){2-9}
  \small{\rnamfpt}   & \DIFdelbeginFL \DIFdelFL{1      }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$      }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5683 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5683$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7945 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7945$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5060 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5060$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5110 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5110$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5204 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5204$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5280 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5280$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.4472 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.4472$ }\DIFaddendFL \\
  \small{\rnaeq}   & \DIFdelbeginFL \DIFdelFL{0.5798 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5798$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1      }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$      }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7814 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7814$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7043 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7043$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7025 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7025$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5080 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5080$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5979 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5979$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.6820 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6820$ }\DIFaddendFL \\
  \small{\kinfold}       & \DIFdelbeginFL \DIFdelFL{0.7933 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7933$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7507 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7507$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1      }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$      }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7312 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7312$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7358 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7358$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.6241 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6241$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.6328 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6328$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.6445 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6445$ }\DIFaddendFL \\
  \small{\fftmfpt}       & \DIFdelbeginFL \DIFdelFL{0.6035 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6035$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7935 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7935$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7608 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7608$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1      }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$      }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.9980 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.9980$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5485 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5485$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.8614 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.8614$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.9589 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.9589$ }\DIFaddendFL \\
  \small{\rnatwofold}     & \DIFdelbeginFL \DIFdelFL{0.6076 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6076$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7919 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7919$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7655 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7655$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.9983 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.9983$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1      }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$      }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5584 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5584$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.8538 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.8538$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.9515 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.9515$ }\DIFaddendFL \\
  \small{\fftbor}        & \DIFdelbeginFL \DIFdelFL{0.5416 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5416$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5218 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5218$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.6241 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6241$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5748 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5748$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.5855 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5855$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1      }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$      }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.3450 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.3450$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.4229 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.4229$ }\DIFaddendFL \\
  \DIFdelbeginFL %DIFDELCMD < \small{\ms{BarriersEq}} %%%
\DIFdelendFL \DIFaddbeginFL \small{\barrierseq} \DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.6346 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6346$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.6578 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6578$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.6328 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6328$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.8310 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.8310$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.8217 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.8217$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.3450 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.3450$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1      }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$      }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.9149 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.9149$ }\DIFaddendFL \\
  \small{\ffteq}         & \DIFdelbeginFL \DIFdelFL{0.5614 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.5614$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.7916 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.7916$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.6966 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.6966$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.9670 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.9670$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.9590 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.9590$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.4757 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.4757$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{0.8940 }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$0.8940$ }\DIFaddendFL & \DIFdelbeginFL \DIFdelFL{1      }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$1$      }\DIFaddendFL \\
  \bottomrule
 \end{tabularx}
 \caption[Table of Pearson correlation coefficients for various methods to compute or approximate
 RNA secondary structure folding kinetics]{Table of Pearson correlation coefficients for
 various methods to compute or approximate RNA secondary structure folding kinetics.
 Lower [resp. upper] triangular entries are with [resp. without] the Hastings correction
 for Markov chain probability matrices. The methods are: \rnamfpt (\mfpt, computed by matrix
 inversion for the Markov chain consisting of all secondary structures, with move allowed between
 structures differing by one base pair), \rnaeq (\eqt, computed by spectral decomposition of a
 rate matrix comprising all secondary structures to compute population fraction $P(t)$ at time $t$),
 \kinfold (an implementation of Gillespie's Algorithm to approximate refolding pathways using an
 event-based Monte Carlo simulation), \fftmfpt (\mfpt for Markov chain consisting of \DIFdelbeginFL \DIFdelFL{``}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{`}\DIFaddendFL grid point\DIFdelbeginFL \DIFdelFL{'' }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{'
 }\DIFaddendFL states $(x,y)$ with probability \DIFdelbeginFL \DIFdelFL{$P(x,y)=\sum_S exp(-E(S)/RT)/Z$}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$P(x,y)=\sum_\str \frac{\boltzf{\str}}{\fullZ}$}\DIFaddendFL , computed by \ffttwo,
 where the sum is taken over structures having \bpd $x$ to the empty structure and $y$ to the MFE
 structure, \rnatwofold (\mfpt, computed as previously explained, but using \rnatwofold in place
 of \ffttwo to compute $P(x,y)$), \fftbor (\mfpt, computed for the Markov chain consisting of
 states $0,1,\dots,n$, for which \DIFdelbeginFL \DIFdelFL{$P(x) = \sum_S \exp(-E(S)/RT)/Z$}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{$P(x) = \sum_\str \frac{\boltzf{\str}}{\fullZ}$}\DIFaddendFL , where the
 sum is taken over all secondary structures whose \bpd is $x$ from the MFE structure), \DIFdelbeginFL %DIFDELCMD < \ms{BarriersEq} %%%
\DIFdelendFL \DIFaddbeginFL \barrierseq
 \DIFaddendFL (\eqt, computed using spectral decomposition on the Markov process consisting of \DIFdelbeginFL \DIFdelFL{``}\DIFdelendFL \DIFaddbeginFL \DIFaddFL{`}\DIFaddendFL grid point\DIFdelbeginFL \DIFdelFL{'' }\DIFdelendFL \DIFaddbeginFL \DIFaddFL{'
 }\DIFaddendFL states output from \DIFdelbeginFL %DIFDELCMD < \ms{Barriers}%%%
\DIFdelendFL \DIFaddbeginFL \barriers\DIFaddendFL ), and \ffteq (\eqt, computed in the same fashion as \DIFdelbeginFL %DIFDELCMD < \ms{BarriersEq} %%%
\DIFdelendFL \DIFaddbeginFL \barrierseq
 \DIFaddendFL using a Markov process derived from the energy landscape output by \ffttwo).}
\label{table:correlationHermes}
 \end{table}
 \end{landscape}

%DIF <  \input{Chapters/Wrnap}
%!TEX root = ../main.tex

\chapter{Discussion}
\label{ch:disc}

\lhead{Discussion}

Over the course of this thesis, we have described a collection of tools aimed
at facilitating the computational analysis of RNA. The work in \Chref{rfinder}
aims to address the open problem of determining full \rb sequences, along
with their `on' and `off' conformational states in an efficient and generalized
manner. \Rbs are an imporant regulatory motif that can influence
transcription by the introduction of an intrinsic terminator, an extended stem-loop
followed by a series of uracil residues which cause RNA polymerase to prematurely
terminate transcription \citep{gusarov:1999uu,yarnell:1999wt}. Alternately,
\rbs can regulate translation, generally by sequestration of the ribosome
binding site in response to the presence or absence of a corresponding ligand
\citep{barrick:2007gw}. This degree of control makes \rbs well suited for
the regulation of metabolic pathways, and are generally found within bacterial
genomes. While the aptamer portion of the \rb necessarily exhibits high
sequence and structural conservation---necessary to retain a high binding affinity
for the corresponding ligand---the downstream expression platform is much more
variable, and difficult to detect using covariance models such as \infernal
\citep{infernal}. For this reason, databases such as Rfam \citep{nawrocki:2014uy}
only contain alignments for the conserved aptamer domain; rarely enough to
understand the structural characteristics of the \rb `states' in the greater
context of the mRNA. \rfinder aims to provide a holistic description of
\rbs by coupling powerful aptamer-detection tools (such as \infernal) with software
designed to detect intrinsic terminators (such as \tthp), and use
experimentally-derived structural information from the literature in conjunction
with constrained folding methods to predict full gene `on' and `off' conformations
in a general manner.

The remainder of the thesis follows a natural progression. In \Chref{fftbor} we
present the algorithm \fftbor, an efficient approach to compute the
{\em parameterized} Boltzmann probabilities $\pk = \frac{\bfZ{k}{}}{\fullZ}$ of
those sequences having \bpd $k$ from an arbitrary input structure
\strSt. By using the \fft along with \nRoUs, \fftbor operates in \On{4} speed,
an order of magnitude faster than its predecessor. This approach of applying the
\fft to compute parameterized partition functions was put to use in
\citep{ding:2014ex} as well, where we develop programs \ms{FFThairpin}
[resp. \msresp{FFTmultiloop}] to describe the probability distribution of structures
having exactly $k$ hairpins [resp. multiloops]. Though not discussed here, by
training a support vector machine (SVM) on the probability distributions output
by these
programs we were able to classify Rfam families using five-fold cross validation
with excellent `area under curve' (AUC) values.

\Chref{ffttwo} uses the \fft to compute the \DIFdelbegin \DIFdel{2D }\DIFdelend \DIFaddbegin \twoD \DIFaddend probability distribution, where
given two compatible input secondary structures \strA and \strB, position
$(x,y)$ on the discrete energy landscape corresponds to the Boltzmann
probability for those structures \str which have \bpd $x$ [resp. $y$]
from \strA [resp. \strBresp]. Compared against its closest relative \rnatwofold,
\ffttwo is two orders of magnitude faster, operating in \On{5} speed and only
\On{2} space. This opens the door for using \ffttwo in a high-throughput fashion,
where the probability landscapes can be used as a starting point for kinetic
analysis.

With \hermes in \Chref{hermes}, we do just that; representing the output of
\ffttwo as a graph, we can create a corresponding Markov chain [resp. Markov
process] to estimate \mfpt [resp. equilibrium time] with correlation to the
mathematically-exact values as shown in Table \ref{table:correlationHermes}.
The speed of these approximate tools within \hermes---namely \fftmfpt and
\ffteq---present what we believe to be the first suite of kinetic analysis
tools for RNA sequences that are suitable for high throughput usage, something
we believe to be of interest in the field of synthetic design.

Finally, though not discussed in the body of this thesis, we have also
developed a collection of packages for working with command-line tools for
computational RNA analysis in a more streamlined fashion. Two of these programs,
\href{https://github.com/evansenter/wrnap}{\wrnap} and
\href{https://github.com/evansenter/rbfam}{\rbfam} are among the first of their
type for the Ruby programming langage, a popular interpreted language that
enjoys wide adoption in the web development community. In an era of increasingly
cloud-focused software, we believe that researchers will need a robust set of
tools for working with their data in a web-oriented environment, something
\wrnap and \rbfam aim to provide.


%----------------------------------------------------------------------------------------
%	THESIS CONTENT - APPENDICES
%----------------------------------------------------------------------------------------
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend % \addtocontents{toc}{\vspace{2em}} % Add a gap in the Contents, for aesthetics

\appendix % Cue to tell LaTeX that the following 'chapters' are Appendices

%DIF <  % Include the appendices of the thesis as separate files from the Appendices folder
%DIF <  % Uncomment the lines as you write the Appendices
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend %!TEX root = ../main.tex

\chapter{FFTbor Appendix}
\label{ch:fftborapp}

\lhead{FFTbor Appendix}

\section{Full recursions for \texorpdfstring{\emZ{i,j}}{}
for the Turner energy model}
\label{sec:fftbor:turner}

To compute $\emZ{} = \emZ{1,n}$ given input structure \str, we use the recursions

\begin{align}
\label{eq:fftborapp:zx}
\emZ{i,j} = \emZ{i,j-1} \cdot x^{d_0} +
\sum_{\mathclap{\substack{(s_k,s_j) \in \bpSet, \\ i \leq k < j}}}\enspace
\Big( \emZ{i,k-1} \cdot \emZB{k,j} \cdot \boltzE{E_d} \cdot x^{d_1} \Big),
\end{align}

where $d_0 = 1$ if $j$ is base paired
in $\str_{[i,j]}$ and $0$ otherwise, $d_1 =
\dBP{\str_{[i,j]}}{\str_{[i,k-1]} \cup \str_{[k,j]}}$, and $E_d$
is the energy contribution due to dangling ends (energy
contributions from single bases stacking on adjacent base pairs) and
closing AU base pairs (since a non GC base pair closing a stem has a
destabilizing effect).  The sum is taken over all possible
base pairs $(k,j)$ with $i \leq k < j$.

We compute \emZB{} using the recursion

\begin{align}
\label{eq:fftborapp:zbx}
\begin{split}
\emZB{i,j} &= \boltzE{EH(i,j)} \cdot x^{d_2} \\
&+ \sum_{\mathclap{\substack{(s_k,s_l) \in \bpSet, \\ i < k < l < j }}}\enspace
\emZB{k,l} \cdot \boltzE{EI(i,j,k,l)} \cdot x^{d_3} \\
&+ \sum_{\mathclap{\substack{(s_k,s_l) \in \bpSet, \\ i < k < l < j }}}\enspace
\Big( \emZM{i+1,k-1} \cdot \emZB{k,l} \cdot \boltzE{(a+b+c(j-l-1))}
\cdot x^{d_4} \Big)
\end{split}
\end{align}

where $d_2 = \dBP{\str_{[i,j]}}{\{(i,j)\}}$,
$EH(i,j)$ is the energy of the hairpin loop with closing base
pair $(i,j)$, $EI(i,j,k,l)$ is the energy of the stack, bulge or
interior loop with the closing base pair $(i,j)$ and the interior
base pair $(k,l)$, $d_3 = \dBP{\str_{[i,j]}}{\str_{[k,l]} \cup
\{(i,j)\}}$, and $d_4 = \dBP{\str_{[i,j]}}{\str_{[i+1,k-1]} \cup
\str_{[k,l]} \cup\{(i,j)\}}$. The first term in the
recursion takes care of the case where $(i,j)$ is the only base pair
in $[i,j]$, i.e. $(i,j)$ closes a hairpin loop. The second term
handles the case where there is an interior loop (or a bulge or a
stack) closed by $(i,j)$ and $(k,l)$. The third term takes care of
all the structures where $(i,j)$ closes a multiloop. To reduce
complexity of the algorithm, the interior and bulge loop size can be
limited to a maximum size of $L$ (taken by default to be \DIFdelbegin \DIFdel{30}\DIFdelend \DIFaddbegin \DIFadd{$30$}\DIFaddend ),
by requiring that $l>j-L$ in the above recursion.

The final recursion, for computing \emZM{}, is

\begin{align}
\label{eq:fftborapp:zmx}
\begin{split}
\emZM{i,j} &= \emZM{i,j-1} \cdot \boltzE{c} \cdot x^{d_0} \\
&+ \sum_{\mathclap{\substack{(s_k,s_j) \in \bpSet, \\ i \leq k < j }}}\enspace
\Big( \emZB{k,j} \cdot \boltzE{(b+c(k-i))} \cdot x^{d_5} \\
&+ \emZM{i,k-1} \cdot \emZB{k,j} \cdot \boltzE{b} \cdot x^{d_6} \Big)
\end{split}
\end{align}

where $d_5 = \dBP{\str_{[i,j]}}{\str_{[k,j]}}$ and $d_6
=\dBP{\str_{[i,j]}}{\str_{[i,k-1]} \cup \str_{[k,j]}}$.
Note that since $\emZM{i,j}$ computes the partition function
contribution under the assumption that $[i,j]$ is part of a
multiloop, there will be exactly one stem-loop structure in this
region (the \emZB{} term) or
more than one (the \emZM{}--\emZB{} term).
Justification of recursions (\ref{eq:fftborapp:zx}),
(\ref{eq:fftborapp:zbx}), and
(\ref{eq:fftborapp:zmx})
follow by induction, as in the proof of Theorem \ref{thm:fftbor:recursions}.

%!TEX root = ../main.tex

\chapter{FFTbor2D Appendix}
\label{ch:ffttwoapp}

\DIFdelbegin %DIFDELCMD < \lhead{FFTbor2D Appendix}
%DIFDELCMD < %%%
\DIFdelend \DIFaddbegin \lhead{FFTbor\twoD Appendix}
\DIFaddend 

\section{Full recursions for \texorpdfstring{\emZ{i,j}}{}
for the Turner energy model}
\label{sec:ffttwo:turner}

To compute $\emZ{} = \emZ{1,n}$ given input structures \strAB,
we use the recursions

\begin{align}
\label{eq:ffttwoapp:zx}
\emZ{i,j} = \emZ{i,j-1} \cdot x^{\ab{0}} +
\sum_{\mathclap{\substack{(s_k,s_j) \in \bpSet, \\ i \leq k < j}}}\enspace
\Big( \emZ{i,k-1} \cdot \emZB{k,j} \cdot \boltzE{E_d} \cdot
x^{\ab{1}} \Big),
\end{align}

where
$\alpha_0 = 1$ if $j$ is base paired in $\strA_{[i,j]}$ and $0$ otherwise,
$\beta_0 = 1$ if $j$ is base paired in $\strB_{[i,j]}$ and $0$ otherwise,
$\alpha_1 = \dBP{\strA_{[i,j]}}{\strA_{[i,k-1]} \cup \strA_{[k,j]}}$,
$\beta_1 = \dBP{\strB_{[i,j]}}{\strB_{[i,k-1]} \cup \strB_{[k,j]}}$,
and $E_d$ is the energy contribution due to dangling ends (energy
contributions from single bases stacking on adjacent base pairs) and
closing AU base pairs (since a non GC base pair closing a stem has a
destabilizing effect).  The sum is taken over all possible
base pairs $(k,j)$ with $i \leq k < j$.

We compute \emZB{} using the recursion

\begin{align}
\label{eq:ffttwoapp:zbx}
\begin{split}
\emZB{i,j} &= \boltzE{EH(i,j)} \cdot x^{\ab{2}} \\
&+ \sum_{\mathclap{\substack{(s_k,s_l) \in \bpSet, \\ i < k < l < j }}}\enspace
\emZB{k,l} \cdot \boltzE{EI(i,j,k,l)} \cdot x^{\ab{3}} \\
&+ \sum_{\mathclap{\substack{(s_k,s_l) \in \bpSet, \\ i < k < l < j }}}\enspace
\Big( \emZM{i+1,k-1} \cdot \emZB{k,l} \cdot \boltzE{(a+b+c(j-l-1))}
\cdot x^{\ab{4}} \Big)
\end{split}
\end{align}

where
$EH(i,j)$ is the energy of the hairpin loop with closing base pair $(i,j)$,
$\alpha_2 = \dBP{\strA_{[i,j]}}{\{(i,j)\}}$,
$\beta_2 = \dBP{\strB_{[i,j]}}{\{(i,j)\}}$,
$EI(i,j,k,l)$ is the energy of the stack, bulge or
interior loop with the closing base pair $(i,j)$ and the interior
base pair $(k,l)$,
$\alpha_3 = \dBP{\strA_{[i,j]}}{\strA_{[k,l]} \cup \{(i,j)\}}$,
$\beta_3 = \dBP{\strB_{[i,j]}}{\strB_{[k,l]} \cup \{(i,j)\}}$,
$\alpha_4 = \dBP{\strA_{[i,j]}}{\strA_{[i+1,k-1]} \cup
\strA_{[k,l]} \cup\{(i,j)\}}$, and
$\beta_4 = \dBP{\strB_{[i,j]}}{\strB_{[i+1,k-1]} \cup
\strB_{[k,l]} \cup\{(i,j)\}}$.
The first term in the
recursion takes care of the case where $(i,j)$ is the only base pair
in $[i,j]$, i.e. $(i,j)$ closes a hairpin loop. The second term
handles the case where there is an interior loop (or a bulge or a
stack) closed by $(i,j)$ and $(k,l)$. The third term takes care of
all the structures where $(i,j)$ closes a multiloop. To reduce
complexity of the algorithm, the interior and bulge loop size can be
limited to a maximum size of $L$ (taken by default to be \DIFdelbegin \DIFdel{30}\DIFdelend \DIFaddbegin \DIFadd{$30$}\DIFaddend ),
by requiring that $l>j-L$ in the above recursion.

The final recursion, for computing \emZM{}, is

\begin{align}
\label{eq:ffttwoapp:zmx}
\begin{split}
\emZM{i,j} &= \emZM{i,j-1} \cdot \boltzE{c} \cdot x^{\ab{0}} \\
&+ \sum_{\mathclap{\substack{(s_k,s_j) \in \bpSet, \\ i \leq k < j }}}\enspace
\Big( \emZB{k,j} \cdot \boltzE{(b+c(k-i))} \cdot x^{\ab{5}} \\
&+ \emZM{i,k-1} \cdot \emZB{k,j} \cdot \boltzE{b} \cdot x^{\ab{6}} \Big)
\end{split}
\end{align}

where
$\alpha_5 = \dBP{\strA_{[i,j]}}{\strA_{[k,j]}}$,
$\beta_5 = \dBP{\strB_{[i,j]}}{\strB_{[k,j]}}$,
$\alpha_6 =\dBP{\strA_{[i,j]}}{\strA_{[i,k-1]} \cup \strA_{[k,j]}}$, and
$\beta_6 =\dBP{\strB_{[i,j]}}{\strB_{[i,k-1]} \cup \strB_{[k,j]}}$.
Note that since $\emZM{i,j}$ computes the partition function
contribution under the assumption that $[i,j]$ is part of a
multiloop, there will be exactly one stem-loop structure in this
region (the \emZB{} term) or
more than one (the \emZM{}--\emZB{} term).
Justification of recursions (\ref{eq:ffttwoapp:zx}),
(\ref{eq:ffttwoapp:zbx}), and
(\ref{eq:ffttwoapp:zmx})
follow by induction, as in the proof of Theorem \ref{thm:ffttwo:recursions}.

\section{Proof for Theorem \ref{thm:ffttwo:recursions}}
\label{sec:ffttwo:recursions}

\begin{proof}
Recall that if $F$ is an arbitrary
polynomial [resp. analytic] function, then $[x^{rn+s}]F(x)$
denotes the coefficient of monomial $x^{rn+s}$ in the
Taylor expansion of $F(x)$. For instance, in
\eqnref{ffttwo:pOfX}, $[x^{rn+s}]p(x) = p_{rn+s}$, and in
\eqnref{ffttwo:zOfX}, $[x^{rn+s}]\fullZx = z_{rn+s}$.

By definition, it is clear that $\emZ{i,j}=1$ if $i \leq j \leq i + \theta$,
where we recall that $\theta = 3$ is the minimum number of unpaired bases in
a hairpin loop. For $j > i + \theta$, we have

\begin{align}
\begin{split}
[x^{rn+s}] \emZ{i,j} &= z_{rn+s}(i,j) = \bfZ{rn+s}{i,j} \\
&= \bfZ{(r-\alpha_0)n+(s-\beta_0)}{i,j-1} \\
&+
\sum_{k=i}^{j-1}\enspace
\sum_{u_0+u_1=r-\alpha(k)}\enspace
\sum_{v_0+v_1=s-\beta(k)}
\left( \boltzNuss{k,j}
\cdot \bfZ{u_0 n+v_0}{i,k-1} \cdot \bfZ{u_1 n+v_1}{k+1,j-1} \right) \\
&= [x^{(r-\alpha_0)n+(s-\beta_0)}] \emZ{i,j-1} \\
&+ \sum_{k=i}^{j-1}\enspace
\sum_{u_0+u_1=r-\alpha(k)}\enspace
\sum_{v_0+v_1=s-\beta(k)}
\left( \boltzNuss{k,j} \right. \\
&\left. \cdot \left( [x^{u_0 n+v_0}] \emZ{i,k-1} \right) \cdot
\left( [x^{u_1 n+v_1}] \emZ{k+1,j-1} \right)
\vphantom{\boltzNuss{k,j}} \right) \\
&= [x^{(r-\alpha_0)n+(s-\beta_0)}] \emZ{i,j-1} \\
&+ \sum_{k=i}^{j-1}\enspace
\sum_{u_0+u_1=r-\alpha(k)}\enspace
\sum_{v_0+v_1=s-\beta(k)}
\left( \boltzNuss{k,j} \right. \\
&\left. \cdot [x^{(u_0+u_1)n+(v_0+v_1)}] \left( \emZ{i,k-1} \cdot \emZ{k+1,j-1} \right)
\vphantom{\boltzNuss{k,j}} \right) \\
&= [x^{(r-\alpha_0)n+(s-\beta_0)}] \emZ{i,j-1} \\
&+ \sum_{k=i}^{j-1} \left( \boltzNuss{k,j}
\cdot [x^{(r-\alpha(k))n+(s-\beta(k))}] \left( \emZ{i,k-1} \cdot \emZ{k+1,j-1}
\right) \right) \\
&= [x^{rn+s}] \left( \emZ{i,j-1} \cdot x^{\ab{0}} \right) \\
&+ \sum_{k=i}^{j-1} \left( \boltzNuss{k,j}
\cdot [x^{rn+s}] \left( \emZ{i,k-1} \cdot \emZ{k+1,j-1} \cdot
x^{\alpha(k)n+\beta(k)} \right) \right) \\
&= [x^{rn+s}] \left( \vphantom{\sum_{k=i}^{j-1}} \emZ{i,j-1} \cdot x^{\ab{0}} \right. \\
&\left. + \sum_{k=i}^{j-1} \left( \boltzNuss{k,j}
\cdot \emZ{i,k-1} \cdot \emZ{k+1,j-1} \cdot
x^{\alpha(k)n+\beta(k)} \right) \right) \\
\end{split}
\end{align}
\end{proof}

\section{Proof for Lemma \ref{lem:ffttwo:lemma3}}
\label{sec:ffttwo:lemma3proof}

\begin{proof}
The lemma states that if the \bpd $d_0$ between reference
structures \strAB is even, then $\emZof{}{-\alpha}=\emZof{}{\alpha}$, while if
the distance is odd, then $\emZof{}{-\alpha}=-\emZof{}{\alpha}$.
Suppose first that $d_0$ is even. By Lemma \ref{lem:ffttwo:lemma1},
$\emZ{} = z_0 + z_2 x^2 + z_4 x^4 + \dots +
z_{M-2} x^{M-2)}$, and so $\emZof{}{-\alpha}=\emZof{}{\alpha}$.
Suppose now that
$d_0$ is odd. By Lemma \ref{lem:ffttwo:lemma1},
$Z(x) = z_1 x^1 + z_3 x^3 + z^5 x^5 \dots +
z_{M-1} x^{M-1}$, and so $\emZof{}{-\alpha}=-\emZof{}{\alpha}$.
\end{proof}

\section{Proof for Lemma \ref{lem:ffttwo:lemma4}}
\label{sec:ffttwo:lemma4proof}

\begin{proof}
Recall Euler's formula in complex analysis:
$\exp(ix) = \cos(x) + i \sin(x)$. As well, recall that
$\sin(\pi)=0$, $\cos(\pi)=-1$, and the trigonometric
addition formulas:

\begin{align}
\begin{split}
\cos(\alpha-\beta) &= \cos(\alpha) \cos(\beta) + \sin(\alpha) \sin(\beta) \\
\sin(\alpha-\beta) &= \sin(\alpha) \cos(\beta) - \sin(\beta) \cos(\alpha).
\end{split}
\end{align}

Then we have

\begin{align}
\begin{split}
\nu^{M_0-k} &= \exp\left(\frac{2 \pi i (M_0-k)}{M}\right) \\
&= \cos\left(\frac{2 \pi (M_0-k)}{M}\right) + i
\sin\left(\frac{2 \pi (M_0-k)}{M}\right) \\
&= \cos\left(\pi - \frac{2 \pi k}{M}\right) + i
\sin\left(\pi - \frac{2 \pi k}{M}\right) \\
&=
\left[
\cos(\pi) \cos\left(\frac{2 \pi k}{M}\right) +
\sin(\pi) \sin\left(\frac{2 \pi k}{M}\right)
\right] \\
&+
i \left[
\sin(\pi) \cos\left(\frac{2 \pi k}{M}\right) -
\sin\left(\frac{2 \pi k}{M}\right) \cos(\pi)
\right] \\
&=
-\cos\left(\frac{2 \pi k}{M}\right) + i
\sin\left(\frac{2 \pi k}{M}\right) \\
&=
-1 \left[ \cos\left(\frac{2 \pi k}{M}\right) - i
\sin\left(\frac{2 \pi k}{M}\right) \right] \\
&=
-1 \cdot \overline{\cos\left(\frac{2 \pi k}{M}\right) + i
\sin\left(\frac{2 \pi k}{M}\right)}  \\
&=
-1 \cdot \overline{\exp\left(\frac{2 \pi i k}{M}\right)} = - \overline{\nu^k} . \\
\end{split}
\end{align}

It follows that
$\nu^{M_0-k} = -\overline{\nu^k}$, so
$\nu^{k} = \overline{- \nu^{(M_0-k)}} = -\overline{\nu^{(M_0-k)}}$.

This completes the proof of the lemma.
\end{proof}


%DIF <  %\input{Appendices/AppendixB}
%DIF <  %\input{Appendices/AppendixC}
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend % \addtocontents{toc}{\vspace{2em}} % Add a gap in the Contents, for aesthetics

\backmatter

%----------------------------------------------------------------------------------------
%	BIBLIOGRAPHY
%----------------------------------------------------------------------------------------
\DIFdelbegin %DIFDELCMD < 

%DIFDELCMD < %%%
\DIFdelend \label{Bibliography}

\setstretch{1.3}

\lhead{\emph{Bibliography}}

\bibliographystyle{ieeetr}

\begin{thebibliography}{100}

\bibitem{crick:1970wb}
F.~Crick, ``{Central dogma of molecular biology},'' {\em Nature}, 1970.

\bibitem{coffin:1997ws}
J.~M. Coffin, S.~H. Hughes, H.~E. Varmus, and A.~B. Rabson, ``{Synthesis and
  processing of viral RNA},'' 1997.

\bibitem{berretta:2009tq}
J.~Berretta and A.~Morillon, ``{Pervasive transcription constitutes a new level
  of eukaryotic genome regulation},'' {\em EMBO reports}, 2009.

\bibitem{clark:2011cc}
M.~B. Clark, P.~P. Amaral, F.~J. Schlesinger, M.~E. Dinger, R.~J. Taft, J.~L.
  Rinn, C.~P. Ponting, P.~F. Stadler, K.~V. Morris, A.~Morillon, J.~S.
  Rozowsky, M.~B. Gerstein, C.~Wahlestedt, Y.~Hayashizaki, P.~Carninci, T.~R.
  Gingeras, and J.~S. Mattick, ``{The reality of pervasive transcription.},''
  {\em PLoS biology}, vol.~9, pp.~e1000625--discussion e1001102, July 2011.

\bibitem{jensen:2013vb}
T.~H. Jensen, A.~Jacquier, and D.~Libri, ``{Dealing with pervasive
  transcription},'' {\em Molecular cell}, 2013.

\bibitem{costa:2005ug}
F.~F. Costa, ``{Non-coding RNAs: new players in eukaryotic biology},'' {\em
  Gene}, 2005.

\bibitem{fire:1998tv}
A.~Fire, S.~Q. Xu, M.~K. Montgomery, S.~A. Kostas, and S.~E. Driver, ``{Potent
  and specific genetic interference by double-stranded RNA in Caenorhabditis
  elegans},'' {\em nature}, 1998.

\bibitem{mccaffrey:2002tf}
A.~P. McCaffrey, L.~Meuse, T.~Pham, and D.~S. Conklin, ``{Gene expression: RNA
  interference in adult mice},'' {\em Nature}, 2002.

\bibitem{hannon:2002vn}
G.~J. Hannon, ``{RNA interference},'' {\em Nature}, 2002.

\bibitem{he:2004uk}
L.~He and G.~J. Hannon, ``{MicroRNAs: small RNAs with a big role in gene
  regulation},'' {\em Nature Reviews Genetics}, 2004.

\bibitem{nudler:2004vm}
E.~Nudler and A.~S. Mironov, ``{The riboswitch control of bacterial
  metabolism},'' {\em Trends in biochemical sciences}, 2004.

\bibitem{mandal:2004vh}
M.~Mandal and R.~R. Breaker, ``{Gene regulation by riboswitches},'' {\em Nature
  Reviews Molecular Cell Biology}, 2004.

\bibitem{kruger:1982wk}
K.~Kruger, P.~J. Grabowski, A.~J. Zaug, J.~Sands, D.~E. Gottschling, and T.~R.
  Cech, ``{Self-splicing RNA: autoexcision and autocyclization of the ribosomal
  RNA intervening sequence of Tetrahymena.},'' {\em Cell}, vol.~31,
  pp.~147--157, Nov. 1982.

\bibitem{cech:1990tn}
T.~R. Cech, ``{Self-splicing of group I introns},'' {\em Annual review of
  biochemistry}, 1990.

\bibitem{doherty:2001wq}
E.~A. Doherty and J.~A. Doudna, ``{Ribozyme structures and mechanisms},'' {\em
  Annual review of biophysics and {\ldots}}, 2001.

\bibitem{gilbert:1986td}
W.~Gilbert, ``{Origin of life: The RNA world},'' {\em Nature}, 1986.

\bibitem{yakovchuk:2006bm}
P.~Yakovchuk, E.~Protozanova, and M.~D. Frank-Kamenetskii, ``{Base-stacking and
  base-pairing contributions into thermal stability of the DNA double
  helix.},'' {\em Nucleic acids research}, vol.~34, no.~2, pp.~564--574, 2006.

\bibitem{fresco:1960vg}
J.~R. Fresco, B.~M. Alberts, and P.~Doty, ``{Some molecular details of the
  secondary structure of ribonucleic acid.},'' {\em Nature}, vol.~188,
  pp.~98--101, Oct. 1960.

\bibitem{waterman:1978va}
M.~S. Waterman, ``{Secondary structure of single-stranded nucleic acids},''
  {\em Adv math suppl studies}, 1978.

\bibitem{nussinov1980}
R.~Nussinov and A.~B. Jacobson, ``{Fast algorithm for predicting the secondary
  structure of single-stranded RNA.},'' {\em Proceedings of the National
  Academy of Sciences}, vol.~77, pp.~6309--6313, Nov. 1980.

\bibitem{bellman:1952vza}
{\em {On the theory of dynamic programming}}, 1952.

\bibitem{zuker:1981tf}
M.~Zuker and P.~Stiegler, ``{Optimal computer folding of large RNA sequences
  using thermodynamics and auxiliary information.},'' {\em Nucleic acids
  research}, vol.~9, pp.~133--148, Jan. 1981.

\bibitem{zuker:1989im}
M.~Zuker, ``{On finding all suboptimal foldings of an RNA molecule},'' {\em
  Science (New York, N.Y.)}, vol.~244, pp.~48--52, Apr. 1989.

\bibitem{mccaskill}
J.~McCaskill, ``The equilibrium partition function and base pair binding
  probabilities for {RNA} secondary structure,'' {\em Biopolymers}, vol.~29,
  pp.~1105--1119, 1990.

\bibitem{turner}
D.~Matthews, J.~Sabina, M.~Zuker, and D.~Turner, ``Expanded sequence dependence
  of thermodynamic parameters improves prediction of {RNA} secondary
  structure,'' {\em J. Mol. Biol.}, vol.~288, pp.~911--940, 1999.

\bibitem{turner:2009vy}
D.~H. Turner and D.~H. Mathews, ``{NNDB: the nearest neighbor parameter
  database for predicting stability of nucleic acid secondary structure},''
  {\em Nucleic Acids Research}, 2009.

\bibitem{infernal}
E.~P. Nawrocki, D.~L. Kolbe, and S.~R. Eddy, ``{Infernal 1.0: inference of
  {R}{N}{A} alignments},'' {\em Bioinformatics}, 2009.

\bibitem{nawrocki:2013hk}
E.~P. Nawrocki and S.~R. Eddy, ``{Infernal 1.1: 100-fold faster RNA homology
  searches},'' {\em Bioinformatics}, vol.~29, pp.~2933--2935, Oct. 2013.

\bibitem{ermolaeva:2000cl}
M.~D. Ermolaeva, H.~G. Khalak, O.~White, H.~O. Smith, and S.~L. Salzberg,
  ``{Prediction of transcription terminators in bacterial genomes},'' {\em
  Journal of molecular biology}, vol.~301, pp.~27--33, Aug. 2000.

\bibitem{lorenz.amb11}
R.~Lorenz, S.~H. Bernhart, C.~Honer Zu~Siederdissen, H.~Tafer, C.~Flamm, P.~F.
  Stadler, and I.~L. Hofacker, ``Viennarna {Package} 2.0.,'' {\em Algorithms.
  Mol. Biol.}, vol.~6, p.~26, 2011.

\bibitem{gorodkin:1997tr}
J.~Gorodkin, L.~J. Heyer, and G.~D. Stormo, ``{Finding the most significant
  common sequence and structure motifs in a set of RNA sequences.},'' {\em
  Nucleic Acids Research}, vol.~25, pp.~3724--3732, Sept. 1997.

\bibitem{havgaard:2007ca}
J.~H. Havgaard, E.~Torarinsson, and J.~Gorodkin, ``{Fast Pairwise Structural
  RNA Alignments by Pruning of the Dynamical Programming Matrix},'' {\em PLoS
  computational biology}, vol.~3, no.~10, p.~e193, 2007.

\bibitem{bengert2004}
P.~Bengert and T.~Dandekar, ``{Riboswitch finder--a tool for identification of
  riboswitch RNAs},'' {\em Nucleic Acids Research}, vol.~32, pp.~W154--W159,
  July 2004.

\bibitem{chang:2009de}
T.-H. Chang, H.-D. Huang, L.-C. Wu, C.-T. Yeh, B.-J. Liu, and J.-T. Horng,
  ``{Computational identification of riboswitches based on RNA conserved
  functional sequences and conformations.},'' {\em RNA (New York, N.Y.)},
  vol.~15, pp.~1426--1430, July 2009.

\bibitem{abreugoodger:2005hb}
C.~Abreu-Goodger and E.~Merino, ``{RibEx: a web server for locating
  riboswitches and other conserved bacterial regulatory elements.},'' {\em
  Nucleic Acids Research}, vol.~33, pp.~W690--2, July 2005.

\bibitem{yao2006}
Z.~Yao, Z.~Weinberg, and W.~L. Ruzzo, ``{CMfinder--a covariance model based RNA
  motif finding algorithm},'' {\em Bioinformatics}, vol.~22, pp.~445--452, Feb.
  2006.

\bibitem{darty:2009gt}
K.~Darty, A.~Denise, and Y.~Ponty, ``{VARNA: Interactive drawing and editing of
  the RNA secondary structure.},'' {\em Bioinformatics (Oxford, England)},
  vol.~25, pp.~1974--1975, Aug. 2009.

\bibitem{fusy:2012ka}
{\'E}.~Fusy and P.~Clote, ``{Combinatorics of locally optimal RNA secondary
  structures},'' {\em Journal of mathematical biology}, vol.~68, pp.~341--375,
  Dec. 2012.

\bibitem{voss:2006iq}
B.~Vo{\ss}, R.~Giegerich, and M.~Rehmsmeier, ``{Complete probabilistic analysis
  of RNA shapes.},'' {\em BMC biology}, vol.~4, no.~1, p.~5, 2006.

\bibitem{Lorenz:2008gz}
W.~A. Lorenz, Y.~Ponty, and P.~Clote, ``{Asymptotics of RNA Shapes},'' {\em
  Journal of Computational Biology}, vol.~15, no.~1, pp.~31--63, 2008.

\bibitem{mandal2003}
M.~Mandal, B.~Boese, J.~E. Barrick, W.~C. Winkler, and R.~R. Breaker,
  ``{Riboswitches Control Fundamental Biochemical Pathways in Bacillus subtilis
  and Other Bacteria},'' {\em Cell}, vol.~113, pp.~577--586, May 2003.

\bibitem{janssen:2015cq}
S.~Janssen and R.~Giegerich, ``{The RNA shapes studio.},'' {\em Bioinformatics
  (Oxford, England)}, vol.~31, pp.~423--425, Feb. 2015.

\bibitem{mandal:2004ja}
M.~Mandal and R.~R. Breaker, ``{Gene regulation by riboswitches},'' {\em Nature
  Reviews Molecular Cell Biology}, vol.~5, pp.~451--463, June 2004.

\bibitem{mathews:1999jw}
D.~H. Mathews, J.~Sabina, M.~Zuker, and D.~H. Turner, ``{Expanded sequence
  dependence of thermodynamic parameters improves prediction of RNA secondary
  structure},'' {\em Journal of molecular biology}, vol.~288, pp.~911--940, May
  1999.

\bibitem{mandalboesebarrickwinklerbreaker}
M.~Mandal, B.~Boese, J.~Barrick, W.~Winkler, and R.~Breaker, ``Riboswitches
  control fundamental biochemical pathways in {B}acillus subtilis and other
  bacteria,'' {\em Cell}, vol.~113(5), pp.~577--586, 2003.

\bibitem{serganov:2004dq}
A.~Serganov, Y.-R. Yuan, O.~Pikovskaya, A.~Polonskaia, L.~Malinina, A.~T. Phan,
  C.~H{\"o}bartner, R.~Micura, R.~R. Breaker, and D.~J. Patel, ``{Structural
  basis for discriminative regulation of gene expression by adenine- and
  guanine-sensing mRNAs.},'' {\em Chemistry \& biology}, vol.~11,
  pp.~1729--1741, Dec. 2004.

\bibitem{rice:2000wr}
P.~Rice, I.~Longden, and A.~Bleasby, ``{EMBOSS: the European molecular biology
  open software suite},'' {\em Trends in Genetics}, 2000.

\bibitem{sankoff:1985wc}
D.~Sankoff, ``{Simultaneous solution of the RNA folding, alignment and
  protosequence problems},'' {\em SIAM Journal on Applied Mathematics}, 1985.

\bibitem{wilkinson:2006vd}
K.~A. Wilkinson, E.~J. Merino, and K.~M. Weeks, ``{Selective
  2{\textasciiacute}-hydroxyl acylation analyzed by primer extension (SHAPE):
  quantitative RNA structure analysis at single nucleotide resolution},'' {\em
  Nature protocols}, 2006.

\bibitem{mandal2004a}
M.~Mandal and R.~R. Breaker, ``{Adenine riboswitches and gene activation by
  disruption of a transcription terminator},'' {\em Nature structural \&
  molecular biology}, vol.~11, pp.~29--35, Dec. 2003.

\bibitem{freyhult.b07}
E.~Freyhult, V.~Moulton, and P.~Clote, ``Boltzmann probability of {RNA}
  structural neighbors and riboswitch detection.,'' {\em Bioinformatics},
  vol.~23, pp.~2054--2062, August 2007.

\bibitem{cloteloulorenz}
P.~Clote, F.~Lou, and W.~Lorenz, ``Maximum expected accuracy structural
  neighbors of an {RNA} secondary structure,'' {\em BMC Bioinformatics},
  vol.~13, p.~S6, April 2012.

\bibitem{ding.nar03}
Y.~Ding and C.~E. Lawrence, ``A statistical sampling algorithm for {RNA}
  secondary structure prediction.,'' {\em Nucleic. Acids. Res.}, vol.~31,
  pp.~7280--7301, 2003.

\bibitem{senter.po12}
E.~Senter, S.~Sheikh, I.~Dotu, Y.~Ponty, and P.~Clote, ``Using the fast fourier
  transform to accelerate the computational search for {RNA} conformational
  switches.,'' {\em PLoS. One.}, vol.~7, no.~12, p.~e50506, 2012.

\bibitem{haas:1991vl}
E.~S. Haas, D.~P. Morse, J.~W. Brown, F.~J. Schmidt, and N.~R. Pace,
  ``{Long-range structure in ribonuclease P RNA.},'' {\em Science (New York,
  N.Y.)}, vol.~254, pp.~853--856, Nov. 1991.

\bibitem{xia:RNA}
T.~Xia, J.~J.~SantaLucia, M.~Burkard, R.~Kierzek, S.~Schroeder, X.~Jiao,
  C.~Cox, and D.~Turner, ``Thermodynamic parameters for an expanded
  nearest-neighbor model for formation of {RNA} duplexes with {W}atson-{C}rick
  base pairs,'' {\em Biochemistry}, vol.~37, pp.~14719--35, 1999.

\bibitem{Freyhult.ab05}
E.~Freyhult, V.~Moulton, and P.~Gardner, ``Predicting {RNA} structure using
  mutual information.,'' {\em Appl. Bioinformatics}, vol.~4, no.~1, pp.~53--59,
  2005.

\bibitem{nussinovjacobson}
R.~Nussinov and A.~B. Jacobson, ``Fast algorithm for predicting the secondary
  structure of single stranded {RNA},'' {\em Proceedings of the National
  Academy of Sciences, USA}, vol.~77, no.~11, pp.~6309--6313, 1980.

\bibitem{highambarycentricinterpolation}
N.~Higham, ``The numerical stability of barycentric {L}agrange interpolation,''
  {\em IMA J. Numer. Anal.}, vol.~24, pp.~547--556, 2004.

\bibitem{cormen}
T.~Cormen, C.~Leiserson, and R.~Rivest, {\em Algorithms}.
\newblock McGraw-Hill, 1990.
\newblock 1028 pages.

\bibitem{wolynes.ptam05}
P.~G. Wolynes, ``Energy landscapes and solved protein-folding problems.,'' {\em
  Philos. Transact. A. Math. Phys. Eng. Sci.}, vol.~363, pp.~453--464, February
  2005.

\bibitem{bryngelson.p95}
J.~D. Bryngelson, J.~N. Onuchic, N.~D. Socci, and P.~G. Wolynes, ``Funnels,
  pathways, and the energy landscape of protein folding: a synthesis.,'' {\em
  Proteins.}, vol.~21, pp.~167--195, March 1995.

\bibitem{blast}
S.~F. Altschul, W.~Gish, W.~Miller, E.~W. Myers, and D.~J. Lipman, ``Basic
  local alignment search tool.,'' {\em J. Mol. Biol.}, vol.~215, pp.~403--410,
  October 1990.

\bibitem{Gardner.nar11}
P.~P. Gardner, J.~Daub, J.~Tate, B.~L. Moore, I.~H. Osuch, S.~Griffiths-Jones,
  R.~D. Finn, E.~P. Nawrocki, D.~L. Kolbe, S.~R. Eddy, and A.~Bateman, ``Rfam:
  {Wikipedia}, clans and the "decimal" release.,'' {\em Nucleic. Acids. Res.},
  vol.~39, pp.~D141--D145, January 2011.

\bibitem{flamm}
C.~Flamm, W.~Fontana, I.~Hofacker, and P.~Schuster, ``{RNA} folding at
  elementary step resolution,'' {\em RNA}, vol.~6, pp.~325--338, 2000.

\bibitem{altschulErikson:dinucleotideShuffle}
S.~Altschul and B.~Erikson, ``Significance of nucleotide sequence alignments:
  {A} method for random sequence permutation that preserves dinucleotide and
  codon usage,'' {\em Mol. Biol. Evol}, vol.~2(6), pp.~526--538, 1985.

\bibitem{gerdes.arg97}
K.~Gerdes, A.~P. Gultyaev, T.~Franch, K.~Pedersen, and N.~D. Mikkelsen,
  ``Antisense {RNA}-regulated programmed cell death.,'' {\em Annu. Rev.
  Genet.}, vol.~31, pp.~1--31, 1997.

\bibitem{lecuyercrothers}
K.~Lecuyer and D.~Crothers, ``The {L}eptomonas collosoma spliced leader {RNA}
  can switch between two alternate structural forms,'' {\em Biochemistry},
  vol.~32(20), pp.~5301--5311, 1993.

\bibitem{lorenz}
R.~Lorenz, C.~Flamm, and I.~L. Hofacker, ``{2D Projections of RNA Folding
  Landscapes.},'' {\em GCB}, 2009.

\bibitem{senter.jmb14}
E.~Senter, I.~Dotu, and P.~Clote, ``{RNA} folding pathways and kinetics using
  {2D} energy landscapes,'' {\em J Math Biol}, 2014.

\bibitem{hofacker:RNAbor2D}
R.~Lorenz, C.~Flamm, and I.~Hofacker, ``2{D} projections of {RNA} folding
  landscapes,'' in {\em German Conference on Bioinformatics 2009} (I.~Grosse,
  S.~Neumann, S.~Posch, F.~Schreiber, and P.~Stadler, eds.), vol.~157 of {\em
  Lecture Notes in Informatics}, pp.~11--20, 2009.

\bibitem{hochrein.jacs13}
L.~M. Hochrein, M.~Schwarzkopf, M.~Shahgholi, P.~Yin, and N.~A. Pierce,
  ``Conditional {Dicer} substrate formation via shape and sequence transduction
  with small conditional {RNAs}.,'' {\em J. Am. Chem. Soc.}, vol.~135,
  pp.~17322--17330, November 2013.

\bibitem{wachsmuth.nar13}
M.~Wachsmuth, S.~Findeiss, N.~Weissheimer, P.~F. Stadler, and M.~Morl, ``De
  novo design of a synthetic riboswitch that regulates transcription
  termination.,'' {\em Nucleic. Acids. Res.}, vol.~41, pp.~2541--2551, February
  2013.

\bibitem{synthetichammerheads}
I.~Dotu, J.~Garcia-Martin, B.~Slinger, V.~Mechery, M.~Meyer, and P.~Clote,
  ``Complete {RNA} inverse folding: computational design of functional
  hammerhead ribozymes,'' {\em Nucleic Acids Res.}, 2014.
\newblock in press.

\bibitem{zadeh.jcc11}
J.~N. Zadeh, B.~R. Wolfe, and N.~A. Pierce, ``Nucleic acid sequence design via
  efficient ensemble defect optimization.,'' {\em J. Comput. Chem.}, vol.~32,
  pp.~439--452, February 2011.

\bibitem{gruber08}
A.~Gruber, R.~Lorenz, S.~Bernhart, R.~Neubock, and I.~Hofacker, ``The vienna
  rna websuite,'' {\em Nucleic Acids Research}, vol.~36, pp.~70--74, 2008.

\bibitem{garcia.jbcbb13}
J.~Garcia-Martin, P.~Clote, and I.~Dotu, ``{RNAiFold}: A constraint programming
  algorithm for {RNA} inverse folding and molecular design,'' {\em J Bioinform
  Comput Biol}, vol.~11, no.~2, p.~1350001, 2013.
\newblock in press.

\bibitem{garciamartin.nar13}
J.~A. Garcia-Martin, P.~Clote, and I.~Dotu, ``{RNAiFold}: a web server for
  {RNA} inverse folding and molecular design.,'' {\em Nucleic. Acids. Res.},
  vol.~41, pp.~W465--W470, July 2013.

\bibitem{bujotzek.jcam11}
A.~Bujotzek, M.~Shan, R.~Haag, and M.~Weber, ``Towards a rational spacer design
  for bivalent inhibition of estrogen receptor.,'' {\em J. Comput. Aided. Mol.
  Des.}, vol.~25, pp.~253--262, March 2011.

\bibitem{fasting.acie12}
C.~Fasting, C.~A. Schalley, M.~Weber, O.~Seitz, S.~Hecht, B.~Koksch,
  J.~Dernedde, C.~Graf, E.~W. Knapp, and R.~Haag, ``Multivalency as a chemical
  organization and action principle.,'' {\em Angew. Chem. Int. Ed. Engl.},
  vol.~51, pp.~10472--10498, October 2012.

\bibitem{senter:2015fq}
E.~Senter and P.~Clote, ``{Fast, Approximate Kinetics of RNA Folding},'' {\em
  Journal of computational biology : a journal of computational molecular cell
  biology}, vol.~22, pp.~124--144, Feb. 2015.

\bibitem{hobartner.jmb03}
C.~Hobartner and R.~Micura, ``Bistable secondary structures of small {RNAs} and
  their structural probing by comparative imino proton {NMR} spectroscopy.,''
  {\em J. Mol. Biol.}, vol.~325, pp.~421--431, January 2003.

\bibitem{vieregg.mp06}
J.~R. Vieregg and I.~Tinoco, ``Modelling {RNA} folding under mechanical
  tension.,'' {\em Mol. Phys.}, vol.~104, pp.~1343--1352, April 2006.

\bibitem{flammhofacker}
C.~Flamm, I.~Hofacker, P.~Stadler, and M.~Wolfinger, ``Barrier trees of
  degenerate landscapes,'' {\em Z. Phys. Chem.}, vol.~216, pp.~155--173, 2002.

\bibitem{huang.bb14}
J.~Huang and B.~Voss, ``Analysing {RNA}-kinetics based on folding space
  abstraction.,'' {\em BMC. Bioinformatics}, vol.~15, p.~60, 2014.

\bibitem{zhao.jcp11}
P.~Zhao, W.~Zhang, and S.~J. Chen, ``Cotranscriptional folding kinetics of
  ribonucleic acid secondary structures.,'' {\em J. Chem. Phys.}, vol.~135,
  p.~245101, December 2011.

\bibitem{morganhiggsbarrier}
S.~Morgan and P.~Higgs, ``Barrier heights between ground states in a model of
  {RNA} secondary structure,'' {\em J. Phys. A: Math. Gen.}, vol.~31,
  pp.~3153--3170, 1998.

\bibitem{flamm.r01}
C.~Flamm, I.~L. Hofacker, S.~Maurer-Stroh, P.~F. Stadler, and M.~Zehl, ``Design
  of multistable {RNA} molecules.,'' {\em RNA.}, vol.~7, pp.~254--265, February
  2001.

\bibitem{shapiro.jmb01}
B.~A. Shapiro, D.~Bengali, W.~Kasprzak, and J.~C. Wu, ``{RNA} folding pathway
  functional intermediates: their prediction and analysis.,'' {\em J. Mol.
  Biol.}, vol.~312, pp.~27--44, September 2001.

\bibitem{kucharik:2014ft}
M.~Kuchar{\'\i}k, I.~L. Hofacker, P.~F. Stadler, and J.~Qin, ``{Basin Hopping
  Graph: a computational framework to characterize RNA folding landscapes.},''
  {\em Bioinformatics}, vol.~30, pp.~2009--2017, July 2014.

\bibitem{thachuk.psb10}
C.~Thachuk, J.~Manuch, A.~Rafiey, L.~A. Mathieson, L.~Stacho, and A.~Condon,
  ``An algorithm for the energy barrier problem without pseudoknots and
  temporary arcs.,'' {\em Pac Symp Biocomput.}, vol.~0, no.~O, p.~O,
  2010:108-19.

\bibitem{wolfingerstadler:kinetics}
M.~Wolfinger, W.~Svrcek-Seiler1, C.~Flamm, and P.~Stadler, ``Efficient
  computation of {RNA} folding dynamics,'' {\em J Phys. A: {Math.} {G}en.},
  vol.~37, pp.~4731--4741, 2004.

\bibitem{wuchtyfontanahofackerschuster}
S.~Wuchty, W.~Fontana, I.~Hofacker, and P.~Schuster, ``Complete suboptimal
  folding of {RNA} and the stability of secondary structures,'' {\em
  Biopolymers}, vol.~49, pp.~145--164, 1999.

\bibitem{tang.jcb05}
X.~Tang, B.~Kirkpatrick, S.~Thomas, G.~Song, and N.~M. Amato, ``Using motion
  planning to study {RNA} folding kinetics.,'' {\em J. Comput. Biol.}, vol.~12,
  no.~6, pp.~862--881, 2005.

\bibitem{flammphd}
C.~Flamm, {\em Kinetic Folding of {RNA}}.
\newblock PhD thesis, University of Vienna, 1998.
\newblock Department of Chemistry.

\bibitem{gillespiestochasticsimulation1}
D.~Gillespie, ``A general method for numerically simulating the stochastic time
  evolution of coupled chemical reactions,'' {\em J Comp Phys}, vol.~22,
  no.~403, pp.~403--434, 1976.

\bibitem{aviram.amb12}
I.~Aviram, I.~Veltman, A.~Churkin, and D.~Barash, ``Efficient procedures for
  the numerical simulation of mid-size {RNA} kinetics.,'' {\em Algorithms. Mol.
  Biol.}, vol.~7, no.~1, p.~24, 2012.

\bibitem{dykeman:2015bxa}
E.~C. Dykeman, ``{An implementation of the Gillespie algorithm for RNA kinetics
  with logarithmic time update},'' {\em Nucleic Acids Research}, vol.~43,
  pp.~5708--5715, July 2015.

\bibitem{xayaphoummine.nar05}
A.~Xayaphoummine, T.~Bucher, and H.~Isambert, ``Kinefold web server for
  {RNA}/{DNA} folding path and structure prediction including pseudoknots and
  knots.,'' {\em Nucleic. Acids. Res.}, vol.~33, pp.~W605--W610, July 2005.

\bibitem{danilova.jbcb06}
L.~V. Danilova, D.~D. Pervouchine, A.~V. Favorov, and A.~A. Mironov,
  ``{RNAKinetics}: a web server that models secondary structure kinetics of an
  elongating {RNA}.,'' {\em J. Bioinform. Comput. Biol.}, vol.~4, pp.~589--596,
  April 2006.

\bibitem{geis.jmb08}
M.~Geis, C.~Flamm, M.~T. Wolfinger, A.~Tanzer, I.~L. Hofacker, M.~Middendorf,
  C.~Mandl, P.~F. Stadler, and C.~Thurner, ``Folding kinetics of large
  {RNAs}.,'' {\em J. Mol. Biol.}, vol.~379, pp.~160--173, May 2008.

\bibitem{lai.r13}
D.~Lai, J.~R. Proctor, and I.~M. Meyer, ``On the importance of
  cotranscriptional {RNA} structure formation.,'' {\em RNA.}, vol.~19,
  pp.~1461--1473, November 2013.

\bibitem{turner.nar10}
D.~H. Turner and D.~H. Mathews, ``{NNDB}: the nearest neighbor parameter
  database for predicting stability of nucleic acid secondary structure.,''
  {\em Nucleic. Acids. Res.}, vol.~38, pp.~D280--D282, January 2010.

\bibitem{anderson.b13}
J.~W. Anderson, P.~A. Haas, L.~A. Mathieson, V.~Volynkin, R.~Lyngso, P.~Tataru,
  and J.~Hein, ``Oxfold: kinetic folding of {RNA} using stochastic context-free
  grammars and evolutionary information.,'' {\em Bioinformatics}, vol.~29,
  pp.~704--710, March 2013.

\bibitem{meyermfpt}
C.~Meyer, ``The role of the group inverse in the theory of finite {M}arkov
  chains,'' {\em SIAM Rev.}, vol.~17, no.~46, pp.~443--464, 1975.

\bibitem{cloteBackofen:book}
P.~Clote and R.~Backofen, {\em Computational Molecular Biology: An
  Introduction}.
\newblock John Wiley \& Sons, 2000.
\newblock 279 pages.

\bibitem{metropolis:MonteCarlo}
N.~Metropolis, A.~Rosenbluth, M.~Rosenbluth, A.~Teller, and E.~Teller,
  ``Equation of state calculations by fast computing machines,'' {\em J. Chem.
  Phys.}, vol.~21, pp.~1087--1092, 1953.

\bibitem{waterman:book}
M.~Waterman, {\em Introduction to Computational Biology: {M}aps, Sequences and
  Genomes}.
\newblock Chapman \& Hall/CRC Press, 1995.

\bibitem{anfinsen}
C.~B. Anfinsen, ``Principles that govern the folding of protein chains,'' {\em
  Science}, vol.~181, pp.~223--230, 1973.

\bibitem{matrixtheory}
J.~N. Franklin, {\em Matrix Theory}.
\newblock Dover Publications, Mineola, New York, 2000.
\newblock 292 pages.

\bibitem{fftw05}
M.~Frigo and S.~G. Johnson, ``The design and implementation of {FFTW3},'' {\em
  Proceedings of the IEEE}, vol.~93, no.~2, pp.~216--231, 2005.
\newblock Special issue on ``Program Generation, Optimization, and Platform
  Adaptation''.

\bibitem{wuchty.b99}
S.~Wuchty, W.~Fontana, I.~L. Hofacker, and P.~Schuster, ``Complete suboptimal
  folding of {RNA} and the stability of secondary structures,'' {\em
  Biopolymers}, vol.~49, pp.~145--165, 1999.

\bibitem{tang.jmb08}
X.~Tang, S.~Thomas, L.~Tapia, D.~P. Giedroc, and N.~M. Amato, ``Simulating
  {RNA} folding kinetics on approximated energy landscapes.,'' {\em J. Mol.
  Biol.}, vol.~381, pp.~1055--1067, September 2008.

\bibitem{gusarov:1999uu}
I.~Gusarov and E.~Nudler, ``{The mechanism of intrinsic transcription
  termination.},'' {\em Molecular cell}, vol.~3, pp.~495--504, Apr. 1999.

\bibitem{yarnell:1999wt}
W.~S. Yarnell and J.~W. Roberts, ``{Mechanism of intrinsic transcription
  termination and antitermination.},'' {\em Science (New York, N.Y.)},
  vol.~284, pp.~611--615, Apr. 1999.

\bibitem{barrick:2007gw}
J.~E. Barrick and R.~R. Breaker, ``{The distributions, mechanisms, and
  structures of metabolite-binding riboswitches.},'' {\em Genome Biology},
  vol.~8, no.~11, p.~R239, 2007.

\bibitem{nawrocki:2014uy}
E.~P. Nawrocki, S.~W. Burge, and A.~Bateman, ``{Rfam 12.0: updates to the RNA
  families database},'' {\em Nucleic acids {\ldots}}, 2014.

\bibitem{ding:2014ex}
Y.~Ding, W.~A. Lorenz, I.~Dotu, E.~Senter, and P.~Clote, ``{Computing the
  Probability of RNA Hairpin and Multiloop Formation},'' {\em Journal of
  computational biology : a journal of computational molecular cell biology},
  vol.~21, pp.~201--218, Mar. 2014.

\end{thebibliography}


\end{document}
