%!TEX root = ../main.tex

\chapter{FFTbor}

\label{FFTbor}

\lhead{Chapter X. \emph{FFTbor}}

Using complex roots of unity and the Fast Fourier Transform, we
design a new thermodynamics-based algorithm, \fftbor, that
computes the Boltzmann probability that secondary structures differ
by $k$ base pairs from an arbitrary initial structure of a given RNA
sequence. The algorithm, which runs in quartic time $O(n^4)$ and quadratic space
$O(n^2)$, is used to determine the
correlation between kinetic folding speed and the {\em ruggedness}
of the energy landscape, and to predict the location of riboswitch
expression platform candidates. A web server is available at
\url{http://bioinformatics.bc.edu/clotelab/FFTbor/}.

\section*{Introduction}

In \citep{Freyhult.b07}, we developed a dynamic programming algorithm,
{\tt RNAbor}, pronounced {\em RNA neighbor}, which simultaneously computes for
each integer $k$, the Boltzmann probability $p_k = \frac{Z_k}{Z}$
of the subensemble of structures
whose base pair distance to a given {\em initial}, or
{\em reference}, structure \strSt is $k$.\footnote{As later
explained, $Z$ denotes the partition function, defined as the sum of
all Boltzmann factors $\exp(-E(S)/RT)$, over all secondary structures $S$
of a given RNA sequence, and $R$ denotes the  universal
gas constant and $T$ absolute temperature. Similarly $Z_k$ denotes the
sum of all Boltzmann factors of all structures $S$, whose base pair distance
to the initial structure \strSt is exactly $k$.}
{\tt RNAbor} stores the value of the (partial)
partition functions $Z_k(i,j)$ for all $1 \leq i \leq j \leq n$ and
$0 \leq k \leq n$, each of which requires quadratic time to compute.
Thus it follows that {\tt RNAbor} runs in time $O(n^5)$ and space
$O(n^3)$, which severely limits its applicability to genomic annotation.
This restriction is somewhat mitigated by the fact that
in \citep{CloteLouLorenz}, we showed how to use sampling
\citep{Ding.nar03} to efficiently approximate
{\tt RNAbor} in cubic time $O(n^3)$ and quadratic space $O(n^2)$,
{\em provided} that the starting structure \strSt is the minimum free
energy (MFE) structure. We expect that a more efficient version of
{\tt RNAbor} could be used in applications in genomics and synthetic
biology, to detect potential conformational switches --
RNA sequences containing two or more (distinct) metastable structures.

In this paper, we describe a radically different algorithm, \fftbor,
prounounced {\em FFT neighbor},
that uses polynomial interpolation to compute the
coefficients $p_0,\ldots,p_{n-1}$ of the polynomial
\begin{eqnarray}
\label{eq:PofX}
p(x) = p_0 + p_1 x + p_2 x^2 + \cdots + p_{n-1} x^{n-1},
\end{eqnarray}
where $p_k$ is defined by $p_k = \frac{Z_k}{Z}$.
Due to severe numerical instability issues in both the Lagrange
interpolation formula and in Gaussian elimination, we employ
the Fast Fourier Transform (FFT) to compute the inverse Discrete Fourier
Transform (DFT) on values $y_0,\ldots,y_{n-1}$, where $y_k = p(\omega^k)$ and
$\omega = e^{2 \pi i/n}$ is the principal $n$th complex root of unity and
$p(x)$ is defined in (\ref{eq:PofX}). This
gives rise to an improved version of {\tt RNAbor}, denoted \fftbor,
which runs in time $O(n^4)$ and space $O(n^2)$.
Once two metastable structures $S_1,S_2$ are identified, we can
subsequently evaluate the feasibility of transition between
structures $S_1$ and $S_2$,
by computing the {\em barrier energy} using algorithms, such as that
described in Dotu et al.  \citep{Dotu.nar10} or Flamm et al.
\citep{Flamm.r01}.





\subsection*{Background}

Let $\seq = s_1,\ldots,s_n$ denote an RNA sequence, i.e. a sequence
of letters in the alphabet of nucleotides $\{A,C,G,U\}$. A
secondary structure $S$ is a set of base pairs
$(i,j)$, where $1 \le i \le i+\theta < j \le n$ and $\theta \ge 0$
represents the minimum number of unpaired nucleotides in a hairpin loop
(due to steric constraints, $\theta$ is usually taken to be $3$),
such that if $(i,j)$ and $(x,y)$ both belong to $S$, then $i=x \iff j=y$
(a nucleotide is involved in at most one base pair)
and $i<x<j \iff i<y<j$ (no pseudoknots are allowed).

The secondary structure $S$ is {\em compatible}
with \seq if for every base pair $(i,j)$ in $S$, the pair
$(s_i,s_j)$ is contained in the set
$\mathbb{B} = \{(A,U), (U,A), (G,C), (C,G), (G,U), (U,G)\}$
of six Watson-Crick and wobble base pairs. Often we write that
$S$ is a secondary structure {\em on} \seq, or equivalently,
a secondary structure {\em of} \seq,
in place of stating that $S$ is compatible with \seq.
Throughout this paper, by {\em structure}, we always mean a
secondary structure which is compatible with an arbitrary, but fixed
RNA sequence $\seq=s_1,\ldots,s_n$.

Given two secondary structures $S,T$ on \seq, we
define the base pair distance $d_{BP}$ between $S$ and $T$
to be the number of base pairs that they have that are not in
common, i.e.
\begin{equation}
d_{BP}(S,T) = |S \cup T| - |S \cap T|.
\end{equation}
Structures $S,T$ are said to be $k$-{\em neighbors} if $d_{BP}(S,T) = k$.


For $1\leq i \leq j \leq n$, let
$S_{[i,j]}$ denote the restriction of $S$ to interval
$[i,j]$ of \seq, i.e. the set of base pairs
%
$S_{[i,j]} = \{ (x,y) \,:\, i \leq  x < y \leq j,  (x,y) \in S \}$.
%
The notion of $k$-neighbor can also be applied to restrictions of secondary
structures; i.e. a secondary structure $T_{[i,j]}$ is a
{\em $k$-neighbor} of $S_{[i,j]}$ if
\[
d_{BP}(S_{[i,j]},T_{[i,j]})= |\{ (x,y): i \leq x<y\leq j,
(x,y) \in S-T \mbox{ or } (x,y) \in T-S \}| = k.
\]
In the following,
we often omit the sequence \seq and initial secondary structure \strSt in our
notation, since these are arbitrary, but fixed.
In particular, we write $\z{k}{i}{j} = \z{k}{i}{j}(\seq,S^*)$ --
see following paragraph for definitions.


Given an RNA sequence $\seq=s_1,\ldots,s_n$ and compatible secondary structure
\strSt, let ${\bf Z}^k$ denote the sum
of the Boltzmann factors $\exp(-E(S)/RT)$ of all $k$-neighbors $S$ of  \strSt;
i.e.
\[
{\bf Z}^k = \z{k}{1}{n} =
\sum_{\substack{S \text{ such that }\\ d_{BP}(S,S^*)=k}}
e^{\frac{-E(S)}{RT}}
\]
where $E(S)$ denotes the Turner (nearest neighbor)
energy \citep{turner,xia:RNA}
of $S$, $R = 0.00198$ kcal/mol denotes the universal
gas constant and $T$ denotes absolute temperature.
Since the maximum base pair distance between a given initial
structure \strSt and any other structure $S$ on RNA sequence
$\seq = s_1,\ldots,s_n$ must satisfy
\begin{eqnarray}
\label{eq:upperBound}
d_{BP}(S,S^*) \leq |S^*| + \lfloor \frac{n-\theta}{2} \rfloor \leq n
\end{eqnarray}
it follows that the full partition function
\begin{eqnarray}
\label{eq:mcCaskillPartitionFun}
{\bf Z}=\z{}{1}{n} = \sum_{k=0}^n \z{k}{1}{n}.
\end{eqnarray}
Moreover, since $\theta=3$,
we need to compute at most the values
$\z{0}{1}{n},\ldots,\z{n-1}{1}{n}$ -- this observation will later prove useful.
The Boltzmann probability
$P[d_{BP}(S,S^*) = k]$
that a secondary structure $S$ has base pair distance $k$ from the initial
structure \strSt can be defined from the partition function by
\[
%p(k) :=\frac{{\bf Z}^k}{{\bf Z}}.
p(k) :=\frac{\z{k}{1}{n}}{\z{}{1}{n}}.
\]
By graphing the probabilities $p_k$ as a function of $k$,
we expect to see one or more peaks at base pair distance $k$ when
there is a meta-stable (low energy) structure $S$
at base pair distance $k$ from \strSt.
See Figure 1 for an illustration.

\subsection*{Recursions for structural neighbors}

For the rest of the paper, we consider both \seq as well as the
secondary structure \strSt on \seq to be fixed. We now recall the
recursions from Freyhult et al. \citep{Freyhult.ab05} to determine
the partition function $\z{k}{i}{j}$ with
respect to the Nussinov-Jacobson
energy $E_0$ model \citep{nussinovJacobson}, defined by
$-1$ times the number of base pairs; i.e. $E_0(S) = -1 \cdot |S|$.
Although we describe here the recursions for the Nussinov-Jacobson
model, for the sake of
simplicity of exposition, both {\tt RNAbor}
\citep{Freyhult.ab05} as well as our current software \fftbor,
concern the Turner energy model, consisting of free energy parameters for
stacked bases, hairpins, bulges, internal loops and multiloops. The full
recursions for \fftbor are described for the
the Turner energy model in the appendix.

The base case for $\z{k}{i}{j}$ is given by
\begin{equation}
  \z{0}{i}{j} = 1,\, \mbox{ for $i \le j$,}
  \label{eq:Zinit1}
\end{equation}
since the only $0$-neighbor to a structure \strSt
is the structure \strSt itself, and
\begin{equation}
  \z{k}{i}{j} = 0,\, \mbox{for $k > 0$, $i \le j \leq
    i+\theta$},
  \label{eq:Zinit2}
\end{equation}
since the empty structure is the only possible structure for a
sequence shorter than $\theta + 2$ nucleotides, and so there are no
$k$-neighbors for $k>0$. The recursion used to compute
$\z{k}{i}{j}$ for $k > 0$ and $j > i+\theta$ is
\begin{equation}
  \z{k}{i}{j} = \z{k -b_0}{i}{j-1} +
  \hspace{-5pt} \sum_{\substack{ (s_r,s_j) \in \mathbb{B}, \\
  i \leq r < j }} \sum_{w+w'=k - b(r)} \hspace{-10pt}
  \exp(-E_0(r,j)/RT) \cdot \z{w}{i}{r-1} \z{w'}{r+1}{j-1},
  \label{eq:N}
\end{equation}
where $E_0(r,j) = -1$ if positions $r,j$ can pair in sequence \seq,
and otherwise $E_0(r,j) = +\infty$. Additionally,
$b_0 = 1$ if $j$ is base-paired
in $S^*_{[i,j]}$ and $0$ otherwise, and
$b(r)=d_{BP}(S^*_{[i,j]}, S^*_{[i,r-1]} \cup S^*_{[r+1,j-1]} \cup\{(r,j)\})$.
This holds since in a secondary
structure $T_{[i,j]}$ on $s_i,\ldots,s_j$ that is a $k$-neighbor of
$S^*_{[i,j]}$,
either nucleotide $j$ is unpaired in $[i,j]$ or it is
paired to a nucleotide $r$ such that $i \leq r < j$. In this
latter case it is enough to study the smaller sequence segments
$[i,r-1]$ and $[r+1,j-1]$ noting that, except for $(r,j)$,
base pairs outside of these regions are not allowed, since there
are no pseudoknots. In addition,
for $d_{BP}(S^*_{[i,j]},T_{[i,j]}) = k$ to hold,
it is necessary for $w+w' = k -b(r)$ to hold, where $w =
d_{BP}(S^*_{[i,r-1]},T_{[i,r-1]})$ and $w' =
d_{BP}(S^*_{[r+1,j-1]},T_{[r+1,j-1]})$, since $b(r)$ is the
number of base pairs that differ between $S^*_{[i,j]}$ and a
structure $T_{[i,j]}$, due to the introduction of the base pair
$(r,j)$.

\section*{Methods}

Given RNA sequence \seq and compatible initial structure \strSt,
we define the {\em polynomial}
\begin{eqnarray}
\label{eq:polynomialYann}
\Z(x) = \sum_{k=0}^n c_k x^k
\end{eqnarray}
where coefficients $c_k=\z{k}{1}{n}$. Moreover, because of
(\ref{eq:upperBound}) and the fact that the minimum number of
unpaired bases in a hairpin loop $\theta$ is $3$, we know that $c_n=0$,
so that $\Z(x)$ is a polynomial of degree strictly less than $n$.
If we evaluate the polynomial $\Z(x)$ for $n$ distinct values
\begin{eqnarray}
\label{eq:ZofX}
\Z(a_1) = y_1,\ldots, \Z(a_{n}) = y_{n},
\end{eqnarray}
then the Lagrange polynomial interpolation formula guarantees that
$\Z(x) = \sum_{k=1}^n y_k P_k(x)$, where the polynomials $P_k(x)$ have degree
at most $n-1$ and are given by the Lagrange formula
\begin{eqnarray}
\label{eq:LagrangeInterpolation}
P_k(x) = \frac{\prod_{i\ne k} (x-x_i)}{\prod_{i \ne k} (x_k-x_i)}.
\end{eqnarray}
Since the polynomials $P_k(x)$ can be explicitly computed, it follows that
we can compute the coefficients $c_k$ of polynomial $\Z(x)$. As we describe
below, the evaluation of $\Z(x)$ for a fixed value of $x$ can be done in
time $O(n^3)$ and space $O(n^2)$.  It follows that the coefficients
$c_k=\z{k}{1}{n}$ can be computed after
$n$ evaluations of $\Z(x)$, where the space for each evaluation of $\Z(x)$
is re-used; hence these evaluations can be performed in time $O(n^4)$ and space
$O(n^2)$. Finally,
Lagrange interpolation is clearly computable in time $O(n^3)$.
Although this approach is theoretically sound, there are severe
numerical stability issues related to the interpolation method
\citep{HighamBarycentricInterpolation},
the choice of values $a_1,\ldots,a_{n}$ in the interpolation,
and floating point arithmetic (round-off error) related to the
astronomically large values of the partition functions
$\z{k}{1}{n}$, for $0 \leq k < n$. After many unsuccessful
approaches including scaling (see Supplementary Information),
we obtained excellent results by
interpolating the polynomial $p(x)$, defined in equation~(\ref{eq:PofX}),
rather than the polynomial $\Z(x)$, defined in equation~(\ref{eq:ZofX}),
and performing interpolation with the Fast Fourier Transform (FFT) \citep{cormen}
where $\alpha_0,\ldots,\alpha_{n-1}$ are
chosen to be $n$th complex roots of unity,
$\alpha_k = e^{\frac{2 \pi k}{n}}$.
One
advantage of the FFT is that interpolation can be performed in $O(n \log n)$
time, rather than the cubic time required by using the Lagrange formula
(\ref{eq:LagrangeInterpolation}) or by Gaussian elimination. Fewer
numerical operations implies increased numerical stability in our application.
Details now follow.

\subsection*{Recursions to compute the polynomial $\Z_{i,j}(x)$}
\label{section:recursionsForPolynomialZij}

Given an initial secondary structure \strSt of a
given RNA sequence \seq, our goal is to compute
\begin{align}
\label{eq:defZk}
\z{k}{1}{n} &= \sum_{\substack{S \text{ such that }\\ d_{BP}(S,S^*)=k}}
e^{\frac{-E_0(S)}{RT}}
\end{align}
where $S$ can be any structure compatible with \seq.
As previously mentioned, the recurrence relation for {\tt RNAbor}
with respect to the Nussinov energy model $E_0$ is
\begin{align}
\label{eq:RNAborNussRecursion}
\z{k}{i}{j} &= \z{k-b_0}{i}{j-1} + \sum_{\substack{s_r s_j \in \mathbb{B},\\i\le r<j}}\left(e^{\frac{-E_0(r,j)}{RT}}\sum_{w+w'=k-b(r)}
\z{w}{i}{r-1}  \z{w'}{r+1}{j-1} \right)
\end{align}
where $E_0(r,j)=-1$ if $r$ and $j$ can base-pair and otherwise
$+\infty$, and
$b_0 = 1$ if $j$ is base paired in $S^*_{[i,j]}$ and $0$ otherwise, and
$b(r)=d_{BP}(S^*_{[i,j]}, S^*_{[i,r-1]} \cup S^*_{[r+1,j-1]} \cup\{(r,j)\})$.
The following theorem shows that an analogous recursion can be used to compute
the {\em polynomial} $\Z_{i,j}(x)$ defined by
\begin{eqnarray}
\label{eq:defZijpolynomial}
\Z_{i,j}(x) = \sum_{k=0}^n c_k(i,j) x^k
\end{eqnarray}
where
\[ c_k(i,j)= \z{k}{i}{j} =
\sum_{\substack{S \text{ such that }\\ d_{BP}(S,S^*_{[i,j]})=k}}
e^{\frac{-E_0(S)}{RT}}.
\]
Here, in the summation, $S$ runs over structures on $s_i,\ldots,s_j$, which
are $k$-neighbors of the restriction $S^*_{[i,j]}$ of initial structure
\strSt to interval $[i,j]$, and
$E_0(S)=-1 \cdot |S|$ denotes the Nussinov-Jacobson energy of $S$.
\medskip

\noindent
{\sc Theorem 1:} Let $s_1,\ldots,s_n$ be a given RNA sequence.
For any integers $1 \leq i \leq j \leq n$, let
\begin{eqnarray*}
\Z_{i,j}(x) = \sum_{k=0}^n c_k x^k
\end{eqnarray*}
where
\[ c_k(i,j)= \z{k}{i}{j}.
\]
Then  for $i\leq j \leq i+\theta$, $\Z_{i,j}(x)=1$ and for
$j>i+\theta$ we have the recurrence relation
\begin{eqnarray}
\label{eq:polynomialRecursionNussJac}
\Z_{i,j}(x) &= \Z_{i,j-1}(x) \cdot x^{b_0} +
\sum_{\substack{s_r s_j \in \mathbb{B},\\i\le r<j}}
\left(e^{\frac{-E_0(r,j)}{RT}}\cdot
\Z_{i,r-1}(x)\cdot\Z_{r+1,j-1}(x)\cdot x^{b(r)} \right).
\end{eqnarray}
\medskip
where $b_0 = 1$ if $j$ is base-paired in $S^*_{[i,j]}$ and $0$ otherwise, and
$b(r)=d_{BP}(S^*_{[i,j]}, S^*_{[i,r-1]} \cup S^*_{[r+1,j-1]} \cup\{(r,j)\})$.

\noindent
{\sc Proof:}
First, some notation is necessary. Recall that if $F$ is an arbitrary
polynomial [resp. analytic] function, then $[x^k] F(x)$
denotes the coefficient of $x^k$ [resp. the $k$th Taylor coefficient in the
Taylor expansion of $F$] -- for instance, in equation~(\ref{eq:PofX}),
$[x^k]p(x) = p_k$, and in equation~(\ref{eq:ZofX}), $[x^k]\Z(x) = c_k(i,j)$.

By definition, it is clear that $\Z_{i,j}(x)=1$ if $i\leq j \leq i+\theta$,
where we recall that $\theta = 3$ is the minimum number of unpaired bases in
a hairpin loop.  For $j>i+\theta$, we have
\begin{eqnarray*}
[x^k] \Z_{i,j}(x) &=& c_k(i,j) = \z{k}{i}{j}\\
&=&\z{k-b_0}{i}{j-1} + \sum_{r=i}^{j-1} \sum_{k_0+k_1 = k-b(r)}
e^{\frac{-E_0(r,j)}{RT}} \cdot \z{k_0}{i}{r-1} \cdot \z{k_1}{r+1}{j-1} \\
&=&
[x^{k-b_0}] \Z_{i,j-1}(x) +
\sum_{r=i}^{j-1} \sum_{k_0+k_1 = k-b(r)}
e^{\frac{-E_0(r,j)}{RT}} \cdot \left\{ [x^{k_0}] \Z_{i,r-1}(x) \right\} \cdot
\left\{ [x^{k_1}] \Z_{r-1,j-1}(x) \right\} \\
&=&
[x^{k-b_0}] \Z_{i,j-1}(x) +
\sum_{r=i}^{j-1} \sum_{k_0+k_1 = k-b(r)}
e^{\frac{-E_0(r,j)}{RT}} \cdot [x^{k_0+k_1}] \Z_{i,r-1}(x) \Z_{r-1,j-1}(x) .\\
\end{eqnarray*}
By induction, the proof of the theorem now follows. $\Box$
\medskip

Notice that if one were to compute all terms of the polynomial $\Z_{1,n}(x)$
by explicitly performing polynomial multiplications,
then the computation would require $O(n^5)$ time and $O(n^3)$ space.
Instead of explicitly performing polynomial expansion in {\em variable} $x$,
we instantiate $x$ to a fixed complex number $\alpha \in \mathbb{C}$, and apply
the following recursion for this instantiation:
\begin{eqnarray}
\label{eq:RNAborNussRecursionInstantiated}
\Z_{i,j}(\alpha) &= &\Z_{i,j-1}(\alpha) \cdot \alpha^{b_0} +
\sum_{\substack{(s_r,s_j) \in \mathbb{B},\\i\le r<j}}
\left(e^{\frac{-E_0(r,j)}{RT}}\cdot
\Z_{i,r-1}(\alpha)\cdot\Z_{r+1,j-1}(\alpha)\cdot \alpha^{b(r)} \right).
\end{eqnarray}
In this fashion, we can compute $\Z(\alpha)=\Z_{1,n}(\alpha)$ in
$O(n^3)$ time and $O(n^2)$ space. For $n$ distinct complex values
$\alpha_0,\ldots,\alpha_{n-1}$, we can compute and save only the
values $\Z(\alpha_0),\ldots, \Z(\alpha_{n-1})$, each time re-using the
$O(n^2)$ space for the next computation of $\Z(\alpha_k)$. It follows that
the computation resources used to determine the (column) vector
\begin{eqnarray}
\label{eq:defY}
{\bf Y} = (y_0,\ldots,y_{n-1})^T =
\left(
\begin{array}{l}
y_0\\
y_1\\
\vdots \\
y_{n-1}\\
\end{array}
\right)
\end{eqnarray}
where
$y_0=\Z(\alpha_0),\ldots, y_{n-1}=\Z(\alpha_{n-1})$ is thus
$O(n^4)$ are quartic time $O(n^4)$ and quadratic space $O(n^2)$.

\subsection*{Polynomial interpolation using the FFT}
Let $\omega = e^{2 \pi i /n}$ be the principal $n$th complex root of unity.
Recall that the
Vandermonde matrix $V_{n}$ is defined to be the
$n \times n$ matrix, whose $i,j$ entry is $\omega^{i \cdot j}$;
i.e.
$$
V_{n} = \left(
\begin{array}{rrrrr}
1&1&1&\cdots&1\\
1&\omega&\omega^2&\cdots&\omega^{n-1}\\
1&\omega^2&\omega^4&\cdots&\omega^{2(n-1)}\\
1&\omega^3&\omega^6&\cdots&\omega^{3(n-1)}\\
\vdots& \vdots& \vdots& \vdots& \vdots\\
1&\omega^{n-1}&\omega^{2(n-1)}&\cdots&\omega^{(n-1)(n-1)}\\
\end{array}
\right)
$$
The Fast Fourier Transform (FFT) is defined to be the $O(n \log n)$
algorithm to compute the Discrete Fourier Transform (DFT), defined
as the matrix product ${\bf Y} = V_{n} {\bf A}$:
$$
\left(
\begin{array}{l}
y_0\\
y_1\\
y_2\\
\vdots \\
y_{n-1}\\
\end{array}
\right)
= V_n \cdot
\left(
\begin{array}{l}
a_0\\
a_1\\
a_2\\
\vdots \\
a_{n-1}\\
\end{array}
\right)
$$
On page 837 of  \citep{cormen}, it is shown that the
$(i,j)$ entry of $V_n^{-1}$ is $\frac{\omega^{-j i}}{n}$
and that
\begin{eqnarray}
\label{eq:ajdef}
a_j &=&\frac{1}{n} \sum_{k=0}^{n-1} y_k \omega^{-kj}
\end{eqnarray}
for $j=0,\ldots,n-1$.

Since we defined $\bf Y$ in (\ref{eq:defY}) by ${\bf Y} =
(y_0,\ldots,y_{n-1})^T$, where
$y_0=\Z(\alpha_0),\ldots, y_{n-1}=\Z(\alpha_{n-1})$ and $\alpha_k = \omega^k
\exp(\frac{k \cdot 2\pi i}{n})$, it follows that the coefficients
$c_k=\z{k}{1}{n}$ in the polynomial
$\Z(x) = c_0 + c_1 x + \cdots + c_{n-1} x^{n-1}$ defined in
(\ref{eq:polynomialYann}) can be computed, at least in principle,
by using the FFT. It turns out, however, that the values of
$\z{k}{1}{n}$ are so astronomically large, that the ensuing numerical
instability makes even this approach infeasible for values of $n$
that exceed $56$ (data not shown).
Nevertheless, our approach can be modified as follows.
Define $\bf Y$ by ${\bf Y} = (y_1,\ldots,y_n)^T$, where
$y_1=\frac{\Z(\alpha_1)}{Z},\ldots, y_{n}=\frac{\Z(\alpha_{n})}{Z}$, and
$Z$ is the partition function defined in (\ref{eq:mcCaskillPartitionFun}).
Using the FFT to compute the inverse DFT, it follows from
(\ref{eq:ajdef}) that we can compute the probabilities $p_0,\ldots,p_{n-1}$
that are coefficients of the polynomial
$p(x)=p_0+p_1 x + \cdots + p_{n-1}x^{n-1}$
defined in equation~(\ref{eq:PofX}). For genomics applications, we are
only interested in the $m$ most significant digits of each $p_k$, as described
in the pseudocode below.
\medskip

\noindent
{\sc Algorithm} for \fftbor\hfill\break
This pseudocode computes the $m$ most significant digits
of probabilities $p_k = \frac{\z{k}{1}{n}}{{\bf Z}}$. \hfill\break
{\sc Input:} RNA sequence $\seq=s_1,\ldots,s_n$, and initial secondary
structure \strSt of \seq, and integer $m$. \hfill\break
{\sc Output:} Probabilities $p_k=\z{k}{1}{n}/{\bf Z}$ to $m$ significant digits for $k=0,\ldots,n-1$. \hfill\break
\smallskip
\mverbatim
1.  generate roots of unity $\omega^k$ for $k=0,\ldots,n-1$, where $\omega=\exp(\frac{2 \pi i}{n})$ and $i=\sqrt{-1}$
2.  note that the partition function ${\bf Z}=y_0=\Z(\omega^0)$
3.  for $k=0$ to $n-1$
4.    compute $y_k = \Z(\omega^k)$ using recursion (\ref{eq:RNAborNussRecursionInstantiated})
5.    $y_k = 10^m \cdot \frac{y_k}{{\bf Z}}$ //normalize $y_k$
6.  compute $P = (p_0,\ldots,p_{n-1})^T$ where $p_j =\frac{\sum_{k=0}^{n-1} a_k \omega^{-kj}}{n}$ by using FFT in (\ref{eq:ajdef})
7.  for $k=0$ to $n-1$
8.    $p_k = \lfloor 10^m \cdot p_k \rfloor \cdot \frac{1}{10^m}$
9.    //truncate to $m$ most significant digits
|mendverbatim
\medskip

\subsubsection*{Speed-up in our implementation of \fftbor}

In this subsection, we show that we need only evaluate the polynomial
$\Z(x)$, as defined in
equation~(\ref{eq:polynomialYann}), for $n/2$ of the complex $n$th roots
of unity. It is first necessary to recall the definition of complex
conjugate.
Recall that the complex conjugate of $z$ is denoted by $\overline{z}$;
i.e. if $z=a+bi$ where $a,b \in \mathbb{R}$ are real numbers and
$i = \sqrt{-1}$,  then $\overline{z} = a-bi$.
\smallskip

\noindent
{\sc Lemma 1:} If $\Z(x)$ is the complex polynomial defined in
equation~(\ref{eq:polynomialYann}), then for any complex $n$th root of
unity $\alpha$, it is the case that $\Z(\overline{\alpha}) =
\overline{\Z(\alpha)}$. In other words, if $\alpha$ is a complex $n$th root
of unity of the form $a+bi$, where $a,b \in \mathbb{R}$ and $b>0$, and
if $\Z(a+bi) = A + Bi$ where $A,B \in \mathbb{R}$, then it is the case that
\[
\Z(a - bi) = A - Bi.
\]
\smallskip

\noindent
{\sc Proof:}
Letting $i = \sqrt{-1}$, if  $\theta = \frac{2 \pi}{n}$, then
$\omega = e^{i \theta} = \cos(\theta) + i \sin(\theta)$
is the principal $n$th complex root of unity, and
$1=\omega^{0},\ldots,e^{(n-1)\cdot i \theta}=\omega^{n-1}$ together
constitute the complete collection of all
$n$th complex roots of unity -- i.e. the $n$ unique solutions of
of the equation $x^n -1 = 0$ over the field $\mathbb{C}$ of complex numbers.
Clearly, for any $1 \leq r < n$,
$e^{-i r \theta} = 1 \cdot e^{-i r \theta} =
e^{2 \pi i} \cdot e^{-i r \theta} = e^{i(2 \pi - r \theta)} =
e^{i(n \theta - r \theta)} = e^{i \theta (n - r)}$.
Moreover, if $\omega^r = e^{i r \theta} = a + b i$ where
$b>0$, then we have $e^{-i r \theta} = a-bi$. It follows that for any
$n$th root of unity of the form $a+bi$, where $b>0$, the number $a-bi$
is also an $n$th root of unity.

Recall that $\Z(x) = \sum_{k=0}^n c_k x^k$, where
$c_k\in \mathbb{R}$ are real numbers representing the partition function
$\z{k}{1}{n}$ over
all secondary structures of a given RNA sequence $s_1,\ldots,s_n$,
whose base pair distance from initial structure
\strSt is $k$. Thus, in order to prove the lemma, it suffices to show
that for all values $k=0,\ldots,n-1$, if $a+bi$ is a complex $n$th
root of unity, where $a,b \in \mathbb{R}$
and $b>0$, and if $(a+bi)^k = C+Di$ where $C,D \in \mathbb{R}$, {\em then}
$(a-bi)^k = C-Di$. Indeed, we have the following.

\begin{eqnarray*}
(a+bi)^m &=& \sum_{k=0}^m {m \choose k} a^{m-k}\cdot  (bi)^k \\
(bi)^k  &=& \left\{
\begin{array}{ll}
b^k &\mbox{if $k \equiv 0 \bmod 4$}\\
i b^k &\mbox{if $k \equiv 1 \bmod 4$}\\
-b^k &\mbox{if $k \equiv 2 \bmod 4$}\\
-i b^k &\mbox{if $k \equiv 3 \bmod 4$}\\
\end{array} \right.
\end{eqnarray*}
\begin{eqnarray*}
(a-bi)^m &=& \sum_{k=0}^m {m \choose k} a^{m-k} \cdot (-bi)^k \\
(-bi)^k &=& \left\{
\begin{array}{ll}
b^k &\mbox{if $k \equiv 0 \bmod 4$}\\
-ib^k &\mbox{if $k \equiv 1 \bmod 4$}\\
-b^k &\mbox{if $k \equiv 2 \bmod 4$}\\
ib^k &\mbox{if $k \equiv 3 \bmod 4$}\\
\end{array} \right.
\end{eqnarray*}
It follows that each term of the form
$a^{m-k} \cdot (bi)^k$, for $k=0,\ldots,m$, is the complex conjugate of
$a^{m-k} \cdot (-bi)^k$, and thus $(a+bi)^m$ is the complex conjugate of
$(a-bi)^m$. Since $\Z(a+bi)$ is a sum of terms of the form $c_k (a+bi)^k$,
it follows that $\Z(a-bi)$ is the complex conjugate of $\Z(a+bi)$.
This completes the proof of the lemma. \hfill  $\Box$
\medskip

Lemma 1 immediately entails that we need only evaluate $\Z(x)$ on $n/2$
many of the complex $n$th roots of unity -- namely, those of the form
$a+bi$, where $b \geq 0$. The remaining values of $\Z(x)$ are obtained by
taking conplex conjugates of the first $n/2$ values. This, along with a
precomputation of powers of the complex $n$th roots of unity, leads to an
enormous performance speed-up in our implementation of \fftbor.

\section*{Results}
\subsection*{Applications of \fftbor}
\label{section:applications}

In this section, we consider two applications of \fftbor:
{\em (i)} correlation between kinetic folding speed and the {\em ruggedness}
of the energy landscape near the minimum free energy structure,
{\em (ii)} computational detection of riboswitch
expression platform candidates.

\subsubsection*{Kinetic folding speed and energy landscape ruggedness}

The output of \fftbor, as shown in
Figure 2, is a probability distribution,
where the $x$-axis represents the base pair distance from an arbitrary,
but fixed secondary structure \strSt, and the $y$-axis represents the
Boltzmann probability $p(k) = \frac{Z_k}{Z}$ that a secondary structure
has base pair distance $k$ from \strSt. Arguably, this probability distribution
is an accurate one-dimensional projection of the rugged, high dimensional energy
landscape near structure \strSt,
of the sort artistically rendered in the well-known
energy landscape depicted in Figure 1  of \citep{Wolynes.ptam05}.
In the sequel, we may call the \fftbor probability distribution a
{\em structural neighbor profile}, or simply {\em structural profile}
\strSt.
A hypothesis behind theoretical work in biomolecular folding theory in
\citep{Bryngelson.p95}
is that kinetic folding slows down as the energy landscape becomes more
{\em rugged}. This is borne out in our computational experiments for RNA
using \fftbor, as reported
in Figure 2.

We randomly chose two TPP riboswitch
aptamers from the seed alignment for
Rfam family RF00059. The first sequence has EMBL accession code
BX842649.1/277414-277318 and is comprised of the 97 nt sequence
ACCUGACGCU AGGGGUGUUG GUGAAUUCAC CGACUGAGAA UAACCCUUUG AACCUGAUAG
AGAUAAUGCU CGCGCAGGGA AGCAAGAAUA GAAAGAU, while the second sequence
has EMBL accession code AACY022101973.1/389-487 and is comprised of the 99
nt sequence
UAUAAGUCCA AGGGGUGCCA AUUGGCUGAG AUGGUUUUAA CCAAUCCCUU
UGAACCUGAU CCGGUUAAUA CCGGCGUAGG AAUGGAUUUU CUCUACAGC.
Rfam consensus and minimum free energy structures for both sequences are
depicted in Figure 3.
Despite the fact that there is no sequence homology according to
pairwise BLAST \citep{BLAST}, this figure clearly demonstrates that
consensus and
minimum free energy structures closely resemble each other, and that the
structures of both TPP riboswitch aptamers are quite similar, with the
exception of the leftmost hairpin loop [resp. multiloop].
The MFE structures differ
from the consensus structures principally by the addition of base pairs not
determined by covariation in the Rfam alignment.
Indeed, if we let $S_0,S_1$
denote the Rfam consensus structure resp. MFE structure for the 97 nt
sequence with EMBL accession code BX842649.1/277414-277318, then
$S_0 \setminus S_1$ has $4$ base pairs, and $S_1\setminus S_0$ has $7$
base pairs. If we let $T_0,T_1$
denote the Rfam consensus structure resp. MFE structure for the 99 nt
sequence with EMBL accession code
AACY022101973.1/389-487, then
$T_0 \setminus T_1$ has $1$ base pair, and $T_1\setminus T_0$ has $5$
base pairs.

We ran \fftbor on each of the TPP riboswitch aptamer
sequences, with the MFE structure of each
sequence taken as the initial structure \strSt for that sequence.  For the
first sequence, BX842649.1/277414-277318, the \fftbor output
suggests that there are low energy structures
at a distance from the MFE structure, which might compete with the MFE
structure and hence slow the kinetics of folding. In contrast, for the
second sequence, AACY022101973.1/389-487, the \fftbor output suggests
that there are no such competing low energy structures, hence
the second sequence should fold more quickly than the first.

To test the hypothesis that folding is slower for rugged energy landscapes,
we ran the kinetic folding software, \kinfold \citep{flamm},
on each of the two TPP riboswitch aptamer sequences,
BX842649.1/277414-277318 and AACY022101973.1/389-487,
to determine the {\em mean first passage time} (MFPT) to
fold into the MFE structure, when starting from the empty structure.
In this computational
experiment, we took MFPT to be the average number of Monte Carlo steps
taken by \kinfold, each step consisting of the addition or removal
of a single base pair (or shift -- see \citep{flamm}), to fold the
empty structure into the MFE
structure, where the average was taken over $30$ runs, with an absolute
maximum number of Monte Carlo steps taken to be $500,000$.
The first sequence, BX842649.1/277414-277318, converged within $500,000$
steps only for 20 out of 30 runs. Assigning the maximum step count of
$500,000$ for the 10 runs that did not converge, we found a mean first
passage time of $311,075.06$ steps for this sequence.
The second sequence, AACY022101973.1/389-487, converged within $500,000$
steps in 29 out of 30 runs, and we found a mean first passage time of
$61,575.69$ steps for this sequence. From computational experiments of this
type, it is suggestive that \fftbor may prove useful in synthetic
biology,
where one would like to design rapidly folding RNA molecules that
fold into a designated target structure.\footnote{See
\citep{Shetty.jbe08,Knight.msb05,Waldminghaus.bc08,Zadeh.jcc11} for more
on synthetic biology.} In particular, one could use
{\tt RNAinverse} \citep{hofacker:ViennaWebServer},
{\tt RNA-SSD} \citep{Andronescu04},
{\tt INFO-RNA} \citep{Busch06},
or our recent constraint programming exact
solution of RNA inverse folding, {\tt RNAiFold}
(submitted, \url{http://bioinformatics.bc.edu/clotelab/RNAiFold/}),
to output a list of sequences,
whose minimum free energy structure is a designated target structure.
Subsequently, using \fftbor, one could prioritize sequences in terms
of \fftbor structural profile, on the grounds that sequences with a profile
similar to the right panel of Figure 2
are likely to fold more rapidly than those whose profile resembles
the left panel of Figure 2.

In order to more systematically determine the relation between kinetic
folding speed and the ruggedness of an energy landscape near the MFE structure,
we need to numerically quantify ruggedness. To this end, in the following
we define the notion of {\em expected base pair distance} to a designated
structure. Let \strSt be an arbitrary secondary structure of the RNA sequence
$\seq = a_1,\dots,a_n$.
The expected base pair distance to \strSt is defined by
\begin{eqnarray}
\label{eq:expectedBasePairDist}
E[ \{ d_{\mbox{\sc bp}}(S,S^*): S \in \mathbb{S}(a_1,\ldots,a_n)\} ]  =
\sum_{S} P(S) \cdot d_{\mbox{\sc bp}}(S,S^*)
\end{eqnarray}
where
$\mathbb{S}(a_1,\ldots,a_n)$ denotes the set of secondary structures for
$\seq = a_1,\dots,a_n$, $P(S) = \frac{\exp(-E(S)/RT)}{Z}$ is the Boltzmann
probability of $S$, and
$d_{\mbox{\sc bp}}(S,S^*)$ denotes base pair distance between $S$ and \strSt.
If we run \fftbor on an input sequence \seq and secondary structure
\strSt, then clearly
$E[ \{ d_{\mbox{\sc bp}}(S,S^*): S \in \mathbb{S}(a_1,\ldots,a_n)\} ]  =
\sum_{k} k \cdot p(k)$, where $p(k)=\frac{Z_k}{Z}$, obtained from the
program output.  If \strSt is the empty structure, then \fftbor output
is simply the probability distribution of the number of base pairs per
secondary structure, taken over the Boltzmann ensemble of all structures.

For the benchmarking assay, we took all 61 selenocysteine insertion sequence
(SECIS) sequences from the seed alignment of Rfam family RF00031
\citep{Gardner.nar11}. Average length was $64.32 \pm 2.83$ nt.
For each sequence, we ran both \fftbor and a Monte Carlo
folding algorithm, developed by E. Freyhult and P. Clote (unpublished).
Using the Monte Carlo algorithm, we
determined the mean first passage time (MFPT), defined as the average
taken over 50 runs, of the number of Monte Carlo steps taken to fold
the empty structure into the MFE structure, where an absolute upper bound
of 5 million steps was allowed in the simulation.  After unsuccessful
attempts due to {\em ruggedness} of the energy landscape near the MFE structure,
by using the Hartigan-Hartigan {\em dip} test of unimodality
\citep{hartiganDipTest},
expected base pair distance from MFE structure, total variation distance
between FFTbor output and the exponential distribution estimated by
the method of moments \citep{zar:biostatistics}, etc.,
we ran \fftbor
when starting from the empty structure (rather than the MFE structure)
as initial structure.
As mentioned above, in this case, \fftbor output is simply
the probability distribution
for the number of base pairs per structure, taken over the ensemble
of all secondary structure for the input RNA
sequence. Surprisingly, we found that there is a significant
correlation of  $0.48436192$ with one-tailed
$p$-value of $0.00018249$ between the
standard deviation of the \fftbor output (when starting from the
empty structure) and logarithm base $10$ of the mean first passage time.
Table~1 and Figure 4 explain this phenomenon in detail.

In the right panel of Figure 4, we
applied \fftbor to each of the two randomly chosen TPP riboswitch
aptamers BX842649.1/277414-277318 and AACY022101973.1/389-487, starting
from the empty reference structure $S^*=\emptyset$.
The mean for the \fftbor structural profile near the empty
structure is $mu_1=23.0203$  [resp. $\mu_2=27.5821$], the
standard deviation $\sigma$ for the \fftbor structural profile
is $\sigma_1=2.22528791$  [resp. $\sigma_2=1.98565959$], and the \kinfold MFPT is
$311,075.06$ [resp. $61,575.69$] for the TPP riboswitch aptamer
AB030643.1/4176-4241 [resp.  AL645723.11/192421-192359]. This anecdotal evidence supports the hypothesis that small standard deviation in \fftbor distribution is correlated with fast folding.

Additionally, in following a suggestion of one of the anonymous referees, we randomized the TPP riboswitches BX842649.1/277414-277318 and AACY022101973.1/389-487 by using our implementation of the Altschul-Erikson dinucleotide shuffle algorithm
\citep{altschulErikson:dinucleotideShuffle}, and then applied \fftbor to these sequences, starting from the empty structure.  The mean $\mu_1$ and standard deviation $\sigma_1$ for the \fftbor distribution for randomized BX842649 are respectively $\mu_1=19.93$ and $\sigma_1=2.88$, while those for randomized AACY022101973 are $\mu_2= 24.39$ and $\sigma_2=24.00$. Running Kinfold, with a maximum of 500,000 steps with 30 replicates (as explained in the text), we found that for randomized BX842649, all 30 runs converged yielding a mean first passage time (MFPT) of 13022.58  with standard deviation of 15221.78. In contrast for randomized AACY022101973, only 15 out of 30 runs converged within 500,000 steps, and discounting these nonconvergent data, we obtain an average mean first passage time (MFPT) of 94446.93  with standard deviation of 157107.43. This additional test provides more anecdotal evidence supporting our hypothesis that small standard deviation $\sigma$ in FFTbor probability density is correlated with fast folding, as measured by MFPT.
%
%The dinucleotide shuffle for BX842649 was
%AAUGAGCCAAGAAAUGAUUAUAUUGAAAUUCUGUAGAACGCACUGACCCAUGGGAAAGCUGUAGCGACUGGAAGAGCCCGAGUCGGGAGGACUACUU
%In contrast, the dinucleotide shuffle for AACY022101973 was
%UACAUAAGAUCCAUUAUGAGGGUGCCGGAUGCUGCCGGGUGAACUUUCCUACGAGGGGUCUACUGAAAUUAUUAAAAGUCCAUUCAAGGUUUUUGCCCC
%

\subsubsection*{Riboswitch expression platform prediction}

A bacterial riboswitch is a portion of the $5'$
untranslated region (UTR) of messenger RNA,
that performs gene regulation by undergoing a conformational change upon
binding with a ligand, such as guanine, thiamine pyrophosphate, lysine,
etc.  \citep{breaker:Riboswitch2}.
This conformational change may either turn  on or off
the corresponding gene by either transcriptional or translational regulation
of the messenger RNA \citep{tuckerBreaker:review}, depending on the
particular riboswitch. The common feature shared by all riboswitches
is that a gene is regulated by conformational change upon ligand binding.
Bacterial riboswitches are often found upstream
of operons, regulating groups of genes, as in purine {\em de novo}
synthesis and salvage \citep{mandalBoeseBarrickWinklerBreaker}.

A riboswitch consists of two equally important parts: an upstream
{\em aptamer}, capable of highly discriminative binding to a particular
ligand, and a downstream {\em expression platform}, capable of undergoing
a radical conformational change upon binding of a ligand with the
discriminating aptamer.
Since aptamers have been under strong evolutionary pressure to
bind with high affinity
(e.g. $K_D \approx 5$ nM for guanine \citep{Mandal.nsmb04}),
there is strong
sequence conservation found in the aptameric region of orthologous
riboswitches.
In contrast, while secondary structure is conserved in the
terminator loop of the expression platform in purine riboswitches,
there is relatively low sequence conservation (data not shown).
While a number of methods exist to computationally predict
riboswitch aptamers
\citep{Bengert2004Riboswitch,Chang2009Computational,AbreuGoodger2005RibEx,Singh2009Riboswitch,Bergig.isb04}
(and especially {\tt INFERNAL} \citep{INFERNAL}, which latter is used
to predict riboswitch aptamers in Rfam), it is an important
biological problem to determine the expression platform, since the structure
of the expression platform can suggest whether there is
transcriptional regulation
via a terminator loop or translational regulation via the sequestration of
the Shine-Dalgarno sequence \citep{tuckerBreaker:review}. Determination of
the precise location and structure of the expression platform is difficult
due to low conserved sequence identity (in-house computations, data not shown).
Although this problem remains open, we report here how
\fftbor may provide help to biologists in the selection and
prioritization of riboswitch candidates.

Figure 5 depicts the gene OFF structure of the
xpt G-box purine riboswitch in {\em B. subtilis}, as determined by
inline-probing -- this structure was taken from
Figure 1 of \citep{breaker:Riboswitch2}.
Note that this structure is only partial, since there are regions
with no base pairs depicted, despite the fact that additional
base pairs could be added. By using
{\tt blastn}, it is found that this 161 nt purine riboswitch
can be found on the complement strand of GenBank accession number
CP002906.1/c2165302-2165142 in
complete genome of {\em B. subtilis subsp. subtilis RO-NN-1}.
Figure 6 depicts the result of
three computational experiments with FFTbor. The left panel displays
the expected base pair distance to the following secondary structure
\begin{small}
\begin{eqnarray}
\label{eq:xpt1}
%ACACUCAUAUAAUCGCGUGGAUAUGGCACGCAAGUUUCUACCGGGCACCGUAAAUGUCCGACUAUGGGUGAGCAAUGGAACCGCACGUGUACGGUUUUUUGUGAUAUCAGCAUUGCUUGCUCUUUAUUUGAGCGGGCAAUGCUU
\bf
.((((((((...(((((((.......)))))))........((((((.......))))))..))))))))........(((((........)))))............((((((((((((((.......)))))))))))))).
\end{eqnarray}
\end{small}
as a function of window offset, where window size equals the size of this
target structure. This structure was
obtained by removing all leading and trailing unpaired positions
from the structure depicted in Figure 5,
except for the leftmost [resp. rightmost] unpaired position
adjacent to the leftmost [resp. rightmost] base-paired position.
The reason for removal of the leading and trailing unpaired positions was
that the structure of \citep{breaker:Riboswitch2}, depicted in
Figure 5, is clearly only partial, as earlier
mentioned.  The center panel displays
the expected base pair distance to the following secondary structure
\begin{small}
\begin{eqnarray}
\label{eq:xpt2}
%ACACUCAUAUAAUCGCGUGGAUAUGGCACGCAAGUUUCUACCGGGCACCGUAAAUGUCCGACUAUGGGUGA
\bf
.((((((((...(((((((.......)))))))........((((((.......))))))..)))))))).
\end{eqnarray}
\end{small}
as a function of window offset, where window size equals the size of this
target {\em aptamer} structure. Similarly, the right panel displays
the expected base pair distance to the following secondary structure
\begin{small}
\begin{eqnarray}
\label{eq:xpt3}
%GAACCGCACGUGUACGGUUUUUUGUGAUAUCAGCAUUGCUUGCUCUUUAUUUGAGCGGGCAAUGCUU
\bf
.(((((........)))))............((((((((((((((.......)))))))))))))).
\end{eqnarray}
\end{small}
as a function of window offset, where window size equals the size of this
target {\em expression platform} structure.
Figure 6 determines the precise location of the
xpt riboswitch, both aptamer and expression platform.

If the biologically functional target structure is unknown, one can
instead attempt a similar moving window computation, where the target
structure is taken to be the minimum free energy structure of the
current window contents. In this case, one may hope to determine a
bimodal distribution, as displayed in Figure 7.
Given an input RNA sequence, or genomic region, the
web server \url{http://bioinformatics.bc.edu/clotelab/FFTbor}
creates a movie as follows, described here for the xpt riboswitch
previously discussed.  We extracted the downstream region of
Taking the
161 nt xpt G-box purine riboswitch described in Figure 5,
with GenBank accession number CP002906.1/c2165302-2165142, we extended
to a sequence of length 200 nt, by appending flanking
downstream genomic nucleotides.  Running \fftbor on all
prefixes of the resulting sequence of lengths
$70,72,74,\ldots,200$, we produced a movie, displayed on the webserver
\url{http://bioinformatics.bc.edu/clotelab/FFTbor}.
Figure 5 displays the output of \fftbor on the
166 nt prefix, clearly showing a bimodal distribution. Attempting to
automate the identification of non-unimodal \fftbor output, we have
applied the Hartigan-Hartigan dip-test \citep{hartiganDipTest},
implemented in {\tt R}; however,
the dip-test appears to be too sensitive, in that a probability distribution
is reported to be non-unimodal, even when visual inspection indicates that
it appears overwhelmingly to be unimodal (data not shown). It is for this
reason that the web server
\url{http://bioinformatics.bc.edu/clotelab/FFTbor} produces a movie of
prefixes, where the user can start/stop the movie,
move forward/backward, or download all raw data output by \fftbor.

%---------------NOT USED ------------------------------
%In a separate computational experiment, we used
%{\tt HMMER} 1.8.5 \citep{eddy:HMMER} to train on purine riboswitch
%aptamers
%from the RF00167 Rfam family seed alignment, after removal from RF00167
%of all {\em B. subtilis} sequences. We then ran the trained hidden Markov
%model on all 400 nt sequences upstream of a start codon in the {\em B. subtilis}
%genome (GenBank accession number NC000964),
%including the complement strand. The six top hits had {\tt HMMER} scores of
%49.12, 49.12, 62.34, 63.09, 65.81, and 70.07, and were located respectively in
%regions NC000964 complement 625839..626239, NC000964 625768..626168,
%NC000964 693610..694010, NC000964 697553..697953,
%NC000964 complement 2319222..2319622, NC000964 4004369..4004769.
%We found that \fftbor output was clearly bimodal on only three of the six
%regions. Figure
%---------------------------------------------

\subsection*{Benchmarking results}

\subsubsection*{Total variation distance for density of states}

Recall that the {\em total variation distance}
between two probability distributions
$P = \{ p_{x} : x \in \Omega\}$ and
$Q = \{ q_{x} : x \in \Omega\}$, defined on the same sample space $\Omega$, is
defined by
\[
\delta(P,Q) = \frac{\sum_{x \in \Omega} |p_x - q_x|}{2}.
\]
The {\em density of states} for an RNA sequence \seq with respect to
an initial structure \strSt of \seq is defined to be the probability
distribution
$P = (p_0,\ldots,p_{n-1})$ where $p_k = \z{k}{1}{n}/{\bf Z}$.
In all our tests, for RNA of length up to 400 nt, we found the total
variation distance between $P$, as computed to 6 decimal places by
{\tt RNAbor} and by \fftbor, to be $0$. It follows that \fftbor
can reliably be used in place of {\tt RNAbor} to determine Boltzmann
probabilities $p(k) = P\left( d_{BP}(S,S^*)=k \right)$.

\subsubsection*{Run time comparison of {\tt RNAbor} and \fftbor}

As visible from the defining recursions, the algorithmic time complexity of
{\tt RNAbor} is $O(n^5)$ and space complexity is $O(n^3)$, where $n$ is
the length of input RNA sequence. In contrast, the time complexity of
\fftbor is $O(n^4)$ and space complexity is $O(n^2)$.
Figure 8 displays run time curves for both
{\tt RNAbor} and \fftbor, when the initial structure \strSt is
taken to be either the empty structure or the minimum free energy
(MFE) structure.

Here, we compare the run time of {\tt RNAbor} \citep{Freyhult.b07} and
the (unparallelized version of) \fftbor, using
a Dell Power Edge 1950, 2 x Intel Xeon E5430 Quad
core with 2.80 GHz and 16 GB RAM. For $n = 20,40,60,\ldots, 300$, in step
size of 20 nt, we generated $n$ random RNA sequences of length $n$ with equal
probability for each nucleotide A,C,G,U (i.e. a $0$th order Markov chain).
For values of $n \leq 200$, $100$ random sequences of length
$n$ were generated, while for values of $220 \leq n \leq 300$, only
$10$ sequences of length $n$ were generated.
RNA sequences larger than 300 nt were not tested,
due to $O(n^3)$ memory constraints required by {\tt RNAbor}.
For each RNA sequence, {\tt RNAbor} and \fftbor were both run,
each starting with empty initial structure \strSt, and also
with initial sequence \strSt taken to be the MFE structure.
Each data point in the table comprises the average run time for three
independent evaluations.

\subsubsection*{OpenMP parallelization of \fftbor}

OpenMP is a simple and flexible
multi-platform shared-memory parallel programming environment, that supports
parallelizations of C/C++ code -- see \url{http://openmp.org/}.
Using OpenMP primitives, we created multiple threads to evaluate the polynomial
$\Z(x)$ on different complex $n$th roots of unity. The table in the left
panel of Figure 9
and Table~2 together present benchmarks,
executed on
a 24-core AMD Opteron 6172 with 2.10GHz and 64GB RAM, for the speedup
of \fftbor as a function of the number of cores.
The table in Figure 9 describes average
run time in seconds ($\pm$ one standard deviation) for running \fftbor
on random RNA of length 200,250,300,400,450,500 with either 1 or 2 cores.
Table~2 presents similar data for running
\fftbor on 2,3,6,4,12,15,20 cores.
Although \fftbor clearly has quartic $O(n^4)$ run time as a function of
RNA sequence length, least-squares fit
of run times from
Table~2 instead
shows a quadratic
run time for RNA sequences of length up to $500$ nt. There appears to be
a {\em power law} dependence of \fftbor
speedup, as a function of number of cores. For instance, for random RNA of
length 200 nt, least-squares fit of the data from the table yields
a run time of $105.29 x^{-0.923}$ with  $R^2$ value of
$0.99782$. A power law behavior is demonstrated, with similarly high
$R^2$ values, for each fixed sequence length in
Table~2, with different
coefficients of variable $x$ but with approximately the
same exponent of $x$ (data not shown, but easily computable from data in
Table~2).

\section*{Conclusion and Discussion}

In this paper, we have used a dynamic programming computation to
evaluate the polynomial
\begin{eqnarray}
\Z(x) = \sum_{k=0}^n c_k x^k
\end{eqnarray}
on the complex $n$th roots of unity $1,e^{2 \pi i/n},\ldots,
e^{2 \pi i (n-1)/n}$,
where the coefficients $c_k=\z{k}{1}{n}$ are equal to the sum of
Boltzmann factors over all secondary structures of a given RNA
sequence, whose base pair distance to a given initial structure \strSt
is $k$.  Recall the definition of polynomial
\begin{eqnarray}
p(x) = \frac{\Z(x)}{\z{}{1}{n}} = \sum_{k=0}^n p_k x^k
\end{eqnarray}
obtained from $\Z(x)$, whose coefficients are Boltzmann
probabilities $p_k=\frac{\z{k}{1}{n}}{\z{}{1}{n}}$ that a secondary
structure has base pair distance $k$ to \strSt.
By using the fast Fourier transform to compute the inverse
discrete transform, we can approximate  to $m$ decimal places the coefficients
$p_k=\frac{\z{k}{1}{n}}{\z{}{1}{n}} =\frac{c_k}{\z{}{1}{n}}$ of $p(x)$, and
thus the $m$ most significant positions of
$c_k=\z{k}{1}{n}=p_k \cdot \z{}{1}{n}$. Interpolation is performed for
$p(x)$, rather than $\Z(x)$, due to issues concerning numerical instability.
The computational advantage of \fftbor over its predecessor
{\tt RNAbor} \citep{Freyhult.b07} is that
the new algorithm runs in quartic time $O(n^4)$ and quadratic space
$O(n^2)$, in contrast to the $O(n^5)$ run time and $O(n^3)$ space
required by {\tt RNAbor}. We have additionally provided a
parallelization of \fftbor using OpenMP primitives.
Additionally, we have
described applications of \fftbor to determine the
correlation between kinetic folding speed and the {\em ruggedness}
of the energy landscape, and to predict the location of riboswitch
expression platform candidates.

It is important to point out that
the algorithm and software {\tt RNAbor} is more general than
that of \fftbor -- in particular, {\tt RNAbor} not only computes the
partition function values $\z{k}{1}{n}$, for all $0 \leq k \leq n$, but as
well as computes the
structures $S_k$, defined to be the minimum free energy structure over all
$k$-neighbors of initial structure \strSt.
In contrast, \fftbor only computes the $m$ most significant digits of
the probabilities $p_k = \z{k}{1}{n}/{\bf Z}$, for $0 \leq k \leq n$, where
by multiplication of $p_k$ by the partition function $Z=\z{}{1}{n}$,
one obtains an approximation of the partition function values $\z{k}{1}{n}$.
There is no possibility that \fftbor can compute the structures $S_k$,
nor can at present we see how to use \fftbor to sample structures from
the Boltzmann ensemble of structures having base pair distance $k$ from \strSt.

In \citep{CloWalBehSte05,Waldispuhl.pcb08}, we introduced the a related
{\em parametric} RNA structure algorithm, {\tt RNAmutants}, which computes
the partition function $\z{k}{1}{n}$ and minimum free energy
structure $MFE(k)$ over all secondary structures of all $k$-point mutants
of a given RNA sequence $\seq=s_1,\ldots,s_n$. In  \citep{WaPo2011},
{\tt RNAmutants} was extended to sample low energy structures over
$k$-point mutants within a certain range of GC-content. Some of the ideas
in \citep{WaPo2011} foreshadowed the results of this paper, and in the
future, we intent to apply interpolation and the FFT to similarly provide a
more efficient version of {\tt RNAmutants}. Nevertheless, this future,
more efficient version will be incapable of efficiently sampling low
energy structures over $k$-point mutants, analogous to the current
differences between {\tt RNAbor} and \fftbor.


\section*{Acknowledgements}

\fftbor depends heavily on the
use of the Fast Fourier Transform implementation FFTW of Frigo and Johnson
\citep{FFTW05} at \url{http://www.fftw.org/}.

%As well, \fftbor accesses
%free energy parameters from Vienna RNA Package \citep{hofacker:ViennaWebServer}.
%Comparison of the output of
%\fftbor and {\tt paRNAss} \citep{Voss.b04} was compared for the set
%of conformational switches made available at
%\url{http://bibiserv.techfak.uni-bielefeld.de/parnass/examples.html}.


%Funding for this research was provided by
%National Science Foundation grants DMS-1016618 and DMS-0817971 to PC.
%Any opinions, findings,
%and conclusions or recommendations expressed in this material are
%those of the authors and do not necessarily reflect the views of the
%National Science Foundation.



\newpage
%\bibliographystyle{plain}
\bibliographystyle{plos2009}
%\bibliography{/Users/clote/text/BIBdir/clote}
\bibliography{biblio}

\hfill\break
\hfill\break
\newpage

\noindent
{\bf\Large Figure Legends}


\begin{figure}[!ht]
\begin{center}
%\includegraphics[width=0.3\textwidth]{figure1A}
%\includegraphics[width=0.3\textwidth]{figure1B}
%\includegraphics[width=0.3\textwidth]{figure1C}
\caption{\fftbor output for the RNA attenuator for the
phenylalanyl-tRNA synthetase (pheST) operon in
{\em E. coli} K-12 substr. DH10B, located adjacent to the
phenylalanyl-tRNA synthetase operon leader, with GenBank accession code
CP000948.1/1887748-1887820 (complement).
%------NOT USED-----
% http://ecocyc.org/ECOLI/NEW-IMAGE?type=GENE&object=EG11272&detail-level=2
%This 73 nt RNA sequence is
%AUCCAGGAGG CUAGCGCGUG AGAAGAGAAA CGGAAAACAG CGCCUGAAAG
%CCUCCCAGUG GAGGCUUUUU UUG, with GenBank accession code
%CP000948.1/1887748-1887820 (complement).
%-------------------
The $x$-axis represents
base pair distance to the minimum free energy structure \strSt; $y$-axis
represents Boltzmann probability $p(k)$ that a structure has base pair
distance $k$ to \strSt.
{\em (Left)} Probability
$P\left( d_{BP}(S,S^*) = k \right)$
that base pair distance to MFE structure is $k$.
{\em (Center)} Cumulative probability
$P\left( d_{BP}(S,S^*) \leq k \right)$
that base pair distance to MFE structure is at most $k$.
{\em (Right)} Finite difference (Derivative)
$P\left( k \leq d_{BP}(S,S^*) = k+1 \right)$
of probability that base pair distance to MFE structure is $k$.
}
\label{fig:ecoliAttenuator}
\end{center}
\end{figure}


\begin{figure*}[!ht]
\centering
% \includegraphics[width=0.45\textwidth]{figure2A}
%\includegraphics[width=0.45\textwidth]{figure2B}
%
\caption{Output from \fftbor on two randomly selected
thiamine pyrophosphate riboswitch (TPP) aptamers, taken from the Rfam database
\citep{Gardner.nar11}. The $x$-axis represents base pair distance from the
minimum free energy structure for each given sequence; the $y$-axis represents
Boltzmann probabilities $p(k) = \frac{Z_k}{Z}$, where
$Z_k$ denotes the sum of Boltzmann factors or all secondary structures,
whose base pair distance from the MFE structure is exactly $k$.
{\em (Left)}
The 97 nt sequence BX842649.1/277414-277318 appears to have a rugged energy
landscape near its minimum free energy structure, with distinct
low energy structures that may compete with the MFE structure during the
folding process.
{\em (Right)}
The 99 nt sequence, AACY022101973.1/389-487 appears to have a smooth energy
landscape near its MFE structure, with no distinct low energy structures
to might compete with the MFE structure.
Based on the \fftbor output or {\em structural profile} near MFE
structure \strSt, one might expect
folding time for the first sequence to increase due to competition from
metastable structures, while one might expect the second sequence to have
rapid folding time.
%
Computational Monte Carlo folding experiments bear out this fact.
\kinfold \citep{flamm} simulations clearly show that the second
sequence folds
at least four times more quickly than the first sequence. See text for
details.}
\label{fig:FFTborOutputForTPPriboswitches}
\end{figure*}


\begin{figure*}[!ht]
\centering
\begin{minipage}[c]{0.24\textwidth}
\centering
%\includegraphics[width=\textwidth]{figure3A}
\mbox{Subfigure A}
\label{fig:tppSecStrA}
\end{minipage}
%
\centering
\begin{minipage}[c]{0.24\textwidth}
\centering
%\includegraphics[width=\textwidth]{figure3B}
\mbox{Subfigure B}
\label{fig:tppSecStrB}
\end{minipage}
%
\centering
\begin{minipage}[c]{0.24\textwidth}
\centering
%\includegraphics[width=\textwidth]{figure3C}
\mbox{Subfigure C}
\label{fig:tppSecStrC}
\end{minipage}
%
\centering
\begin{minipage}[c]{0.24\textwidth}
\centering
%\includegraphics[width=\textwidth]{figure3D}
\mbox{Subfigure D}
\label{fig:tppSecStrD}
\end{minipage}
\caption{Rfam consensus structures (Rfam) and minimum free energy (MFE)
secondary structures for two thiamine pyrophosphate (TPP) riboswitch aptamers,
chosen at random from RF00059 Rfam family seed alignment
\citep{Gardner.nar11}. Using pairwise BLAST \citep{BLAST}, there is no
sequence similarity, although the secondary structures are very similar,
as shown in this figure.
{\em (A)} Rfam consensus structure for BX842649.1/277414-277318.
{\em (B)} MFE structure for BX842649.1/277414-277318.
{\em (C)} Rfam consensus structure for AACY022101973.1/389-487.
{\em (D)} Rfam consensus structure for AACY022101973.1/389-487.
}
\label{fig:tppSecStr}
\end{figure*}



\begin{figure*}[!ht]
\centering
%\includegraphics[width=0.30\textwidth]{figure4A}
%\includegraphics[width=0.30\textwidth]{figure4B}
%\includegraphics[width=0.30\textwidth]{figure4C}
%
\caption{This figure represents the
graphical output of \fftbor, when the empty structure is chosen as
initial structure \strSt.
The $x$-axis represents the number of base pairs per structure,
taken over the ensemble of all secondary structures for the given RNA
sequence; the $y$-axis represents Boltzmann probability $p(k) = \frac{Z_k}{Z}$,
where $Z_k$ is the partition function for all secondary structures
having exactly $k$ base pairs.
{\em (Left)}
For the selenocysteine (SECIS) element AB030643.1/4176-4241 from Rfam family
RF00031, the standard deviation $\sigma$ of the number of base pairs,
taken over the ensemble of all secondary structures, is
$0.727618$, while the logarithm base 10 of the mean first passage time (logMFPT)
is $4.75$.
{\em (Center)}
For the selenocysteine (SECIS) element
AL645723.11/192421-192359 from Rfam family
RF00031, the standard deviation $\sigma$ of the number of base pairs,
taken over the ensemble of all secondary structures, is
$2.679446$, while logMFPT is $5.69$.
Among the 61 sequences in the seed alignment of RF00031,
AB030643.1/4176-4241 was the fastest folder, while
AL645723.11/192421-192359 was the slowest folder.
{\em (Right)}
Superimposition of output of \fftbor for two TPP riboswitch aptamers: the
97 nt sequence BX842649.1/277414-277318 and the
99 nt sequence AACY022101973.1/389-487, both obtained when
taking the empty structure for the initial structure \strSt.
The mean $\mu$ for the \fftbor structural profile near the empty
structure is $23.0203$  [resp. $27.5821$], the
standard deviation $\sigma$ for the \fftbor structural profile
is $2.22528791$  [resp. $1.98565959$], and the \kinfold MFPT is
$311,075.06$ [resp. $61,575.69$] for the TPP riboswitch aptamer
AB030643.1/4176-4241 [resp.  AL645723.11/192421-192359].
The right panel of this figure should be compared with
Figure 2.
These anecdotal results bear up the correlation between standard deviation
$\sigma$ and logMFPT described in Table~1.
}
\label{fig:correlationFFTborEmpty}
\end{figure*}



\begin{figure*}[!ht]
\centering
%\includegraphics[width=0.45\textwidth]{figure5}
%
\caption{Gene OFF secondary structure of the
xpt G-box purine riboswitch in {\em B. subtilis}; structure taken
from that in Figure 1A of \citep{breaker:Riboswitch2}.
}
\label{fig:xptSecStr}
\end{figure*}


\begin{figure*}[!ht]
\centering
%\includegraphics[width=0.30\textwidth]{figure6A}
%\includegraphics[width=0.30\textwidth]{figure6B}
%\includegraphics[width=0.30\textwidth]{figure6C}
%
\caption{Graph of the expected distance from target secondary structure, as
a function of window offset position in the $5'$ untranslated region (UTR)
of the xpt gene of {\em B. subtilis}; i.e. GenBank accession code
CP002906.1/c2165402-2165042 {\em B. subtilis} subsp. subtilis RO-NN-1.
In a moving window application, \fftbor computed the Boltzmann
probability $p(k)$ that secondary structures of the
current window contents have base pair distance $k$ from the target
(or initial) structure \strSt. In each case, the size of the window was
set to equal the length of \strSt.
{\em (Left)} Target structure \strSt comprises the entire secondary
of the xpt riboswitch, as depicted in Figure 5, with the
exception that the leading and trailing unpaired positions were removed,
as explained in the text -- see displayed dot bracket structure in
(\ref{eq:xpt1}).
{\em (Center)} Target structure \strSt comprises only the aptamer
secondary structure, as displayed in dot bracket structure in
(\ref{eq:xpt2}).
{\em (Right)} Target structure \strSt comprises only the expression platform
secondary structure, as displayed in dot bracket structure in
(\ref{eq:xpt3}). The number of points displayed on the $x$-axis differs
in each case, since the window size differs, as explained above.
The very well-defined minimum in each panel corresponds to the exact
location of the entire riboswitch (left panel), aptamer (center panel)
and expression platform (right panel). Note that the base line value for
the expected base pair distance in the left panel (entire riboswitch) is
approximately $70$, while that for both the center panel (aptamer) and
right panel (expression platform) is approximately $35$.
}
\label{fig:xptFFTborSearch}
\end{figure*}


\begin{figure*}[!ht]
\centering
%\includegraphics[width=0.8\textwidth]{figure7}
%------------------ UNUSED FIGURE ---------------------------
%\vskip 0.5cm
%\includegraphics[width=0.35\textwidth]{FIGURES/scatterPlotSecis}
%---------------------------------------------
%
\caption{The 161 nt xpt G-box purine riboswitch
described in Figure 1A of \citep{breaker:Riboswitch2},
found on the complement strand of GenBank accession number
CP002906.1/c2165302-2165142 in
complete genome of {\em B. subtilis subsp. subtilis RO-NN-1}. We extended
this 161 nt sequence to a sequence of length 200 nt, by appending flanking
downstream genomic nucleotides. The web site
{\tt http://bioinformatics.bc.edu/clotelab/FFTbor} displays a movie
of all prefixes of the resulting 200 nt sequence, where prefix lengths range
from $70,72,74,\ldots,200$.
}
\label{fig:xptFFTborOutput}
\end{figure*}



\begin{figure*}[!ht]
\centering
\begin{center}
%\includegraphics[width=0.85\textwidth]{figure8}
\end{center}
\caption{Run times in seconds for {\tt RNAbor} and \fftbor, on random RNA
of length $20,40,60,\ldots,300$ in step size of 20 nt. Each algorithm
was run with the empty initial structure \strSt, see rows
{\tt RNAbor} (empty), \fftbor (empty), and with the minimum free
energy structure as the initial structure \strSt, see rows
{\tt RNAbor} (MFE) and \fftbor (MFE). Note that for both {\tt RNAbor}
and \fftbor, the run time increases when \strSt is the MFE structure,
rather than the empty structure. Notice the radical improvement in the
run time of \fftbor over that of {\tt RNAbor}.
}
\label{fig:evanBenchmark}
\end{figure*}


\begin{figure*}[!ht]
\centering
\begin{center}
%
%\includegraphics[width=0.45\textwidth]{figure9A}
%\includegraphics[width=0.45\textwidth]{figure9B}
\end{center}
\caption{{\em (Left)} Table showing parallel run times in seconds
for \fftbor, using
OpenMP {\tt http://openmp.org/}. Column headers 1,2 indicate
the number of cores used in the computational experiment.
For each sequence length $200,\ldots,500$,
five random RNAs were generated using equal probability for each nucleotide
A,C,G,U. Run time in seconds, plus or minus one standard deviation, are
given for a 24-core
AMD Opteron 6172 with 2.10GHz and 64GB RAM, with only 1 (resp. 2) cores
used.
{\em (Right)} Graph showing parallel run time of \fftbor on
an AMD Opteron 6172 with 2.10GHz and 64GB RAM, using respectively
1,2,3,4,6,9,12,15,20 cores.
}
\label{table:benchmarkParallelOneTwo}
\end{figure*}

\hfill\break
\hfill\break
\newpage
\hfill\break
\hfill\break
\newpage

\noindent
{\bf\Large Tables}

\begin{table}[!ht]
\caption{Pearson correlation between various aspects of selenocysteine
insertion sequences from the seed alignment of Rfam family
RF00031 \citep{Gardner.nar11}.
For each of the 61 RNA sequences, we ran \fftbor, starting
from empty initial structure \strSt, and we ran a Monte Carlo
folding algorithm, developed by E. Freyhult and P. Clote (unpublished).
Using the Monte Carlo algorithm, we
determined the mean first passage time (MFPT), defined as the average
taken over 50 runs, of the number of Monte Carlo steps taken to fold
the empty structure into the MFE structure, where an absolute upper bound
of 5 million steps was allowed in the simulation.  From the output of
\fftbor, we computed
{\em (1)}
the mean number ($\mu$) of base pairs per structure, taken over
the ensemble of all secondary structures for the given sequence,
{\em (2)}
the standard deviation ($\sigma$) of the number of base pairs per structure,
{\em (3)}
the coefficient of variation $\frac{\sigma}{\mu}$,
{\em (4)}
the RNA sequence length,
and
{\em (5)}
the minimum free energy (MFE).  Additionally, we computed the
logarithm base 10 of mean first passage
time (log10MFPT), taken over 50 Monte Carlo runs per sequence
(log base 10 of the standard deviation of number of Monte Carlo
steps per run was approximately
9\% of log10MFPT on average). The table shows the correlation between each of
these aspects.
Some correlations are obvious --
for example,
{\em (i)}
the standard deviation $\sigma$ is highly correlated with the
coefficient of variation $\frac{\sigma}{\mu}$;
{\em (ii)}
the mean $\mu$ is negatively correlated with the
coefficient of variation $\frac{\sigma}{\mu}$;
{\em (iii)}
the mean $\mu$ is negatively correlated with the
minimum free energy (MFE) -- if most low energy structures in the ensemble
have many base pairs, then it is likely that the minimum free energy is very
low (i.e. since MFE is negative, the absolute value of MFE increases);
{\em (iv)}
sequence length is negatively correlated with MFE -- as sequence length
increases, the minimum free energy (MFE) decreases.
However, it may appear surprising that
{\em (v)} the
mean $\mu$ number of base pairs per structure is independent of MFPT
(correlation $-0.036291124$), although
{\em (vi)} MFE is correlated with MFPT
(correlation $0.399015556$) -- i.e.  from {\em (iii)},
lower MFE is correlated with a larger average $\mu$ number of base pairs per
structure, from {\em (vi)}
higher MFE is correlated with longer folding time, but
from {\em (v)} the average $\mu$  number of base pairs per structure is
independent of folding time.
The most important insight from this table is that
{\em (vii)}
standard deviation $\sigma$ is correlated with mean first passage
time -- the correlation is statistically significant, with one-tailed
$p$-value of $0.00018249$.}
\begin{center}
\begin{tabular}{|l|rrrrrr|}
\hline
~ &$\mu$ &$\sigma$& $\frac{\sigma}{\mu}$&len& MFE & logMFPT\\
\hline
$\mu$& 1 & & & & &\\
$\sigma$&-0.43722448& 1  & & & & \\
$\frac{\sigma}{\mu}$&-0.691411183&0.943650913&1 & & &\\
len&0.707683898&-0.158951202&-0.364591789&1& & \\
MFE&-0.569474125&0.739515083&0.759622716&-0.368485646&1 & \\
logMFPT&-0.036291124&0.48436192&0.376230235&0.405865529&0.399015556& 1 \\
\hline
\end{tabular}
\label{table:correlationFFTborEmpty}
\end{center}
\end{table}



\begin{table*}[!ht]
\caption{Table showing parallel run times of \fftbor, using
OpenMP {\tt http://openmp.org/}. Column headers 2,3, etc. indicate
the number of cores used in the computational experiment.
For each sequence length $200,\ldots,500$,
five random RNAs were generated using equal probability for each nucleotide
A,C,G,U. Run time in seconds, plus or minus one standard deviation, are
given for a 24-core
AMD Opteron 6172 with 2.10GHz and 64GB RAM.
Least-squares fit of the data indicates a quadratic dependency of run time
on sequence length (despite the obvious $O(n^4)$ theoretical run time),
and a power law dependence of approximately $x^{-0.99}$ on the number of
cores $x$.}
\begin{tiny}
\begin{center}
\begin{tabular}{|l| r|r|r|r|r|r|r|r|}
\hline
Len & 2     & 3     & 4     & 6     & 9     & 12      & 15      & 20      \\
\hline
200 & 61.8  $\pm$ 8.0 & 41.6  $\pm$ 6.0 & 31.6  $\pm$ 4.2 & 21.1  $\pm$ 2.9 & 15.0  $\pm$ 2.3 & 11.3  $\pm$ 1.1 & 9.6 $\pm$ 1.3 & 7.6 $\pm$ 1.5 \\
250 & 166.1 $\pm$ 13.7  & 111.5 $\pm$ 8.7 & 84.1  $\pm$ 7.1 & 56.9  $\pm$ 4.0 & 38.8  $\pm$ 3.8 & 30.3  $\pm$ 2.6 & 24.6  $\pm$ 1.6 & 18.9  $\pm$ 2.7 \\
300 & 365.2 $\pm$ 30.1  & 246.4 $\pm$ 20.5  & 184.5 $\pm$ 14.9  & 125.1 $\pm$ 9.6 & 85.5  $\pm$ 6.6 & 64.9  $\pm$ 6.5 & 53.6  $\pm$ 5.9 & 42.1  $\pm$ 5.2 \\
350 & 698.4 $\pm$ 46.9  & 470.1 $\pm$ 32.6  & 352.0 $\pm$ 23.0  & 242.6 $\pm$ 15.8  & 163.3 $\pm$ 12.4  & 125.1 $\pm$ 6.7 & 104.2 $\pm$ 8.9 & 76.2  $\pm$ 3.9 \\
400 & 1,129.5 $\pm$ 104.3 & 757.5 $\pm$ 68.9  & 571.6 $\pm$ 53.9  & 391.2 $\pm$ 36.4  & 265.1 $\pm$ 24.2  & 207.2 $\pm$ 18.4  & 165.6 $\pm$ 14.5  & 125.9 $\pm$ 14.9  \\
450 & 1,980.9 $\pm$ 126.5 & 1,326.3 $\pm$ 85.1  & 1,000.0 $\pm$ 59.0  & 688.9 $\pm$ 44.8  & 469.2 $\pm$ 29.1  & 355.1 $\pm$ 25.4  & 289.8 $\pm$ 21.2  & 223.1 $\pm$ 18.2  \\
500 & 3,389.8 $\pm$ 788.4 & 2,067.9 $\pm$ 99.2  & 1,555.0 $\pm$ 72.2  & 1,074.3 $\pm$ 53.7  & 728.1 $\pm$ 40.9  & 548.3 $\pm$ 24.0  & 451.5 $\pm$ 25.7  & 338.1 $\pm$ 22.7  \\
\hline
\end{tabular}
\end{center}
\end{tiny}
\label{table:parallelBenchmark}
\end{table*}
