%!TEX root = ../main.tex

\newcommand{\mathbbP}{\mbox{\bf P}}
\newcommand{\mathbbt}{\mbox{\bf t}}
\newcommand{\hermes}{\mbox\hermes\xspace}
\newcommand{\fftmfpt}{\mbox\fftmfpt\xspace}
\newcommand{\ffteq}{\mbox\ffteq\xspace}
\newcommand{\fftbor}{\mbox\fftbor\xspace}
\newcommand{\ffttwo}{\mbox\ffttwo\xspace}
\newcommand{\rnamfpt}{\mbox{\tt RNAmfpt}\xspace}
\newcommand{\rnaeq}{\mbox{\tt RNAeq}\xspace}
\newcommand{\rnasub}{\mbox\rnasub\xspace}

\chapter{Hermes} % Main chapter title

\label{Hermes} % Change X to a consecutive number; for referencing this chapter elsewhere, use \ref{ChapterX}

\lhead{Chapter X. \emph{Hermes}} % Change X to a consecutive number; this is for the header on each page - perhaps a shortened title

In this paper, we introduce the software suite, \hermes, which
provides fast, novel algorithms for RNA secondary structure kinetics.
Using the fast Fourier transform to efficiently compute the Boltzmann
probability that a secondary structure $S$ of a given RNA sequence has
base pair distance $x$ [resp. $y$] from reference structure \strA [resp.
\strB], \hermes computes the exact kinetics of folding from \strA to
\strB in this coarse-grained model. In particular, \hermes computes
the mean first passage time from the transition probability matrix by
using matrix inversion, and also computes the equilibrium time from the
rate matrix by using spectral decomposition.
Due to the model granularity and the
speed of \hermes, it is capable of determining secondary structure
{\em refolding kinetics} for large RNA sequences, beyond the range of
other methods.  Comparative benchmarking of \hermes with other
methods indicates that
%despite the model granularity,
\hermes provides
refolding kinetics of accuracy suitable for use in
computational design of RNA, an important area of synthetic
biology. Source code and documentation for \hermes are available
at \url{http://bioinformatics.bc.edu/clotelab/Hermes/}.

\section{Introduction} \label{section:introduction}

Remarkable results in RNA synthetic biology have recently been
obtained by various groups. In \cite{Hochrein.jacs13}, small
conditional RNAs have been engineered to silence a gene Y by using the
RNA interference machinery, only if a gene X is transcribed. In
\cite{Wachsmuth.nar13} a novel theophylline riboswitch has been
computationally designed to transcriptionally regulate a gene in {\em
E. coli}, and in \cite{syntheticHammerheads} a purely computational
approach was used to design functionally active hammerhead ribozymes.
Computational design of synthetic RNA molecules invariably uses some
form of thermodynamics-based algorithm; indeed, {\tt NUPACK-Design}
\cite{Zadeh.jcc11} was used to design small conditional RNAs
\cite{Hochrein.jacs13}, Vienna RNA Package \cite{Gruber08} was used in
the design of the synthetic theophylline riboswitch, and the {\tt
RNAiFold} inverse folding software
\cite{garcia.JBCBB13,GarciaMartin.nar13} was used to design the
synthetic hammerhead ribozymes. The next step in the computational
design of synthetic RNA molecules is to control the kinetics of
folding -- such control could be important in engineering
conformational switches. Kinetics is already used as a design feature
for synthetic design in proteins \cite{Bujotzek.jcam11,Fasting.acie12}.

In this paper, we introduce a new software suite, called \hermes,
which efficiently computes RNA secondary structure
folding kinetics by using a coarse-grained
method to model RNA transitions that add or remove a single base pair.
Since our motivation in developing \hermes is to provide a new
tool to aid in engineering synthetic RNA molecules with desired
kinetic properties, \hermes does not model {\em co-transcriptional
folding}, but only the {\em refolding} of RNA sequences.

There is a long history of both experimental and computational work on
RNA folding pathways and kinetics, so much so, that a review of
previous work is beyond the scope of this paper. Here we cite only the
most relevant related work on RNA secondary structure
kinetics. Experimental approaches to
determine the kinetics of RNA folding include temperature jump
experiments \cite{lecuyerCrothers}, using fluorophores
\cite{Hobartner.jmb03}, using mechanical tension at the single
molecule level \cite{Vieregg.mp06}, etc. and will not be further
discussed. In this paper we focus exclusively on computational
approaches to folding kinetics, where stepwise transitions between
secondary structures involve the addition or removal of a single base
pair, as first considered in \cite{flammHofacker}.
Nevertheless, it should be noted that there are a number of
methods that concern the addition or removal of an entire helical
region, e.g. \cite{Huang.bb14,Zhao.jcp11}.

Most computational approaches involve either (1) algorithms to
determine optimal or near-optimal folding pathways, (2) explicit
solutions of the master equation, (3) repeated simulations to fold an
initially empty secondary structure to the target minimum free energy
(MFE) structure. Examples of methods to determine optimal or
near-optimal folding pathways include the greedy approach of Morgan
and Higgs \cite{morganHiggsBarrier}, the exact, optimal, exponential
time program {\tt barriers} \cite{flammHofacker}, the program {\tt
Findpath} \cite{Flamm.r01}, which uses bounded look-ahead
breadth-first search, a genetic algorithm \cite{Shapiro.jmb01}, the
program {\tt RNAtabupath} that uses local search to determine
near-optimal folding pathways, etc. The program {\tt barriers}
\cite{flammHofacker} is the only current method guaranteed to
produce optimal folding pathways. Since it has been shown that
the problem of determining optimal folding pathways is
NP-complete \cite{Thachuk.psb10}, it is now understandable why
{\tt barriers} can take exponential time to converge, depending
on the RNA sequence. For this reason, near-optimal solutions
provided by heuristic methods, such as {\tt RNAtabupath}, are very useful.

Methods that employ the master equation include {\tt Treekin}
\cite{wolfingerStadler:kinetics}, which uses the programs {\tt
RNAsubopt} \cite{wuchtyFontanaHofackerSchuster} and {\tt barriers}
\cite{flammHofacker} to determine {\em macrostates}, defined as basins
of attraction near a locally optimal structure. The resulting
coarse-grained Markov chain is then sufficiently small to allow explicit
solution of the master equation. In \cite{Tang.jcb05}, a moderate
number of RNA structures were sampled according to different
strategies, from which a robotic motion planning graph was defined
to connect each sampled structure to its $k$ nearest sampled
neighbors. Again, the resulting coarse-grained Markov chain is
sufficiently small for an explicit solution of the master equation to
be given.

We now come to simulation approaches to estimate RNA folding kinetics.
The program \kinfold \cite{flammPhD,flamm} is an implementation
of Gillespie's algorithm \cite{gillespieStochasticSimulation1},
directly related to the master equation, hence is considered by many
to be the gold standard for RNA kinetics. A recent extension of the
\kinfold algorithm was reported in \cite{Aviram.amb12}. {\tt
Kinefold} \cite{Xayaphoummine.nar05} uses stochastic simulations of
the nucleation and dissociation of helical regions to predict
secondary structure and folding pathways. In contrast to the
previously mentioned methods, both {\tt RNAKinetics}
\cite{Danilova.jbcb06} and {\tt Kinwalker} \cite{Geis.jmb08} model
co-transcriptional folding, known to be necessary when simulating
{\em in vivo} folding of long RNA molecules \cite{Lai.r13}. As well,
{\tt Kinefold} can simulate both refolding and co-transcriptional folding
pathways. Finally, unlike all the previous simulation results, which
depend on thermodynamic free energy parameters \cite{Turner.nar10},
the program {\tt Oxfold} \cite{Anderson.b13} performs kinetic folding
of RNA using stochastic context-free grammars and evolutionary
information.

In contrast to the previous methods, \hermes computes the mean
first passage time (MFPT) and {\em equilibrium time} for a
coarse-grained Markov chain consisting of the ensemble of all
secondary structures having base pair distance $x$ [resp. $y$] from
reference structures \strA [resp. \strB] of a given RNA sequence.
Mean first passage time (MFPT) is computed exactly by matrix inversion,
and equilibrium time is computed using spectral decomposition of the rate
matrix for the coarse-grained master equation.
\hermes is written in C/C++, and
makes calls to the Fast Fourier Transform implementation {\tt FFTW}
\cite{FFTW05} \url{http://www.fftw.org/}, the Gnu Scientific Library
(GSL) \url{http://www.gnu.org/software/gsl/}, and the free energy
functions from Vienna RNA Package \cite{Lorenz.amb11}. The plan of the
paper is as follows.

Section~\ref{section:algo} gives background definitions and results
concerning Markov chains, Markov processes, mean first passage time (MFPT),
and equilibrium time (ET). This section can be skipped for those readers
who are familiar with kinetics. For sufficiently small RNA sequences, MFPT and
ET are computed by \hermes using the methods described in this section,
and so can provide a {\em gold} and {\em platinum} standard against which
other kinetics methods can be compared. Other methods include
simulation methods such as the Monte Carlo algorithm and the Gillespie
algorithm, and computations of MFPT and ET for coarse-grained Markov
chains that partition secondary structures into disjoint {\em macrostates}.
Section~\ref{section:software} presents an overview of the software suite
\hermes, which consists of three C/C++ software packages
(\ffttwo, {\tt RNAmfpt}, {\tt RNAeq}) and two C-driver programs
(\fftmfpt, \ffteq).
Section~\ref{section:results} presents the benchmarking results,
which compare \hermes with
existent RNA kinetics software by using a collection of 1000
random 20-mers, for which accurate mean first passage times and equilibrium
times can be exactly computed. Our objective here is to show that kinetics
obtained by using coarse-grained methods from \hermes correlate well
with the gold and platinum standards, yet are sufficiently fast to be used
in synthetic biology applications. Slower methods having finer granularity
may subsequently of great use in detailed kinetics studies of one or
a small number of sequences.
Section~\ref{section:discussion} summarizes the contribution
of \hermes, and the Appendix presents detailed descriptions of
all the kinetics methods and parameters used in the benchmarking results
reported in Section~\ref{section:results}.



\section{Preliminaries}
\label{section:algo}

To better understand the underlying algorithms behind the software,
we describe two traditional approaches in kinetics.

\subsection{Markov chains and mean first passage time}

Consider a physical process,
which when monitored over time, yields the stochastic sequence
$q_0,q_1,q_2,\ldots$ of discrete, observed states. If the transition from
state $q_t$ to $q_{t+1}$ depends only on $q_t$ at time
$t$ and not on the historical sequence of prior states visited,
as often assumed in the case of protein or
RNA folding, then a Markov chain provides a reasonable
mathematical model to simulate the process.

A (first-order, time-homogeneous)
{\em Markov chain} $\mathbb{M} = (Q,\pi,P)$ is given by a finite
set $Q = \{ 1,\ldots,n \}$ of states, an initial probability distribution
$\pi = (\pi_1,\ldots,\pi_n)$, and the $n \times n$
{\em transition probability matrix} $P = (p_{i,j})$.
At time $t=0$, the initial state of the system is
$q_t=i$ with probability $\pi_i$, and at discrete time
$t=1,2,3,\ldots$, the system makes a transition from state
$i$ to state $j$ with probability $p_{i,j}$; i.e.
the conditional probability $Pr[ q_{t+1} = j | q_t=i] = p_{i,j}$.
Define the {\em population occupancy frequency} of visiting state
$i$ at time $t$ by $p_{i}(t) = Pr[q_t=i]$.  Denote
$p_{i,j}^{(t)} = Pr[q_t = j | q_0 = i]$ and notice that
the $(i,j)$-th entry of the $t$-th power $P^t$ of matrix $P$  equals
$p_{i,j}^{(t)}$.


The {\em mean first passage time} (MFPT) or
{\em hitting time} for the Markov chain $\mathbb{M}$,
starting from initial state $x_0$ and proceeding
to the target state $x_{\infty}$, is defined
as the sum, taken over all paths $\rho$
from $x_0$ to $x_{\infty}$, of the path length $len(\rho)$ times the probability
of path $\rho$, where $len(\rho_0,\ldots,\rho_n)$ is defined to
be $n$. In other words, $MFPT = \sum_{\rho} Pr[\rho] \cdot len(\rho)$, where
the sum is taken over sequences $\rho=\rho_0,\ldots,\rho_n$ of states
where $\rho_0=x_0$ and $\rho_n=x_{\infty}$, and no state is visited more
than once in the path $\rho$.

Given the target state $x_{\infty}$, MFPT can be exactly
determined by computing the inverse
$(I - P^{-}_{x_{\infty}})^{-1}$, where $I$ is the $(n-1)\times (n-1)$ identity
matrix and $P^{-}_{x_{\infty}}$
denotes the $(n-1)\times (n-1)$ matrix obtained from the
Markov chain transition probability matrix $P$,
by deleting the row and column with index $x_{\infty}$.
Letting ${\bf e}$ denote the
$(n-1) \times 1$ column vector consisting entirely of $1$'s, it can be
shown that mean first passage time from state $x_0$ to state $x_{\infty}$
is the $x_0$-th coordinate
of column vector
$(I - P^{-}_{x_{\infty}})^{-1} \cdot {\bf e}$ \cite{meyerMFPT}.


The {\em stationary} probability distribution $\pSt = (\pSt_1,\ldots,\pSt_n)$
is a row vector such that $\pSt \cdot P = \pSt$; i.e.
$\pSt_j = \sum_{i}^n \pSt_i \cdot p_{i,j}$, for all $1\leq j \leq n$.
It can be shown that the stationary probability $\pSt_i$ is the limit,
as $m$ tends to infinity, of the frequency of visiting state $i$ in a
random walk of length $m$ on Markov chain $\mathbb{M}$.
It is well-known that the stationary distribution exists and is unique
for any finite aperiodic irreducible Markov chain \cite{cloteBackofen:book}.

The Metropolis Monte Carlo algorithm \cite{metropolis:MonteCarlo} can
be used to simulate a random walk from initial state $x_0$ to target state
$x_{\infty}$, when energies are associated with the states, as is the case in
macromolecular folding, where free energies can be determined for
protein and RNA conformations from mean field theory, quantum theory,
or experimental measurements. In such cases, a {\em move set}
defines the set $N_x$ of conformations reachable in unit time
from conformation $x$, and the transition probability matrix
$P = (p_{x,y})$ is defined as follows:
\begin{align}
\label{eq:transitionProb1}
p_{x,y} = \left\{
\begin{array}{ll}
\frac{1}{|N_x|} \cdot \min\left(1,\exp(-\frac{E(y)-E(x)}{RT})\right)
&\mbox{if $y \in N_x$}\\
1 - \sum_{u \in N_x} p_{x,u} &\mbox{if $x=y$}\\
0 &\mbox{if $y \not\in N_x$, and $x \ne y$}
\end{array} \right.
\end{align}

If $\pSt_x \cdot p_{x,y} = \pSt_y \cdot p_{y,x}$ holds for all distinct
$x,y \in Q$, then
{\em detailed balance} is said to hold, or equivalently the Markov
chain $\mathbb{M}$ is said to be {\em reversible}.
If transitional probabilities
are defined as in (\ref{eq:transitionProb1}), and if neighborhood
size is constant ($|N_x|=|N_y|$ for all $x,y$), then it is well-known
that the stationary probability distribution
$\pSt=(\pSt_1,\ldots,\pSt_n)$ is the Boltzmann distribution; i.e.
\begin{align}
\label{eq:stationaryBoltzmannProb}
\pSt_i = \frac{\exp(-E(i)/RT)}{Z}
\end{align}
where $E(k)$ is the energy of conformation $k$ at temperature $T$,
$R$ is the universal gas constant,
$T$ is absolute temperature, and the {\em partition function}
$Z = \sum_{i=1}^n \exp(-E(i)/RT)$ is a normalization constant
\cite{waterman:book,cloteBackofen:book}. If neighborhood size is not
constant, as in the case where states are RNA secondary structures and transitions
are restricted to the addition or removal of a single base pair, then
by {\em Hasting's trick}, an equivalent Markov chain can be defined which
satisfies detailed balance -- see equation (\ref{eq:MFPTwithHastings}).

Following Anfinsen's experimental work on the denaturation and
refolding of bovine pancreatic ribonuclease \cite{anfinsen},
the native conformation is assumed to be the ground state having
minimum free energy. These results justify the use of the Monte Carlo
Algorithm \ref{algo:MonteCarlo} in macromolecular kinetics and
structure prediction.


% \begin{algorithm}[Metropolis Monte Carlo algorithm] \hfill\break
% \label{algo:MonteCarlo}
\mverbatim
 1. procedure Metropolis(initialState $x_0$, targetState $x_{\infty}$, maxTime $T_{\max}$)
 2. //time-driven simulation of random walk from $x_0$ to $x_{\infty}$
 3.   $t=0$ (time); $x=x_0$ (initial state)
 4.   while $x \ne x_{\infty}$ and $t< T_{\max}$
 5.     choose a random neighbor $y \in N_x$
 6.     if $E(y)<E(x)$  //greedy step
 7.       x = y         //update x
 8.     else            //Metropolis step
 9.       choose random $z \in (0,1)$
10.       if $z < \exp\left(-\frac{E(y)-E(x)}{RT}\right)$
11.         x = y      //update x
12.     t += 1  //update time regardless of whether x is modified
13.   return x
|mendverbatim
% \end{algorithm}

The mean first passage time from state $x$ to state $y$ can
be approximated by repeated runs of the Monte Carlo algorithm.
In particular, \v{S}ali, Shakhnovich, and Karplus used such Monte Carlo
simulations to investigate the Levinthal paradox of how a protein
can fold to its native state within milliseconds to seconds.
By repeated Monte Carlo simulations using a protein lattice model,
\v{S}ali et al. observed that a large energy difference between
the ground state
and the first misfolded state appears to be correlated with fast folding.

\subsection{Markov processes and equilibrium time}



A continuous time {\em Markov process} $\mathbb{M}=(Q,\pi,P(t))$
is given by a finite set $Q= \{1,\ldots,n\}$ of states, the
initial probability distribution $\pi$,
and the $n\times n$ matrix $P(t)=(p_{i,j}(t))$ of
probability transition functions.\footnote{Only in the
current section does $P(t)$ denote the Markov process
transition probability matrix.  In all later sections of the paper,
$P(t) = (P_1(t),\ldots,P_n(t))$ will denote the population occupancy
frequency function, defined in (\ref{eq:populationOccupancyEquation}).
In Markov processes, the rate matrix $R$ is used rather than the transition
probability matrix $P(t)$, so there will be no ambiguity in later
reserving $P(t)$ to denote the population occupancy function.}
Letting $q_t$ denote the state at (continuous)
time $t$, the probability that the initial state $q_0$ at time $0$ is
$k$ is $\pi_k$, while
\begin{align}
\label{eq:MarkovProcess}
p_{i,j}(t) = Pr[q_{t} = j| q_0 = i] .
\end{align}
The matrix $P'(t)$ of derivatives, defined by
\begin{displaymath}
  P'(t) = \left(
\begin{array}{ccc}
\frac{d p_{1,1}}{d t}(t) & \ldots  & \frac{d p_{1,n}}{d t}(t)\\
\vdots & \ddots & \vdots\\
 \frac{d p_{n,1}}{d t}(t) & \ldots  & \frac{d p_{n,n}}{d t}(t)\\
\end{array}
\right),
\end{displaymath}
can be shown to satisfy
\begin{displaymath}
P'(t) = P(t) \cdot R
\end{displaymath}
where $R = (r_{i,j})$ is an $n \times n$ {\em rate matrix} with the
property that each diagonal entry is $-1$ times the row sum
\begin{displaymath}
r_{i,i} = - \sum_{j\ne i} r_{i,j}.
\end{displaymath}
Define the {\em population occupancy} distribution
$p(t) = (p_1(t),\ldots,p_n(t))$ by
\begin{align}
\label{eq:markovProcessPopulationFreq}
p_i(t) = Pr[q(t) = i] = \sum_{k=1}^n\pi_k p_{k,i}(t)
\end{align}
where $q(t)$ denotes the state of the Markov process
at (continuous) time $t$.

In the case of macromolecular folding, where Markov process states
are molecular conformations and conformational energies are available,
it is typical to define the rate matrix
$R = (r_{x,y})$ as follows:
\begin{align}
\label{eq:transitionProb2}
r_{x,y} = \left\{
\begin{array}{ll}
\min\left(1,\exp(-\frac{E(y)-E(x)}{RT})\right)
&\mbox{if $y \in N_x$}\\
- \sum_{u \in N_x} p_{x,u} &\mbox{if $x=y$}\\
0 &\mbox{if $y \not\in N_x$, and $x \ne y$}.
\end{array} \right.
\end{align}
The {\em master equation} is defined by the matrix differential equation
\begin{align}
\label{eq:masterEquationMatrix}
\frac{d p(t)}{dt} = p(t) \cdot R
\end{align}
or equivalently, for all $1 \leq x \leq n$,
\begin{align}
\label{eq:masterEquation}
\frac{d p_x(t)}{dt} = \sum_{y=1}^n (p_y(t) \cdot r_{y,x} - p_x(t) \cdot r_{x,y})
  = \sum_{x \ne y} (p_y(t) \cdot r_{y,x} - p_x(t) \cdot r_{x,y}) .
\end{align}

As in the case of Markov chains,
$\pSt = (\pSt_1,\ldots,\pSt_n)$ is defined to be
the stationary distribution if
$\pSt \cdot P(t)= \pSt$; i.e.  $\pSt_k = Pr[ q(0) = k ]$ implies that
$Pr[ q(t) = k] = \pSt_k$ for all $t \in \mathbb{R}$ and
$1 \leq k \leq n$.
Define the {\em equilibrium distribution} $\pSt = (\pSt_1,\ldots,\pSt_n)$
to be the unique solution for $p(t)$, when the master equation
(\ref{eq:masterEquation}) is set to equal zero; i.e.
\begin{align}
\label{eq:equilibriumTime}
\sum_{x \ne y} \pSt_x \cdot r_{x,y} = \sum_{x \ne y} \pSt_y \cdot r_{y,x}.
\end{align}

If the equilibrium distribution exists, then necessarily it is equal to
the stationary distribution. A Markov process is said
to satisfy detailed balance if $\pSt_x \cdot r_{x,y} = \pSt_y \cdot r_{y,x}$,
for all $1 \leq x,y \leq n$, where the rate matrix $R = (r_{x,y})$.

The rate equation $R$ for is usually defined
as in (\ref{eq:transitionProb2}) for
Markov processes which model macromolecular folding,
hence it is easy to see that
such Markov processes satisfy detailed balance and moreover that
the equilibrium distribution is the
Boltzmann distribution; i.e.  $\pSt_x = \exp(-E(x)/RT)$ for
all $1 \leq x \leq n$.
Since detailed balance ensures that the eigenvalues of the rate matrix $R$ are
real, one can solve the matrix differential equation
(\ref{eq:masterEquation}) by diagonalizing the rate matrix, and thus obtain
the solution
\begin{align}
\label{eq:populationOccupancyEquation}
p(t) = \sum_{k=1}^n c_k {\bf v}_k e^{\lambda_k t}
\end{align}
where $p(t) = (p_1(t),\ldots,p_n(t))$, and the
values $c_k$ are determined by the initial population occupancy
distribution $p(0)$ at time $0$. Here
${\bf v}_k$ denotes the $k$th eigenvector and
$\lambda_k$ the $k$-th eigenvalue. In particular,
$c_k = \left(p(0) \cdot T^{-1})\right)_k$, where the $j$-th row of $T$
is the $j$-th left eigenvector of $R$,
and $p(0)$ is the population occupancy distribution
at time $t=0$. If the eigenvalues are labeled in decreasing
order, then $\lambda_1 \geq \lambda_2 \geq \cdots \geq \lambda_n$,
and the largest eigenvalue $\lambda_1=0$ has eigenvector $\pSt$, corresponding
to the equilibrium population occupancy distribution, which in this case
is the Boltzmann distribution. The
remaining $n-1$ eigenvalues are negative, and their corresponding eigenvectors
correspond to nonequilibrium kinetic {\em relaxation modes}.

In our software \hermes, we prefer to work with column vectors and
right eigenvectors, and so the population occupancy frequency
$p(t)$ is defined to be the column vector
$p(t)=(p_1(t),\ldots,p_n(t))^T$. Let
$\mathbbt^1,\ldots,\mathbbt^n$ be the right eigenvectors and
$\lambda_1,\ldots,\lambda_n$ be the corresponding right eigenvalues of the
transpose $R^T$ of the rate matrix.
Letting $T$ be the $n\times n$ matrix, whose columns are
$\mathbbt^1,\ldots,\mathbbt^n$, using standard matrix algebra
\cite{matrixTheory}, it can be shown that
\begin{align}
\label{eq:solutionMasterEquation}
p(t) &=& \sum_{j=1}^n (T^{-1}
\mathbbP(0))_j \cdot \mathbbt^j \cdot e^{\lambda_j t}
\end{align}
The {\em equilibrium time} can be directly computed by using a
nonlinear solver to solve for $t$ in
\begin{align}
\label{eq:solutionEquilibriumTime}
\pSt &=& \sum_{j=1}^n (T^{-1}
\mathbbP(0))_j \cdot \mathbbt^j \cdot e^{\lambda_j t}
\end{align}
where $\pSt= (\pSt_1,\ldots,\pSt_n)^T$ and $\pSt_k = \frac{\exp(-E(k)/RT)}{Z}$.
However, we have found it more
expedient to compute the {\em equilibrium time} as the smallest $t_0$,
such that for $t \in \{t_0+1,t_0+2,t_0+3,t_0+4\}$, the absolute difference
$|p(t)[x_{\infty}] - p(t_0)[x_{\infty}]| < \epsilon$, for $\epsilon =
10^{-4}$, where $x_{\infty}$ is the target RNA structure (usually taken
to be the minimum free energy structure, though this is not necessary for
the software). We also considered defining the equilibrium time to be the
smallest $t_0$, such that for $t \in \{t_0+1,t_0+2,t_0+3,t_0+4\}$, the
absolute difference $|p(t)[x] - p(t_0)[x]| < \epsilon$ for all $x \in
Q$; however, results suggest that this definition is inferior
to that just given, perhaps due to numerical instability issues when
$Q$ is taken to be the set of all secondary structures for sequences in
the benchmarking set described later.


In \cite{gillespieStochasticSimulation1} Gillespie described
a very influential algorithm to simulate a finite Markov process.
The pseudocode, is given in Algorithm~\ref{algo:Gillespie}.
Though Gillespie's original application was for chemical kinetics,
Flamm et al.  adapted the method for the kinetics of RNA secondary structure
folding, as implemented in \kinfold \cite{flammPhD,flamm}.


% \begin{algorithm}[Gillespie algorithm] \hfill\break
% \label{algo:Gillespie}
\mverbatim
 1. procedure Gillespie(initialState $x$, targetState $w$, maxTime $T_{\max}$)
 2.   x = initial state; time $t=0$
 3.   while $x \ne w$ and $t< T_{\max}$ \{
 4.     $\Phi=0$ //$\Phi$ is flux (not probability) out of $x$
 5.     for all $y \in N_x$ do
 6.       $r_{x,y} = \min\left(1,\exp(-\frac{E(y)-E(x)}{RT})\right)$
 7.       $\Phi += r_{x,y}$
 8.     for all $y \in N_x$ do $r_{x,y} = \frac{r_{x,y}}{\Phi}$
 9.     choose random $z_1 \in (0,1)$
10.     t += $-\frac{1}{\Phi} \ln(z_1)$  //update time
11.     choose random $z_2 \in (0,1)$
12.     sum = 0
13.     for $y \in N_x$
14.       sum += $r_{x,y}$
15.       if $z_2 \leq sum$ then
16.         x=y; break //use roulette wheel
17.    \}
18.   return z
|mendverbatim
% \end{algorithm}


\section{Methods}
\label{section:software}

The \hermes software package was developed on the Macintosh OS X
operating system (10.9.2 and 10.10) and should work with any Unix-like platform
(Ubuntu, Debian, and CentOS were tested). We make the source code freely
available under the MIT License in two locations.
Our lab hosts the latest stable version of the code at
\url{http://bioinformatics.bc.edu/clotelab/Hermes} and a fully
version-controlled copy at \url{https://github.com/evansenter/hermes}.
The data and figures presented in this article were generated with the
source code hosted at the first URL, and we make no guarantee as to
the stability of development branches in our Git repository.

External dependencies for the software include a C (resp. C++)
compiler supporting the GNU99 language specification (resp. C++98),
FFTW implementation of Fast Fourier Transform \cite{FFTW05} ($\geq
3.3.4$) \url{http://www.fftw.org/}, Gnu Scientific Library GSL ($\geq
1.15$) \url{http://www.gnu.org/software/gsl/}, Vienna RNA Package
\cite{Lorenz.amb11} ($\geq 2.0.7$) \url{http://www.viennarna.at}, and
any corresponding sub-packages included with the aforementioned
software. For a more detailed explanation of both external
dependencies and installation instructions, refer to the
`DOCS.pdf' file at the web site
outlining the configuration and compilation process for the \hermes
suite.

\hermes is organized into three independent directories: (1)
\ffttwo, (2) \rnamfpt, and (3) \rnaeq (see
Figure \ref{fig:organizationHermes}). These packages compile into both
standalone executables and archive files. The archives provide a DRY
API which allow the development of novel applications using source
from across the \hermes package without having to copy-and-paste
relevant functions. We provide two such examples of this in the {\tt
ext} subdirectory: \fftmfpt and \ffteq. These applications
are simple C drivers that use functions from \ffttwo, \rnamfpt and
\rnaeq to replicate a pipeline of executable calls without having to
deal with intermediary data transformation, I/O between calls or
slow-down due to a scripting language driver such as Python or R.

\subsection{Programs}

What follows is a brief overview of the three core applications
(\ffttwo, \rnamfpt, \rnaeq), as well as the two extension
programs (\fftmfpt, \ffteq). For an exhaustive list of
command-line options, please see the corresponding help files.

Given an RNA sequence \seq of length $n$ and two reference
structures $A,B$, the algorithm \ffttwo \cite{Senter.jmb14}
computes for each $0 \leq x,y \leq n$ the Boltzmann probability
$P(x,y) = \frac{Z(x,y)}{Z}$ of all structures having base pair
distance $x$ to \strA [resp. $y$ to \strB]. Although the algorithm for the
software \ffttwo is unchanged from that of \cite{Senter.jmb14},
the software described in this
paper has been heavily refactored to provide a user-facing API,
homogeneous with the rest of the \hermes suite. As with the rest of
\hermes, \ffttwo makes use of either the Turner 1999 \cite{turner}
or the later Turner 2004 energy parameters
\cite{Turner.nar10}, taken respectively
from the files rna\_turner1999.par and  rna\_turner2004.par in Vienna
RNA Package. \ffttwo takes as input an RNA sequence \seq and two
secondary structures $A,B$ of the input RNA sequence -- both given
in dot-bracket
notation. Alternatively, \ffttwo can also process a file of RNAs,
where data for each RNA appears in the form: FASTA comment, sequence,
structure \strA, structure \strB -- each on a separate line. In return,
for each RNA, the program returns as output a 2D probability grid
whose value at
grid position $(x,y)$ is the Boltzmann probability of all secondary
structures compatible with \seq having base pair distance $x$ (resp.
$y$) from \strA (resp. \strB).

\rnamfpt computes the {\em mean first passage time} (MFPT), sometimes
referred to as the {\em hitting time} of a Markov chain, by using
matrix inversion \cite{meyerMFPT} -- see
Section~\ref{section:algo}. The program takes as input a
comma separated value (CSV) file containing the non-zero positions and
values of a 2D probability grid; i.e. a CSV format file having columns {\em i},
{\em j}, and {\em p}. The first two columns, {\em i} and {\em j}
correspond to the {\em 0-indexed} row-ordered position in the rate
matrix and the final column {\em p} is the probability $p_{i,j}$
of a transition from {\em i} to {\em j}.  From this input, the
mean first passage time is constructed by matrix inversion.
Because this program was designed with the original intent
of handling 2D-probability grids, all vertexes are uniquely identified
by index tuples (which conceptually correspond to positioning in a 2D
array). However, it is trivial to use this code with both
1D-probability grids such as those produced by \fftbor
\cite{Senter.po12} or arbitrary transition matrices without any change
to the underlying implementation. The software additionally provides
many options for defining the structure of the graph underlying the
Markov chain. Some of these include the option to force a fully
connected graph (useful in cases where there is no non-zero path
between the start / end state) or to enforce detailed balance.
Finally, \rnamfpt also accepts as input the probability transition matrix,
a stochastic matrix with row sums equal to $1$, and computes the mean
first passage time for the corresponding Markov chain.

\rnaeq computes the population proportion of a user-provided structure
over arbitrary time units. Like \rnamfpt, this program takes as input a
comma separated value (CSV) file containing the non-zero positions and
values of a 2D probability grid. From this input a rate matrix
is constructed for the underlying Markov process. Alternatively,
\rnaeq can accept as input an arbitrary rate matrix. Performing
spectral decomposition of the column-ordered rate matrix that
underlies the corresponding Markov process, \rnaeq computes either
the equilibrium time or population occupancy frequencies.
Additionally, \rnaeq can call the
Vienna RNA Package program \rnasub \cite{Wuchty.b99}, with a
user-specified upper bound to the energy difference with the minimum
free energy. With this option, the rate matrix is
constructed for the Markov process, whose states consist of all
the structures returned by \rnaeq, and the equilibrium time or
population occupancy frequencies are computed. Due to the time
and memory required for this option, we do not expect it to be used except for
small sequences.

Beyond these three core applications, \hermes includes two
additional programs located in the {\tt ext} directory, which fulfill
more specific roles and highlight the modular nature of the codebase.

\fftmfpt approximates the mean first passage time of a given RNA
sequence folding from input structure A to B, by {\em exactly}
computing the mean first passage time from state $(0,d_0)$ to state
$(d_0,0)$ in the 2D probability grid obtained from running
\ffttwo. Here, $d_0$ is the base pair distance between structures
$A,B$, and the MFPT is computed for the Markov chain, whose states are
the non-empty 2D probability grid points, and whose transition
probabilities are defined by $p_{(x,y),(x',y')} =
\frac{P(x',y')}{P(x,y)}$. As we report in this paper, given an RNA
sequence \seq, if \strA is the empty structure and \strB the MFE
structure of \seq, then \fftmfpt output is well correlated with the
exact MFPT in folding the empty structure to the MFE structure, where
transitions between structures involve the addition or removal of a
single base pair.

\ffteq allows an investigator to efficiently estimate population
kinetics for a sequence folding between two arbitrary, but fixed,
structures. The transition rate matrix underlying the Markov process
necessary for eigendecomposition is derived from the 2D-energy
landscape. Vertices in the rate matrix represent a subset of
structures compatible with the input sequence as modeled by
\ffttwo, which makes the graph size more tractable than structural
sampling with \rnasub, even with constraints.


%\subsection{Algorithms Used} \label{section:algorithmsUsed}
%
%\subsection*{Mean first passage time (MFPT)} Given a discrete time
%Markov chain with states $1,\ldots,n$, a stopping state $k$, and
%transition probability matrix $M = (p_{i,j})$, the $MFPT$ to reach
%state $k$ can be computed as follows. Let $M^{-}_{(k)}$ denote the
%matrix obtained from $M$ by removing both the $k$th row and $k$th
%column. Compute the column vector $T = (I - M^{-}_{(k)})^{-1} \cdot
%{\bf e}$, where $I$ denotes the $(n-1) \times (n-1)$ identity matrix,
%and ${\bf e}$ denotes the column vector composed of $n-1$ many ones.
%It follows that the mean first passage time to reach state $k$, when
%starting in state $1 \leq i < k$ [resp. $k< i \leq n$] is the $i$th
%[resp. $(i-1)$th coordinate of $T$ \cite{meyerMFPT}.
%
%\subsection*{Computing equilibrium time for a Markov process}
%
%Given a continuous time Markov process $\mathbb{M}$ with rate matrix
%$R=(r_{x,y})$, the master equation is
%\begin{align}
%\label{eq:masterEquation} \frac{dP_x(t)}{dt} = \sum_{y} (\pSt_y \cdot
%r_{y,x} - \pSt_x \cdot r_{x,y}) = \sum_{x \ne y} (\pSt_y \cdot r_{y,x} -
%\pSt_x \cdot r_{x,y})
%\end{align}
%where if $n=|X|$ then $P(t)=(P_1(t),\ldots,P_n(t))$ denotes the
%population (row) vector at time $t$. Equilibrium occurs when the
%master equation equals 0; i.e. $\frac{dP(t)}{dt} = \pSt \cdot R = 0$.
%Now let $P(t)$ denote the column vector
%$P(t)=(P_1(t),\ldots,P_n(t))^T$, and let
%$\mathbbt^1,\ldots,\mathbbt^n$ be the eigenvectors and
%$\lambda_1,\ldots,\lambda_n$ be the corresponding eigenvalues of the
%rate matrix. Letting $T$ be the $n\times n$ matrix, whose columns are
%$\mathbbt^1,\ldots,\mathbbt^n$, using standard matrix algebra
%\cite{matrixTheory}, it can be shown that
%\begin{align}
%\label{eq:solutionMasterEquation} P(t) &=& \sum_{j=1}^n (T^{-1}
%\mathbbP(0))_j \cdot \mathbbt^j \cdot e^{\lambda_j t}
%\end{align}
%The {\em equilibrium time} can be directly computed by using a
%nonlinear solver to solve for $t$ in
%(\ref{eq:solutionMasterEquation}); however, we have found it more
%expedient to compute the {\em equilibrium time} as the smallest $t_0$,
%such that for $t \in \{t_0+1,t_0+2,t_0+3\}$, the absolute difference
%$|P(t)[x_{\infty}] - P(t_0)[x_{\infty}]| < \epsilon$, for $\epsilon =
%10^{-4}$. We also considered defining the equilibrium time to be the
%smallest $t_0$, such that for $t \in \{t_0+1,t_0+2,t_0+3\}$, the
%absolute difference $|P(t)[x] - P(t_0)[x]| < \epsilon$ for all $x \in
%\str$; however, results suggest that this definition is inferior
%to that just given, perhaps due to numerical instability issues. This
%alternative definition remains available through the command-line
%options, but due to its weak performance we have omitted the data
%below. Gnu Scientific Library (GSL) routines are used to determine the
%spectral decomposition of the rate matrix.



\begin{figure*}[!th]
\begin{center}
%\includegraphics[width= \textwidth]{FIGURES/hermes_layout}
\includegraphics[width= \textwidth]{figure1.eps}
\caption{Overall organization of
\hermes. \ffttwo, \rnamfpt, and \rnaeq are three distinct software
packages we have developed, which compile into both standalone
executables and archive files, providing a DRY API that allow novel
applications development using source from each of the packages,
without having to copy-and-paste relevant functions. The applications
\fftmfpt and \ffteq are C drivers that use data structures and functions from
\ffttwo, \rnamfpt and \rnaeq. \fftmfpt computes the mean first
passage time (MFPT) for an RNA secondary structure to fold from an
initial structure, such as the empty structure or a given metastable
structure, into a target structure, such as the minimum free energy
(MFE) structure or possibly the Rfam \cite{Gardner.nar11} consensus
structure. \ffteq uses spectral decomposition to compute the equilibrium time and the
fraction of the population of RNA structures that are
equal to a given target structure, as a function of time.
 } \label{fig:organizationHermes}
\end{center}
\end{figure*}


\section{Benchmarking results}
\label{section:results}

\subsection{Kinetics methods}

In this section, we briefly describe the kinetics algorithms used in our
comparative study using a benchmarking set of 1,000 small RNAs. Each RNA from
the benchmarking set was folded from the initial empty structure to the
target minimum free energy structure.  Detailed
explanations of all kinetics programs and parameters are given
in the Appendix.

The kinetics methods benchmarked in Table~\ref{table:correlation}
are \rnamfpt,
\rnaeq,
\fftmfpt,
\fftbor,
\ffteq -- all computed using \hermes, together with,
\kinfold \cite{flammHofacker},
{\tt BarrierBasins} \cite{wolfingerStadler:kinetics},
and \rnatwofold \cite{hofacker:RNAbor2D}. The {\em exact} data \rnamfpt
[resp. \rnaeq] were computed using the \hermes software {\tt RNAmfpt}
[resp. {\tt RNAeq}] in the following fashion.

The gold standard, \rnamfpt, is obtained by computing the
mean first passage time by matrix inverse $(I - P^{-}_{x_{\infty}})^{-1}$
from the transition probability matrix $P$, defined by equation
(\ref{eq:transitionProb1}) from the ensemble of all secondary structures
for each 20-mer in the benchmarking set described in the next section.
The platinum standard, \rnaeq, is obtained from the
population occupancy frequencies
(\ref{eq:solutionEquilibriumTime}), determined by spectral decomposition
of the rate matrix, the latter defined by
(\ref{eq:transitionProb2}) from the ensemble of all secondary structures
for each 20-mer in the benchmarking set.

\fftmfpt computes the mean first passage time
for the coarse-grained model with macrostates $M(x,y)$, consisting of
all secondary structures having base pair distance $x$ from reference
structure \strA and distance $y$ from reference structure \strB.
\ffttwo is used in \hermes to compute the Boltzmann probabilities
$p(x,y) = \frac{Z(x,y)}{Z}$ for the coarse-grained model consisting of
macrostates $M(x,y)$. Transition probabilities are computed by
equation (\ref{eq:transitionProb1}), and matrix inversion is used to
determine mean first passage time for the coarse-grained model.
For comparison purposes, the method \fftbor calls the
program \fftbor \cite{Senter.po12} to compute the Boltzmann probabilities
$p(x) = \frac{Z(x)}{Z}$ for the coarse-grained model consisting of
macrostates $M(x)$ of all secondary structures having base pair distance
$x$ to a given target structure. Mean first passage time is then computed
for this model, which is even more coarse-grained than that of \fftmfpt.
\ffteq computes the equilibrium time
for the coarse-grained model with macrostates $M(x,y)$, consisting of
all secondary structures having base pair distance $x$ from reference
structure \strA and distance $y$ from reference structure \strB. As in
\fftmfpt, the program
\ffttwo is used to compute the Boltzmann probabilities
$p(x,y) = \frac{Z(x,y)}{Z}$ for the coarse-grained model consisting of
macrostates $M(x,y)$. The rate matrix is computed by
equation (\ref{eq:transitionProb2}), then population occupancy
frequencies and equilibrium time
are computed by equation (\ref{eq:markovProcessPopulationFreq})
using spectral decomposition of the rate matrix.


\begin{table*}
\begin{tabular}
  {|l|llllllll|} \hline \small{Hastings (Yes\textbackslash No)} & \small{MFPT} & \small{Equilibrium} & \small{Kinfold} & \small{FFTmfpt} & \small{RNA2Dfold} & \small{FFTbor} & \small{BarrierBasins} & \small{FFTeq} \\
  \hline \small{MFPT}   & 1      & 0.5683 & 0.7945 & 0.5060 & 0.5110 & 0.5204 & 0.5280 & 0.4472 \\
  \small{Equilibrium}   & 0.5798 & 1      & 0.7814 & 0.7043 & 0.7025 & 0.5080 & 0.5979 & 0.6820 \\
  \small{Kinfold}       & 0.7933 & 0.7507 & 1      & 0.7312 & 0.7358 & 0.6241 & 0.6328 & 0.6445 \\
  \small{FFTmfpt}       & 0.6035 & 0.7935 & 0.7608 & 1      & 0.9980 & 0.5485 & 0.8614 & 0.9589 \\
  \small{RNA2Dfold}     & 0.6076 & 0.7919 & 0.7655 & 0.9983 & 1      & 0.5584 & 0.8538 & 0.9515 \\
  \small{FFTbor}        & 0.5416 & 0.5218 & 0.6241 & 0.5748 & 0.5855 & 1      & 0.3450 & 0.4229 \\
  \small{BarrierBasins} & 0.6346 & 0.6578 & 0.6328 & 0.8310 & 0.8217 & 0.3450 & 1      & 0.9149 \\
  \small{FFTeq}         & 0.5614 & 0.7916 & 0.6966 & 0.9670 & 0.9590 & 0.4757 & 0.8940 & 1      \\
  \hline
 \end{tabular}

 \caption{Table of Pearson correlation coefficients for various methods to compute or approximate RNA secondary structure folding kinetics. Lower [resp. upper] triangular entries are with [resp. without] the Hastings correction for Markov chain probability matrices. The methods are: \rnamfpt (mean first passage time, computed by matrix inversion for the Markov chain consisting of all secondary structures, with move allowed between structures differing by one base pair), \rnaeq (equilibrium time, computed by spectral decomposition of a rate matrix comprising all secondary structures to compute population fraction $P(t)$ at time $t$), \kinfold (an implementation of Gillespie's Algorithm to approximate refolding pathways using an event-based Monte Carlo simulation), \fftmfpt (mean first passage time for Markov chain consisting of ``grid point'' states $(x,y)$ with probability $P(x,y)=\sum_S exp(-E(S)/RT)/Z$, computed by \ffttwo, where the sum is taken over structures having base pair distance $x$ to the empty structure and $y$ to the MFE structure, \rnatwofold (mean first passage time, computed as previously explained, but using \rnatwofold in place of \ffttwo to compute $P(x,y)$), \fftbor (mean first passage time, computed for the Markov chain consisting of states $0,1,\ldots,n$, for which $P(x) = \sum_S \exp(-E(S)/RT)/Z$, where the sum is taken over all secondary structures whose base pair distance is $x$ from the MFE structure), {\tt BarrierBasins} (equilibrium time, computed using spectral decomposition on the Markov process consisting of ``grid point'' states output from {\tt Barriers}), and \ffteq (equilibrium time, computed in the same fashion as {\tt BarrierBasins} using a Markov process derived from the energy landscape output by \ffttwo). }
\label{table:correlation}
\end{table*}


Kinetics algorithms developed by other groups and benchmarked in our study
are the following: \kinfold \cite{flamm},
{\tt BarriersBasin} \cite{flammHofacker,wolfingerStadler:kinetics},
{\tt Kinwalker} \cite{Geis.jmb08}. Below, we give a brief overview of these methods.

\kinfold \cite{flamm} is an implementation of Gillespie's
Algorithm \ref{algo:Gillespie}.
For benchmarking purposes, we determined the mean first passage time
from $10^4$ \kinfold simulations, each with an upper bound of
$10^8$ Monte Carlo steps in order to ensure convergence of each run.
{\tt BarriersBasin} is our name for the method described by Wolfinger
et al. \cite{wolfingerStadler:kinetics}, which first runs
\rnasub \cite{flammHofacker} to obtain an energy-sorted list of
secondary structures that contain the empty structure,
then computes locally optimal structures, saddle points and macrostate
{\em basins of attraction}  using the program {\tt barriers}
\cite{wolfingerStadler:kinetics}. Subsequently, we use \hermes to
determine the equilibrium time for the resulting Markov process whose
rate matrix is returned by {\tt barriers}.\footnote{Since
the rate matrix returned by {\tt barriers}
does not necessarily have the property that diagonal entries are negative
row sums, as required by equation (\ref{eq:transitionProb2}), we modify
diagonal entries of the {\tt barriers} rate matrix to satisfy this property.}
\footnote{At the time we did the benchmarking, we were unaware of the
program {\tt treekin} from Vienna RNA Package, which can be combined with
the rate matrix from {\tt barriers} to determine population occupation
frequency output in a file. It is necessary for the user to subsequently
write a program to determine the equilibrium time from the output of
{\tt treekin}. Since \hermes determines equilibrium time directly
from any input rate matrix, we used \hermes to postprocess the
output rate matrix from {\tt barriers}.}
Since {\tt Kinwalker} \cite{Geis.jmb08} implements co-transcriptional folding,
we found that its correlation with the gold and platinum standards was
extremely poor, hence we do not further report the results from our tests
with {\tt Kinwalker}.
Programs for the remaining kinetics methods, such as the motion planning
method of Tang et al. \cite{Tang.jmb08}, the stochastic simulation
{\tt Kinefold} \cite{Xayaphoummine.nar05}, etc.
from the introduction were either not available, or else it
appeared to be difficult/impossible to extract
mean first passage times or equilibrium times to compare with the
gold and platinum standards.


Finally, Table~\ref{table:correlation}
also includes the method we call here by \rnatwofold.
Given reference structures $A,B$, the \rnatwofold  program
\cite{hofacker:RNAbor2D,Lorenz.amb11} computes for each integer
$x,y$, the minimum free energy structure and partition function
among all secondary structures having base pair distance $x$ to \strA
and $y$ to \strB.\footnote{\rnatwofold generalizes an earlier algorithm
{\tt RNAbor} \cite{Freyhult.b07,Freyhult.nar07}, which computes the
minimum free energy structure and partition function among all structures
having base pair distance $x$ to reference structure \strA.}
\ffttwo \cite{Senter.jmb14} is a much faster program, using the
fast Fourier transform, which
computes the partition function (but not the minimum free energy
structure) over all structures having base pair distance $x$ to \strA and
$y$ to \strB. By the kinetics method dubbed
\rnatwofold, we mean that we applied
\rnatwofold from Vienna RNA Package 2.1.7
\cite{hofacker:RNAbor2D,Lorenz.amb11} in order
to obtain probabilities $p(x,y) = \frac{Z(x,y)}{Z}$,
and then applied \hermes to compute the mean first passage time for
the resulting coarse-grained Markov chain, as in \fftmfpt.

For more detailed description of each kinetics method, please see
the Appendix.

\subsection{Benchmarking data}

In this section, we describe a
benchmarking set of 1,000 small RNAs used to benchmark the previously
described kinetics methods in a comparative study. To ensure that mean
first passage time can be computed from
$(I - P^{-}_{x_{\infty}})^{-1} \cdot {\bf e}$ by using matrix
inversion, that spectral decomposition of the rate matrix is possible,
and to ensure that \kinfold simulations would provide sufficient
statistics, we generated a collection of 1,000 random RNA sequences of
length 20 nt, each having expected compositional frequency of $1/4$
for A,C,G,U, and each having at most 2,500 distinct secondary
structures, such that the minimum free energy is less than or equal to
$-5.5$ kcal/mol.

For example, one of the 1,000 sequences is ACGCGACGUGCACCGCACGU with
minimum free energy structure {\tt .....((((((...))))))} having free
energy of $-6.4$ kcal/mol. Statistics for the free energies of the
2,453 secondary structures of this 20-mer are the following: mean is
$10.695$, standary deviation is $4.804$, maximum is $25.00$, minimum
is $-6.40$. A histogram for the free energy of all secondary
structures of ACGCGACGUGCACCGCACGU is depicted in
%Figure \ref{fig:freeEnergyHistogramSecStrACGCGACGUGCACCGCACGU}.
the left panel of Figure \ref{fig:PLMVd}. The right panel of the
same figure depicts the minimum free energy structure of the
54 nt hammerhead type III ribozyme from Peach Latent Mosaic Viroid
(PLMVd), discussed later. This secondary structure is identical
to the consensus structure from Rfam 11.0 \cite{Gardner.nar11}.


\begin{figure*} [!ht]
\begin{center}
%\includegraphics[width=0.45\textwidth]{FIGURES/freeEnergyHistogramSecStrACGCGACGUGCACCGCACGU}
\includegraphics[width=0.45\textwidth]{figure2a.eps}
%\includegraphics[width=0.45\textwidth]{FIGURES/plmvd_positional_entropy}
\includegraphics[width=0.45\textwidth]{figure2b.eps}
\caption{\small
{\em (Left)} Histogram of free energies of secondary structures of
ACGCGACGUGCACCGCACGU, which range from $-6.5$ to $+25$ kcal/mol, with
mean of $10.695$ kcal/mol.
{\em (Right)} Minimum free energy structure of the 54 nt Peach Latent Mosaic
Viroid (PLMVd) AJ005312.1/282-335, which is identical to the consensus
structure from Rfam 11.0 \cite{Gardner.nar11}. {\tt RNAfold} from
Vienna RNA Package 2.1.7 with energy parameters from the Turner 1999
model were used, since the minimum free energy structure determined by
the more recent Turner 2004 energy parameters
does {\em not} agree with the Rfam consensus structure -- see
\cite{syntheticHammerheads}. Positional entropy, a measure
of divergence in the base pairing status at each positions for the
low energy ensemble of structures, is indicated by color, using the
RNA Vienna Package utility script {\tt relplot.pl}.
}
\label{fig:PLMVd}
\end{center}
\end{figure*}


Figure \ref{fig:meanStdevKinfoldRuns1000sequences} displays the mean
and standard deviation for \kinfold simulations of folding time
for each of the 1,000 RNA sequences from our benchmarking data. For
each sequence, the mean and standard deviation of the time required to
fold the empty structure to the MFE structure were computed from
10,000 \kinfold runs, each run with an upper bound of $10^8$
Monte Carlo steps, thus ensuring that all simulations converged. The
sequences were then sorted by increasing folding time mean. Standard
deviation exceeded the mean in $83.9\%$ of the 1,000 cases, indicating
the enormous variation between separate \kinfold runs, even for
20 nt RNA sequences having at most 2,500 secondary structures. In our
opinion, \kinfold is an expertly crafted implementation of
Gillespie's algorithm for an event driven Monte Carlo simulation of
one-step RNA secondary structure folding. From the standpoint of
biophysics and physical chemistry, there is no more reliable
simulation method, except of course the exact computation of mean
first passage time using linear algebra. Nevertheless, the enormous
time required for reliable \kinfold estimations and the large
standard deviations observed point out the need for a faster method to
approximate folding time.


\begin{figure*}
\centering
%\includegraphics[width=0.45\textwidth]{FIGURES/kinfoldFoldingTimeforSeq1Cropped}
\includegraphics[width=0.45\textwidth]{figure3a.eps}
%\includegraphics[width=0.45\textwidth]{FIGURES/kinfoldMeanStdevCropped}
\includegraphics[width=0.45\textwidth]{figure3b.eps}
\caption{{\em (Left)} Histogram of \kinfold folding times for 20-mer CCGAUUGGCG AAAGGCCACC. The mean [resp. standard deviation] of 10,000 runs of \kinfold for this 20-mer is 538.37 [resp. 755.65]. Note the close fit to the exponential distribution, {\em (Right)} Mean minus standard deviation ($\mu -\sigma$), mean ($\mu$), and mean plus standard deviation ($\mu + \sigma$) of the logarithm of \kinfold folding times, taken over 10,000 runs for each of the 1,000 sequences from the benchmarking set of 20-mers. For graphical illustration, we have sorted the log folding times in increasing order. } \label{fig:meanStdevKinfoldRuns1000sequences}
\end{figure*}



\subsection{Correlations}

In this section, we display the correlation between (1) the {\em gold
standard} method \rnamfpt, both with and without the Hastings
modification using equations (\ref{eq:MFPTwithHastings}) and
(\ref{eq:MFPTwithoutHastings}), (2) the {\em platinum standard} method {\tt
Equilibrium}, (3) the {\em silver standard} method \kinfold, (4) {\tt
FFTmfpt} with and without the Hastings modification using equations
(\ref{eq:transitionProbabilityFFTbor2DwithHastings}) and
(\ref{eq:transitionProbabilityFFTbor2DwithoutHastings}), (5) {\tt
FFTeq} which computes equilibrium time for the 2D-grid, (6) {\tt
RNA2Dfold } with and without the Hastings modification using equations
(\ref{eq:transitionProbabilityFFTbor2DwithHastings}) and
(\ref{eq:transitionProbabilityFFTbor2DwithoutHastings}). Correlations
with [resp. without] the Hastings modification are summarized in the
lower [resp. upper] triangular portion of
Table~\ref{table:correlation}. It is clear that correlations between
the mathematically exact methods \rnamfpt, \rnaeq, and
approximation methods \kinfold, \fftmfpt, \ffteq, {\tt
RNA2Dfold} are improved when using the Hastings correction.

Figures~\ref{fig:scatterplot_single_bp_moves_for_kinfold_vs_rnaeq},
\ref{fig:scatterplot_single_bp_moves_for_actual_mfpt_vs_fftbor2d},
\ref{fig:scatterplot_single_bp_moves_for_fftbor2d_vs_rnapopulation}
depict scatterplots for kinetics obtained by some of the algorithms
above. The left panel of
Figure \ref{fig:scatterplot_single_bp_moves_for_kinfold_vs_rnaeq}
shows a scatter plots for gold standard \rnamfpt versus platinum
standard \rnaeq, with correlation value 0.5652. The right
panel of the same figure shows a scatter plot for \kinfold versus
\rnaeq, with correlation 0.7814. Note the persence of two
clusters in this and some of the other scatter plots. Cluster A
consists of RNA sequences whose folding time, as determined by {\tt
MFPT} or \rnaeq, is rapid -- specifically, the natural
logarithm of the MFPT is at most 7.5. Cluster B consists of the
remaining RNA sequences, whose folding time is longer than that of
cluster A. There are no significant differences between RNA sequences
in clusters A and B with respect to GC-content, sequence logo, minimum
free energy, number of secondary structures, etc.  The left panel of
Figure \ref{fig:scatterplot_single_bp_moves_for_actual_mfpt_vs_fftbor2d}
shows the scatter plot for \rnamfpt versus \kinfold, with
correlation 0.7933, and the right panel shows the scatter plot for
\rnamfpt versus \fftmfpt, with correlation 0.6035.
Figure \ref{fig:scatterplot_single_bp_moves_for_fftbor2d_vs_rnapopulation}
shows scatter plots for \fftmfpt versus \kinfold (Left) and
for \fftmfpt versus \ffteq (Right), with respective
correlation values 0.7608 and 0.9589. \kinfold obviously provides
a better correlation with the exact value of mean first passage time;
however, since the standard deviation of \kinfold runs is as
large as the mean,\footnote{It follows from spectral decomposition that
equilibrium time follows an exponential distribution (or sum of
exponential distributions). Exponential distributions have the property
that the mean is equal to the standard deviation, hence
it is not surprising that \kinfold
kinetics have this property.} accurate kinetics estimates
from \kinfold require prohibitively large computational time -- indeed, in
\cite{wolfingerStadler:kinetics} reliable kinetics for phe-tRNA from
yeast were obtained by 9,000 \kinfold simulations, each for $10^8$
steps, requiring 3 months of CPU time on an Intel Pentium 4 running at
2.4 GHz under Linux. Although the correlation value of 0.6035 between
\rnamfpt and \fftmfpt is much less than that obtained by {\tt
Kinfold}, the runtime required by our method \fftmfpt is measured
in seconds, even for moderate to large RNAs. For this reason, we
advocate the use of \fftmfpt in synthetic biology screens to
design RNA sequences having certain desired kinetic properties. Once
promising candidates are found, it is possible to devote additional
computational time to \kinfold simulations for more accurate
kinetics.


\begin{figure}
\centering
%\includegraphics[width=0.45 \textwidth]{FIGURES/scatterplot_for_rnamfpt_with_hastings_vs_rateeq_mfe_only_without_hastings}
\includegraphics[width=0.45 \textwidth]{figure4a.eps}
%\includegraphics[width=0.45 \textwidth]{FIGURES/scatterplot_for_rateeq_mfe_only_without_hastings_vs_kinfold_lp_global}
\includegraphics[width=0.45 \textwidth]{figure4b.eps}
\caption{ Scatter plots of the natural logarithm of times from {\tt
MFPT} versus \rnaeq (Left) and for \kinfold versus
\rnaeq (Right). }
\label{fig:scatterplot_single_bp_moves_for_kinfold_vs_rnaeq}
\end{figure}


\begin{figure}
\centering
%\includegraphics[width=0.45 \textwidth]{FIGURES/scatterplot_for_rnamfpt_with_hastings_vs_kinfold_lp_global}
\includegraphics[width=0.45 \textwidth]{figure5a.eps}
%\includegraphics[width=0.45 \textwidth]{FIGURES/scatterplot_for_rnamfpt_with_hastings_vs_fftbor2d_with_hastings}
\includegraphics[width=0.45 \textwidth]{figure5b.eps}
\caption{ Scatter plots of the natural logarithm of times from {\tt
MFPT} versus \kinfold (Left) and for \rnamfpt versus {\tt
FFTmfpt} (Right). }
\label{fig:scatterplot_single_bp_moves_for_actual_mfpt_vs_fftbor2d}
\end{figure}


\begin{figure}
\centering
%\includegraphics[width=0.45 \textwidth]{FIGURES/scatterplot_for_kinfold_lp_global_vs_fftbor2d_with_hastings}
\includegraphics[width=0.45 \textwidth]{figure6a.eps}
%\includegraphics[width=0.45 \textwidth]{FIGURES/scatterplot_for_fftbor2d_without_hastings_vs_ffteq_mfe_only_without_hastings}
\includegraphics[width=0.45 \textwidth]{figure6b.eps}
\caption{ Scatter plots of the natural logarithm of times from {\tt
Kinfold} versus \fftmfpt (Left) and for \fftmfpt versus {\tt
FFTeq} (Right). }
\label{fig:scatterplot_single_bp_moves_for_fftbor2d_vs_rnapopulation}
\end{figure}



\section{Discussion}
\label{section:discussion}

In this paper, we have introduced
the fast, novel RNA kinetics software suite, {\tt
Hermes}, which consists of the three software packages, \ffttwo,
{\tt RNAmfpt}, and {\tt RNAeq},  together with two C drivers,
\fftmfpt and \ffteq.

Given an RNA sequence \seq,  \ffttwo computes the Boltzmann probability
$p(x,y) = \frac{Z(x,y)}{Z}$ of secondary structures of \seq, whose base
pair distance from initial structure \strA is $x$ and from target structure
\strB is $y$. \ffttwo is an enhancement of the algorithm
described in \cite{Senter.jmb14}, shown there
to run both in theory and in practice much faster than \rnatwofold
\cite{hofacker:RNAbor2D},
since \ffttwo uses the fast Fourier transform to
compute Boltzmann probabilities $p(x,y)$ by polynomial interpolation.


{\tt RNAmfpt} computes the mean first passage time for the
coarse-grained Markov chain,
consisting of the macrostates $M(x,y)$ of all secondary structures,
whose base pair distance is $x$ [resp. $y$] from initial structure \strA
[resp. target structure \strB].
Given an initial state $A=(x_0,y_0)$, a
target state $B=(x_1,y_1)$, and the 2D grid of positions $(x,y)$
and probabilities $p(x,y)$ as computed by \ffttwo and
illustrated in
Figure \ref{fig:checkerboardGridPointsSplicedLeader},
{\tt RNAmfpt} computes the {\em mean first passage time}, taken
over all paths from $(x_0,y_0)$ to  $(x_1,y_1)$.  Alternatively,
{\tt RNAmfpt} can take as input the probability transition matrix $P$ for
an arbitary finite Markov chain $\mathbb{M}$, and subsequently
determines the {\em mean first passage time} by computing the inverse
$(I - P^{-}_{x_{\infty}})^{-1} = \displaystyle\sum_{k=0}^{\infty}
(P^{-}_{x_{\infty}})^k$.


As in the case of {\tt RNAmfpt}, {\tt RNAeq}
computes the equilibrium time for the coarse-grained Markov process,
consisting of the macrostates $M(x,y)$ of all secondary structures,
whose base pair distance is $x$ [resp. $y$] from initial structure \strA
[resp. target structure \strB].
Given an initial state $A=(x_0,y_0)$, a
target state $B=(x_1,y_1)$, and the 2D grid of positions $(x,y)$
and probabilities $p(x,y)$ as returned by \ffttwo and
illustrated in
Figure \ref{fig:checkerboardGridPointsSplicedLeader},
{\tt RNAeq} computes the population occupancy frequencies
$P(t) = (P_1(t),\ldots,P_n(t))$ by using spectral decomposition, and
then determines the {\em equilibrium time}.
Alternatively, given an input RNA sequence {\tt RNAeq}
calls \rnasub \cite{wuchtyFontanaHofackerSchuster} to generate
structures within a user-specified energy bound (or to sample a user-specified
number of structures from the low energy ensemble). {\tt RNAeq} subsequently
computes the rate matrix and computes population occupancy frequencies
and equilibrium time. {\tt RNAeq} can additionally
take as input the rate matrix $R$ for
an arbitary finite Markov process $\mathbb{M}$, and then
compute population occupancy frequencies
and the {\em equilibrium time}.

Given an RNA sequence,
the C-drivers \fftmfpt and \ffteq respectively compute
the mean first passage time (MFPT) and
equilibrium time (ET) for the coarse-grained Markov chain on the
macrostates $M(x,y)$ of all secondary structures whose base pair
distance is $x$ [resp. $y$] from initial structure \strA [resp. target
structure \strB]. This is done by calling \ffttwo, in order to
determine the 2D probabilities, from which the transition probability
matrix or respectively rate matrix are determined, and then calling
{\tt RNAmfpt} or respectively {\tt RNAeq}.

\begin{figure*} [!ht]
\begin{center}
%\includegraphics[width=.45\textwidth]{FIGURES/checkerboardGridPointsSplicedLeader}
\includegraphics[width=.45\textwidth]{figure7a.eps}
%\includegraphics[width=.45\textwidth]{FIGURES/PeterFFTbor2DLcollosoma}
\includegraphics[width=.45\textwidth]{figure7b.eps}
\caption{\small {\em (Left)}
As first pointed out in \cite{hofacker:RNAbor2D}, the
set of grid points $(x,y)$, for which a secondary structure $S$ has
base pair distance $x$ from reference structure \strA and base pair
distance $y$ from reference structure \strB, forms a kind of
checkerboard image. For the 56 nt spliced leader RNA from {\em L.
collosoma} \cite{lecuyerCrothers}, with sequence AACUAAAACA AUUUUUGAAG
AACAGUUUCU GUACUUCAUU GGUAUGUAGA GACUUC, and reference structure \strA
given by {\tt
..((...((((((..(((((.((((...)))).)))))..))).)))..)).....}, and
reference structure \strB given by {\tt
.......................((((((((((((.....)))))..)))))))..}, we obtain
the possible grid points depicted in this figure.
The $x$-axis (resp. $y$-axis)
represents base pair distance between initial structure \strA (resp. target
structure \strB).
Consideration of the triangle inequality and parity condition
(see text) ensure that partition function values $Z(x,y)$ and Boltzmann
probabilities $p(x,y) = \frac{Z(x,y)}{Z}$ are zero for all
{\em non}-grid points $(x,y)$.
{\em (Right)}
2D projection of energy landscape for Spliced Leader (SL) RNA
from {\em Leptomonas collosoma}, in which
the $z$-axis represents the ensemble free energy $-RT \log Z(x,y)$,
where $Z(x,y)$ is computed in \ffttwo by $Z_{x,y}= p(x,y) \cdot Z$.
Low energy positions $(x,y)$ correspond to high Boltzmann probability positions.
Image taken from \cite{Senter.jmb14}.
}
\end{center}
\label{fig:checkerboardGridPointsSplicedLeader}
\end{figure*}

Using functions from Gnu Scientific Library (GSL)
\url{http://www.gnu.org/software/gsl/}, mean first passage time
(MFPT) is computed by matrix inversion
$(I - P^{-}_{x_{\infty}})^{-1}$ from the transition probability matrix
$P$, while equilibrium time (ET) is computed by spectral
decomposition of the rate matrix. \hermes makes available a
variety of novel RNA folding kinetics methods: \rnamfpt (via {\tt RNAmpft}), {\tt
Equilibrium} (via {\tt RNAeq}), \fftmfpt, and \ffteq.
Using a collection of 1,000 randomly generated 20-mers having at most
2,500 secondary structures, using \hermes the gold standard
mean first passage time and platinum standard equilibrium time were
computed for each sequence.
Pearson correlation was then computed
between the folding times of all the methods of \hermes, along
with the current state-of-the-art kinetics programs \kinfold
\cite{flamm}, {\tt BarrierBasins} \cite{wolfingerStadler:kinetics},
{\tt Kinwalker} \cite{Geis.jmb08}.

\kinfold is an implementation
of the Gillespie algorithm \cite{gillespieStochasticSimulation1},
which essentially samples from an exponential distribution of folding
times, hence \kinfold mean and standard deviation are
approximately equal -- see
Figure \ref{fig:meanStdevKinfoldRuns1000sequences}. Elementary
considerations from statistics indicate that for our benchmarking set
of 20-mers, the minimum sample size $n = \left( \frac{z_{\alpha/2}
\cdot \sigma}{E} \right)^2$ ranges from 937,712 to 23,289,310 for us
to have a confidence level of 95\% that the average of $n$ {\tt
Kinfold} runs differs from the real folding time by at most 100 steps.
Since this is the case for tiny sequences of 20 nt, it follows that
only enormous computation time can provide reliable folding times for
larger sequences when using \kinfold.

Since the state-of-the-art software \kinfold, which implements
Gillespie's algorithm, cannot return statistically significant folding
times unless significant computation time is allotted, it is natural
to turn to {\em coarse-grained} models, as done by
Wolfinger et al. \cite{wolfingerStadler:kinetics} and by
Tang et al. \cite{Tang.jmb08}.  The software of Tang et al. appears not
to be available. Concerning the
method of Wolfinger et al. (called {\tt BarrierBasins} in our benchmarking),
there is now a web server available, which runs \rnasub
\cite{Wuchty.b99} to generate all secondary structures within a user-specified
energy range, then runs {\tt barriers} \cite{flammHofacker} to generate
basins of attraction around a user-specified number of locally optimal
structures, and then runs {\tt treekin} on the output of {\tt barriers}.
The program {\tt treekin} performs some of the same operations as \hermes,
by computing population occupation frequencies by spectral decomposition.
Nevertheless it would require a user to write scripts and perform
several manual steps, in order to determine the equilibrium time for
an input RNA sequence, with respect to the macrostate Markov process
of \cite{wolfingerStadler:kinetics}. In addition, because {\tt barriers} computes
basins of attraction by utilizing the output of \rnasub, estimating kinetics for the
refolding of an RNA molecule from the empty structure requires exhaustive enumeration
of all suboptimal structures having non-positive free energy.
Figure \ref{fig:populationOccupancySplicedLeader} depicts some population
occupancy curves for the 56 nt spliced leader RNA from
{\em L. collorosoma}, a known conformational switch experimentally
investigated in \cite{lecuyerCrothers}, a determined by
\hermes (left panel) and {\tt barriers} with {\tt treekin}
(right panel).

In contrast, \hermes is a fully automated software suite which
computes mean first passage time and equilibrium time with respect to
the coarse-grained model that consists of macrostates $M(x,y)$ of all
structures having base pair distance $x$ to initial structure \strA and
$y$ to target structure \strB. Additionally, \hermes provides
a number of other functionalities useful for RNA kinetics. Since
\hermes includes \ffttwo, which uses the fast Fourier transform
to compute probabilities $p(x,y)$ of macrostates $M(x,y)$, the resulting
coarse-grained kinetics computations are extremely fast. \hermes
can provide approximate refolding kinetics for RNA sequences that are
larger than those which other software can handle. As well, the
correlation values shown in Table~\ref{table:correlation}
demonstrate that \hermes can be used for fast, approximate and relatively
accurate RNA kinetics in a pipeline to select promising candidates in
the design of RNA sequences.


\begin{figure*}
\centering
%\includegraphics[width=0.45 \textwidth]{FIGURES/giegerich_spliced_leader_37c_kinetics}
\includegraphics[width=0.45 \textwidth]{figure8a.eps}
%\includegraphics[width=0.45 \textwidth]{FIGURES/splicedLeaderTreekinNoDangle}
\includegraphics[width=0.45 \textwidth]{figure8b.eps}
\caption{
{\em (Left)}
Population occupancy curves computed with \ffteq for the
56 nt conformational switch {\em L. collosoma} spliced leader RNA,
with sequence AACUAAAACA AUUUUUGAAG AACAGUUUCU GUACUUCAUU GGUAUGUAGA
GACUUC. The dot bracket format for the MFE structure, as computed by
version 2.1.7 of {\tt RNAfold -d0}, is {\tt
.......................((((((((((((.....)))))..)))))))..} with free
energy $-8.6$ kcal/mol, while that of the the alternate suboptimal
structure is {\tt
..((...((((((..(((((.((((...)))).)))))..))).)))..)).....} with free
energy $-7.5$ kcal/mol. In the case of the MFE structure, the
equilibrium occupancy $P(t_{\infty})$, which \hermes approximates
as $0.17893806$ should equal the Boltzmann probability $0.17909227$,
since the MFE structure is the only structure at distance $x_0$ [resp.
$y_0$] from the reference structures \strA (empty
structure) [resp. \strB (MFE structure)]. As well, if there are few
other low energy structures at the same base pair distance $x_1$
[resp. $y_1$] from \strA [resp. \strB] as that of the alternate suboptimal
structure, then we expect that the occupancy probability $0.03003426$
for the $(x_1,y_1)$ be approximately the Boltzmann probability
$0.03005854$ of the alternate structure.
{\em (Right)} Population occupancy curves computed
for {\em L. collosoma} spliced leader RNA, computed using
RNA Vienna Package with the following commands:
{\tt RNAsubopt -d0 -s -e 17  > foo.sub},
{\tt barriers -M noShift --max=1000 --rates --minh 0.1 < foo.sub  > foo.bar},
{\tt treekin --method I  --p0 940=1 --ratesfile rates.out  < foo.bar > out.bar},
where the value 940 indicates the empty structure appearing as the 940\textsuperscript{th}
structure in the sorted file {\tt foo.bar} of 1,000 locally optimal structures.
%{\em (Right)} Population occupancy curves computed
%for {\em L. collosoma} spliced leader RNA, computed using
%RNA Vienna Package {\tt barriers} server
%{\tt http://rna.tbi.univie.ac.at/cgi-bin/barriers.cgi}, where 1000
%locally optimal structures were requested. After manually determining
%that the empty structure is in position 206 of the output from
%\rnasub, 206 is entered as the initial structure whose population
%density is set to 100\%, and the figure depicted here is generated.
The red curve corresponds to the {\em basin of attraction} around the
minimum free energy structure. At equilibrium, the Boltzmann probability
of this basin is between 0.3 and 0.4, a value substantially greater than
the Boltzmann probability of the minimum free energy structure itself,
which is $\approx 0.1791$ as shown in the left panel.}
\label{fig:populationOccupancySplicedLeader}
\end{figure*}

Finally, we mention an application of \hermes to recent data from
Dotu et al. \cite{syntheticHammerheads}, which describes the
first purely computational design of functional synthetic hammerhead type III
ribozymes, experimentally shown to cleave {\em in vitro}. For the tiny
data set of 10 synthetic ribozymes, we ran \fftmfpt to determine the
log(MFPT) of folding from the empty structure to the target consensus
structure of Peach Mosaic Latent Viroid (PMLVd),
shown in Figure \ref{fig:PLMVd}. We determined a
Pearson correlation of $0.4796$ [resp. $0.5322$]  between the logarithm of the
{\em mean first passage time} [resp. {\em equilibrium time}], as computed by
\fftmfpt [resp. {\tt RNAeq}] and the cleavage rate, using the
experimentally measured cleavage rate of 10 synthetic hammerhead ribozymes
(data from \cite{syntheticHammerheads}).
These correlation values seem surprising, since we had expected that
rapidly folding hammerheads (low values of MFPT and ET) might be
associated with fast cleavage. The correlation values suggest instead
that perhaps the longer a hammerhead requires to fold into its functional
secondary structure, shown in Figure \ref{fig:PLMVd}, the more rapid its
cleavage. If this observation can be experimentally validated on a large set
of hammerheads, then the insight could be potentially important for
RNA design, an area of synthetic biology.

\section{Acknowledgements}

We would like to thank Ivan Dotu, for the suggestion of checking the
folding kinetics of the synthetic hammerheads in \cite{syntheticHammerheads}.
Some of the work described in this paper was done while PC visited
Niles Pierce at the California Institute of Technology, Knut Reinert
at the Free University of Berlin and Martin Vingron at the Max Planck
Institute for Molecular Genetics. Warm thanks are due to all three
persons. The research was funded by National Science Foundation grant
DBI-1262439, the Guggenheim Foundation and Deutscher Akademischer
Austauschdienst. Any opinions, findings, and conclusions or
recommendations expressed in this material are those of the authors
and do not necessarily reflect the views of the National Science
Foundation.



\clearpage
%\hfill\break\newpage

\section*{Appendix}
\label{section:appendix}

\subsection*{Kinetics methods in benchmarking}

In this section, we describe the kinetics algorithms used in our
comparative study on a benchmarking set of 1000 small RNAs. The
kinetics algorithms we consider are the following: \rnamfpt, {\tt
Equilibrium}, \kinfold \cite{flamm}, \fftmfpt, \ffteq,
\rnatwofold, {\tt FFTbor1D}, {\tt BarrierBasins}
\cite{wolfingerStadler:kinetics}, {\tt Kinwalker} \cite{Geis.jmb08}.
Our software suite \hermes introduces the new methods \rnamfpt,
\rnaeq, \fftmfpt, \ffteq (along with {\tt
RNA2Dfold} and {\tt FFTbor1D} through the utility programs {\tt
RNAmfpt} and {\tt RateEq}). Although all these methods allow one to
compute folding kinetics from an arbitrary initial structure to an
arbitrary final structure, in the benchmarking comparison, we consider
only folding from the empty structure to the MFE structure.
\begin{description}
\item[MFPT]: Given RNA sequence \seq and secondary structure $x$,
let $N(x)$ denote the set of neighboring secondary structures of
\seq, whose base pair distance with $x$ is one. Define the Markov
chain $\mathcal{M}(\seq)=(\str,P)$, where $\str$ denotes the
set of all secondary structures of \seq, $\pSt(x)$ is the stationary
probability of structure $x$, defined by $\pSt(x)=\exp(-E(x)/RT)/Z$,
$Z=\sum_{x} \exp(-E(x)/RT)$, and the transition probability matrix $P
= ( p_{x,y} )$ is defined either with or without the Hastings
modification as follows.

{\em With} the Hastings modification,
\begin{align}
\label{eq:MFPTwithHastings} p_{x,y} = \left\{
\begin{array}{ll}
\frac{1}{|N(x)|} \cdot \min(1, \frac{\pSt(y)}{\pSt(x)} \cdot
\frac{N(x)}{N(y)}) &\mbox{if $y \in N(x)$}\\
0 &\mbox{if $x \ne y, y \not\in N(x)$}\\
1 - \sum_{\substack{z \in N(x)}} p_{x,z} &\mbox{if $x=y$}\\
\end{array}
\right.
\end{align}

{\em Without} the Hastings modification,
\begin{align}
\label{eq:MFPTwithoutHastings} p_{x,y} = \left\{
\begin{array}{ll}
\frac{1}{|N(x)|} \cdot \min(1, \frac{\pSt(y)}{\pSt(x)} ) &\mbox{if $y
\in N(x)$}\\
0 &\mbox{if $x \ne y, y \not\in N(x)$}\\
1 - \sum_{\substack{z \in N(z)}} p_{x,z} &\mbox{if $x=y$}\\
\end{array}
\right.
\end{align}
The exact value of mean first passage time (MFPT) can be computed as
follows. Let $x_0$ [resp. $x_{\infty}$] denote the empty structure
[resp. MFE structure] for sequence \seq (here we have implicitly
identified integer indices with secondary structures). Let
$\widetilde{P}_{x_{\infty}}$ be the matrix obtained from $P$ by
removal of the row and column with index $x_{\infty}$, and $I$ denote
the $(n-1)\times(n-1)$ identity matrix, where $n=|\str|$ is the
number of secondary structures of \seq. Let $e$ denote the vector of
size $n-1$, each of whose coordinates is $1$. It is well-known
\cite{meyerMFPT} that for each $k\ne x_{\infty}$, the $k$th coordinate
of the vector $(I - \widetilde{P}_{x_{\infty}})^{-1} \cdot {\bf e}$ is
exactly equal to the mean first passage time from the structure with
index $k$ to the target structure $x_{\infty}$. In particular, the
MFPT from the empty structure to the MFE structure is computable by
applying matrix inversion from GSL. Since this computation of the
mean first passage time is mathematically exact, we consider that MFPT
to be the {\em gold standard} value for RNA folding kinetics.

\item[Equilibrium]: Define the continuous Markov process
$\mathcal{M}=(\str,R)$, where $R$ is the {\em rate matrix} defined
by
\begin{align}
\label{eq:rateMatrix} k_{x,y} = \left\{
\begin{array}{ll}

%\frac{1}{|N(x)|} \cdot \min(1, \frac{\pSt(y)}{\pSt(x)} )
\min(1, \frac{\pSt(y)}{\pSt(x)} ) &\mbox{if $y \in N(x)$}\\
0 &\mbox{if $x \ne y, y \not\in N(x)$}\\
- \sum_{\substack{z \in N(z)}} p_{x,z} &\mbox{if $x=y$}\\
\end{array}
\right.
\end{align}
Clearly the rate matrix satisfies {\em detailed balance}; i.e. $\pSt(x)
\cdot k_{x,y} = \pSt(y) \cdot k_{y,x}$ for all distinct $x,y \in
\str$. In fact, the rate matrix for Markov processes is usually
defined as above, precisely to ensure detailed balance, which then
implies that all eigenvalues of the rate matrix are real, thus
permitting explicit solution of the population occupancy frequency for
all states.\footnote{We additionally considered a Hastings correction
for the rate matrix, where $k_{x,y} = \min(1, \frac{\pSt(y)}{\pSt(x)}
\cdot \frac{N(x)}{N(y)})$. The correlation in
Table~\ref{table:correlation} for equilibrium time computed from this
modified rate matrix is somewhat better than without the Hastings
correction. However, the Hastings correction is never used for rate
matrices, hence we only consider the usual definition of rate matrix
given in equation (\ref{eq:rateMatrix}).}

\item[Kinfold]: Version 1.3 of \kinfold \cite{Flamm:00a}, from
the Vienna RNA Package (version 2.1.7) was run with the following
flags:
\begin{quote}
{\tt Kinfold --dangle 0 --met --noShift --logML --num=10000
--time=100000000}

%< infile.txt >\& /dev/null
\end{quote}
\kinfold 1.3 uses the Turner2004 energy model, and the flags in
the command line indicate that dangles are not considered, that the
Metropolis rule is employed (rather than the Kawasaki rule), that
moves consist only of the addition or removal of a single base pair,
that the number of repeated Monte Carlo runs per sequence is $10^4$,
and that each run has an upper bound of $10^8$ steps. For the RNA
sequences of 20 nt used in our benchmarking set, with these flags,
every run for every sequence properly terminated in the target minimum
free energy structure. For each sequence, the mean (and standard
deviation) of $10^4$ runs was determined; this mean is deemed the mean
first passage time, as estimated by \kinfold. Implemented in the
Vienna RNA Package 2.1.7, \kinfold is what may be considered the
{\em silver standard} value for folding kinetics, since exact
determination of mean first passage time, using matrix inversion is
impossible except for sequences not having more than several thousand
secondary structures. However, due to the stochastic nature of the
Gillespie algorithm implemented by \kinfold, very many runs with
very long run times are required to obtain a reasonable approximation
of the MFPT.

\item[FFTmfpt]: Given an RNA sequence \seq, the algorithm {\tt
FFTbor2D} \cite{Senter.jmb14} computes the probabilities $p(x,y) =
\frac{Z(x,y)}{Z}$ of secondary structures of \seq to have base pair
distance $x$ [resp. $y$] to secondary structure \strA [resp. \strB], where
\strA is taken to be the empty structure and \strB is the MFE structure.
Let $d_{BP}(S,T)$ denote the base pair distance between structures
$S,T$. Then $Z(x,y) = \sum_S \exp(-E(S)/RT)$, where the sum is over
structures $S$, such that $d_{BP}(A,S)=x$, and $d_{BP}(B,S)=y$; as
well the partition function $Z =\sum_S \exp(-E(S)/RT)$, where the sum
is over all seconddary structures $S$ of the sequence \seq.

Let $d_0= \dBP{\strA}{\strB}$, the base pair distance between initial structure
\strA and target structure \strB.  Let $n$ denote the length of sequence \seq.
Define the Markov chain $\mathcal{M}_1(\seq) = (Q_1,P_1)$, where
\begin{align}
\label{eq:defQ1} Q_1 &=& \{ (x,y) : 0 \leq x,y \leq n, (x+y \bmod 2)
= (d_0 \bmod 2), \\ % Parity condition
& & (d_0 \leq x+y), (x \leq d_0 + y), (y \leq d_0 + x). \nonumber %
Triangle inequality
\end{align}
For reference, we say that the {\em parity condition} holds for
$(x,y)$ if
\begin{align}
\label{eq:parityCondition} (x+y \bmod 2) = (d_0 \bmod 2).
\end{align}
We say that the {\em triangle inequality} holds for $(x,y)$ if
\begin{align}
\label{eq:triangleInequality} (d_0 \leq x+y), (x \leq d_0 + y), (y
\leq d_0 + x)
\end{align}
Figure \ref{fig:checkerboardGridPointsSplicedLeader} displays the
space of possible grid points $(x,y)$ for spliced leader RNA from {\em
L. collosoma} with two reference structures.

Since we allow transitions between secondary structures that differ by
exactly one base pair, Markov chain transitions are allowed to occur
only between states $(x,y),(u,v) \in Q_1$, such that $u=x\pm 1$, $v =
y \pm 1$. However, we have found that for some RNA sequences, there is
no non-zero probability path from $(0,d_0)$ to $(d_0,0)$
(corresponding to a folding pathway from structure \strA to
\strB).\footnote{Since \ffttwo computes probabilities $p(x,y)$ by
polynomial interpolation using the fast Fourier transform, any
probability less than $10^{-8}$ is set to $0$. Also with {\tt
RNA2Dfold}, it may arise that there is no non-zero probability path
from structure $(0,d_0)$ to $(d_0,0)$.}

To address this situation, we proceed as follows. Let $\epsilon =
10^{-8}$ and for all $(x,y) \in Q_1$, modify probabilities $p(x,y)$ by
\begin{align}
\label{eq:modifyNormalizeCellProb} p(x,y) &=& \frac{p(x,y) +
\epsilon/|Q_1|}{1+\epsilon }.
\end{align}
This operation corresponds to adding the negligeable value of
$\epsilon=10^{-8}$ to the total probability, thus ensuring that there
are paths of non-zero probability between any two states. After this
$\epsilon$-modification and renormalization, {\em when using the
Hastings modification}, the transition probabilities $P( (u,v) | (x,y)
)$ are given by
\begin{align}
\label{eq:transitionProbabilityFFTbor2DwithHastings} P( (u,v) |(x,y)
) = \left\{
\begin{array}{ll}
\frac{1}{|N(x,y)|} \cdot \min(1, \frac{p(u,v)}{p(x,y)} \cdot
\frac{N(x,y)}{N(u,v)}) &\mbox{if $(u,v) \in N(x,y)$}\\
0 &\mbox{if $(x,y) \ne (u,v) \mbox{ and } (u,v) \not\in N(x,y)$}\\
1 - \sum_{\substack{(u,v) \in N(x,y)}} P( (u,v) | (x,y) ) &\mbox{if
$(x,y) =(u,v)$}\\
\end{array}
\right.
\end{align}
Here the set $N(x,y)$ of adjacent neighbors is defined by $N(x,y) = \{
(u,v) \in Q_1 : u = x \pm 1, v = y \pm 1 \}$, and the stationary
probability $p(x,y)$ is obtained from \ffttwo.

{\em Without the Hastings modification}, the transition probabilities
$P( (u,v) | (x,y) )$ are instead given by
\begin{align}
\label{eq:transitionProbabilityFFTbor2DwithoutHastings} P( (u,v)
|(x,y) ) = \left\{
\begin{array}{ll}
\frac{1}{|N(x,y)|} \cdot \min(1, \frac{p(u,v)}{p(x,y)}) &\mbox{if
$(u,v) \ne (x,y)$}\\
1 - \sum_{\substack{(u,v) \ne (x,y)}} P( (u,v) | (x,y) ) &\mbox{if
$(x,y) =(u,v)$}\\
\end{array}
\right.
\end{align}

\item[FFTeq]: This method consists of computing the equilibrium time
from the master equation for the coarse-grain Markov process
$\mathcal{M}=(Q_1,R)$, where $Q_1$ is defined in equation
(\ref{eq:defQ1}), and the rate matrix $R = ( k( (x,y), (u,v) ) )$ is
defined by
\begin{align}
\label{eq:rateProbabilityFFTbor2DwithoutHastings} k( (x,y), (u,v) ) =
\left\{
\begin{array}{ll}
\min(1, \frac{p(u,v)}{p(x,y)}) &\mbox{if $(u,v) \ne (x,y)$}\\
1 - \sum_{\substack{(u,v) \ne (x,y)}} P( (u,v) | (x,y) ) &\mbox{if
$(x,y) =(u,v)$}\\
\end{array}
\right.
\end{align}
Equilibrium time is then computed for this Markov process.

\item[RNA2Dfold]: The program \rnatwofold of Vienna RNA Package
2.1.7 was run to compute the probabilities $p(x,y) = \frac{Z(x,y)}{Z}$
of secondary structures of \seq to have base pair distance $x$
[resp. $y$] to secondary structure \strA [resp. \strB], where \strA is the
empty structure and \strB is the minimum free energy structure. As in
the case for \ffttwo,
Figure \ref{fig:checkerboardGridPointsSplicedLeader} displays the
checkerboard image of possible grid points $(x,y)$ for spliced leader
RNA from {\em L. collosoma} with two reference structures.

The probabilities $p(x,y)$ were then $\epsilon$-modified and
normalized, as in the description above for \ffteq, and the mean
first passage time for the Markov chain $\mathcal{M}_2(\seq) =
(Q_1,P_2)$ is computed, where $P_2$ is the transition probability
matrix, as in \ffttwo, with the exception that $p(x,y)$ is
initially obtained from \rnatwofold, rather than \ffttwo.

When working with {\tt
RNA2Dfold}, it is more accurate to compute transition probabilities of
the form $$ P( (u,v) | (x,y) ) = \frac{1}{N(x,y)} \cdot \min(1,
\exp(-\frac{E(u,v)-E(x,y)}{RT})) $$ instead of the mathematically
equivalent form $$ P( (u,v) | (x,y) ) = \frac{1}{N(x,y)} \cdot \min(1,
\frac{p(u,v)}{p(x,y)}) $$ for reasons due solely to numerical
precision. However, since it is necessary to perform
$\epsilon$-modification and renormalization of transition
probabilities,\footnote{Otherwise, as earlier explained, there may be
no non-zero probability path from $(0,d_0)$ to $(d_0,0)$.} one is led
to modify the free energy in the following manner.

Letting $G$ denote ensemble free energy $-RT \ln Z$, as computed by
{\tt RNAfold} and \rnatwofold,
\begin{align*}
p &=& \exp(-E/RT)/\exp(-G/RT) \\
&=& \exp(-[ E-G] /RT)
\end{align*}
hence it follows that $-RT \ln p = E-G$ and $E = G-RT \ln p$.

By adding $\epsilon=10^{-8}/|Q_1|$ to every probability in every state
$(x,y) \in Q_1$, the $\epsilon$-modified and renormalized probability
corresponding to $p$ is $\frac{p+\epsilon/|Q_1|}{1+ \epsilon}$ and so
the corresponding new energy term is
\begin{align}
\label{eq:modifiedEnergyTerm} Enew &=& G- RT \ln( \frac{p+\epsilon}{1
+ \epsilon \cdot |Q_1|} ).
\end{align}
It follows that we are forced to use probabilities in any case, and so
we encounter the above-mentioned loss of accuracy.

\item[BarrierBasins]: Using the {\tt --rate} flag for the program {\tt
barriers} \cite{flammHofacker}, Wolfinger et al.
\cite{wolfingerStadler:kinetics} computed the rate matrix for a Markov
process, whose states are the basins of attraction around locally
optimal secondary structures. Subsequently, using {\tt Treekin}, the
authors displayed population occupancy curves for certain macrostates
(basins of attraction) containing specific structures of interest.
Since this method was not given a name by the authors, we let {\tt
BarrierBasins} denote the equilibrium time for the Markov process just
described.

\item[Kinwalker]: {\tt Kinwalker} was used with default settings.
However, since {\tt Kinwalker} models co-transcriptional folding of
RNA, its correlation was extremely poor with all other methods, which
do model refolding, rather than co-transcriptional folding (data not
shown). For this reason, we do not display the results with {\tt
Kinwalker}.
\end{description}
