%!TEX root = ../main.tex

\newcommand{\ob}{\,\mbox{\bf\texttt{[}}\,}
\newcommand{\cb}{\,\mbox{\bf\texttt{]}}\,}
\newcommand{\unafold}{\mbox{\tt UNAFold}\xspace}
\newcommand{\mfold}{\mbox{\tt mfold}\xspace}
\newcommand{\rnahairpinml}{\mbox{\tt RNAmloopHP}\xspace}
\newcommand{\rnamloop}{\mbox{\tt RNAmloop}\xspace}
\newcommand{\rnamlnumber}{\mbox{\tt RNAmloopNum}\xspace}
\newcommand{\rnamlorder}{\mbox{\tt RNAmloopOrder}\xspace}
\newcommand{\rnahairpin}{\mbox{\tt RNAhairpin}\xspace}
\newcommand{\rnaz}{\mbox{\tt RNAz}\xspace}

\chapter{RNAparametric} % Main chapter title

\label{RNAparametric} % Change X to a consecutive number; for referencing this chapter elsewhere, use \ref{ChapterX}

\lhead{Chapter X. \emph{RNAparametric}} % Change X to a consecutive number; this is for the header on each page - perhaps a shortened title

We describe four novel algorithms,
{\rnahairpin}, {\rnamlnumber}, {\rnamlorder}, {\rnahairpinml},
which compute
the Boltzmann partition function for global {\em structural constraints} --
respectively for the number of hairpins, the number of multiloops,
maximum order (or depth) of multiloops, and the {\em simultaneous}
number of hairpins and of multiloops.  Given an RNA sequence of length $n$
and a user-specified integer $0 \leq K \leq n$,
{\rnahairpin} [resp. {\rnamlnumber} resp. {\rnamlorder}] computes
the partition functions $Z(k)$ for each $0 \leq k \leq K$ in
time $O(K^2 n^3)$ and space $O(K n^2)$. The program {\rnahairpinml}
simultaneously computes the partition functions $Z(m,h)$ for each
possible number $m$ of multiloops and $h$ of hairpins, for
$0 \leq m \leq M$ and $0 \leq h \leq H$, with run time
$O(M^2 H^2 n^3)$ and space $O(MH n^2)$. In addition, programs
{\rnahairpin} [resp. {\rnahairpinml}] sample from the low energy
ensemble of structures having exactly $h$ hairpins [resp. $m$ multiloops
and $h$ hairpins], for user-specified values of $h,m$.
Moreover, by using the fast Fourier transform (FFT),
{\rnahairpin} and {\rnamlnumber} have been improved to run in
time $O(n^4)$ and space $O(n^2)$, although this improvement is not possible
for {\rnamlorder}.

We present two applications of the novel algorithms. First, we show that
for many Rfam families of RNA, structures sampled from {\rnahairpinml}
are more accurate than the minimum free energy structure; for instance,
sensitivity improves by almost 24\% for transfer RNA, while for
certain ribozyme families, there is an improvement of around 5\%.
Second, we show that the probabilities $p(k) = Z(k)/Z$
of forming $k$ hairpins [resp. multiloops] provide discriminating
novel features for a support vector machine or relevance vector machine
binary classifier for Rfam families of RNA.
Our data suggests that multiloop order does
not provide any significant discriminatory power over that of hairpin
and multiloop number, and since these probabilities
can be efficiently computed using the FFT, hairpin and multiloop formation
probabilities could be added to other features in existent
noncoding RNA gene finders.
Our programs, written in C/C++, are publicly available at
{\tt http://bioinformatics.bc.edu/clotelab/RNAparametric}.

\section{Introduction}

It has recently emerged that RNA plays surprising and previously unsuspected
roles in many biological processes, including retranslation of the genetic code
(selenocysteine insertion \citep{boeckForschhammer}, ribosomal frameshift
\citep{denise:frameshift}), gene regulation by allostery
(riboswitches) \citep{mandalBoeseBarrickWinklerBreaker}
and by the RISC complex (microRNAs) \citep{burgeBartel:miRNAscience},
regulation of heat shock protein expression by
temperature sensitive conformational switches \citep{ROSEswitch,tuckerBreaker:review},
pointwise editing of messenger RNA (guide RNA)
\citep{haeseler:Cryptogenes}, chemical
modification of specific nucleotides in the ribosome (small nucleolar RNAs)
\citep{loweEddy:snoRNAinArchaea},
regulation of alternative splicing \citep{Cheah.n07},
regulation of chromatin remodeling (small interfering RNAs)
\citep{Cam.c09} etc. RNA can encode
genomic information (e.g. HIV and hepatitis C) and with no assistance
from proteins can catalyze reactions such as peptidytransferase
(at ribosomal P-site) \citep{weinger:peptidyltransferase}
and cleavage
of RNA phosphodiester bonds at specific sites (group I introns)
\citep{intron:reviewCech}.
Since RNA plays various unsuspected regulatory and catalytic roles, and
since it is known from the {\sc encode} consortium report
\citep{encodeConsortium} that
the human genome is ``pervasively transcribed'',
most of whose RNA transcripts have completely unknown structure
and function, it is clear that the development of
noncoding RNA gene finders
remains an important open problem, despite significant
advances with tools such as {\rnaz} \citep{Gruber.nar07},
{\foldalign} \citep{Havgaard.nar05}, etc. The current paper
provides novel computable features that could prove useful
in enriching features sets for noncoding RNA gene finders.

In this paper, we present four novel
thermodynamics-based algorithms to compute parametric structural aspects of the
Boltzmann ensemble of low energy structures for a given RNA sequence.
Specifically, given an RNA sequence $\seq =a_1,\ldots,a_n$ and
optionally an upper bound $K$, {\rnahairpin} computes, for each value of
parameter $k$, for $0 \leq k \leq K \leq n$, the Boltzmann partition
function $Z^h(k)$ and
Boltzmann probability $p_h(k) = Z^h(k)/Z$ of all structures  of \seq having
exactly $k$ hairpins. Here $Z^h(k)$ designates the sum of Boltzmann factors
\boltzf{\str}, taken over all
secondary structures $S$ of \seq that have exactly $k$ hairpins; the
partition function $Z$ denotes the sum of all Boltzmann factors, where the
sum is taken over all secondary structures of \seq. Analogously,
{\rnamlnumber} computes, for each value of $0 \leq k \leq K \leq n$,
the Boltzmann partition function $Z^m(k)$ and probability
$p_m(k) = Z^m(k)/Z$ of all structures, that have exactly $k$
multiloops.
The program {\rnamlorder} computes the
Boltzmann partition function $Z^d(k)$ and probability
$p_d(k) = Z^d(k)/Z$ of all structures, having multiloops of order
$k$ but none of larger order, where {\em multiloop order} is
is the maximum depth of multiloop nesting.
(See Section~\ref{section:multiloopOrderPartitionFunction} for
formal definition.)
Finally, {\rnahairpinml} simultaneously computes the
Boltzmann partition function $Z(m,h)$ and probability
$p(m,h) = Z(m,h)/Z$ of all structures, having $m$ multiloops and
$h$ hairpins. Since our preliminary work showed that structures sampled
from {\rnahairpin} could improve structure prediction for certain Rfam
families, the program {\rnahairpinml} also supports sampling.

Other groups have shown an interest in global structural features
of RNA families. Here we cite four specific examples.
First, Hofacker et al.  \citep{hofacker99a} determined the asymptotic
number of hairpins, multiloops, and other structural features for
random RNA, using the homopolymer model first introduced by
Stein and Waterman \citep{steinWaterman}.
Second, Giegerich et al. \citep{giegerich:shapesBioinf} developed the
program {\tt RNAshapes}, which computes the minimum free energy structure
for various {\em shapes}; for instance, the cloverleaf shape of tRNA is
$\ob \ob \cb \ob \cb \ob \cb \cb$.
Third, the {\sc rna strand} database \citep{Andronescu.bb08}
consists of 4666 RNA secondary
structures collected from other databases, including
the Nucleic Acid Database \citep{berman03}, the Protein Data Bank
\citep{Berman.acdb02}, Sprinzl's tRNA database
\citep{sprinzl}, Gutell's database \citep{gutell:ribosomalRNA}, etc.
%(see \citep{Andronescu.bb08} for
%citation of original data sources).
{\sc rna strand} provides frequency analysis for sequence length,
number of stems, hairpin loops, bulges, internal loops, multiloops,
pseudoknots, etc., which can be generated for a
class of RNAs selected by the user from a set of predefined RNA classes,
such as 16S ribosomal RNA, 23S ribosomal RNA, 5S ribosomal RNA,
7SK RNA, ciliate telomerase RNA, cis-regulatory element, group I intron,
etc.
Fourth, Kazan et al. \citep{Kazan.pcb10} presented a machine learning
algorithm {\tt RNAcontext}, which used sequence profiles (sequence LOGOS)
as well as local secondary structure profiles (structure LOGOS) to
predict RNA nucleotides that bind to a particular riboprotein. Here,
a structural profile computes the frequency, for each $k$ in the putative
binding region,  that nucleotide position $k$ is located in a hairpin,
bulge/internal loop, multiloop, or base pair (frequencies are obtained
by counting instances from {\tt Sfold} samples).

Additionally, a number of groups have developed
algorithms to compute the minimum free energy structure and
partition function by integrating base pairing constraints.
These constraints may be {\em hard}, in the sense that
certain nucleotides are required to pair with certain other
nucleotides, while other nucleotides may be required to be unpaired.
Alternatively, constraints may be {\em soft}, in the sense that
certain nucleotides are more likely to be paired or unpaired.
Since chemical and enzymatic probing data (SHAPE, in-line probing,
PARS) is not binary 0/1, such soft constraints permit a better
mathematical integration of such footprinting data in structure
prediction. For instance, the methods of Deigan et al.  \citep{Deigan.pnas09}
and Zarringhalam et al.  \citep{Zarringhalam.po12} obtain accuracies
of $96-100\%$ for RNA structure prediction of moderate size.
See the papers of  Mathews et al.  \citep{mathewsConstraints},
Deigan et al. \citep{Deigan.pnas09},
Zarringhalam et al. \citep{Zarringhalam.po12}, and
Washietl et al. \citep{Washietl.nar12}.

Our contribution in this paper is to extend such constrained structure
prediction to more global features, such as requiring secondary structures
to have a certain number of hairpins, a certain number of multiloops and
multiloops of a certain maximum order.  In addition to computing the
number of structures having $k$ hairpins and the partition function $Z^h(k)$
for each $0 \leq k \leq K \leq n$, the program {\rnahairpin} can
additionally sample a user-specified number of low energy structures having a
user-specified number of hairpins.  Similarly, the program {\rnahairpinml}
samples low energy structures simultaneously having $m$ multiloops and
$h$ hairpins, for user-specified values of $m,h$.  In future work,
we hope to extend
{\rnamlnumber} and {\rnamlorder} to sample low energy structures having
a user-specified number or order of multiloops, and to extend all algorithms
to compute parametric minimum free energy structures -- for instance, in the
case of {\rnahairpinml}, to compute the minimum free energy structure over
all structures having $m$ multiloops and $h$ hairpins.

The following is the plan of the paper.
Section~\ref{section:definitions} introduces standard definitions and
notation to be used. Since our algorithms derive from
McCaskill's algorithm \citep{mcCaskill} to compute the partition function
$Z= \sum_{S} \exp(-E(S)/RT)$, for the benefit of the reader, we present
that algorithm in Section~\ref{section:McCaskill}.
Sections~\ref{section:hairpinPartitionFunction},
\ref{section:multiloopNumberPartitionFunction}, and
\ref{section:multiloopOrderPartitionFunction}
respectively describe the algorithms to compute the partition functions
$Z^h(k)$, $Z^m(k)$, $Z^d(k)$ for formation of $k$ hairpins
$k$ multiloops and (maximum) order $k$ multiloops, for all $k$.
Section~\ref{section:applications} describes two applications of the
new algorithms, and Section~\ref{section:discussion} presents a
discussion and conclusion of the paper. In the Appendix, we describe
how the fast Fourier transform is used to speed up the computations of
{\rnahairpin} and {\rnamlnumber}.


%Boltzmann probabilities for number of hairpins and the number/order of
%multiloops can be {\em approximated} by using {\sfold} \citep{Ding.nar04}
%to sample a large number of structures from the Boltzmann ensemble.
%Nevertheless, {\sfold} cannot approximate the {\em parametric partition
%functions} $Z^h,Z^m,Z^d$,
%computed by our software, and the frequencies computed by
%{\sfold} are not as precise as our {\em exact} method.
%For instance, by sampling 1000 structures from the low energy ensemble,
%using the command
%\begin{quote}
%{\tt RNAsubopt -d0 -p 1000}
%\end{quote}
%for the Vienna RNA Package implementation of {\sfold}, we obtain
%hairpin formation probabilities
%$p^3=0.913$, $p^4=0.087$, $p^5=0.000$, although the real probabilities
%obtained using {\rnahairpin} are
%$p^3=0.90663$, $p^4=0.09296$, $p^5=0.00040$.
%These are only slight differences, but since we are using hairpin,
%multiloop and order probabilities as features for a support vector machine,
%there is an advantage to computing the exact value.


\section{Basic definitions}
\label{section:definitions}

In this section, we introduce some notation and definitions used in the
description of the algorithms {\rnahairpin}, {\rnamlnumber} and
{\rnamlorder}.
Let $a = a_1,\ldots,a_n$ be an arbitrary
RNA sequence, and let $a[i,j]$ denote the subsequence $a_i,\ldots,a_j$.
Given an RNA sequence $a = a_1,\ldots,a_n$, a secondary structure is
a set of ordered pairs corresponding to
base pair positions, which satisfies the following requirements.
\begin{enumerate}
\item
{\em Watson-Crick or GU wobble pairs:}
If $(i,j)$ belongs to $S$, then pair $(a_i,a_j)$ must be one of the following
canonical base pairs:
$(A,U)$, $(U,A)$, $(G,C)$, $(C,G)$, $(G,U)$, $(U,G)$.
\item
{\em Threshold requirement:}
If $(i,j)$ belongs to $S$, then $j-i > \theta$.
\item
{\em Nonexistence of pseudoknots:}
If $(i,j)$ and $(k,\ell)$ belong to $S$, then it is not the case that
$i<k<j<\ell$.
\item
{\em No base triples:}
If $(i,j)$ and $(i,k)$ belong to $S$, then $j=k$;
if $(i,j)$ and $(k,j)$ belong to $S$, then $i=k$.
\end{enumerate}
Following standard convention,
the threshold $\theta$, or minimum number of unpaired
bases in a hairpin loop, is taken to be $3$.

Secondary structures are
often portrayed in {\em dot bracket notation}, consisting of a balanced
parenthesis expression with dots. Positions $i,j$ occupied by left and
right parenthesis correspond to the base pair $(i,j)$, while positions
occupied by a dot correspond to an unpaired position $i$. The dot bracket
notation for the minimum free energy (MFE) structure for
the selenocysteine insertion element {\tt fdhA} is
\begin{quote}
\begin{tiny}
\mverbatim
CGCCACCCUGCGAACCCAAUAAUAAAAUAUACAAGGGAGCAAGGUGGCG
(((((((.(((...(((.................))).))).)))))))
|mendverbatim
\end{tiny}
\end{quote}
with free energy -20.53 kcal/mol.
A {\em pseudoknot} (not considered in our software
{\rnahairpin}, {\rnamlnumber}), and {\rnamlorder}
consists of two {\em unnested} base
pairs, $(i,j)$, $(k,\ell)$, where $i<k<j<\ell$.

In defining multiloops below, we will
have recourse to the notion of {\em component}, defined as follows.
For $1 \leq i \leq \ell \leq r \leq j \leq n$, the base pair $(\ell,r)$
is an {\em exterior} base pair in the interval $[i,j]$, if there is no
base pair $(\ell',r')$ with $i \leq \ell'<\ell < r < r' \leq j$.
When the interval $i=1$ and $j=n$, then we drop mention of the interval
$[i,j]$ and simply speak of {\em exterior} base pair.
If $S$ is a secondary structure on RNA sequence $a_1,\ldots,a_n$ and
$1 \leq i\leq j \leq n$, then the number of exterior base pairs in
the interval $[i,j]$ is said to be the number of components of $S$ in $[i,j]$.

\subsubsection*{Free energy parameters}
Following the pioneering work of I.Tinoco, Jr.,
Freier et al. \citep{Freier.pnas86} measured the free energy and
enthalpy of numerous RNA hybridization duplexes, such as
$5'$-{\tt GAACGUUC}-$3'$ with its reverse complement. By least-squares
fitting, {\em base stacking} free energies were determined. By similar
methods, the Turner Lab \citep{turner,xia:RNA}
has extended and refined base stacking free energies,
loop free energies for hairpins, bulges, internal loops, multiloops, and
{\em dangles}, which latter are
stacked single-stranded nucleotides adjacent to a
{\em canonical} $5'$ or $3'$ base pair. In this paper, we use the
energy parameters from the Turner 1999 model \citep{turner,xia:RNA}
as implemented in Vienna RNA Package 1.8.5
\citep{hofacker:ViennaWebServer}, except that we do not consider
dangles. In future work, we plan to extend to the algorithms to
the Turner 2004 energy model with dangles \citep{Turner.nar09}.

\section{McCaskill's partition function}
\label{section:McCaskill}

Since our work extends McCaskill's algorithm \citep{mcCaskill}, for the
paper to be self-contained, we give a brief presentation
of McCaskill's algorithm. This presentation follows the very lucid
account given by Bompfunewerer et al.  in \citep{Bompfunewerer.jmb08}.

Given RNA nucleotide sequence $a_1,\ldots,a_n$, we will use the standard
notation $\mathcal{H}$ to denote the free energy of a hairpin,
$\mathcal{I}$ to denote the free energy of an internal loop
(combining the cases of stacked base pair, bulge and proper internal
loop), while the
free energy for a multiloop containing $N_b$ base pairs and $N_u$ unpaired
bases is given by the affine approximation $a+bN_b+cN_u$.

For RNA sequence
$a_1,\ldots,a_n$, for all $1 \leq i \leq j \leq n$, the
McCaskill partition function  $Z(i,j)$ is defined by
$\sum_S e^{-E(S)/RT}$, where
the sum is taken over all secondary structures $S$ of $a[i,j]$,
$E(S)$ is the free energy of secondary structure $S$,
$R$ is the universal gas constant with value
$R=1.987$ cal/mol$^{-1}$ K$^{-1}$, and $T$ is absolute temperature.

\begin{definition}[McCaskill's partition function]
\label{def:partitionFunctionDefMcCaskill} \hfill\break
\begin{itemize}
\item
$Z(i,j)$: partition function over all secondary structures of
   $a[i,j]$.
\item
$Z^B(i,j)$:
 partition function over all secondary structures of
   $a[i,j]$, which contain the base pair $(i,j)$.
\item
$Z^M(i,j)$:
 partition function over all secondary structures of
   $a[i,j]$, subject to the constraint that
   $a[i,j]$ is part of a multiloop and has {\em at least} one component.
\item
$Z^{M1}(i,j)$:
 partition function over all secondary structures of
  $a[i,j]$, subject to the constraint that
   $a[i,j]$ is part of a multiloop and has at {\em exactly}
   one component. Moreover, it is {\em required} that $i$ base-pair
   in the interval $[i,j]$; i.e. $(i,r)$ is a base pair, for some
   $i<r\leq j$.
\end{itemize}
\end{definition}


With this, we have the unconstrained partition function
\begin{eqnarray}
Z(i,j) &= &Z(i,j-1) + \sum_{r=i}^{j-\theta-1} Z(i,r-1) \cdot Z^B(r,j).
\end{eqnarray}
The constrained partition function closed by base pair $(i,j)$ is
given by
$Z^B(i,j)$ equals
\begin{eqnarray}
&e^{-\mathcal{H}(i,j)/RT} +
\displaystyle\sum_{i \leq \ell \leq r \leq j}
e^{-\mathcal{I}(i,\ell,r,j)/RT}\cdot Z^B(\ell,r) +\\
& e^{-(a+b)/RT} \cdot \left( \sum_{r=i+1}^{j-\theta-2} Z^M(i+1,r-1)
\cdot Z^{M1}(r,j-1) \right). \nonumber
\end{eqnarray}
The multiloop partition function with a single component and where
position $i$ is required to base-pair in the interval $[i,j]$ is given
by
\begin{eqnarray}
Z^{M1}(i,j) &= &
\displaystyle\sum_{r=i+\theta+1}^j Z^B(i,r) \cdot
e^{-c(j-r)/RT} .
\end{eqnarray}
Finally, the multiloop partition function with one or more components,
having no requirement that position $i$ base-pair in the interval $[i,j]$
is given by $Z^{M}(i,j)$ equals
\begin{eqnarray}
&
\displaystyle\sum_{r=i}^{j-\theta-1}  Z^{M1}(r,j) \cdot
e^{-(b+c(r-i))/RT}  + \\
&\displaystyle\sum_{r=i+\theta+1}^{j-\theta-1}  Z^{M}(i,r-1) \cdot
Z^{M1}(r,j) \cdot e^{-b/RT}  \nonumber
\end{eqnarray}
See \Figref{feynmanDiagram} for a pictorial representation
of the recursions of McCaskill's (original) algorithm \citep{mcCaskill};
note that the recursions are are not quite the same
as those given in \citep{hofacker:FastFolding}.
We now turn to our parametric versions of the partition function.

\begin{figure}[htp]
%\centerline{\includegraphics[width=0.6\textwidth]{feynmanDiagram}}
\centerline{\includegraphics[width=0.6\textwidth]{figure1}}
\caption{Feynman diagram of original recursions from McCaskill's
algorithm \citep{mcCaskill} to compute the partition function.
Dashed lines present intervals of unpaired bases, and shaded
arcs represent structures in which $i$ and $j$ will not necessarily pair.
}
\label{fig:feynmanDiagram}
\end{figure}
%Figure 1

%\noindent
%================ FIGURE 1 GOES ABOUT HERE ===================

\section{Hairpins}
\label{section:hairpinPartitionFunction}



We begin by defining some abbreviations for the partition function for
hairpins
\[
ZH(i,j) =
\left\{ \begin{array}{ll}
0 &\mbox{if $j-i \leq \theta$}\\
e^{-\mathcal{H}(i,j)/RT} &\mbox{else}
\end{array} \right.
\]
and internal loops having $h$ hairpins
\[
ZI^h(i,j) =  \displaystyle\sum_{i \leq k \leq \ell \leq j}
e^{-\mathcal{I}(i,j;h,k)/RT} \cdot ZB^h(k,\ell)
\]
where the sum is over $k,\ell$ such that $(k-i)+(j-\ell)>0$. This combines the
treatment of both left and right bulges with proper internal loops.

For $h\geq 0$, define the base cases $Z^h(i,i)=1$,
$ZB^h(i,i)= ZM^h(i,i)= ZM1^h(i,i)=0$.
The unconstrained partition function for secondary structures
restricted to the interval $[i,j]$ that contain $h$ hairpins is given by
\[
Z^h(i,j) =
\left\{ \begin{array}{ll}
1 &\mbox{if $h=0$}\\
Z^h(i,j-1) + \sum_{r=i}^{j-\theta-1} \sum_{h_0+h_1=h}
Z^{h_0} ZB^{h_1}(r,j) &\mbox{if $h>0$}.
\end{array} \right.
\]
The partition function for secondary structures restricted to the
interval $[i,j]$ that contain $h$ hairpins and are closed by
the base pair $(i,j)$ is given by
$ZB^h(i,j) = 0$, if $h=0$;
$ZB^h(i,j) = ZH(i,j) + ZI^h(i,j)$, if $h=1$; and for $h\geq 2$,
$ZB^h(i,j)$ is defined equal to
\[
ZI^h(i,j) + \sum_{r=i+\theta+2}^{j-\theta-2} \sum_{k=1}^{h-1}
     ZM^{k}(i+1,r-1) \cdot ZM1^{h-k}(r,j-1) \cdot e^{-(a+b)/RT}.
\]

The multiloop partition function with a single component and where
position $i$ is required to base-pair in the interval $[i,j]$ is given by
\begin{eqnarray}
ZM1^{h}(i,j) &= &
\displaystyle\sum_{r=i+\theta+1}^j ZB^h(i,r) \cdot
e^{-c(j-r)/RT} .
\end{eqnarray}
Finally, the multiloop partition function with one or more components,
having no requirement that position $i$ base-pair in the interval $[i,j]$
is given by defining $ZM^{h}(i,j)$ to equal
\begin{eqnarray}
&\displaystyle\sum_{r=i}^{j-\theta-1}  ZM1^{h}(r,j) \cdot
e^{-(b+c(r-i))/RT}  + \\
&\displaystyle\sum_{r=i+\theta+1}^{j-\theta-1} \displaystyle\sum_{k=1}^{h-1}
ZM^{k}(i,r-1) \cdot ZM1^{h-k}(r,j) \cdot e^{-b/RT}  \nonumber
\end{eqnarray}

\section{Number of multiloops}
\label{section:multiloopNumberPartitionFunction}



As before, define the abbreviations for the partition function for
hairpins
\[
ZH(i,j) =
\left\{ \begin{array}{ll}
0 &\mbox{if $j-i \leq \theta$}\\
e^{-\mathcal{H}(i,j/RT} &\mbox{else}
\end{array} \right.
\]
and internal loops having $k$ multiloops
\[
ZI^m(i,j) =  \displaystyle\sum_{i \leq k \leq \ell \leq j}
e^{-\mathcal{I}(i,j;h,k)/RT} \cdot ZB^m(k,\ell)
\]
where the sum is over $k,\ell$ such that $(k-i)+(j-\ell)>0$. As
in the previous section, this combines the
treatment of both left and right bulges with proper internal loops.

Define $Z^0(i,i)=1$, and for $m> 0$, define $Z^m(i,i)=0$. For the remaining
base cases, define $ZB^m(i,i)= ZM^m(i,i)= ZM1^m(i,i)=0$.
The unconstrained partition function for secondary structures
restricted to the interval $[i,j]$ that contain $m$ multiloops is given by
\[
Z^m(i,j) = Z^m(i,j-1)+
\sum_{r=i}^{j-\theta-1} \sum_{k=0}^m Z^k(i,r-1) \cdot ZB^{m-k}(r,j)
\]
The partition function for secondary structures restricted to the
interval $[i,j]$ that contain $m$ multiloops and are closed by
the base pair $(i,j)$ is given by
$ZB^m(i,j) = ZI^m(i,j)$, if $m=0$, and in the case that $m>0$ and
$i,j$ can form a base pair,
$ZB^m(i,j)$ equals
\begin{eqnarray*}
&ZI^m(i,j) + \displaystyle\sum_{r=i+\theta+2}^{j-\theta-2} \sum_{k=0}^{m-1}
     ZM^{k}(i+1,r-1) \\
&\cdot ZM1^{m-k-1}(r,j-1) \cdot e^{-(a+b)/RT}.
\end{eqnarray*}


The multiloop partition function with a single component and where
position $i$ is required to base-pair in the interval $[i,j]$ is given by
\begin{eqnarray}
ZM1^{m}(i,j) &= &
\displaystyle\sum_{r=i+\theta+1}^j ZB^m(i,r) \cdot
e^{-c(j-r)/RT} .
\end{eqnarray}
Finally, the multiloop partition function with one or more components,
having no requirement that position $i$ base-pair in the interval $[i,j]$
is given by $ZM^{m}(i,j)$
\begin{eqnarray}
&\displaystyle\sum_{r=i}^{j-\theta-1}  ZM1^{m}(r,j) \cdot
e^{-(b+c(r-i))/RT}  + \\
&\displaystyle\sum_{r=i+\theta+1}^{j-\theta-1} \sum_{k=1}^{m-1}
ZM^{k}(i,r-1) \cdot ZM1^{m-k-1}(r,j) \cdot e^{-b/RT}  \nonumber
\end{eqnarray}


\begin{figure}[tbhp]
\centering
%\includegraphics[scale=0.3]{haipinProfile}
%\includegraphics[scale=0.3]{MLNumber}
%\includegraphics[scale=0.3]{MLOrder}
\includegraphics[scale=0.3]{figure2a}
\includegraphics[scale=0.3]{figure2b}
\includegraphics[scale=0.3]{figure2c}
\caption{{\em (Left)}
Hairpin profile of Rfam families:
U2 spliceosomal RNA (RF00004),
transfer RNA (tRNA, RF00005) and
U4 spliceosomal RNA (RF00015).
{\em (Center)}
Multiloop number profile of
Rfam families: RNaseP (RF00010), transfer messenger RNA (tmRNA, RF00023),
and Rev response element of HIV env gene (RF00036).
{\em (Right)}
Multiloop order (or depth) profile of
Rfam families: RNaseP (RF00010), transfer messenger RNA (tmRNA, RF00023),
and Rev response element of HIV env gene (RF00036).
Notice that we chose Rfam families consisting of long RNA sequences
for multiloop number/order profiles, since multiloops are energetically
unfavorable, hence are not generally present in small RNA.
}
\label{fig:RfamProfiles}
\end{figure}
%Figure 2
%\noindent
%================ FIGURE 2 GOES ABOUT HERE ===================

\section{Multiloop order}
\label{section:multiloopOrderPartitionFunction}

%first defined
%by Waterman \citep{waterman:book} and extensively investigated by
%Nebel \citep{nebel:JCB2002},

The {\em order} (or {\em depth}) of a secondary structure is
the maximum {\em depth} of nesting of its multiloops.
Formally, multiloop order
can be defined via a finite analogue of the Cantor-Bendixson topological
derivative \citep{clote:CantorBendixson,kechris}.
The {\em derivative} $D(\strS)$ of secondary
structure $\strS$ is equal to the set of base pairs $(i,j) \in \strS$,
within which there is an internal branching; i.e.
\begin{eqnarray*}
D(\strS) &=& \{ (i,j) \in \strS : \mbox{there exist distinct
$(x,y),(u,v) \in \strS$} \\
&&\mbox{such that $i<x<y<u<v<j$} \}.
\end{eqnarray*}
The {\em order} of a secondary structure $\strS$ is now
defined to be $n-1$, where $n$ is the least integer such that
$D(\strS)=\emptyset$. For readers familiar with the notion of
RNA {\em shape} \citep{giegerich:shapesNAR},
it follows that the order of a helix is
zero, with shape $\ob \cb$, while the order of a tRNA cloverleaf secondary
structure is one, with shape $\ob \ob \cb \ob \cb \ob \cb \cb$.
Examples of order $2$ secondary structures, with shape
$\ob \ob \ob \cb \ob \cb \ob \cb \cb \ob \cb \cb$, are furnished by
certain RNase P RNA molecules, such as
{\sc strand} database \citep{Andronescu.bb08} sequence
ASE00001 from {\em Acidianus ambivalens}, and by some transfer-messenger
RNA, such as {\sc strand} database sequence
TMR00040, from {\em Azos.oryz.}.

For typographic reasons, we denote the multiloop partition function
by $Z^d$, rather than $Z^o$.
As before, define the partition function for hairpins
\[
ZH(i,j) =
\left\{ \begin{array}{ll}
0 &\mbox{if $j-i \leq \theta$}\\
e^{-\mathcal{H}(i,j)/RT} &\mbox{else}
\end{array} \right.
\]
and internal loops having multiloop order or depth $d$
\[
ZI^d(i,j) =  \displaystyle\sum_{i \leq k \leq \ell \leq j}
e^{-\mathcal{I}(i,j;k,\ell)/RT} \cdot ZB^d(k,\ell)
\]
where the sum is over $k,\ell$ such that $(k-i)+(j-\ell)>0$.
Define $Z^0(i,i)=1$ and for $d> 0$, define $Z^d(i,i)=0$.
For $d\geq 0$, define $ZB^d(i,i)= ZM^d(i,i)= ZM1^d(i,i)=0$.
The unconstrained partition function for secondary structures of multiloop
order $d$, when
restricted to the interval $[i,j]$, is given by
$Z^d(i,j)$ equals
\[
Z^d(i,j-1) + \sum_{r=i}^{j-\theta-1} \quad
    \sum_{0 \leq k,\ell \leq d, \max(k,\ell)=d}
    Z^k(i,r-1) \cdot ZB^\ell(r,j)
\]
The partition function for secondary structures of multiloop order $d$
when restricted to the
interval $[i,j]$  and are closed by
the base pair $(i,j)$ is given as follows.
For $d=0$, let
\[
ZB^d(i,j) =
ZH(i,j) + ZI^d(i,j)
\]
while for $d>0$, define
\begin{eqnarray*}
ZB^d(i,j) &=&
ZI^d(i,j) + \sum_{r=i+\theta+1}^{j-\theta-1}
    \sum_{0 \leq k,\ell \leq d, \max(k,\ell)=d} \\
&&    ZM^k(i,r-1)  \cdot
ZM1^\ell(r,j)  \cdot e^{-(a+b)/RT}.
\end{eqnarray*}


The multiloop partition function with a single component and where
position $i$ is required to base-pair in the interval $[i,j]$ is given by
\begin{eqnarray}
ZM1^{d}(i,j) &= &
\displaystyle\sum_{r=i+\theta+1}^j ZB^d(i,r) \cdot
e^{-c(j-r)/RT}.
\end{eqnarray}
Finally, the multiloop partition function with one or more components,
having no requirement that position $i$ base-pair in the interval $[i,j]$
is given by
\begin{eqnarray}
ZM^{d}(i,j) &=&
%\displaystyle\sum_{r=i}^{j-\theta-1}  ZM1^{d}(r,j) \cdot
\sum_{r=i}^{j-\theta-1}  ZM1^{d}(r,j) \cdot
e^{-(b+c(r-i))/RT}  + \\
&& \sum_{r=i+\theta+1}^{j-\theta-1}
  \sum_{0 \leq k,\ell \leq d, \max(k,\ell)=d} ZM^{k}(i,r-1) \cdot
\nonumber\\
&&ZM1^{\ell}(r,j) \cdot e^{-b/RT}.  \nonumber
\end{eqnarray}

\section{Simultaneous multiloop number and hairpin number}
\label{section:RNAhairpinml}

Given the algorithms described in Sections
\ref{section:hairpinPartitionFunction} and
\ref{section:multiloopNumberPartitionFunction}, it is straightforward
to design the algorithm {\rnahairpinml}, which computes the partitiion
function $Z(m,h)$ simultaneously for $m$ multiloops and $h$ hairpins.
Sampling low energy structures is done by a straightforward variation of
the sampling method introduced by Ding and Lawrence \citep{Ding.nar03}.
For purposes of brevity, further details of the partition function and
sampling will not be discussed, though the interested reader can study
our publicly available source code.

\section{Applications}
\label{section:applications}

In this section, we mention two main applications of the new algorithms,
though first we mention that {\rnahairpin} presents a novel method to
generate suboptimal secondary structures.


In the literature, there are a number of approaches to compute
{\em suboptimal} secondary structures. Historically, the first
method was due to Zuker \citep{Zuk89a}, as implemented in {\tt mfold}
{\tt mfold} \citep{Zuk89a} and {\tt Unafold} \citep{Markham.mmb08},
who for certain base pairs $(i,j)$ computed the minimum
free erergy structure containing $(i,j)$ that was sufficiently distinct
from previously generated suboptimal structures.
Next, the program \rnasub of Wuchty et al.
\citep{wuchtyFontanaHofackerSchuster} generated all secondary structures
within a user-specified energy above the minimum free energy.
In contrast the program {\tt Sfold} \citep{Ding.nar03} samples from
the low energy Boltzmann ensemble of structures; indeed, our implementation
of sampling in {\rnahairpin} is a modification of the method of
Ding and Lawrence \citep{Ding.nar03}. (Note that the {\tt Sfold} algorithm
is implemented in the Vienna RNA Package program
\rnasub with flag {\tt -p}; as well the program
{\tt RNAstructure} \citep{mathewsConstraints} also supports sampling.)
The program {\tt RNAshapes} of Steffen et al.
\citep{giegerich:shapesBioinf} computes the minimum free energy structure
from each shape class.
The program \rnabor of Freyhult et al. \citep{freyhult.b07}
computes, for each $k$,
the minimum free energy structure $MFE(k)$ having base pair
distance $k$ from a user-specified reference structure, while the
program \rnatwofold of Lorenz et al. \citep{hofacker:RNAbor2D}
computes, for each $k,\ell$, the minimum free energy structure
$MFE(k,\ell)$ having base pair
distance $k$ [resp. $\ell$] from a first [resp. second]
user-specified reference structure.
The program {\tt RNAlocopt} of Lorenz and Clote  \citep{RNAlocopt}
samples low  locally optimal secondary structures, where a locally
optimal structure has the property that its free energy cannot be lowered
by the addition or removal of a single base pair. The program
{\tt RNAbormea} of Lou and Clote \citep{Clote.bb12} determines
for each $k$, the maximum expected accuracy structure among all structures
having base pair distance $k$ from a user-specified reference structure.
To this list of previous methods, {\rnahairpin} generates suboptimal
secondary structures in a manner that seems orthogonal to previous methods.

\subsection{Improved structure prediction for certain RNA families}


Certain RNAs have a characteristic structure that is important for
their function. For instance, the cloverleaf structure of transfer RNA
generally has three hairpins,  which then form an L-shaped tertiary
structure by additional pseudoknots. Transfer RNAs usually contain
a small number of chemically modified nucleotides, making their structure
at times difficult to predict using minimum free energy structure methods.
In such cases, {\rnahairpin} [resp. and expecially {\rnahairpinml}]
can improve structure prediction by
sampling low energy structures that are required to have a specific
number of hairpins [resp. number $m$ of multiloops and $h$ of hairpins].

\begin{table*}[!t]
\begin{tabular}{|l|rrrrr|}
\hline
RNA family & H & {\tt RNAhairpin} $\mu \pm \sigma$ & {\tt RNAfold}
$\mu \pm \sigma$ & avg len & num seq\\
\hline
RF00001 & 2 & $0.6213 \pm 0.2667$ & $0.6332 \pm 0.2721$ & 116.6 & 712\\
RF00004 & 5 & $0.7548 \pm 0.1840$ & $0.7104 \pm 0.2058$ & 190.5 & 208 \\
RF00005 & 3 & $0.7345 \pm 0.2313$ & $0.5370 \pm 0.1992$ & 73.4 &960  \\
RF00008 & 2 & $0.9565 \pm 0.1284$ & $0.9154 \pm 0.1894$ & 55.4 & 84\\
RF00031 & 1 & $0.7679 \pm 0.1748$ & $0.7657 \pm 0.1945$ & 64.5 & 61\\
RF00045 & 4 & $0.4420 \pm 0.2983$ & $0.4205 \pm 0.3274$ & 202.6 & 66 \\
RF00094 & 2 & $0.3080 \pm 0.2131$ & $0.3604 \pm 0.2091$ & 91.1 & 33\\
RF00167 & 2 & $0.8113 \pm 0.2301$ & $0.8568 \pm 0.2290$ & 100.8 & 133 \\
RF00375 & 2 & $0.8278 \pm 0.3060$ & $0.8814 \pm 0.2044$ & 99.0 & 130\\
RF00504 & 2 & $0.5940 \pm 0.2711$ & $0.5603 \pm 0.2895$ & 100.9 & 44\\
RF00635 & 4 & $0.3024 \pm 0.1127$ & $0.3707 \pm 0.1204$ & 117.9 & 13\\
RF01055 & 4 & $0.5821 \pm 0.2725$ & $0.5787 \pm 0.2641$ & 142.0 & 160 \\
\hline
\end{tabular}
\caption{Comparison between {\tt RNAhairpin} and {\tt RNAfold}
of the average sensitivity (ratio of
number of correctly predicted base pairs in Rfam structure over
number of base pairs in Rfam structure) for various Rfam families.
{\rnahairpin}
was used to sample a single secondary structure having H many hairpins,
and the average sensitivity of {\rnahairpin} and {\tt RNAfold}
was computed over all sequences in the seed alignment of the following
Rfam families:
RF00001 (5S rRNA),
RF00004 (splicesomal U2 RNA),
RF00005 (tRNA),
RF00008 (type III hammerhead ribozyme),
RF00031 (selenocysteine insertion sequence I),
RF00045 (snoRNA),
RF00094 (HDV ribozyme),
RF00167 (purine riboswitch),
RF00375 (HIV primer binding site),
RF00504 (glycine riboswitch),
RF00635 (HAR1A),
RF01055 (moco RNA motif).
}
\label{table:sensitivity}
\end{table*}

\begin{table*}[!b]
\begin{tabular}{|l|rrrrrr|}
\hline
RNA family & M & H & {\tt RNAhairpin} $\mu \pm \sigma$ & {\tt RNAfold}
$\mu \pm \sigma$ & avg len & num seq\\
\hline
RF00001 & 1 & 2 & $0.6308 \pm 0.2571$ & $0.6332 \pm 0.2721$ & 116.6 & 712\\
RF00004 & 0 & 5 & $0.6980 \pm 0.1780$ & $0.7104 \pm 0.2058$ & 190.5 & 208 \\
RF00005 & 1 & 3 & $0.7740 \pm 0.1946$ & $0.5370 \pm 0.1992$ & 73.4 &960  \\
RF00008 & 1 & 2 & $0.9582 \pm 0.1005 $ & $0.9154 \pm 0.1894$ & 55.4 & 84\\
RF00031 & 0 & 1 & $0.7679 \pm 0.1748$ & $0.7657 \pm 0.1945$ & 64.5 & 61\\
RF00045 & 1 & 4 & $0.4456 \pm 0.2977$ & $0.4205 \pm 0.3274$ & 202.6 & 66 \\
RF00094 & 0 & 2 & $0.3464 \pm 0.1951$ & $0.3604 \pm 0.2091$ & 91.1 & 33\\
RF00167 & 1 & 2 & $0.8511 \pm 0.1726$ & $0.8568 \pm 0.2290$ & 100.8 & 133 \\
RF00375 & 1 & 2 & $0.8283 \pm 0.3063$ & $0.8814 \pm 0.2044$ & 99.0 & 130\\
RF00504 & 1 & 2 & $0.6101 \pm 0.264$ & $0.5603 \pm 0.2895$ & 100.9 & 44\\
RF00635 & 1 & 3 & $0.2930 \pm 0.1059$ & $0.3707 \pm 0.1204$ & 117.9 & 13\\
RF01055 & 1 & 4 & $0.60170 \pm 0.277$ & $0.5787 \pm 0.2641$ & 142.0 & 160 \\
\hline
\end{tabular}
\caption{Comparison between {\rnahairpinml} and {\tt RNAfold}
of the average sensitivity for the same Rfam families, as in
Table~\ref{table:sensitivity}. By now sampling a single secondary structure
having simultaneously M many multiloops and H many hairpins, the
average sensitivity improved over that of {\rnahairpin}
in essentially all cases.
%Improvement over {\rnahairpin}: over 2\% in RF00005 (tRNA),
%4\% in RF00094 (HDV ribozyme),
%over 4\% in RF00167 (purine riboswitch),
%about 1.5\% in RF00504  (glycine riboswitch),
%about 2\% in RF01055 (moco RNA motif).
Moreover, {\rnahairpinml} provides more accurate structure
prediction (sensitivity) than {\tt RNAfold} for a number of
Rfam families. There is an improvement
of almost $approx 24\%$ in RF00005 (tRNA),
over $4\%$ in RF00008 (type III hammerhead ribozyme),
$2.5\%$ in RF00045 (snoRNA),
$5\%$ in RF00504  (glycine riboswitch),
over $2\%$ in RF01055 (moco RNA motif).
On the other hand, {\rnahairpinml} has significantly lower
sensitivity than {\tt RNAfold} in the following two cases, where
the difference is over $5\%$ for RF00375 (HIV primer binding site),
and $8\%$ for RF00635 (HAR1A).
Insignificant differences, such as
$0.6308$ for {\rnahairpinml} versus $0.6332$ in RF00001 (5S rRNA)
are likely to be due to the stochastic nature of sampling low energy
structures, rather than computing the MFE structure having a specified
number of multiloops and hairpins.
}
\label{table:sensitivityMultiloopHairpin}
\end{table*}

Table~\ref{table:sensitivity} presents a comparison of {\rnahairpin}
and {\tt RNAfold} statistics for sequences taken from the seed alignments
of several families from Rfam 11.0 \citep{Burge.nar13}
(August 2012, 2208 families).  For each sequence,
we sampled only one low energy structure having H hairpins. For a given
sequence and structure computed either by {\rnahairpin} or {\tt RNAfold},
the sensitivity, or true positive rate, is computed, defined as
the ratio of number of correctly predicted base pairs in the
Rfam structure over the number of base pairs in the Rfam structure.
The average and standard deviation of sensitivity is given, for each
Rfam family of the table, for both {\rnahairpin} and {\tt RNAstructure}.
For these computations, version 1.8.5 of {\tt RNAfold} was used without
dangles, so that both {\rnahairpin} and {\tt RNAfold} employed the same
energy model. In future work, we plan to lift {\rnahairpin} to
the Turner 2004 energy model and implement dangles, which then would
support the same energy model as version 2.0 and higher of
{\tt RNAfold} \citep{Lorenz.amb11}.


In the case of tRNA, there is more than 20\% improvement in sensitivity
of {\rnahairpin} over {\tt RNAfold}; {\rnahairpin} has greater sensitivity
than {\tt RNAfold} for other instances, such as
in the case of the hammerhead ribozyme (around 4\% improvement).
On the other hand, {\tt RNAfold} has greater
sensitivity than {\rnahairpin} for several classes, including
HIV primer binding site RF00375 (over 5\% improvement), and
purine riboswitch aptamers RF00167 (around 4.5\% improvement).
Clearly {\rnahairpin} is not a better structure prediction tool than
{\tt RNAfold}; however, for particular classes of functional RNA,
which require certain hairpin structures for function, {\rnahairpin}
may provide a useful tool. See Section~\ref{section:discussion} for
more discussion.

The program {\rnahairpinml}, which samples low energy structures having
$m$ multiloops and $h$ hairpins,  improves the structure prediction
accuracy of {\rnahairpin} (e.g. an improvement of over 4\% for RF000167
purine riboswitches), and also outperforms {\tt RNAfold} for a larger number
of cases on the previously described Rfam families. For instance,
there is an improvement
of almost $approx 24\%$ in RF00005 (tRNA),
over $4\%$ in RF00008 (type III hammerhead ribozyme),
$5\%$ in RF00504  (glycine riboswitch), etc.
On the other hand, {\rnahairpinml} has significantly lower
sensitivity than {\tt RNAfold} in the following two cases, where
the difference is over $5\%$ for RF00375 (HIV primer binding site),
and $8\%$ for RF00635 (HAR1A). The consensus structures for these
Rfam families have large loop regions, which may in fact be base-paired,
which could explain the poorer performance of {\rnahairpinml}.
(Recall that Rfam consensus structures are determined by covariation found
in multiple alignments, thus loop regions in consensus structures could
indeed by base-paired and involve additional hairpins and/or multiloops.)
In any case, we do not propose the use of {\rnahairpinml} in place of
minimum free energy structure software, such as {\tt RNAfold}; instead,
if a biologist has knowledge or intuition about the existence of a certain
number of multiloops and hairpins, then {\rnahairpinml} may prove to be a
useful tool.

\begin{table*}
\begin{tabular}{|l|rrrrrr|}
\hline
Family name and description & H & HM & HMO & num seq& avg len & avg GC \% \\
\hline
RF00004 U2 spliceosomal RNA & 0.9217  & 0.9282  & 0.9328  & 208 & 204.26 &48.0\% \\
RF00005 tRNA &  0.6367  & 0.9038  & 0.9017  & 959 & 73.4& 47.0\%\\
RF00008 hammerhead III &  0.9191  & 0.9705  & 0.9562  & 84 & 55.4 & 48.4\% \\
RF00027 let 7 microRNA precursor &  0.8338  & 0.8766  & 0.8617  & 67 & 79.6 & 43.7\% \\
RF00031 SECIS 1 & 0.7917  & 0.8361  & 0.7941  & 61 & 64.5 & 49.0\% \\
RF00045 SNORA73 & 0.6306  & 0.6515  & 0.6609  & 66 & 202.6 & 53.1\% \\
RF00167 purine riboswitch & 0.6508  & 0.8608  & 0.8529  & 136 & 100.8 & 38.1\% \\
\hline
\end{tabular}
\caption{Area under curve (AUC) for receiver operating characteristic
(ROC) curves for
seven Rfam families, each family tested under 5-fold cross-validation with
support vector machines (SVM) using a radial basis kernel with
cost $C=1$ and $\gamma$ equal to the inverse of the number of features.
In the case of H (hairpin number), there were 21 hairpin formation
probabilities $p(0),\ldots,p(20)$ taken as features,
(though in
most cases all but a very small number of these probabilities were zero);
in the case  of HM (hairpin and multiloop number), there were
27=21+6 hairpin and multiloop formation probabilities taken as
features, and in the case  of HMO
(hairpin and multiloop number with maximum multiloop order), there were
27=21+6 hairpin and multiloop formation probabilities taken as features
along with 6 multiloop maximum order probabilities, hence altogether
33=21+6+6 features.
The R packages {\tt e1071} \citep{RpackageForSVM} and {\tt pROC} \citep{pROC}
were used with {\tt libSVM} \citep{libSVM}.
}
\label{table:AUC}
\end{table*}

\begin{table*}
\begin{tabular}{|l|rrrrrrr|}
\hline
Ratio SVM/RVM &RF00004& RF00005& RF00008& RF00027& RF00031& RF00045& RF00167\\
\hline
HP&      0.9874&  1.0657&  0.9874&  1.4234&  1.1965&  0.9895&  1.1894\\
HP/M&    0.9863&  0.9798&  0.9977&  1.0625&  1.0954&  0.9808&  1.0153\\
HP/M/O&  0.9818&  0.9855&  1.0025&  0.8986&  1.2324&  1.0237&  1.0031\\
\hline
\end{tabular}
\caption{Ratio of ROC area under curve values for two types of machine
learning methods: support vector machines (SVM) and relevance vector
machines (RVM), using the same seven Rfam families that were considered
in Table~\ref{table:AUC}. In 11 out of 21 tests, AUC values for SVMs
were greater than those for RVMs. In the case of RF00027, it is interesting
to note that when using only hairpin features, SVM AUC was much higher than
RVM AUC (SVM/RVM $1.4234$), while for the same class, when using the larger
feature set for hairpins, multiloop number and multiloop order, SVM AUC
was lower than RVM AUC (SVM/RVM $0.8986$). At present, the reason for
this surprising result is unclear.
The R packages {\tt e1071} \citep{RpackageForSVM} and {\tt pROC} \citep{pROC}
were used for SVM and RVM computations;
for SVM, the radial basis kernel (rbfkernel) was employed with
default parameters, while for RVM, rvmbinary
rbfdot kernel was used with default parameters and 1000 iterations.
}
\label{table:RF00027aucSVMversusRVM}
\end{table*}

\subsection{Support vector machine results}

In this section, we describe receiver operating characteristic
(ROC) curves, computed by 5-fold cross-validation, where in each
case, the positive instances were taken to be sequences from the
seed alignment of a given Rfam family, and negative instances were
taken to be random sequences having the same number of dinucleotides,
as computed by our implementation of the Altschul-Erikson algorithm
\citep{altschulErikson:dinucleotideShuffle}.
(Similar results were obtained, when we took negative instances to be
sequences the seed alignments of all other Rfam families -- data not
shown.)

For each positive instance, we generated 10 random negative instances.
Using {\tt libSVM} \citep{libSVM}, we performed a {\em stratified training}
by selecting one-fifth of the
positive instances together with an equal number of negative instances
(one of the 10 negative instances was selected for each positive instance)
for training. The remaining four-fifths of the positive sequences, together
with all corresponding negative instances, constituted the test set
(note that in testing, there were 10 negative instances per positive sequence).
A radial basis kernel was chosen with cost $C=1$, and parameter $\gamma$
taken to be the reciprocal of the number of features.

We considered three features sets: HP, HP/M, and HP/M/O,
where HP features were the 21 probabilities of hairpin formation
$p_h(0),\ldots,p_h(20)$, where M features were the 6 probabilities of
multiloop formation $p_m(0),\ldots,p_m(6)$, and where O features
were the 6 probabilities of multiloop order (or depth) $p_d(0),\ldots,p_d(6)$.
Thus the SVM binary classifier HP (hairpins) has 21 features, though in
most cases all but a small number of the features are $0$;
the classifier HP/M (hairpin and multiloop number) has
27=21+6 features; the classifier HP/M/O (hairpin, multiloop number,
multiloop order) has 33=21+6+6 features.
The R packages {\tt e1071} \citep{RpackageForSVM} and {\tt pROC} \citep{pROC}
were used with {\tt libSVM} \citep{libSVM}.

Table~\ref{table:AUC} summarizes the area under curve (AUC) values for
ROC curves for the three different SVM classifiers HP, HP/M, HP/M/O,
while \Figref{SVMforU2andPurineRiboswitch} depicts the corresponding
ROC curves. Note that in all cases, inclusion of multiloop order probabilities
as features does not add any discriminatory power, and even in certain cases
reduces the AUC. This is fortunate, since {\rnamlorder} cannot be
sped up by using the fast Fourier transform, unlike {\rnahairpin} and
{\rnamlnumber}. The results of this table and figure indicate that,
although hairpin and multiloop formation probabilities may not be
sufficient to be used solely as the feature set of a noncoding RNA gene
finder, we believe that, when added, these features could lead to improvements
in performance of existent RNA gene finders. Moreover, to the best of our
knowledge, current noncoding RNA gene finders do not take into account
global propensity to form hairpins or multiloops.

Table~\ref{table:RF00027aucSVMversusRVM} presents the ratio of ROC  area
under curve values for support vector machines (SVM) over that for relevance
vector machines (RVM). A value greater [resp. less] than unity in the
table indicates that SVM outperforms [resp. underperforms] RVM using the
same features.
\Figref{RF00027aucSVMversusRVM} shows an unexpected situation
for the 5-fold (stratified) cross-validation experiments of the Rfam family
RF00027, Using the feature set consisting of only 21 hairpin
formation probabilities $p_h(0),\ldots,p_h(20)$, the ratio of SVM/RVM AUC
is $1.4234$, indicating that SVM far outperforms RVM for this family using these
features, while for the full feature set of hairpin formation probabilities
$p_h(0),\ldots,p_h(20)$, multiloop number probabilities $p_h(0),\ldots,p_h(6)$,
and multiloop maximum order (depth) probabilities $p_h(0),\ldots,p_h(6)$,
the ratio of SVM/RVM AUC is $0.8986$,
indicating that RVM outperforms SVM.


\begin{figure*}[tbhp]
\centering
%\includegraphics[width=0.45 \textwidth]{svm_rf00004_roc_curve_overlay}
\includegraphics[width=0.45 \textwidth]{figure3a}
\hskip 1cm
%\includegraphics[width=0.45 \textwidth]{svm_rf00167_roc_curve_overlay}
\includegraphics[width=0.45 \textwidth]{figure3b}
\caption{
Receiver operating characteristic (ROC) curves for the performance of
support vector machine binary classification using a
feature set consisting of probabilities
$p(0),\ldots,p(20)$ for the number hairpins (HP),
probabilities $p(0),\ldots,p(5)$ for the
number of multiloops (M), and
probabilities $p(0),\ldots,p(5)$ for the
maximum order of multiloops (O).
In the case of HP (hairpin number), there were 21 features, though in
most cases all but at most 6-8 features had the value $0$;
in the case  of HP/M (hairpin and multiloop number), there were
27=21+6 features, and in the case  of HP/M/O
(hairpin and multiloop number with maximum multiloop order), there were
33=21+6+6 features.
The R packages {\tt e1071} \citep{RpackageForSVM} and {\tt pROC} \citep{pROC}
were used with {\tt libSVM} \citep{libSVM}. A radial basis kernel was used
in each case with cost $C=1$; parameter $\gamma$ was taken to be the
inverse of the number of features, i.e. for
HP, $\gamma = 1/21 = 0.0476$, for HP/M, $\gamma = 1/27 = 0.0370$, for
HP/M/O, $\gamma = 1/33 = 0.0303$.
%
As shown in this figure, accounting for multiloop
order did not improve classification ROC curves, and data presented in
Table~\ref{table:AUC} shows that in some cases, ROC area under curve
is lessened by taking into account maximum multiloop order. This is
in fact fortunate, since the fast Fourier transform can be applied to
reduce time and space requirements for {\rnahairpin} and
{\rnamlnumber}, but not {\rnamlorder}.
{\em (Left)}
Rfam family RF00008 (U2 spliceosomal RNA).
{\em (Right)}
Rfam family RF00167 (purine riboswitch).
}
\label{fig:SVMforU2andPurineRiboswitch}
\end{figure*}

\begin{figure*}[tbhp]
\centering
%\includegraphics[width=0.45 \textwidth]{roc_rf00027_h_roc_curve_overlay}
\includegraphics[width=0.45 \textwidth]{figure4a}
\hskip 1cm
%\includegraphics[width=0.45 \textwidth]{roc_rf00027_hmo_roc_curve_overlay}
\includegraphics[width=0.45 \textwidth]{figure4b}
\caption{Receiver operating characteristic (ROC) curves for
5-fold cross-validation for sequences from the seed alignment of
RF00027.
The left panel shows an overlay of support vector machine (SVM) and
relevance vector machine (RVM) for the feature set consisting of
21 hairpin formation probabilities $p_h(0),\ldots,p_h(20)$,
while the right panel presents an overlay of SVM and RVM for the
full feature set of hairpin formation probabilities $p_h(0),\ldots,p_h(20)$,
multiloop number probabilities $p_h(0),\ldots,p_h(6)$, and
multiloop maximum order (depth) probabilities $p_h(0),\ldots,p_h(6)$.
As explained in the caption of Table~\ref{table:RF00027aucSVMversusRVM},
it seems unusual that SVM outperforms RVM using only hairpin probability
features, while the reverse is true when using the full feature set.
}
\label{fig:RF00027aucSVMversusRVM}
\end{figure*}


\section{Discussion and conclusion}
\label{section:discussion}

We terminate the paper with a discussion of strengths and shortcomings
of each application shown: improved structure prediction and
support vector machine classification.

\subsection*{Parametric structure prediction}

For benchmarking purposes in Table~\ref{table:sensitivity}, we
sampled only one low energy structure having H many hairpins, where
in most cases H was taken to be the number of hairpins in the Rfam
consensus structure of the first member of the Rfam family. This explains
how it could happen that the {\rnahairpin} sensitivity for a certain
sequence could at times be different than the {\tt RNAhairpin} sensitivity
for the same sequence, even when the sampled structure and the minimum
free energy structure have the same number of hairpins.
Of course, in general, our code {\tt RNAhairpin} will be used to
sample a large number (1,000 or 10,000) of structures per
sequence.

Since the base pairs that appear in Rfam consensus structures are
inferred by covariation observed in a multiple alignment,
many valid base pairs do not appear in the consensus structure. For
this reason, we did not compute {\em positive predictive rate}, defined
as the ratio of the number of correctly predicted base pairs in the
Rfam consensus structure, divided by the number of base pairs in the
predicted structure. This is the reason that Table~\ref{table:sensitivity}
only reports average sensitivity values.

As previously mentioned, we computed the average sensitivity of
the minimum free energy (MFE) structure obtained from Vienna RNA
Package {\tt RNAfold} \citep{hofacker:ViennaWebServer}, version 1.8.5
without dangles, in order to ensure that both {\rnahairpin} and {\tt RNAfold}
employ the same energy model.  In future work, we plan
to extend {\tt RNAhairpin} to the Turner 2004 energy model with dangles
\citep{Turner.nar09}. By the same token, it is not conceptually difficult
to modify the program {\rnamlnumber}, in order to sample low energy structures
having a specified number of multiloops. Such sampled structures could
yield better structure predictions for certain types of RNA. Finally,
it would be possible to combine the algorithms {\rnahairpin} and {\rnamlnumber}
in order sample structures having {\em both} a specified number of hairpins
and a possibly distinct number of multiloops. Nevertheless, such an
algorithm would run in time $O(H^2 M^2 n^3)$ and space $O(H M n^2)$,
where $H$ [resp. $M$] is an upper bound on the number of hairpins
[resp. multiloops]. For relatively small values of $H,M$, such an
algorithm would indeed be feasible, and could prove useful for certain
classes of RNA, whose function is known to depend on certain structural
motifs.

Table~\ref{table:sensitivity} presents examples of Rfam families, where
the average sensitivity of {\tt RNAhairpin} exceeds that of {\tt RNAfold}.
Improvements were obtained for RNA families, where a certain number of
hairpins are known to be functionally important, as in the cloverleaf
tRNA, typically having three hairpins. In this case, the sensitivity of
{\tt RNAhairpin} exceeds that of {\tt RNAfold} by approximately 20\%.
For certain ribozymes, such as type III hammerhead ribozyme (RF00008)
and glycine ribozyme (RF00504), the improvement was over 4\% resp. 3\%.
Not shown in the table are RNA families, where the sensitivity of
{\tt RNAfold} exceeded that of {\tt RNAhairpin} -- for instance,
for 5S rRNA (RF00001), {\tt RNAhairpin} average sensitivity was
$0.621306$ compared with {\tt RNAfold} sensitivity of $0.633189$;
for purine riboswitches (RF00167), {\tt RNAhairpin} obtained had
$0.811327$ compared with {\tt RNAfold} sensitivity of  $0.856764$.
We believe that {\rnahairpin} showed better sensitivity than {\tt RNAfold}
in the case of tRNA, because of two reasons: (1) tRNA has a well-known
cloverleaf structure usually involving 3 hairpins, and (2) there may
be a large sequence and energy difference among especially bacterial
tRNAs. Item (2) could cause minimum free energy structures to be quite
distinct from the usual cloverleaf -- manual investigation
confirms this hypothesis in some randomly chosen instances, while item
(2) ensures that {\rnahairpin} will sample structures having 3 hairpins,
hence more likely to adopt the functional cloverleaf structure.
It could be that similar reasons explain the small improvement of
{\rnahairpin} over {\tt RNAfold} for some of the other examples, including
certain ribozymes. However, along this line of reasoning, we expected
{\rnahairpin} to outperform {\tt RNAfold} for purine riboswitch aptamers,
which have a very well-defined multiloop with two hairpins;
Table~\ref{table:sensitivity} shows this is not the case.
As described in the caption of
Table~\ref{table:sensitivityMultiloopHairpin}, by sampling low energy
structures that simultaneously have a specified number $m$ of multiloops and
number $h$ of hairpins, we substantially improve the prediction accuracy
of {\tt RNAfold}. However, such improvements tend to occur when the Rfam
families show a prounounced common fold, as in the case of tRNA and certain
ribozymes, and when there are no large loop (undefined) regions in the
Rfam consensus structures.  In any case, we believe that
minimum free energy structure prediction algorithms, such as
{\tt RNAfold}, {\unafold}, {\mfold}, {\tt RNAstructure}, remain
the best universal thermodynamics-based tool for structure prediction.


\subsection*{Features for SVM classifiers}

The development of noncoding RNA gene finders is important for the
analysis and classification of the pervasively transcribed
RNA from the human genome, most of which has no previously known
structure or function.  In this paper, we have described
four novel thermodynamics-based algorithms, {\rnahairpin},
{\rnamlnumber}, {\rnamlorder}, and {\rnahairpinml} which
compute global, parametric features of the ensemble
of low energy secondary structures for a
given RNA sequence. For the first three algorithms, we have
shown that there is a significant {\em global} signal, as witnessed
by ROC area under curve, that suggests that probability and multiloop
formation probabilities present useful features that could be added
to existent noncoding RNA gene finders -- note that this remark only
concerns gene finders for specific noncoding RNA families, not general
ncRNA gene finders.\footnote{Since we only recently implemented
{\rnahairpinml}, we have not evaluated whether ROC AUC is better than
for {\rnahairpin} alone, or for {\rnahairpin} combined with {\rnamlnumber}.}

One of our goals in developing these parametric algorithms
was to provide additional discriminatory features that
can be added to other features within the context of
a support vector machine, in order to improve the accuracy of
noncoding RNA gene finders. Indeed, by adding novel features,
it is known that one can improve the accuracy of SVM classifiers.
For instance, the state-of-the-art precursor
microRNA (pre-miRNA) SVM developed by Ng and Mishra
\citep{Ng.b07} uses features
MFEI2, MFEI1, \%G+C, dP, dG, dQ, dD, dF,,zG, zQ, zD, zP, zF, etc.
(see \citep{Ng.b07} for explanation), which outperforms
the simpler
triplet kernel pre-miRNA SVM developed by Xue et al. \citep{Xue.bb05}.
